// @mui/x-charts/ScatterChart/Scatter.mjs
'use client';

import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
const _excluded = ["ownerState"];
import * as React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import useSlotProps from '@mui/utils/useSlotProps';
import { getInteractionItemProps } from "../hooks/useInteractionItemProps.mjs";
import { useStore } from "../internals/store/useStore.mjs";
import { useItemHighlightStateGetter } from "../hooks/useItemHighlightStateGetter.mjs";
import { selectorChartsIsVoronoiEnabled } from "../internals/plugins/featurePlugins/useChartClosestPoint/index.mjs";
import { ScatterMarker } from "./ScatterMarker.mjs";
import { useUtilityClasses } from "./scatterClasses.mjs";
import { useScatterPlotData } from "./useScatterPlotData.mjs";
import { useChartsContext } from "../context/ChartsProvider/index.mjs";

/**
 * @deprecated The `Scatter` component is an internal implementation detail of `ScatterPlot` and will be removed from the public API in v10. Use `ScatterPlot` instead.
 */

/**
 * @deprecated The `Scatter` component is an internal implementation detail of `ScatterPlot` and will be removed from the public API in v10. Use `ScatterPlot` instead.
 */

/**
 * @deprecated The `Scatter` component is an internal implementation detail of `ScatterPlot` and will be removed from the public API in v10. Use `ScatterPlot` instead.
 */
import { jsx as _jsx } from "react/jsx-runtime";
/**
 * @deprecated The `Scatter` component is an internal implementation detail of `ScatterPlot` and will be removed from the public API in v10. Use `ScatterPlot` instead.
 *
 * Demos:
 *
 * - [Scatter](https://mui.com/x/react-charts/scatter/)
 * - [Scatter demonstration](https://mui.com/x/react-charts/scatter-demo/)
 *
 * API:
 *
 * - [Scatter API](https://mui.com/x/api/charts/scatter/)
 */
function Scatter(props) {
  const {
    series,
    xScale,
    yScale,
    colorGetter,
    sizeGetter,
    onItemClick,
    classes: inClasses,
    slots,
    slotProps
  } = props;
  const {
    instance
  } = useChartsContext();
  const store = useStore();
  const isVoronoiEnabled = store.use(selectorChartsIsVoronoiEnabled);
  const skipInteractionHandlers = isVoronoiEnabled;
  const getHighlightState = useItemHighlightStateGetter();
  const scatterPlotData = useScatterPlotData(series, xScale, yScale, instance.isPointInside);
  const Marker = slots?.marker ?? ScatterMarker;
  const _useSlotProps = useSlotProps({
      elementType: Marker,
      externalSlotProps: slotProps?.marker,
      additionalProps: {
        seriesId: series.id
      },
      ownerState: {}
    }),
    markerProps = _objectWithoutPropertiesLoose(_useSlotProps, _excluded);
  const classes = useUtilityClasses({
    classes: inClasses
  });
  return /*#__PURE__*/_jsx("g", {
    "data-series": series.id,
    className: classes.series,
    children: scatterPlotData.map(dataPoint => {
      const highlightState = getHighlightState(dataPoint);
      const isItemHighlighted = highlightState === 'highlighted';
      const isItemFaded = highlightState === 'faded';
      return /*#__PURE__*/_jsx(Marker, _extends({
        className: clsx(classes.marker, markerProps.className),
        dataIndex: dataPoint.dataIndex,
        color: colorGetter(dataPoint.dataIndex),
        size: sizeGetter(dataPoint.dataIndex),
        isHighlighted: isItemHighlighted,
        isFaded: isItemFaded,
        x: dataPoint.x,
        y: dataPoint.y,
        onClick: onItemClick && (event => onItemClick(event, {
          type: 'scatter',
          seriesId: series.id,
          dataIndex: dataPoint.dataIndex
        })),
        "data-highlighted": isItemHighlighted || undefined,
        "data-faded": isItemFaded || undefined
      }, skipInteractionHandlers ? undefined : getInteractionItemProps(instance, dataPoint), markerProps), dataPoint.id ?? dataPoint.dataIndex);
    })
  });
}
process.env.NODE_ENV !== "production" ? Scatter.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  classes: PropTypes.object,
  color: PropTypes.string.isRequired,
  /**
   * Function to get the color of a scatter item given its data index.
   * The data index argument is optional. If not provided, the color for the entire series is returned.
   * If provided, the color for the specific scatter item is returned.
   */
  colorGetter: PropTypes.func.isRequired,
  /**
   * Callback fired when clicking on a scatter item.
   * @param {MouseEvent} event Mouse event recorded on the `<svg/>` element.
   * @param {ScatterItemIdentifier} scatterItemIdentifier The scatter item identifier.
   */
  onItemClick: PropTypes.func,
  series: PropTypes.object.isRequired,
  /**
   * Function to get the marker size of a scatter item given its data index.
   */
  sizeGetter: PropTypes.func.isRequired,
  slotProps: PropTypes.object,
  slots: PropTypes.object,
  xScale: PropTypes.func.isRequired,
  yScale: PropTypes.func.isRequired
} : void 0;
export { Scatter };
