// @mui/x-charts/ScatterChart/ScatterPlot.mjs
'use client';

import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import { styled } from '@mui/material/styles';
import { Scatter } from "./Scatter.mjs";
import { useScatterSeriesContext } from "../hooks/useScatterSeries.mjs";
import { useXAxes, useYAxes } from "../hooks/index.mjs";
import { useZAxes } from "../hooks/useZAxis.mjs";
import { scatterSeriesConfig } from "./seriesConfig/index.mjs";
import getMarkerSize from "./seriesConfig/getMarkerSize.mjs";
import { BatchScatter } from "./BatchScatter.mjs";
import { ScatterAsync } from "./async/ScatterAsync.mjs";
import { useUtilityClasses } from "./scatterClasses.mjs";
import { useChartsContext } from "../context/ChartsProvider/index.mjs";
import { useStore } from "../internals/store/useStore.mjs";
import { selectorShouldUseProgressiveRenderer } from "../internals/plugins/featurePlugins/useProgressiveRendering/index.mjs";
import { jsx as _jsx } from "react/jsx-runtime";
const EMPTY_SERIES_IDS = [];
const ScatterPlotRoot = styled('g', {
  name: 'MuiScatterPlot',
  slot: 'Root'
})();

/**
 * Demos:
 *
 * - [Scatter](https://mui.com/x/react-charts/scatter/)
 * - [Scatter demonstration](https://mui.com/x/react-charts/scatter-demo/)
 *
 * API:
 *
 * - [ScatterPlot API](https://mui.com/x/api/charts/scatter-plot/)
 */
function ScatterPlot(props) {
  const {
    slots,
    slotProps,
    onItemClick,
    renderer,
    className,
    classes: inClasses
  } = props;
  const seriesData = useScatterSeriesContext();
  const {
    xAxis,
    xAxisIds
  } = useXAxes();
  const {
    yAxis,
    yAxisIds
  } = useYAxes();
  const {
    zAxis,
    zAxisIds
  } = useZAxes();
  const classes = useUtilityClasses({
    classes: inClasses
  });
  const {
    instance
  } = useChartsContext();
  const store = useStore();
  const plotId = React.useId();
  const seriesIds = seriesData?.seriesOrder ?? EMPTY_SERIES_IDS;
  const isProgressive = store.use(selectorShouldUseProgressiveRenderer, seriesIds, renderer);
  React.useEffect(() => instance.registerProgressivePlan(plotId, seriesIds, renderer), [instance, plotId, seriesIds, renderer]);
  if (seriesData === undefined) {
    return null;
  }
  const {
    series,
    seriesOrder
  } = seriesData;
  const defaultXAxisId = xAxisIds[0];
  const defaultYAxisId = yAxisIds[0];
  const defaultZAxisId = zAxisIds[0];
  let DefaultScatterItems;
  if (renderer === 'svg-batch') {
    DefaultScatterItems = BatchScatter;
  } else if (isProgressive) {
    DefaultScatterItems = ScatterAsync;
  } else {
    DefaultScatterItems = Scatter;
  }
  const ScatterItems = slots?.scatter ?? DefaultScatterItems;
  return /*#__PURE__*/_jsx(ScatterPlotRoot, {
    className: clsx(classes.root, className),
    children: seriesOrder.map(seriesId => {
      const {
        id,
        xAxisId,
        yAxisId,
        colorAxisId,
        zAxisId,
        sizeAxisId,
        color,
        hidden
      } = series[seriesId];
      if (hidden) {
        return null;
      }
      const colorGetter = scatterSeriesConfig.colorProcessor(series[seriesId], xAxis[xAxisId ?? defaultXAxisId], yAxis[yAxisId ?? defaultYAxisId], zAxis[colorAxisId ?? zAxisId ?? defaultZAxisId]);
      const sizeGetter = getMarkerSize(series[seriesId], zAxis[sizeAxisId ?? defaultZAxisId]);
      const xScale = xAxis[xAxisId ?? defaultXAxisId].scale;
      const yScale = yAxis[yAxisId ?? defaultYAxisId].scale;
      return /*#__PURE__*/_jsx(ScatterItems, _extends({
        xScale: xScale,
        yScale: yScale,
        color: color,
        colorGetter: colorGetter,
        sizeGetter: sizeGetter,
        series: series[seriesId],
        onItemClick: onItemClick,
        slots: slots,
        slotProps: slotProps
      }, slotProps?.scatter), id);
    })
  });
}
process.env.NODE_ENV !== "production" ? ScatterPlot.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  classes: PropTypes.object,
  className: PropTypes.string,
  /**
   * Callback fired when a marker is clicked directly.
   * For closest-point clicks (the `ScatterChart` default), pass `onItemClick` to the chart container instead.
   * @param {MouseEvent} event Mouse event recorded on the `<svg/>` element.
   * @param {ScatterItemIdentifier} scatterItemIdentifier The scatter item identifier.
   */
  onItemClick: PropTypes.func,
  /**
   * The type of renderer to use for the scatter plot.
   * - `svg-single`: Renders every scatter item in a `<circle />` element, synchronously.
   * - `svg-progressive`: Renders every scatter item in a `<circle />` element, in progressive batches that paint over several animation frames to keep the main thread responsive.
   * - `svg-batch`: Batch renders scatter items in `<path />` elements for better performance with large datasets, at the cost of some limitations.
   *                Read more: https://mui.com/x/react-charts/scatter/#performance
   *
   * When not set, defaults to `svg-single`. Enable the `progressiveRendering` experimental feature to auto-select `svg-progressive` above an internal point-count threshold; this will become the default in the next major version.
   * @default 'svg-single'
   */
  renderer: PropTypes.oneOf(['svg-batch', 'svg-progressive', 'svg-single']),
  /**
   * The props used for each component slot.
   * @default {}
   */
  slotProps: PropTypes.object,
  /**
   * Overridable component slots.
   * @default {}
   */
  slots: PropTypes.object
} : void 0;
export { ScatterPlot };
