// @mui/x-charts/ScatterChart/useScatterChartProps.mjs
'use client';

import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
const _excluded = ["xAxis", "yAxis", "zAxis", "series", "axisHighlight", "hitAreaRadius", "disableHitArea", "hideLegend", "width", "height", "margin", "colors", "sx", "grid", "onItemClick", "children", "slots", "slotProps", "loading", "highlightedItem", "onHighlightChange", "className", "showToolbar", "renderer", "brushConfig"];
import * as React from 'react';
import { SCATTER_CHART_PLUGINS } from "./ScatterChart.plugins.mjs";
/**
 * A helper function that extracts ScatterChartProps from the input props
 * and returns an object with props for the children components of ScatterChart.
 *
 * @param props The input props for ScatterChart
 * @returns An object with props for the children components of ScatterChart
 */
export const useScatterChartProps = props => {
  const {
      xAxis,
      yAxis,
      zAxis,
      series,
      axisHighlight,
      hitAreaRadius,
      disableHitArea,
      width,
      height,
      margin,
      colors,
      sx,
      grid,
      onItemClick,
      children,
      slots,
      slotProps,
      loading,
      highlightedItem,
      onHighlightChange,
      className,
      renderer,
      brushConfig
    } = props,
    other = _objectWithoutPropertiesLoose(props, _excluded);
  const seriesWithDefault = React.useMemo(() => series.map(s => _extends({
    type: 'scatter'
  }, s)), [series]);
  const resolvedDisableHitArea = disableHitArea;
  // 'svg-single' (and the undefined default) is the only renderer with per-item DOM events.
  // Every other renderer needs voronoi-based click handling regardless of disableHitArea.
  const useVoronoiOnItemClick = resolvedDisableHitArea !== true || renderer !== undefined && renderer !== 'svg-single';
  const chartsContainerProps = _extends({}, other, {
    series: seriesWithDefault,
    width,
    height,
    margin,
    colors,
    xAxis,
    yAxis,
    zAxis,
    highlightedItem,
    onHighlightChange,
    disableHitArea,
    hitAreaRadius,
    onItemClick: useVoronoiOnItemClick ? onItemClick : undefined,
    plugins: SCATTER_CHART_PLUGINS,
    slots,
    slotProps,
    brushConfig
  });
  const chartsAxisProps = {
    slots,
    slotProps
  };
  const gridProps = {
    vertical: grid?.vertical,
    horizontal: grid?.horizontal
  };
  const scatterPlotProps = {
    onItemClick: useVoronoiOnItemClick ? undefined : onItemClick,
    slots,
    slotProps,
    renderer
  };
  const overlayProps = {
    loading,
    slots,
    slotProps
  };
  const legendProps = {
    slots,
    slotProps
  };
  const axisHighlightProps = _extends({
    y: 'none',
    x: 'none'
  }, axisHighlight);
  const chartsWrapperProps = {
    sx,
    legendPosition: props.slotProps?.legend?.position,
    legendDirection: props.slotProps?.legend?.direction,
    hideLegend: props.hideLegend ?? false,
    className
  };
  return {
    chartsWrapperProps,
    chartsContainerProps,
    chartsAxisProps,
    gridProps,
    scatterPlotProps,
    overlayProps,
    legendProps,
    axisHighlightProps,
    children
  };
};
