// @mui/x-scheduler/internals/components/event/time-grid-event/TimeGridEvent.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TimeGridEvent = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _styles = require("@mui/material/styles");
var _Typography = _interopRequireDefault(require("@mui/material/Typography"));
var _RepeatRounded = _interopRequireDefault(require("@mui/icons-material/RepeatRounded"));
var _calendarGrid = require("@mui/x-scheduler-internals/calendar-grid");
var _eventDragPreview = require("../../../components/event-drag-preview");
var _useFormatTime = require("../../../hooks/useFormatTime");
var _EventCalendarStyledContext = require("../../../../event-calendar/EventCalendarStyledContext");
var _useTimeGridEvent = require("./useTimeGridEvent");
var _TimeGridEventShared = require("./TimeGridEventShared");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["occurrence", "variant", "className"];
const TimeGridEventRoot = (0, _styles.styled)(_calendarGrid.CalendarGrid.TimeEvent, {
  name: 'MuiEventCalendar',
  slot: 'TimeGridEventRoot'
})(({
  theme
}) => (0, _extends2.default)({}, (0, _TimeGridEventShared.getTimeGridEventRootStyles)(theme), {
  padding: theme.spacing(0.5, 1, 0.5, 1),
  '&[data-under-fifteen-minutes="true"]': {
    padding: theme.spacing(0, 1)
  },
  '&:hover': {
    backgroundColor: 'var(--event-surface-subtle-hover)'
  },
  '&[data-editing]': {
    backgroundColor: 'var(--event-surface-selected)',
    color: 'var(--event-on-surface-selected)',
    '&:hover': {
      backgroundColor: 'var(--event-surface-selected-hover)'
    }
  },
  '&[role="button"]': {
    cursor: 'pointer'
  },
  '&::before': {
    content: '""',
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    width: 3,
    borderRadius: '4px 0 0 4px',
    background: 'var(--event-surface-accent)',
    pointerEvents: 'none'
  },
  '&[data-editing]::before': {
    background: 'var(--event-surface-selected)'
  }
}));
const TimeGridEventTitle = (0, _styles.styled)(_Typography.default, {
  name: 'MuiEventCalendar',
  slot: 'TimeGridEventTitle'
})(({
  theme
}) => (0, _extends2.default)({
  margin: 0,
  color: 'var(--event-on-surface-subtle-primary)',
  fontWeight: theme.typography.fontWeightMedium,
  fontSize: 'var(--EventCalendar-fontSize-eventTitle, 0.75rem)',
  lineHeight: 1.43,
  paddingRight: theme.spacing(1.5),
  height: 'fit-content',
  '@container (max-height: 20px)': {
    fontSize: '11px',
    lineHeight: '11px'
  },
  '[data-editing] &': {
    color: 'var(--event-on-surface-selected)'
  }
}, (0, _TimeGridEventShared.linesClampStyles)(1)));
const TimeGridEventTime = (0, _styles.styled)('time', {
  name: 'MuiEventCalendar',
  slot: 'TimeGridEventTime'
})(({
  theme
}) => ({
  color: 'var(--event-on-surface-subtle-secondary)',
  fontWeight: theme.typography.fontWeightRegular,
  fontSize: 'var(--EventCalendar-fontSize-timeText, 0.75rem)',
  lineHeight: 1.43,
  '[data-editing] &': {
    color: 'var(--event-on-surface-selected)'
  },
  '&[data-lines-clamp]': (0, _extends2.default)({}, (0, _TimeGridEventShared.linesClampStyles)(1), {
    paddingInlineEnd: theme.spacing(1.5)
  }),
  '@container (max-width: 50px & max-height: 50px)': {
    display: 'none'
  },
  '@container (max-height: 20px)': {
    fontSize: '11px',
    lineHeight: '11px'
  }
}));
const TimeGridEventRecurringIcon = (0, _styles.styled)(_RepeatRounded.default, {
  name: 'MuiEventCalendar',
  slot: 'TimeGridEventRecurringIcon'
})(({
  theme
}) => ({
  position: 'absolute',
  right: 3,
  bottom: 3,
  padding: theme.spacing(0.25),
  fontSize: 'var(--EventCalendar-fontSize-recurringIcon, 1.25rem)',
  color: 'var(--event-on-surface-subtle-secondary)',
  '[data-editing] &': {
    color: 'var(--event-on-surface-selected)'
  },
  '@container (max-width: 50px)': {
    display: 'none'
  },
  '@container (max-height: 12px)': {
    fontSize: 14,
    lineHeight: 14,
    padding: 0,
    bottom: 0
  }
}));
const TimeGridEventResizeHandler = (0, _styles.styled)(_calendarGrid.CalendarGrid.TimeEventResizeHandler, {
  name: 'MuiEventCalendar',
  slot: 'TimeGridEventResizeHandler'
})({
  position: 'absolute',
  height: 4,
  left: 0,
  right: 0,
  zIndex: 3,
  cursor: 'ns-resize',
  opacity: 0,
  '*:hover > &': {
    opacity: 1
  },
  '&[data-start]': {
    top: 0
  },
  '&[data-end]': {
    bottom: 0
  }
});
const TimeGridEvent = exports.TimeGridEvent = /*#__PURE__*/React.forwardRef(function TimeGridEvent(props, forwardedRef) {
  const {
      occurrence,
      variant,
      className
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const {
    classes
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const {
    isRecurring,
    isDraggable,
    isStartResizable,
    isEndResizable,
    isLessThan30Minutes,
    isBetween30and60Minutes,
    rootDataAttributes,
    rootPositionProps
  } = (0, _useTimeGridEvent.useTimeGridEvent)(occurrence);
  const formatTime = (0, _useFormatTime.useFormatTime)();
  const content = React.useMemo(() => {
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)(React.Fragment, {
      children: [isLessThan30Minutes || isBetween30and60Minutes ? /*#__PURE__*/(0, _jsxRuntime.jsxs)(TimeGridEventTitle, {
        className: classes.timeGridEventTitle,
        children: [occurrence.title, ' ', /*#__PURE__*/(0, _jsxRuntime.jsx)(TimeGridEventTime, {
          className: classes.timeGridEventTime,
          children: formatTime(occurrence.displayTimezone.start.value)
        })]
      }) : /*#__PURE__*/(0, _jsxRuntime.jsxs)(React.Fragment, {
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(TimeGridEventTitle, {
          className: classes.timeGridEventTitle,
          children: occurrence.title
        }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(TimeGridEventTime, {
          className: classes.timeGridEventTime,
          "data-lines-clamp": true,
          children: [formatTime(occurrence.displayTimezone.start.value), " -", ' ', formatTime(occurrence.displayTimezone.end.value)]
        })]
      }), isRecurring && /*#__PURE__*/(0, _jsxRuntime.jsx)(TimeGridEventRecurringIcon, {
        className: classes.timeGridEventRecurringIcon,
        "aria-hidden": "true"
      })]
    });
  }, [isBetween30and60Minutes, isLessThan30Minutes, occurrence.title, occurrence.displayTimezone.start.value, occurrence.displayTimezone.end.value, formatTime, isRecurring, classes]);
  const sharedProps = (0, _extends2.default)({}, rootPositionProps, {
    ref: forwardedRef
  }, other, {
    className: (0, _clsx.default)(className, occurrence.className)
  });
  if (variant === 'placeholder') {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_TimeGridEventShared.TimeGridEventPlaceholder, (0, _extends2.default)({
      "aria-hidden": true
    }, rootDataAttributes, sharedProps, {
      className: (0, _clsx.default)(classes.timeGridEventPlaceholder, sharedProps.className),
      children: content
    }));
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(TimeGridEventRoot, (0, _extends2.default)({
    isDraggable: isDraggable,
    eventId: occurrence.id,
    occurrenceKey: occurrence.key,
    renderDragPreview: parameters => /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventDragPreview.EventDragPreview, (0, _extends2.default)({}, parameters))
  }, rootDataAttributes, sharedProps, {
    className: (0, _clsx.default)(classes.timeGridEvent, sharedProps.className),
    children: [isStartResizable && /*#__PURE__*/(0, _jsxRuntime.jsx)(TimeGridEventResizeHandler, {
      className: classes.timeGridEventResizeHandler,
      side: "start"
    }), content, isEndResizable && /*#__PURE__*/(0, _jsxRuntime.jsx)(TimeGridEventResizeHandler, {
      className: classes.timeGridEventResizeHandler,
      side: "end"
    })]
  }));
});
if (process.env.NODE_ENV !== "production") TimeGridEvent.displayName = "TimeGridEvent";
