// @mui/x-scheduler/internals/components/event/time-grid-event/TimeGridEventTouch.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TimeGridEventTouch = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _styles = require("@mui/material/styles");
var _Typography = _interopRequireDefault(require("@mui/material/Typography"));
var _calendarGrid = require("@mui/x-scheduler-internals/calendar-grid");
var _eventDragPreview = require("../../../components/event-drag-preview");
var _EventCalendarStyledContext = require("../../../../event-calendar/EventCalendarStyledContext");
var _useTimeGridEvent = require("./useTimeGridEvent");
var _TimeGridEventShared = require("./TimeGridEventShared");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["occurrence", "variant", "className"];
const TimeGridEventTouchRoot = (0, _styles.styled)(_calendarGrid.CalendarGrid.TimeEvent, {
  name: 'MuiEventCalendar',
  slot: 'TimeGridEventTouchRoot'
})(({
  theme
}) => (0, _extends2.default)({}, (0, _TimeGridEventShared.getTimeGridEventRootStyles)(theme), {
  padding: theme.spacing(0.5, 0.7, 0.5, 0.7),
  '&[data-under-fifteen-minutes="true"]': {
    padding: theme.spacing(0, 0.5)
  },
  '&[data-editing]': {
    backgroundColor: 'var(--event-surface-selected)',
    color: 'var(--event-on-surface-selected)'
  }
}));
const TOUCH_TITLE_LINE_HEIGHT_PX = 12;
const TOUCH_TITLE_MAX_LINES = 24;
const touchTitleLineClampSteps = {};
for (let lines = 2; lines <= TOUCH_TITLE_MAX_LINES; lines += 1) {
  const minHeight = lines * TOUCH_TITLE_LINE_HEIGHT_PX;
  touchTitleLineClampSteps[`@container (min-height: ${minHeight}px)`] = {
    WebkitLineClamp: lines,
    maxHeight: `${lines * TOUCH_TITLE_LINE_HEIGHT_PX}px`
  };
}
const TimeGridEventTouchTitle = (0, _styles.styled)(_Typography.default, {
  name: 'MuiEventCalendar',
  slot: 'TimeGridEventTouchTitle'
})(({
  theme
}) => (0, _extends2.default)({
  margin: 0,
  color: 'var(--event-on-surface-subtle-primary)',
  fontWeight: theme.typography.fontWeightMedium,
  fontSize: 'var(--EventCalendar-fontSize-eventTitle, 0.75rem)',
  lineHeight: `${TOUCH_TITLE_LINE_HEIGHT_PX}px`,
  '[data-editing] &': {
    color: 'var(--event-on-surface-selected)'
  }
}, (0, _TimeGridEventShared.linesClampStyles)(1), {
  maxHeight: `${TOUCH_TITLE_LINE_HEIGHT_PX}px`
}, touchTitleLineClampSteps));
const TimeGridEventTouchResizeHandler = (0, _styles.styled)(_calendarGrid.CalendarGrid.TimeEventResizeHandler, {
  name: 'MuiEventCalendar',
  slot: 'TimeGridEventTouchResizeHandler'
})({
  position: 'absolute',
  height: 4,
  left: 0,
  right: 0,
  zIndex: 3,
  cursor: 'ns-resize',
  '&[data-start]': {
    top: 0
  },
  '&[data-end]': {
    bottom: 0
  }
});
const TimeGridEventTouch = exports.TimeGridEventTouch = /*#__PURE__*/React.forwardRef(function TimeGridEventTouch(props, forwardedRef) {
  const {
      occurrence,
      variant,
      className
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const {
    classes
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const {
    isDraggable,
    isStartResizable,
    isEndResizable,
    rootDataAttributes,
    rootPositionProps
  } = (0, _useTimeGridEvent.useTimeGridEvent)(occurrence);
  const content = /*#__PURE__*/(0, _jsxRuntime.jsx)(TimeGridEventTouchTitle, {
    className: classes.timeGridEventTitle,
    children: occurrence.title
  });
  const sharedProps = (0, _extends2.default)({}, rootPositionProps, {
    ref: forwardedRef
  }, other, {
    className: (0, _clsx.default)(className, occurrence.className)
  });
  if (variant === 'placeholder') {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_TimeGridEventShared.TimeGridEventPlaceholder, (0, _extends2.default)({
      "aria-hidden": true
    }, rootDataAttributes, sharedProps, {
      className: (0, _clsx.default)(classes.timeGridEventPlaceholder, sharedProps.className),
      children: content
    }));
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(TimeGridEventTouchRoot, (0, _extends2.default)({
    isDraggable: isDraggable,
    eventId: occurrence.id,
    occurrenceKey: occurrence.key,
    renderDragPreview: parameters => /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventDragPreview.EventDragPreview, (0, _extends2.default)({}, parameters))
  }, rootDataAttributes, sharedProps, {
    className: (0, _clsx.default)(classes.timeGridEvent, sharedProps.className),
    children: [isStartResizable && /*#__PURE__*/(0, _jsxRuntime.jsx)(TimeGridEventTouchResizeHandler, {
      className: classes.timeGridEventResizeHandler,
      side: "start"
    }), content, isEndResizable && /*#__PURE__*/(0, _jsxRuntime.jsx)(TimeGridEventTouchResizeHandler, {
      className: classes.timeGridEventResizeHandler,
      side: "end"
    })]
  }));
});
if (process.env.NODE_ENV !== "production") TimeGridEventTouch.displayName = "TimeGridEventTouch";
