// @mui/x-scheduler/internals/components/event/time-grid-event/TimeGridEventTouch.mjs
'use client';

import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import _extends from "@babel/runtime/helpers/esm/extends";
const _excluded = ["occurrence", "variant", "className"];
import * as React from 'react';
import clsx from 'clsx';
import { styled } from '@mui/material/styles';
import Typography from '@mui/material/Typography';
import { CalendarGrid } from '@mui/x-scheduler-internals/calendar-grid';
import { EventDragPreview } from "../../event-drag-preview/index.mjs";
import { useEventCalendarStyledContext } from "../../../../event-calendar/EventCalendarStyledContext.mjs";
import { useTimeGridEvent } from "./useTimeGridEvent.mjs";
import { getTimeGridEventRootStyles, linesClampStyles, TimeGridEventPlaceholder } from "./TimeGridEventShared.mjs";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const TimeGridEventTouchRoot = styled(CalendarGrid.TimeEvent, {
  name: 'MuiEventCalendar',
  slot: 'TimeGridEventTouchRoot'
})(({
  theme
}) => _extends({}, getTimeGridEventRootStyles(theme), {
  padding: theme.spacing(0.5, 0.7, 0.5, 0.7),
  '&[data-under-fifteen-minutes="true"]': {
    padding: theme.spacing(0, 0.5)
  },
  '&[data-editing]': {
    backgroundColor: 'var(--event-surface-selected)',
    color: 'var(--event-on-surface-selected)'
  }
}));
const TOUCH_TITLE_LINE_HEIGHT_PX = 12;
const TOUCH_TITLE_MAX_LINES = 24;
const touchTitleLineClampSteps = {};
for (let lines = 2; lines <= TOUCH_TITLE_MAX_LINES; lines += 1) {
  const minHeight = lines * TOUCH_TITLE_LINE_HEIGHT_PX;
  touchTitleLineClampSteps[`@container (min-height: ${minHeight}px)`] = {
    WebkitLineClamp: lines,
    maxHeight: `${lines * TOUCH_TITLE_LINE_HEIGHT_PX}px`
  };
}
const TimeGridEventTouchTitle = styled(Typography, {
  name: 'MuiEventCalendar',
  slot: 'TimeGridEventTouchTitle'
})(({
  theme
}) => _extends({
  margin: 0,
  color: 'var(--event-on-surface-subtle-primary)',
  fontWeight: theme.typography.fontWeightMedium,
  fontSize: 'var(--EventCalendar-fontSize-eventTitle, 0.75rem)',
  lineHeight: `${TOUCH_TITLE_LINE_HEIGHT_PX}px`,
  '[data-editing] &': {
    color: 'var(--event-on-surface-selected)'
  }
}, linesClampStyles(1), {
  maxHeight: `${TOUCH_TITLE_LINE_HEIGHT_PX}px`
}, touchTitleLineClampSteps));
const TimeGridEventTouchResizeHandler = styled(CalendarGrid.TimeEventResizeHandler, {
  name: 'MuiEventCalendar',
  slot: 'TimeGridEventTouchResizeHandler'
})({
  position: 'absolute',
  height: 4,
  left: 0,
  right: 0,
  zIndex: 3,
  cursor: 'ns-resize',
  '&[data-start]': {
    top: 0
  },
  '&[data-end]': {
    bottom: 0
  }
});
export const TimeGridEventTouch = /*#__PURE__*/React.forwardRef(function TimeGridEventTouch(props, forwardedRef) {
  const {
      occurrence,
      variant,
      className
    } = props,
    other = _objectWithoutPropertiesLoose(props, _excluded);
  const {
    classes
  } = useEventCalendarStyledContext();
  const {
    isDraggable,
    isStartResizable,
    isEndResizable,
    rootDataAttributes,
    rootPositionProps
  } = useTimeGridEvent(occurrence);
  const content = /*#__PURE__*/_jsx(TimeGridEventTouchTitle, {
    className: classes.timeGridEventTitle,
    children: occurrence.title
  });
  const sharedProps = _extends({}, rootPositionProps, {
    ref: forwardedRef
  }, other, {
    className: clsx(className, occurrence.className)
  });
  if (variant === 'placeholder') {
    return /*#__PURE__*/_jsx(TimeGridEventPlaceholder, _extends({
      "aria-hidden": true
    }, rootDataAttributes, sharedProps, {
      className: clsx(classes.timeGridEventPlaceholder, sharedProps.className),
      children: content
    }));
  }
  return /*#__PURE__*/_jsxs(TimeGridEventTouchRoot, _extends({
    isDraggable: isDraggable,
    eventId: occurrence.id,
    occurrenceKey: occurrence.key,
    renderDragPreview: parameters => /*#__PURE__*/_jsx(EventDragPreview, _extends({}, parameters))
  }, rootDataAttributes, sharedProps, {
    className: clsx(classes.timeGridEvent, sharedProps.className),
    children: [isStartResizable && /*#__PURE__*/_jsx(TimeGridEventTouchResizeHandler, {
      className: classes.timeGridEventResizeHandler,
      side: "start"
    }), content, isEndResizable && /*#__PURE__*/_jsx(TimeGridEventTouchResizeHandler, {
      className: classes.timeGridEventResizeHandler,
      side: "end"
    })]
  }));
});
if (process.env.NODE_ENV !== "production") TimeGridEventTouch.displayName = "TimeGridEventTouch";
