// @mui/x-charts/RadarChart/RadarSeriesPlot/RadarSeriesMarks.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RadarSeriesMarks = RadarSeriesMarks;
exports.getCircleProps = getCircleProps;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _useRadarSeriesData = require("./useRadarSeriesData");
var _radarClasses = require("../radarClasses");
var _useItemHighlightStateGetter = require("../../hooks/useItemHighlightStateGetter");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["seriesId", "onItemClick", "classes"];
function getCircleProps(params) {
  const {
    getHighlightState,
    seriesId,
    classes,
    point,
    fillArea,
    color
  } = params;
  const highlightState = getHighlightState({
    type: 'radar',
    seriesId
  });
  const isItemHighlighted = highlightState === 'highlighted';
  const isItemFaded = highlightState === 'faded';
  return {
    cx: point.x,
    cy: point.y,
    r: 3,
    fill: color,
    stroke: color,
    opacity: fillArea && isItemFaded ? 0.5 : 1,
    className: classes.seriesMark,
    'data-highlighted': isItemHighlighted || undefined,
    'data-faded': isItemFaded || undefined
  };
}
function RadarSeriesMarks(props) {
  const {
      onItemClick,
      classes: inClasses
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const seriesCoordinates = (0, _useRadarSeriesData.useRadarSeriesData)(props.seriesId);
  const classes = (0, _radarClasses.useUtilityClasses)(inClasses);
  const getHighlightState = (0, _useItemHighlightStateGetter.useItemHighlightStateGetter)();
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(React.Fragment, {
    children: seriesCoordinates?.map(({
      seriesId: id,
      points,
      hideMark,
      fillArea,
      hidden
    }) => {
      if (hideMark || hidden) {
        return null;
      }
      return /*#__PURE__*/(0, _jsxRuntime.jsx)("g", {
        "data-series": id,
        children: points.map((point, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)("circle", (0, _extends2.default)({}, getCircleProps({
          seriesId: id,
          point,
          color: point.color,
          fillArea,
          getHighlightState,
          classes
        }), {
          pointerEvents: onItemClick ? undefined : 'none',
          onClick: event => onItemClick?.(event, {
            type: 'radar',
            seriesId: id,
            dataIndex: index
          }),
          cursor: onItemClick ? 'pointer' : 'unset'
        }, other), index))
      }, id);
    })
  });
}
process.env.NODE_ENV !== "production" ? RadarSeriesMarks.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * Override or extend the styles applied to the component.
   */
  classes: _propTypes.default.object,
  /**
   * A CSS class name applied to the root element.
   */
  className: _propTypes.default.string,
  /**
   * Callback fired when a mark is clicked.
   * @param {React.MouseEvent<SVGPathElement, MouseEvent>} event The event source of the callback.
   * @param {RadarItemIdentifier} radarItemIdentifier The radar item identifier.
   */
  onItemClick: _propTypes.default.func,
  /**
   * The id of the series to display.
   * If undefined all series are displayed.
   */
  seriesId: _propTypes.default.string
} : void 0;
