// @mui/x-charts/RadarChart/RadarSeriesPlot/RadarSeriesMarks.mjs
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
const _excluded = ["seriesId", "onItemClick", "classes"];
import * as React from 'react';
import PropTypes from 'prop-types';
import { useRadarSeriesData } from "./useRadarSeriesData.mjs";
import { useUtilityClasses } from "../radarClasses.mjs";
import { useItemHighlightStateGetter } from "../../hooks/useItemHighlightStateGetter.mjs";
import { jsx as _jsx } from "react/jsx-runtime";
export function getCircleProps(params) {
  const {
    getHighlightState,
    seriesId,
    classes,
    point,
    fillArea,
    color
  } = params;
  const highlightState = getHighlightState({
    type: 'radar',
    seriesId
  });
  const isItemHighlighted = highlightState === 'highlighted';
  const isItemFaded = highlightState === 'faded';
  return {
    cx: point.x,
    cy: point.y,
    r: 3,
    fill: color,
    stroke: color,
    opacity: fillArea && isItemFaded ? 0.5 : 1,
    className: classes.seriesMark,
    'data-highlighted': isItemHighlighted || undefined,
    'data-faded': isItemFaded || undefined
  };
}
function RadarSeriesMarks(props) {
  const {
      onItemClick,
      classes: inClasses
    } = props,
    other = _objectWithoutPropertiesLoose(props, _excluded);
  const seriesCoordinates = useRadarSeriesData(props.seriesId);
  const classes = useUtilityClasses(inClasses);
  const getHighlightState = useItemHighlightStateGetter();
  return /*#__PURE__*/_jsx(React.Fragment, {
    children: seriesCoordinates?.map(({
      seriesId: id,
      points,
      hideMark,
      fillArea,
      hidden
    }) => {
      if (hideMark || hidden) {
        return null;
      }
      return /*#__PURE__*/_jsx("g", {
        "data-series": id,
        children: points.map((point, index) => /*#__PURE__*/_jsx("circle", _extends({}, getCircleProps({
          seriesId: id,
          point,
          color: point.color,
          fillArea,
          getHighlightState,
          classes
        }), {
          pointerEvents: onItemClick ? undefined : 'none',
          onClick: event => onItemClick?.(event, {
            type: 'radar',
            seriesId: id,
            dataIndex: index
          }),
          cursor: onItemClick ? 'pointer' : 'unset'
        }, other), index))
      }, id);
    })
  });
}
process.env.NODE_ENV !== "production" ? RadarSeriesMarks.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * Override or extend the styles applied to the component.
   */
  classes: PropTypes.object,
  /**
   * A CSS class name applied to the root element.
   */
  className: PropTypes.string,
  /**
   * Callback fired when a mark is clicked.
   * @param {React.MouseEvent<SVGPathElement, MouseEvent>} event The event source of the callback.
   * @param {RadarItemIdentifier} radarItemIdentifier The radar item identifier.
   */
  onItemClick: PropTypes.func,
  /**
   * The id of the series to display.
   * If undefined all series are displayed.
   */
  seriesId: PropTypes.string
} : void 0;
export { RadarSeriesMarks };
