// @base-ui/react/drawer/index.parts.mjs
export { DrawerBackdrop as Backdrop } from "./backdrop/DrawerBackdrop.mjs";
export { DrawerClose as Close } from "./close/DrawerClose.mjs";
export { DrawerContent as Content } from "./content/DrawerContent.mjs";
export { DrawerDescription as Description } from "./description/DrawerDescription.mjs";
export { DrawerIndent as Indent } from "./indent/DrawerIndent.mjs";
export { DrawerIndentBackground as IndentBackground } from "./indent-background/DrawerIndentBackground.mjs";
export { DrawerPopup as Popup } from "./popup/DrawerPopup.mjs";
export { DrawerPortal as Portal } from "./portal/DrawerPortal.mjs";
export { DrawerProvider as Provider } from "./provider/DrawerProvider.mjs";
export { DrawerRoot as Root } from "./root/DrawerRoot.mjs";
export { DrawerSwipeArea as SwipeArea } from "./swipe-area/DrawerSwipeArea.mjs";
export { DrawerTitle as Title } from "./title/DrawerTitle.mjs";
export { DrawerTrigger as Trigger } from "./trigger/DrawerTrigger.mjs";
export { DrawerViewport as Viewport } from "./viewport/DrawerViewport.mjs";
export { DrawerVirtualKeyboardProvider as VirtualKeyboardProvider } from "./virtual-keyboard-provider/DrawerVirtualKeyboardProvider.mjs";
export { createDialogHandle as createHandle, DialogHandle as Handle } from "../dialog/store/DialogHandle.mjs";
