// @mui/x-charts/internals/plugins/featurePlugins/useProgressiveRendering/useProgressiveRendering.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useProgressiveRendering = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _useEventCallback = _interopRequireDefault(require("@mui/utils/useEventCallback"));
var _useProgressiveRendering = require("./useProgressiveRendering.selectors");
var _useChartCartesianAxis = require("../useChartCartesianAxis");
const EMPTY_PLANS = new Map();

/** Rounds revealed per tick. One round adds a batch to every series at once. */
const REVEAL_ROUNDS_PER_FRAME = 1;

/** Frames skipped between reveal ticks. `0` = every frame; higher = more headroom. */
const REVEAL_FRAMES_SKIPPED = 0;

/** Rounds kept visible during a zoom/pan interaction (first level only). */
const INTERACTION_REVEALED_ROUNDS = 1;

/** Settle delay before the reveal resumes after an interaction ends (ms). */
const RESUME_AFTER_INTERACTION_DELAY = 200;

/**
 * Chart-wide progressive rendering coordinator. Lives on the store so every
 * renderer in the chart shares one scheduler: each registers a plan via
 * `registerProgressivePlan`, and the plugin ramps a global "rounds" counter,
 * one round adding a batch to every series at once.
 */
const useProgressiveRendering = ({
  store
}) => {
  const registerProgressivePlan = (0, _useEventCallback.default)((plotId, seriesIds, renderer) => {
    if (!(0, _useProgressiveRendering.selectorShouldUseProgressiveRenderer)(store.state, seriesIds, renderer)) {
      return undefined;
    }
    const current = store.state.progressiveRendering.plans.get(plotId);
    if (!(0, _useProgressiveRendering.sameSeriesIds)(current, seriesIds)) {
      const nextPlans = new Map(store.state.progressiveRendering.plans);
      nextPlans.set(plotId, seriesIds);
      // A real plan change restarts the progressive paint.
      store.set('progressiveRendering', {
        plans: nextPlans,
        revealedRounds: 0
      });
    }
    return () => {
      if (!store.state.progressiveRendering.plans.has(plotId)) {
        return;
      }
      const nextPlans = new Map(store.state.progressiveRendering.plans);
      nextPlans.delete(plotId);
      store.set('progressiveRendering', (0, _extends2.default)({}, store.state.progressiveRendering, {
        plans: nextPlans
      }));
    };
  });
  const plans = store.use(_useProgressiveRendering.selectorProgressivePlans) ?? EMPTY_PLANS;
  const isZoomInteracting = store.use(_useChartCartesianAxis.selectorChartZoomIsInteracting);
  React.useEffect(() => {
    const startTotal = (0, _useProgressiveRendering.selectorProgressiveTotalRounds)(store.state);
    if (startTotal === 0) {
      return undefined;
    }

    // While interacting, keep only the first level and pause; resets the wave
    // so it restarts from the first level once the interaction ends.
    if (isZoomInteracting) {
      const target = Math.min(INTERACTION_REVEALED_ROUNDS, startTotal);
      if ((0, _useProgressiveRendering.selectorProgressiveRevealedRounds)(store.state) !== target) {
        store.set('progressiveRendering', (0, _extends2.default)({}, store.state.progressiveRendering, {
          revealedRounds: target
        }));
      }
      return undefined;
    }
    if (typeof requestAnimationFrame !== 'function') {
      store.set('progressiveRendering', (0, _extends2.default)({}, store.state.progressiveRendering, {
        revealedRounds: startTotal
      }));
      return undefined;
    }
    let frame = 0;
    let resumeTimeout;
    let cancelled = false;
    // Resume from the rounds already revealed (keeps the first level on screen).
    // Closure var, not a state-updater derivation: StrictMode double-invokes
    // those, scheduling the frame chain twice.
    let revealed = (0, _useProgressiveRendering.selectorProgressiveRevealedRounds)(store.state);
    function scheduleNext() {
      let remaining = REVEAL_FRAMES_SKIPPED;
      const tick = () => {
        if (cancelled) {
          return;
        }
        if (remaining > 0) {
          remaining -= 1;
          frame = requestAnimationFrame(tick);
          return;
        }
        step();
      };
      frame = requestAnimationFrame(tick);
    }
    function step() {
      if (cancelled) {
        return;
      }
      const total = (0, _useProgressiveRendering.selectorProgressiveTotalRounds)(store.state);
      revealed = Math.min(total, revealed + REVEAL_ROUNDS_PER_FRAME);
      store.set('progressiveRendering', (0, _extends2.default)({}, store.state.progressiveRendering, {
        revealedRounds: revealed
      }));
      if (revealed < total) {
        scheduleNext();
      }
    }
    if (revealed >= startTotal) {
      return undefined;
    }

    // `revealed > 0` means resuming after an interaction: wait the settle delay.
    // Initial paint (`revealed === 0`) starts immediately.
    if (revealed > 0) {
      resumeTimeout = setTimeout(() => {
        frame = requestAnimationFrame(step);
      }, RESUME_AFTER_INTERACTION_DELAY);
    } else {
      frame = requestAnimationFrame(step);
    }
    return () => {
      cancelled = true;
      cancelAnimationFrame(frame);
      if (resumeTimeout !== undefined) {
        clearTimeout(resumeTimeout);
      }
    };
  }, [plans, isZoomInteracting, store]);
  return {
    instance: {
      registerProgressivePlan
    }
  };
};
exports.useProgressiveRendering = useProgressiveRendering;
useProgressiveRendering.params = {};
useProgressiveRendering.getInitialState = () => ({
  progressiveRendering: {
    plans: EMPTY_PLANS,
    revealedRounds: 0
  }
});
