// @mui/x-charts/internals/plugins/featurePlugins/useProgressiveRendering/useProgressiveRendering.selectors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRevealedBatchCount = getRevealedBatchCount;
exports.sameSeriesIds = sameSeriesIds;
exports.selectorShouldUseProgressiveRenderer = exports.selectorProgressiveTotalRounds = exports.selectorProgressiveSeriesRevealedBatches = exports.selectorProgressiveRevealedRounds = exports.selectorProgressivePlans = exports.selectorProgressiveBatchSize = exports.selectorProgressiveAggregate = void 0;
var _store = require("@mui/x-internals/store");
var _useChartSeries = require("../../corePlugins/useChartSeries");
var _useChartExperimentalFeature = require("../../corePlugins/useChartExperimentalFeature");
var _useChartCartesianAxis = require("../useChartCartesianAxis");
/** Auto mode switches to the progressive renderer above this total point count. */
const PROGRESSIVE_POINT_THRESHOLD = 20000;

/** Target reveal commits. Progressive wall time ≈ `(nb_commits + 1) / 2` × a sync render. */
const TARGET_PROGRESSIVE_COMMITS = 5;

/** Min per-tick reveal budget (total points). Avoids tiny React-bound commits. */
const MIN_BATCH_TOTAL = 1000;

/** Max per-tick reveal budget (total points). Caps main-thread block per commit. */
const MAX_BATCH_TOTAL = 10000;

/**
 * Per-series points per tick: total budget aims for
 * {@link TARGET_PROGRESSIVE_COMMITS} commits, clamped by
 * {@link MIN_BATCH_TOTAL}/{@link MAX_BATCH_TOTAL}, split evenly across series.
 */
const getEffectiveBatchSize = (nSeries, totalPoints) => {
  const safeSeries = Math.max(1, nSeries);
  const safePoints = Math.max(1, totalPoints);
  const totalPerTick = Math.min(MAX_BATCH_TOTAL, Math.max(MIN_BATCH_TOTAL, Math.ceil(safePoints / TARGET_PROGRESSIVE_COMMITS)));
  return Math.max(1, Math.floor(totalPerTick / safeSeries));
};
const selectorProgressiveState = state => state.progressiveRendering;

/** Map of registered plots → their set of series ids. */
const selectorProgressivePlans = exports.selectorProgressivePlans = (0, _store.createSelector)(selectorProgressiveState, s => s?.plans);

/** Total number of rounds revealed across all plots so far. */
const selectorProgressiveRevealedRounds = exports.selectorProgressiveRevealedRounds = (0, _store.createSelector)(selectorProgressiveState, s => s?.revealedRounds ?? 0);

/** Point count of a series, looked up across every processed series type. */
function getSeriesPointCount(processedSeries, seriesId) {
  for (const type in processedSeries) {
    if (!Object.hasOwn(processedSeries, type)) {
      continue;
    }
    const item = processedSeries[type]?.series[seriesId];
    if (item) {
      return Array.isArray(item.data) ? item.data.length : 0;
    }
  }
  return 0;
}

/**
 * Aggregate of every plan: per-series batch counts, total rounds, and batch
 * size (so every consumer sizes batches the same). Point counts read from the
 * processed series, not the registration.
 */
const selectorProgressiveAggregate = exports.selectorProgressiveAggregate = (0, _store.createSelectorMemoized)(selectorProgressivePlans, _useChartSeries.selectorChartSeriesProcessed, function selectorProgressiveAggregate(plans, processedSeries) {
  const nBatchesBySeries = new Map();
  if (!plans || plans.size === 0) {
    return {
      nBatchesBySeries,
      totalRounds: 0,
      batchSize: MIN_BATCH_TOTAL
    };
  }
  let nSeries = 0;
  let totalPoints = 0;
  const pointCounts = new Map();
  plans.forEach(seriesIds => {
    seriesIds.forEach(seriesId => {
      const nPoints = getSeriesPointCount(processedSeries, seriesId);
      pointCounts.set(seriesId, nPoints);
      nSeries += 1;
      totalPoints += nPoints;
    });
  });
  const batchSize = getEffectiveBatchSize(nSeries, totalPoints);
  let totalRounds = 0;
  pointCounts.forEach((nPoints, seriesId) => {
    const n = Math.max(1, Math.ceil(nPoints / batchSize));
    nBatchesBySeries.set(seriesId, n);
    if (n > totalRounds) {
      totalRounds = n;
    }
  });
  return {
    nBatchesBySeries,
    totalRounds,
    batchSize
  };
});
const selectorProgressiveBatchSize = exports.selectorProgressiveBatchSize = (0, _store.createSelector)(selectorProgressiveAggregate, a => a.batchSize);
const selectorProgressiveTotalRounds = exports.selectorProgressiveTotalRounds = (0, _store.createSelector)(selectorProgressiveAggregate, a => a.totalRounds);

/**
 * Revealed batch count for a series: `revealed` rounds normally, clamped to the
 * first batch while interacting, never above `total`.
 */
function getRevealedBatchCount(total, revealed, isInteracting) {
  const effectiveRevealed = isInteracting ? 1 : revealed;
  return Math.min(effectiveRevealed, total);
}

/**
 * How many of `seriesId`'s batches are revealed, capped at its total batch
 * count. Clamped to the first batch while interacting.
 */
const selectorProgressiveSeriesRevealedBatches = exports.selectorProgressiveSeriesRevealedBatches = (0, _store.createSelector)(selectorProgressiveAggregate, selectorProgressiveRevealedRounds, _useChartCartesianAxis.selectorChartZoomIsInteracting, function selectorProgressiveSeriesRevealedBatches(agg, revealed, isInteracting, seriesId) {
  const total = agg.nBatchesBySeries.get(seriesId) ?? 0;
  return getRevealedBatchCount(total, revealed, isInteracting);
});

/**
 * Whether `seriesIds` render progressively:
 * - `svg-single`/`svg-batch`: never.
 * - `svg-progressive`: always.
 * - unset (auto): only when the experimental feature is on and total points
 *   exceed {@link PROGRESSIVE_POINT_THRESHOLD}.
 */
const selectorProgressiveRenderingEnabled = state => (0, _useChartExperimentalFeature.selectorChartExperimentalFeaturesState)(state, 'progressiveRendering');
const selectorShouldUseProgressiveRenderer = exports.selectorShouldUseProgressiveRenderer = (0, _store.createSelector)(_useChartSeries.selectorChartSeriesProcessed, selectorProgressiveRenderingEnabled, (processedSeries, progressiveRenderingEnabled, seriesIds, renderer) => {
  if (renderer === 'svg-single' || renderer === 'svg-batch') {
    return false;
  }
  if (renderer === 'svg-progressive') {
    return true;
  }
  if (!progressiveRenderingEnabled) {
    return false;
  }
  let totalPoints = 0;
  for (const seriesId of seriesIds) {
    totalPoints += getSeriesPointCount(processedSeries, seriesId);
  }
  return totalPoints > PROGRESSIVE_POINT_THRESHOLD;
});

/** Order-sensitive equality between two registered series-id sets. */
function sameSeriesIds(a, b) {
  if (!a || a.length !== b.length) {
    return false;
  }
  for (let i = 0; i < a.length; i += 1) {
    if (a[i] !== b[i]) {
      return false;
    }
  }
  return true;
}
