// @mui/x-scheduler/internals/components/event/day-grid-event/DayGridEvent.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DayGridEvent = void 0;
var _formatErrorMessage2 = _interopRequireDefault(require("@mui/x-internals/formatErrorMessage"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _styles = require("@mui/material/styles");
var _store = require("@base-ui/utils/store");
var _RepeatRounded = _interopRequireDefault(require("@mui/icons-material/RepeatRounded"));
var _calendarGrid = require("@mui/x-scheduler-internals/calendar-grid");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _eventCalendarSelectors = require("@mui/x-scheduler-internals/event-calendar-selectors");
var _eventUtils = require("../../../utils/event-utils");
var _eventDragPreview = require("../../../components/event-drag-preview");
var _useFormatTime = require("../../../hooks/useFormatTime");
var _tokens = require("../../../utils/tokens");
var _EventCalendarStyledContext = require("../../../../event-calendar/EventCalendarStyledContext");
var _eventCalendarClasses = require("../../../../event-calendar/eventCalendarClasses");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["occurrence", "variant", "style", "className"];
const ARROW_DEPTH = 8; // px - depth of the chevron point
const LEFT_ARROW_CLIP = `polygon(${ARROW_DEPTH}px 0, 100% 0, 100% 100%, ${ARROW_DEPTH}px 100%, 0 50%)`;
const RIGHT_ARROW_CLIP = `polygon(0 0, calc(100% - ${ARROW_DEPTH}px) 0, 100% 50%, calc(100% - ${ARROW_DEPTH}px) 100%, 0 100%)`;
const BOTH_ARROWS_CLIP = `polygon(${ARROW_DEPTH}px 0, calc(100% - ${ARROW_DEPTH}px) 0, 100% 50%, calc(100% - ${ARROW_DEPTH}px) 100%, ${ARROW_DEPTH}px 100%, 0 50%)`;
const DayGridEventBaseStyles = theme => ({
  containerType: 'inline-size',
  borderRadius: theme.shape.borderRadius * 0.75,
  minWidth: 18,
  height: 18,
  cursor: 'pointer',
  position: 'relative',
  zIndex: 1,
  gridRow: 'var(--grid-row)',
  gridColumn: 1,
  padding: `0 ${theme.spacing(0.5)}`,
  display: 'flex',
  alignItems: 'center',
  gap: theme.spacing(1),
  width: `calc(var(--grid-column-span) * 100% + (var(--grid-column-span) - 1) * (2 * ${theme.spacing(0.5)} + 1px))`,
  '&[data-dragging], &[data-resizing]': {
    opacity: 0.5
  },
  variants: (0, _tokens.getPaletteVariants)(theme)
});
const DayGridEventRoot = (0, _styles.styled)(_calendarGrid.CalendarGrid.DayEvent, {
  name: 'MuiEventCalendar',
  slot: 'DayGridEvent'
})(({
  theme
}) => (0, _extends2.default)({}, DayGridEventBaseStyles(theme), {
  '&:focus-visible': {
    outline: '2px solid var(--event-surface-accent)',
    outlineOffset: 1
  },
  '&[data-variant="filled"]': {
    backgroundColor: 'var(--event-surface-bold)',
    '&:active': {},
    '&:hover': {
      backgroundColor: 'var(--event-surface-bold-hover)'
    },
    '&[data-editing]': {
      backgroundColor: 'var(--event-surface-selected)',
      '&:hover': {
        backgroundColor: 'var(--event-surface-selected-hover)'
      }
    },
    [`& .${_eventCalendarClasses.eventCalendarClasses.dayGridEventRecurringIcon}`]: {
      color: 'var(--event-on-surface-bold)'
    },
    '&[data-starting-before-edge]': {
      borderTopLeftRadius: 0,
      borderBottomLeftRadius: 0,
      clipPath: LEFT_ARROW_CLIP,
      paddingLeft: ARROW_DEPTH + 8
    },
    '&[data-ending-after-edge]': {
      borderTopRightRadius: 0,
      borderBottomRightRadius: 0,
      clipPath: RIGHT_ARROW_CLIP,
      paddingRight: ARROW_DEPTH + 8
    },
    '&[data-starting-before-edge][data-ending-after-edge]': {
      clipPath: BOTH_ARROWS_CLIP
    },
    '&[data-starting-before-edge]:focus-visible, &[data-ending-after-edge]:focus-visible': {
      clipPath: 'none',
      borderRadius: theme.shape.borderRadius * 0.75
    }
  },
  '&[data-variant="invisible"]': {
    width: '100%',
    visibility: 'hidden',
    height: 18
  },
  '&[data-variant="compact"]': {
    height: 'fit-content',
    '&:active': {},
    '&:hover': {
      backgroundColor: (theme.vars || theme).palette.action.hover
    },
    '&[data-editing]': {
      backgroundColor: 'var(--event-surface-subtle)',
      '&:hover': {
        backgroundColor: 'var(--event-surface-subtle-hover)'
      }
    }
  }
}));
const DayGridEventPlaceholder = (0, _styles.styled)(_calendarGrid.CalendarGrid.DayEventPlaceholder, {
  name: 'MuiEventCalendar',
  slot: 'DayGridEventPlaceholder'
})(({
  theme
}) => (0, _extends2.default)({}, DayGridEventBaseStyles(theme), {
  minHeight: 18,
  zIndex: 2,
  backgroundColor: 'var(--event-surface-subtle-hover)',
  color: 'var(--event-on-surface-subtle-primary)',
  border: `1px dashed var(--event-on-surface-subtle-secondary)`
}));
const DayGridEventTitle = (0, _styles.styled)('p', {
  name: 'MuiEventCalendar',
  slot: 'DayGridEventTitle'
})(({
  theme
}) => ({
  margin: 0,
  fontWeight: theme.typography.fontWeightMedium,
  fontSize: 'var(--EventCalendar-fontSize-eventTitle, 0.75rem)',
  lineHeight: 1.43,
  flexGrow: 1,
  '[data-variant="filled"] &': {
    color: 'var(--event-on-surface-bold)'
  },
  '[data-variant="compact"] &': {
    color: (theme.vars || theme).palette.text.primary
  },
  '[data-editing] &': {
    color: 'var(--event-on-surface-selected)'
  },
  '[data-variant="compact"][data-editing] &': {
    color: 'var(--event-on-surface-subtle-primary)'
  }
}));
const DayGridEventTime = (0, _styles.styled)('time', {
  name: 'MuiEventCalendar',
  slot: 'DayGridEventTime'
})(({
  theme
}) => ({
  display: 'inline-block',
  color: 'var(--event-on-surface-subtle-secondary)',
  fontWeight: theme.typography.fontWeightRegular,
  fontSize: theme.typography.caption.fontSize,
  lineHeight: 1.43,
  whiteSpace: 'nowrap',
  paddingInlineEnd: theme.spacing(0.5),
  '@container (width < 300px)': {
    paddingInlineEnd: 0,
    display: 'inline',
    '& > span:last-of-type': {
      display: 'none'
    }
  },
  '[data-variant="compact"] &': {
    color: (theme.vars || theme).palette.text.secondary
  },
  '[data-editing] &': {
    color: 'var(--event-on-surface-selected)'
  },
  '[data-variant="compact"][data-editing] &': {
    color: 'var(--event-on-surface-subtle-secondary)'
  }
}));
const DayGridEventRecurringIcon = (0, _styles.styled)(_RepeatRounded.default, {
  name: 'MuiEventCalendar',
  slot: 'DayGridEventRecurringIcon'
})(({
  theme
}) => ({
  color: (theme.vars || theme).palette.text.primary,
  fontSize: '1rem',
  justifySelf: 'flex-end',
  '[data-editing] &': {
    color: 'var(--event-on-surface-selected)'
  },
  '[data-variant="compact"][data-editing] &': {
    color: 'var(--event-on-surface-bold)'
  }
}));
const DayGridEventResizeHandler = (0, _styles.styled)(_calendarGrid.CalendarGrid.DayEventResizeHandler, {
  name: 'MuiEventCalendar',
  slot: 'DayGridEventResizeHandler'
})({
  position: 'absolute',
  width: 4,
  top: 0,
  bottom: 0,
  zIndex: 3,
  cursor: 'ew-resize',
  opacity: 0,
  '*:hover > &': {
    opacity: 1
  },
  '&[data-start]': {
    left: 0
  },
  '&[data-end]': {
    right: 0
  }
});
const DayGridEventCardWrapper = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'DayGridEventCardWrapper'
})(({
  theme
}) => ({
  display: 'flex',
  alignItems: 'center',
  gap: theme.spacing(1),
  flexGrow: 1,
  '@container (width < 300px)': {
    gap: theme.spacing(0.5),
    [`& .${_eventCalendarClasses.eventCalendarClasses.dayGridEventTitle}`]: {
      marginInlineStart: theme.spacing(0.5),
      paddingInlineEnd: theme.spacing(1.5)
    }
  }
}));
const DayGridEventCardContent = (0, _styles.styled)('p', {
  name: 'MuiEventCalendar',
  slot: 'DayGridEventCardContent'
})({
  margin: 0,
  height: 20,
  lineHeight: '20px'
});
const EventColorIndicator = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'EventColorIndicator'
})({
  width: 8,
  height: 8,
  borderRadius: '50%',
  flexShrink: 0,
  backgroundColor: 'var(--event-main)',
  marginTop: 2
});
const DayGridEventLinesClamp = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'DayGridEventLinesClamp'
})({
  display: '-webkit-box',
  WebkitLineClamp: 'var(--number-of-lines)',
  WebkitBoxOrient: 'vertical',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  wordBreak: 'break-word',
  overflowWrap: 'break-word',
  flexGrow: 1
});
const isResizableSelector = (0, _store.createSelector)((state, side, occurrence) => {
  if (!_schedulerSelectors.schedulerEventSelectors.isResizable(state, occurrence.id, side)) {
    return false;
  }
  const view = _eventCalendarSelectors.eventCalendarViewSelectors.view(state);

  // There is only one day cell in the day view
  if (view === 'day') {
    return false;
  }

  // In month view, only multi-day and all-day events can be resized
  if (view === 'month') {
    return (0, _eventUtils.isOccurrenceAllDayOrMultipleDay)(occurrence, state.adapter);
  }
  return true;
});
const DayGridEvent = exports.DayGridEvent = /*#__PURE__*/React.forwardRef(function DayGridEvent(props, forwardedRef) {
  const {
      occurrence,
      variant,
      style: styleProp,
      className
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);

  // Context hooks
  const {
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  // Selector hooks
  const isDraggable = (0, _store.useStore)(store, _schedulerSelectors.schedulerEventSelectors.isDraggable, occurrence.id);
  const isStartResizable = (0, _store.useStore)(store, isResizableSelector, 'start', occurrence);
  const isEndResizable = (0, _store.useStore)(store, isResizableSelector, 'end', occurrence);
  const isRecurring = (0, _store.useStore)(store, _schedulerSelectors.schedulerEventSelectors.isRecurring, occurrence.id);
  const resource = (0, _store.useStore)(store, _schedulerSelectors.schedulerResourceSelectors.processedResource, occurrence.resource);
  const color = (0, _store.useStore)(store, _schedulerSelectors.schedulerEventSelectors.color, occurrence.id);

  // Feature hooks
  const formatTime = (0, _useFormatTime.useFormatTime)();
  const content = React.useMemo(() => {
    switch (variant) {
      case 'invisible':
        {
          return null;
        }
      case 'filled':
      case 'placeholder':
        return /*#__PURE__*/(0, _jsxRuntime.jsxs)(React.Fragment, {
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(DayGridEventLinesClamp, {
            className: classes.dayGridEventLinesClamp,
            style: {
              '--number-of-lines': 1
            },
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(DayGridEventTitle, {
              className: classes.dayGridEventTitle,
              children: occurrence.title
            })
          }), isRecurring && /*#__PURE__*/(0, _jsxRuntime.jsx)(DayGridEventRecurringIcon, {
            className: classes.dayGridEventRecurringIcon,
            "aria-hidden": "true",
            fontSize: "small"
          })]
        });
      case 'compact':
        return /*#__PURE__*/(0, _jsxRuntime.jsxs)(DayGridEventCardWrapper, {
          className: classes.dayGridEventCardWrapper,
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(EventColorIndicator, {
            className: classes.eventColorIndicator,
            role: "img",
            "aria-label": resource?.title ? localeText.resourceAriaLabel(resource.title) : localeText.noResourceAriaLabel
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(DayGridEventCardContent, {
            className: classes.dayGridEventCardContent,
            children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(DayGridEventLinesClamp, {
              className: classes.dayGridEventLinesClamp,
              style: {
                '--number-of-lines': 1
              },
              children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(DayGridEventTime, {
                className: classes.dayGridEventTime,
                children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                  children: formatTime(occurrence.displayTimezone.start.value)
                }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("span", {
                  children: [" - ", formatTime(occurrence.displayTimezone.end.value)]
                })]
              }), /*#__PURE__*/(0, _jsxRuntime.jsx)(DayGridEventTitle, {
                className: classes.dayGridEventTitle,
                as: "span",
                children: occurrence.title
              })]
            })
          }), isRecurring && /*#__PURE__*/(0, _jsxRuntime.jsx)(DayGridEventRecurringIcon, {
            className: classes.dayGridEventRecurringIcon,
            "aria-hidden": "true",
            fontSize: "small"
          })]
        });
      default:
        throw new Error(process.env.NODE_ENV !== "production" ? 'MUI X Scheduler: Unsupported variant provided to DayGridEvent component. ' + 'The DayGridEvent component only supports specific variant values. ' + 'Check the component documentation for supported variants.' : (0, _formatErrorMessage2.default)(208));
    }
  }, [variant, occurrence.title, occurrence.displayTimezone.start.value, occurrence.displayTimezone.end.value, isRecurring, resource?.title, localeText, formatTime, classes]);
  const sharedProps = (0, _extends2.default)({
    start: occurrence.displayTimezone.start,
    end: occurrence.displayTimezone.end,
    ref: forwardedRef,
    'data-variant': variant,
    'data-palette': color,
    style: (0, _extends2.default)({
      '--grid-row': occurrence.position.index,
      '--grid-column-span': occurrence.position.daySpan
    }, styleProp)
  }, other, {
    className: (0, _clsx.default)(className, occurrence.className)
  });
  if (variant === 'placeholder') {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(DayGridEventPlaceholder, (0, _extends2.default)({
      "aria-hidden": true
    }, sharedProps, {
      className: (0, _clsx.default)(classes.dayGridEventPlaceholder, sharedProps.className),
      children: content
    }));
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(DayGridEventRoot, (0, _extends2.default)({
    eventId: occurrence.id,
    occurrenceKey: occurrence.key,
    isDraggable: isDraggable,
    renderDragPreview: parameters => /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventDragPreview.EventDragPreview, (0, _extends2.default)({}, parameters)),
    "aria-hidden": variant === 'invisible'
  }, sharedProps, {
    className: (0, _clsx.default)(classes.dayGridEvent, sharedProps.className),
    children: [isStartResizable && /*#__PURE__*/(0, _jsxRuntime.jsx)(DayGridEventResizeHandler, {
      className: classes.dayGridEventResizeHandler,
      side: "start"
    }), content, isEndResizable && /*#__PURE__*/(0, _jsxRuntime.jsx)(DayGridEventResizeHandler, {
      className: classes.dayGridEventResizeHandler,
      side: "end"
    })]
  }));
});
if (process.env.NODE_ENV !== "production") DayGridEvent.displayName = "DayGridEvent";
