// @mui/x-scheduler/event-calendar/header-toolbar/HeaderToolbar.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HeaderToolbar = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _store = require("@base-ui/utils/store");
var _styles = require("@mui/material/styles");
var _ChevronLeft = _interopRequireDefault(require("@mui/icons-material/ChevronLeft"));
var _ChevronRight = _interopRequireDefault(require("@mui/icons-material/ChevronRight"));
var _MenuOpen = _interopRequireDefault(require("@mui/icons-material/MenuOpen"));
var _Menu = _interopRequireDefault(require("@mui/icons-material/Menu"));
var _Button = _interopRequireDefault(require("@mui/material/Button"));
var _IconButton = _interopRequireDefault(require("@mui/material/IconButton"));
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _eventCalendarSelectors = require("@mui/x-scheduler-internals/event-calendar-selectors");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _internals = require("@mui/x-scheduler-internals/internals");
var _viewSwitcher = require("./view-switcher");
var _preferencesMenu = require("./preferences-menu");
var _EventCalendarStyledContext = require("../EventCalendarStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const HeaderToolbarRoot = (0, _styles.styled)('header', {
  name: 'MuiEventCalendar',
  slot: 'HeaderToolbar'
})(({
  theme
}) => ({
  display: 'flex',
  gap: theme.spacing(2),
  justifyContent: 'space-between',
  alignItems: 'center',
  width: '100%',
  flexWrap: 'wrap'
}));
const HeaderToolbarActions = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'HeaderToolbarActions'
})(({
  theme
}) => ({
  display: 'flex',
  gap: theme.spacing(2)
}));
const HeaderToolbarDateNavigator = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'HeaderToolbarDateNavigator'
})(({
  theme
}) => ({
  display: 'flex',
  gap: theme.spacing(2)
}));
const HeaderToolbarLeftElement = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'HeaderToolbarLeftElement'
})(({
  theme
}) => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'flex-start',
  gap: theme.spacing(2)
}));
const HeaderToolbarLabel = (0, _styles.styled)('p', {
  name: 'MuiEventCalendar',
  slot: 'HeaderToolbarLabel'
})(({
  theme
}) => (0, _extends2.default)({
  margin: 0
}, theme.typography.h6, {
  fontWeight: theme.typography.fontWeightBold,
  lineHeight: 1.4,
  display: 'flex',
  alignItems: 'center',
  gap: theme.spacing(1)
}));
const HeaderToolbarWeekNumber = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'HeaderToolbarWeekNumber'
})(({
  theme
}) => (0, _extends2.default)({
  padding: theme.spacing(0.1, 0.7),
  borderRadius: theme.shape.borderRadius,
  fontSize: '0.875rem',
  fontWeight: theme.typography.fontWeightRegular,
  color: (theme.vars || theme).palette.text.primary,
  textAlign: 'center',
  backgroundColor: (theme.vars || theme).palette.grey[200]
}, theme.applyStyles('dark', {
  backgroundColor: (theme.vars || theme).palette.grey[900]
})));
const HeaderToolbar = exports.HeaderToolbar = /*#__PURE__*/React.forwardRef(function HeaderToolbar(props, forwardedRef) {
  // Context hooks
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const {
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const adapter = (0, _useAdapterContext.useAdapterContext)();

  // Selector hooks
  const views = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarViewSelectors.views);
  const view = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarViewSelectors.view);
  const visibleDate = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.visibleDate);
  const isSidePanelOpen = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarPreferenceSelectors.isSidePanelOpen);
  const showWeekNumber = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarPreferenceSelectors.showWeekNumber);
  const weekStartsOn = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarPreferenceSelectors.weekStartsOn);
  const weekNumber = (0, _internals.getWeekNumber)(adapter, visibleDate, weekStartsOn);
  const showViewSwitcher = views.length > 1;
  const showWeekLabel = showWeekNumber && (view === 'week' || view === 'day');
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(HeaderToolbarRoot, (0, _extends2.default)({
    ref: forwardedRef,
    "data-single-primary-action": !showViewSwitcher
  }, props, {
    className: (0, _clsx.default)(props.className, classes.headerToolbar),
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(HeaderToolbarLeftElement, {
      className: classes.headerToolbarLeftElement,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_IconButton.default, {
        className: classes.headerToolbarSidePanelToggle,
        "aria-label": isSidePanelOpen ? localeText.closeSidePanel : localeText.openSidePanel,
        onClick: event => store.setPreferences({
          isSidePanelOpen: !isSidePanelOpen
        }, event.nativeEvent),
        children: isSidePanelOpen ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_MenuOpen.default, {}) : /*#__PURE__*/(0, _jsxRuntime.jsx)(_Menu.default, {})
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(HeaderToolbarLabel, {
        "aria-live": "polite",
        children: [adapter.format(visibleDate, 'monthFullLetter'), ' ', adapter.format(visibleDate, 'yearPadded'), showWeekLabel && /*#__PURE__*/(0, _jsxRuntime.jsx)(HeaderToolbarWeekNumber, {
          className: classes.headerToolbarWeekNumber,
          children: `${localeText.week} ${weekNumber}`
        })]
      })]
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(HeaderToolbarActions, {
      className: classes.headerToolbarActions,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(HeaderToolbarDateNavigator, {
        role: "navigation",
        className: classes.headerToolbarDateNavigator,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_IconButton.default, {
          className: classes.headerToolbarPreviousButton,
          onClick: store.goToPreviousVisibleDate,
          "aria-label": localeText.previousTimeSpan(view),
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_ChevronLeft.default, {})
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_IconButton.default, {
          className: classes.headerToolbarNextButton,
          onClick: store.goToNextVisibleDate,
          "aria-label": localeText.nextTimeSpan(view),
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_ChevronRight.default, {})
        })]
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Button.default, {
        onClick: store.goToToday,
        children: localeText.today
      }), showViewSwitcher && /*#__PURE__*/(0, _jsxRuntime.jsx)(_viewSwitcher.ViewSwitcher, {
        views: views,
        view: view,
        onViewChange: store.setView
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_preferencesMenu.PreferencesMenu, {})]
    })]
  }));
});
if (process.env.NODE_ENV !== "production") HeaderToolbar.displayName = "HeaderToolbar";
