// @mui/x-charts/LineChart/CircleMarkElement.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CircleMarkElement = CircleMarkElement;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _styles = require("@mui/material/styles");
var _useInteractionItemProps = require("../hooks/useInteractionItemProps");
var _useChartExperimentalFeature = require("../internals/plugins/corePlugins/useChartExperimentalFeature");
var _useStore = require("../internals/store/useStore");
var _animation = require("../internals/animation/animation");
var _lineClasses = require("./lineClasses");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["x", "y", "seriesId", "classes", "color", "dataIndex", "onClick", "skipAnimation", "isFaded", "isHighlighted", "shape", "hidden"];
const Circle = (0, _styles.styled)('circle', {
  slot: 'internal',
  shouldForwardProp: undefined
})({
  [`&.${_lineClasses.lineClasses.markAnimate}`]: {
    transitionDuration: `${_animation.ANIMATION_DURATION_MS}ms`,
    transitionProperty: 'cx, cy, opacity',
    transitionTimingFunction: _animation.ANIMATION_TIMING_FUNCTION
  }
});

/**
 * The line mark element that only render circle for performance improvement.
 *
 * Demos:
 *
 * - [Lines](https://mui.com/x/react-charts/lines/)
 * - [Line demonstration](https://mui.com/x/react-charts/line-demo/)
 *
 * API:
 *
 * - [CircleMarkElement API](https://mui.com/x/api/charts/circle-mark-element/)
 */
function CircleMarkElement(props) {
  const {
      x,
      y,
      seriesId,
      classes: innerClasses,
      color,
      dataIndex,
      onClick,
      skipAnimation,
      isFaded = false,
      isHighlighted = false
      // @ts-expect-error, prevents it from being passed to the svg element
      ,

      hidden
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const store = (0, _useStore.useStore)();
  const enablePositionBasedPointerInteraction = store.use(_useChartExperimentalFeature.selectorChartExperimentalFeaturesState, 'enablePositionBasedPointerInteraction');
  const interactionProps = (0, _useInteractionItemProps.useInteractionItemProps)({
    type: 'line',
    seriesId,
    dataIndex
  });
  const theme = (0, _styles.useTheme)();
  const classes = (0, _lineClasses.useUtilityClasses)({
    skipAnimation,
    classes: innerClasses
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(Circle, (0, _extends2.default)({}, other, enablePositionBasedPointerInteraction ? {} : interactionProps, {
    cx: x,
    cy: y,
    r: 5,
    fill: (theme.vars || theme).palette.background.paper,
    stroke: color,
    strokeWidth: 2,
    className: classes.mark,
    onClick: onClick,
    cursor: onClick ? 'pointer' : 'unset',
    pointerEvents: hidden ? 'none' : undefined,
    "data-highlighted": isHighlighted || undefined,
    "data-faded": isFaded || undefined,
    "data-series-id": seriesId,
    "data-series": seriesId,
    "data-index": dataIndex,
    opacity: hidden ? 0 : 1
  }));
}
process.env.NODE_ENV !== "production" ? CircleMarkElement.propTypes = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  classes: _propTypes.default.object,
  /**
   * The index to the element in the series' data array.
   */
  dataIndex: _propTypes.default.number.isRequired,
  seriesId: _propTypes.default.string.isRequired,
  /**
   * The shape of the marker.
   */
  shape: _propTypes.default.oneOf(['circle', 'cross', 'diamond', 'square', 'star', 'triangle', 'wye']).isRequired,
  /**
   * If `true`, animations are skipped.
   * @default false
   */
  skipAnimation: _propTypes.default.bool
} : void 0;
