// @mui/x-charts/LineChart/LineHighlightPlot.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LineHighlightPlot = LineHighlightPlot;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _formatErrorMessage2 = _interopRequireDefault(require("@mui/x-internals/formatErrorMessage"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _useStore = require("../internals/store/useStore");
var _LineHighlightElement = require("./LineHighlightElement");
var _getValueToPositionMapper = require("../hooks/getValueToPositionMapper");
var _constants = require("../constants");
var _useLineSeries = require("../hooks/useLineSeries");
var _getColor = _interopRequireDefault(require("./seriesConfig/getColor"));
var _ChartsProvider = require("../context/ChartsProvider");
var _useChartCartesianAxis = require("../internals/plugins/featurePlugins/useChartCartesianAxis");
var _useAxis = require("../hooks/useAxis");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["slots", "slotProps"];
/**
 * Demos:
 *
 * - [Lines](https://mui.com/x/react-charts/lines/)
 * - [Line demonstration](https://mui.com/x/react-charts/line-demo/)
 *
 * API:
 *
 * - [LineHighlightPlot API](https://mui.com/x/api/charts/line-highlight-plot/)
 */
function LineHighlightPlot(props) {
  const {
      slots,
      slotProps
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const seriesData = (0, _useLineSeries.useLineSeriesContext)();
  const {
    xAxis,
    xAxisIds
  } = (0, _useAxis.useXAxes)();
  const {
    yAxis,
    yAxisIds
  } = (0, _useAxis.useYAxes)();
  const {
    instance
  } = (0, _ChartsProvider.useChartsContext)();
  const store = (0, _useStore.useStore)();
  const highlightedIndexes = store.use(_useChartCartesianAxis.selectorChartsHighlightXAxisIndex);
  if (seriesData === undefined) {
    return null;
  }
  const {
    series,
    stackingGroups
  } = seriesData;
  const defaultXAxisId = xAxisIds[0];
  const defaultYAxisId = yAxisIds[0];
  const hasStartEndMark = stackingGroups.some(({
    ids
  }) => ids.some(seriesId => series[seriesId].showMark === 'start' || series[seriesId].showMark === 'end'));
  if (highlightedIndexes.length === 0 && !hasStartEndMark) {
    return null;
  }
  const getHighlightedIndex = new Map();
  highlightedIndexes.forEach(({
    axisId,
    dataIndex
  }) => {
    getHighlightedIndex.set(axisId, dataIndex);
  });
  const Element = slots?.lineHighlight ?? _LineHighlightElement.LineHighlightElement;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("g", (0, _extends2.default)({}, other, {
    children: stackingGroups.flatMap(({
      ids: groupIds
    }) => groupIds.flatMap(seriesId => {
      const {
        xAxisId = defaultXAxisId,
        yAxisId = defaultYAxisId,
        visibleStackedData,
        data,
        disableHighlight,
        showMark,
        shape = 'circle',
        hidden
      } = series[seriesId];
      if (hidden) {
        return null;
      }
      const axisIndex = disableHighlight ? undefined : getHighlightedIndex.get(xAxisId);
      const showMarkIndex = showMark === 'start' || showMark === 'end' ? getStartEndMarkIndex(data, showMark) : undefined;
      const highlightedIndex = axisIndex ?? showMarkIndex;
      if (highlightedIndex === undefined || data[highlightedIndex] == null) {
        return null;
      }
      const xScale = (0, _getValueToPositionMapper.getValueToPositionMapper)(xAxis[xAxisId].scale);
      const yScale = yAxis[yAxisId].scale;
      const xData = xAxis[xAxisId].data;
      if (xData === undefined) {
        throw new Error(process.env.NODE_ENV !== "production" ? `MUI X Charts: ${xAxisId === _constants.DEFAULT_X_AXIS_KEY ? 'The first `xAxis`' : `The x-axis with id "${xAxisId}"`} should have a data property to be able to display a line plot. ` + 'The x-axis data defines the positions for each point in the line. ' + 'Provide a data array to the x-axis configuration.' : (0, _formatErrorMessage2.default)(10, xAxisId === _constants.DEFAULT_X_AXIS_KEY ? 'The first `xAxis`' : `The x-axis with id "${xAxisId}"`));
      }
      const x = xScale(xData[highlightedIndex]);
      const y = yScale(visibleStackedData[highlightedIndex][1]); // This should not be undefined since y should not be a band scale
      if (!instance.isPointInside(x, y)) {
        return null;
      }
      const colorGetter = (0, _getColor.default)(series[seriesId], xAxis[xAxisId], yAxis[yAxisId]);
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(Element, (0, _extends2.default)({
        seriesId: seriesId,
        color: colorGetter(highlightedIndex),
        x: x,
        y: y,
        shape: shape
      }, slotProps?.lineHighlight), `${seriesId}`);
    }))
  }));
}

/**
 * Returns the index of the first (`'start'`) or last (`'end'`) non-null item, or `-1` if none.
 */
function getStartEndMarkIndex(data, type) {
  if (type === 'start') {
    const index = data.findIndex(value => value != null);
    return index < 0 ? undefined : index;
  }
  const index = data.findLastIndex(value => value != null);
  return index < 0 ? undefined : index;
}
process.env.NODE_ENV !== "production" ? LineHighlightPlot.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * The props used for each component slot.
   * @default {}
   */
  slotProps: _propTypes.default.object,
  /**
   * Overridable component slots.
   * @default {}
   */
  slots: _propTypes.default.object
} : void 0;
