// @mui/x-charts/LineChart/LineHighlightPlot.mjs
'use client';

import _extends from "@babel/runtime/helpers/esm/extends";
import _formatErrorMessage from "@mui/x-internals/formatErrorMessage";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
const _excluded = ["slots", "slotProps"];
import * as React from 'react';
import PropTypes from 'prop-types';
import { useStore } from "../internals/store/useStore.mjs";
import { LineHighlightElement } from "./LineHighlightElement.mjs";
import { getValueToPositionMapper } from "../hooks/getValueToPositionMapper.mjs";
import { DEFAULT_X_AXIS_KEY } from "../constants/index.mjs";
import { useLineSeriesContext } from "../hooks/useLineSeries.mjs";
import getColor from "./seriesConfig/getColor.mjs";
import { useChartsContext } from "../context/ChartsProvider/index.mjs";
import { selectorChartsHighlightXAxisIndex } from "../internals/plugins/featurePlugins/useChartCartesianAxis/index.mjs";
import { useXAxes, useYAxes } from "../hooks/useAxis.mjs";
import { jsx as _jsx } from "react/jsx-runtime";
/**
 * Demos:
 *
 * - [Lines](https://mui.com/x/react-charts/lines/)
 * - [Line demonstration](https://mui.com/x/react-charts/line-demo/)
 *
 * API:
 *
 * - [LineHighlightPlot API](https://mui.com/x/api/charts/line-highlight-plot/)
 */
function LineHighlightPlot(props) {
  const {
      slots,
      slotProps
    } = props,
    other = _objectWithoutPropertiesLoose(props, _excluded);
  const seriesData = useLineSeriesContext();
  const {
    xAxis,
    xAxisIds
  } = useXAxes();
  const {
    yAxis,
    yAxisIds
  } = useYAxes();
  const {
    instance
  } = useChartsContext();
  const store = useStore();
  const highlightedIndexes = store.use(selectorChartsHighlightXAxisIndex);
  if (seriesData === undefined) {
    return null;
  }
  const {
    series,
    stackingGroups
  } = seriesData;
  const defaultXAxisId = xAxisIds[0];
  const defaultYAxisId = yAxisIds[0];
  const hasStartEndMark = stackingGroups.some(({
    ids
  }) => ids.some(seriesId => series[seriesId].showMark === 'start' || series[seriesId].showMark === 'end'));
  if (highlightedIndexes.length === 0 && !hasStartEndMark) {
    return null;
  }
  const getHighlightedIndex = new Map();
  highlightedIndexes.forEach(({
    axisId,
    dataIndex
  }) => {
    getHighlightedIndex.set(axisId, dataIndex);
  });
  const Element = slots?.lineHighlight ?? LineHighlightElement;
  return /*#__PURE__*/_jsx("g", _extends({}, other, {
    children: stackingGroups.flatMap(({
      ids: groupIds
    }) => groupIds.flatMap(seriesId => {
      const {
        xAxisId = defaultXAxisId,
        yAxisId = defaultYAxisId,
        visibleStackedData,
        data,
        disableHighlight,
        showMark,
        shape = 'circle',
        hidden
      } = series[seriesId];
      if (hidden) {
        return null;
      }
      const axisIndex = disableHighlight ? undefined : getHighlightedIndex.get(xAxisId);
      const showMarkIndex = showMark === 'start' || showMark === 'end' ? getStartEndMarkIndex(data, showMark) : undefined;
      const highlightedIndex = axisIndex ?? showMarkIndex;
      if (highlightedIndex === undefined || data[highlightedIndex] == null) {
        return null;
      }
      const xScale = getValueToPositionMapper(xAxis[xAxisId].scale);
      const yScale = yAxis[yAxisId].scale;
      const xData = xAxis[xAxisId].data;
      if (xData === undefined) {
        throw new Error(process.env.NODE_ENV !== "production" ? `MUI X Charts: ${xAxisId === DEFAULT_X_AXIS_KEY ? 'The first `xAxis`' : `The x-axis with id "${xAxisId}"`} should have a data property to be able to display a line plot. ` + 'The x-axis data defines the positions for each point in the line. ' + 'Provide a data array to the x-axis configuration.' : _formatErrorMessage(10, xAxisId === DEFAULT_X_AXIS_KEY ? 'The first `xAxis`' : `The x-axis with id "${xAxisId}"`));
      }
      const x = xScale(xData[highlightedIndex]);
      const y = yScale(visibleStackedData[highlightedIndex][1]); // This should not be undefined since y should not be a band scale
      if (!instance.isPointInside(x, y)) {
        return null;
      }
      const colorGetter = getColor(series[seriesId], xAxis[xAxisId], yAxis[yAxisId]);
      return /*#__PURE__*/_jsx(Element, _extends({
        seriesId: seriesId,
        color: colorGetter(highlightedIndex),
        x: x,
        y: y,
        shape: shape
      }, slotProps?.lineHighlight), `${seriesId}`);
    }))
  }));
}

/**
 * Returns the index of the first (`'start'`) or last (`'end'`) non-null item, or `-1` if none.
 */
function getStartEndMarkIndex(data, type) {
  if (type === 'start') {
    const index = data.findIndex(value => value != null);
    return index < 0 ? undefined : index;
  }
  const index = data.findLastIndex(value => value != null);
  return index < 0 ? undefined : index;
}
process.env.NODE_ENV !== "production" ? LineHighlightPlot.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * The props used for each component slot.
   * @default {}
   */
  slotProps: PropTypes.object,
  /**
   * Overridable component slots.
   * @default {}
   */
  slots: PropTypes.object
} : void 0;
export { LineHighlightPlot };
