// @mui/x-charts/LineChart/MarkElement.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MarkElement = MarkElement;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _styles = require("@mui/material/styles");
var _d3Shape = require("@mui/x-charts-vendor/d3-shape");
var _animation = require("../internals/animation/animation");
var _useInteractionItemProps = require("../hooks/useInteractionItemProps");
var _useChartExperimentalFeature = require("../internals/plugins/corePlugins/useChartExperimentalFeature");
var _useStore = require("../internals/store/useStore");
var _getSymbol = require("../internals/getSymbol");
var _lineClasses = require("./lineClasses");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["x", "y", "seriesId", "classes", "color", "shape", "dataIndex", "onClick", "skipAnimation", "isFaded", "isHighlighted", "hidden", "style"];
const MarkElementPath = (0, _styles.styled)('path', {
  name: 'MuiMarkElement',
  slot: 'Root'
})(({
  theme
}) => ({
  fill: (theme.vars || theme).palette.background.paper,
  [`&.${_lineClasses.lineClasses.markAnimate}`]: {
    transitionDuration: `${_animation.ANIMATION_DURATION_MS}ms`,
    transitionProperty: 'transform, transform-origin, opacity',
    transitionTimingFunction: _animation.ANIMATION_TIMING_FUNCTION
  }
}));
/**
 * Demos:
 *
 * - [Lines](https://mui.com/x/react-charts/lines/)
 * - [Line demonstration](https://mui.com/x/react-charts/line-demo/)
 *
 * API:
 *
 * - [MarkElement API](https://mui.com/x/api/charts/mark-element/)
 */
function MarkElement(props) {
  const {
      x,
      y,
      seriesId,
      classes: innerClasses,
      color,
      shape,
      dataIndex,
      onClick,
      skipAnimation,
      isFaded = false,
      isHighlighted = false,
      hidden,
      style
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const store = (0, _useStore.useStore)();
  const enablePositionBasedPointerInteraction = store.use(_useChartExperimentalFeature.selectorChartExperimentalFeaturesState, 'enablePositionBasedPointerInteraction');
  const interactionProps = (0, _useInteractionItemProps.useInteractionItemProps)({
    type: 'line',
    seriesId,
    dataIndex
  });
  const ownerState = {
    seriesId,
    classes: innerClasses,
    isHighlighted,
    isFaded,
    skipAnimation
  };
  const classes = (0, _lineClasses.useUtilityClasses)({
    skipAnimation,
    classes: innerClasses
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(MarkElementPath, (0, _extends2.default)({}, other, enablePositionBasedPointerInteraction ? {} : interactionProps, {
    style: (0, _extends2.default)({}, style, {
      transform: `translate(${x}px, ${y}px)`,
      transformOrigin: `${x}px ${y}px`
    }),
    ownerState: ownerState,
    className: classes.mark,
    d: (0, _d3Shape.symbol)(_d3Shape.symbolsFill[(0, _getSymbol.getSymbol)(shape)])(),
    onClick: onClick,
    cursor: onClick ? 'pointer' : 'unset',
    pointerEvents: hidden ? 'none' : undefined,
    "data-highlighted": isHighlighted || undefined,
    "data-faded": isFaded || undefined,
    "data-series-id": seriesId,
    "data-series": seriesId,
    "data-index": dataIndex,
    opacity: hidden ? 0 : 1,
    strokeWidth: 2,
    stroke: color
  }));
}
process.env.NODE_ENV !== "production" ? MarkElement.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  classes: _propTypes.default.object,
  /**
   * The index to the element in the series' data array.
   */
  dataIndex: _propTypes.default.number.isRequired,
  /**
   * If `true`, the marker is hidden.
   * @default false
   */
  hidden: _propTypes.default.bool,
  /**
   * If `true`, the marker is faded.
   * @default false
   */
  isFaded: _propTypes.default.bool,
  /**
   * If `true`, the marker is highlighted.
   * @default false
   */
  isHighlighted: _propTypes.default.bool,
  seriesId: _propTypes.default.string.isRequired,
  /**
   * The shape of the marker.
   */
  shape: _propTypes.default.oneOf(['circle', 'cross', 'diamond', 'square', 'star', 'triangle', 'wye']).isRequired,
  /**
   * If `true`, animations are skipped.
   * @default false
   */
  skipAnimation: _propTypes.default.bool
} : void 0;
