// @mui/x-charts/RadarChart/RadarGrid/CircularRadarStripes.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CircularRadarStripes = CircularRadarStripes;
var React = _interopRequireWildcard(require("react"));
var _getRingPath = require("../../internals/getRingPath");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * @ignore - internal component.
 */
function CircularRadarStripes(props) {
  const {
    center,
    divisions,
    radius,
    stripeColor,
    classes
  } = props;
  const divisionRadius = Array.from({
    length: divisions
  }, (_, index) => radius * (index + 1) / divisions);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(React.Fragment, {
    children: divisionRadius.map((r, index) => {
      const smallerRadius = divisionRadius[index - 1] ?? 0;
      return /*#__PURE__*/(0, _jsxRuntime.jsx)("path", {
        d: (0, _getRingPath.getRingPath)(center, r, smallerRadius),
        fillRule: "evenodd",
        fill: stripeColor?.(index) ?? 'none',
        fillOpacity: 0.1,
        className: classes?.gridStripe
      }, r);
    })
  });
}
