// @mui/x-charts/RadarChart/RadarGrid/useRadarGridData.mjs
import { useRotationScale } from "../../hooks/useScale.mjs";
import { useRadiusAxes } from "../../hooks/index.mjs";
import { selectorChartPolarCenter } from "../../internals/plugins/featurePlugins/useChartPolarAxis/index.mjs";
import { useChartsContext } from "../../context/ChartsProvider/useChartsContext.mjs";
export function useRadarGridData() {
  const {
    instance,
    store
  } = useChartsContext();
  const rotationScale = useRotationScale();
  const {
    radiusAxis
  } = useRadiusAxes();
  const {
    cx,
    cy
  } = store.use(selectorChartPolarCenter);
  if (!rotationScale || rotationScale.domain().length === 0) {
    return null;
  }
  const metrics = rotationScale.domain();
  const angles = metrics.map(key => rotationScale(key));
  return {
    center: {
      x: cx,
      y: cy
    },
    corners: metrics.map((metric, dataIndex) => {
      const radiusScale = radiusAxis[metric].scale;
      const r = radiusScale.range()[1];
      const angle = angles[dataIndex];
      const [x, y] = instance.polar2svg(r, angle);
      return {
        x,
        y
      };
    }),
    radius: radiusAxis[metrics[0]].scale.range()[1]
  };
}
