// @mui/x-scheduler/standalone-event/StandaloneEvent.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StandaloneEvent = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _styles = require("@mui/material/styles");
var _standaloneEvent = require("@mui/x-scheduler-internals/standalone-event");
var _eventDragPreview = require("../internals/components/event-drag-preview");
var _EventCalendarStyledContext = require("../event-calendar/EventCalendarStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const StandaloneEventRoot = (0, _styles.styled)(_standaloneEvent.StandaloneEvent, {
  name: 'MuiEventCalendar',
  slot: 'StandaloneEvent'
})({});
const StandaloneEvent = exports.StandaloneEvent = /*#__PURE__*/React.forwardRef(function StandaloneEvent(props, forwardedRef) {
  const styledContext = React.useContext(_EventCalendarStyledContext.EventCalendarStyledContext);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(StandaloneEventRoot, (0, _extends2.default)({
    ref: forwardedRef,
    className: styledContext?.classes.standaloneEvent
  }, props, {
    renderDragPreview: parameters => /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventDragPreview.EventDragPreview, (0, _extends2.default)({}, parameters))
  }));
});
if (process.env.NODE_ENV !== "production") StandaloneEvent.displayName = "StandaloneEvent";
process.env.NODE_ENV !== "production" ? StandaloneEvent.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * CSS class applied to the element, or a function that
   * returns a class based on the component's state.
   */
  className: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string]),
  data: _propTypes.default.shape({
    allDay: _propTypes.default.bool,
    className: _propTypes.default.string,
    color: _propTypes.default.oneOf(['amber', 'blue', 'green', 'grey', 'indigo', 'lime', 'orange', 'pink', 'purple', 'red', 'teal']),
    description: _propTypes.default.string,
    draggable: _propTypes.default.bool,
    duration: _propTypes.default.number,
    exDates: _propTypes.default.arrayOf(_propTypes.default.string),
    extractedFromId: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
    id: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]).isRequired,
    readOnly: _propTypes.default.bool,
    resizable: _propTypes.default.oneOfType([_propTypes.default.oneOf(['end', 'start']), _propTypes.default.bool]),
    resource: _propTypes.default.string,
    rrule: _propTypes.default.oneOfType([_propTypes.default.shape({
      byDay: _propTypes.default.arrayOf(_propTypes.default.oneOf(['-1FR', '-1MO', '-1SA', '-1SU', '-1TH', '-1TU', '-1WE', '-2FR', '-2MO', '-2SA', '-2SU', '-2TH', '-2TU', '-2WE', '-3FR', '-3MO', '-3SA', '-3SU', '-3TH', '-3TU', '-3WE', '-4FR', '-4MO', '-4SA', '-4SU', '-4TH', '-4TU', '-4WE', '-5FR', '-5MO', '-5SA', '-5SU', '-5TH', '-5TU', '-5WE', '1FR', '1MO', '1SA', '1SU', '1TH', '1TU', '1WE', '2FR', '2MO', '2SA', '2SU', '2TH', '2TU', '2WE', '3FR', '3MO', '3SA', '3SU', '3TH', '3TU', '3WE', '4FR', '4MO', '4SA', '4SU', '4TH', '4TU', '4WE', '5FR', '5MO', '5SA', '5SU', '5TH', '5TU', '5WE', 'FR', 'MO', 'SA', 'SU', 'TH', 'TU', 'WE']).isRequired),
      byMonth: _propTypes.default.arrayOf(_propTypes.default.number),
      byMonthDay: _propTypes.default.arrayOf(_propTypes.default.number),
      count: _propTypes.default.number,
      freq: _propTypes.default.oneOf(['DAILY', 'MONTHLY', 'WEEKLY', 'YEARLY']).isRequired,
      interval: _propTypes.default.number,
      until: _propTypes.default.string
    }), _propTypes.default.string]),
    timezone: _propTypes.default.string,
    title: _propTypes.default.string.isRequired
  }).isRequired,
  /**
   * Whether the component renders a native `<button>` element when replacing it
   * via the `render` prop.
   * Set to `true` if the rendered element is a native button.
   * @default false
   */
  nativeButton: _propTypes.default.bool,
  /**
   * Callback fired when the event is dropped into the Event Calendar.
   */
  onEventDrop: _propTypes.default.func,
  /**
   * Allows you to replace the component's HTML element
   * with a different tag, or compose it with another component.
   *
   * Accepts a `ReactElement` or a function that returns the element to render.
   */
  render: _propTypes.default.oneOfType([_propTypes.default.element, _propTypes.default.func]),
  /**
   * Style applied to the element, or a function that
   * returns a style object based on the component's state.
   */
  style: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object]),
  /**
   * The system prop that allows defining system overrides as well as additional CSS styles.
   */
  sx: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object, _propTypes.default.bool])), _propTypes.default.func, _propTypes.default.object])
} : void 0;
