// @mui/x-charts/internals/commonNextFocusItem.mjs
import { getPreviousNonEmptySeries } from "./plugins/featurePlugins/useChartKeyboardNavigation/utils/getPreviousNonEmptySeries.mjs";
import { getMaxSeriesLength } from "./plugins/featurePlugins/useChartKeyboardNavigation/utils/getMaxSeriesLength.mjs";
import { getNextNonEmptySeries } from "./plugins/featurePlugins/useChartKeyboardNavigation/utils/getNextNonEmptySeries.mjs";
import { findVisibleDataIndex } from "./plugins/featurePlugins/useChartKeyboardNavigation/utils/findVisibleDataIndex.mjs";
import { seriesHasData } from "./seriesHasData.mjs";
import { selectorChartSeriesProcessed } from "./plugins/corePlugins/useChartSeries/useChartSeries.selectors.mjs";
function isSeriesHidden(processedSeries, type, seriesId) {
  const seriesItem = processedSeries[type]?.series[seriesId];
  return Boolean(seriesItem && 'hidden' in seriesItem && seriesItem.hidden);
}
export function createGetNextIndexFocusedItem(
/**
 * The set of series types compatible with this navigation action.
 */
compatibleSeriesTypes,
/**
 * If true, allows cycling from the last item to the first one.
 */
allowCycles = false,
/**
 * If true, series max index is defined by the current series length and not all series.
 */
useCurrentSeriesMaxLength = false) {
  return function getNextIndexFocusedItem(currentItem, state) {
    const processedSeries = selectorChartSeriesProcessed(state);
    let seriesId = currentItem?.seriesId;
    let type = currentItem?.type;
    if (!type || seriesId == null || !seriesHasData(processedSeries, type, seriesId) || isSeriesHidden(processedSeries, type, seriesId)) {
      const nextSeries = getNextNonEmptySeries(processedSeries, compatibleSeriesTypes, type, seriesId);
      if (nextSeries === null) {
        return null;
      }
      type = nextSeries.type;
      seriesId = nextSeries.seriesId;
    }
    const maxLength = useCurrentSeriesMaxLength ? processedSeries[type]?.series[seriesId]?.data.length ?? 0 : getMaxSeriesLength(processedSeries, compatibleSeriesTypes);
    let dataIndex = currentItem?.dataIndex == null ? 0 : currentItem.dataIndex + 1;
    if (allowCycles) {
      dataIndex = dataIndex % maxLength;
    } else {
      dataIndex = Math.min(maxLength - 1, dataIndex);
    }
    const visibleDataIndex = findVisibleDataIndex({
      processedSeries,
      type,
      seriesId,
      startIndex: dataIndex,
      dataLength: maxLength,
      direction: 1,
      allowCycles
    });
    if (visibleDataIndex === null) {
      return null;
    }
    return {
      type: type,
      seriesId,
      dataIndex: visibleDataIndex
    };
  };
}
export function createGetPreviousIndexFocusedItem(
/**
 * The set of series types compatible with this navigation action.
 */
compatibleSeriesTypes,
/**
 * If true, allows cycling from the last item to the first one.
 */
allowCycles = false,
/**
 * If true, series max index is defined by the current series length and not all series.
 */
useCurrentSeriesMaxLength = false) {
  return function getPreviousIndexFocusedItem(currentItem, state) {
    const processedSeries = selectorChartSeriesProcessed(state);
    let seriesId = currentItem?.seriesId;
    let type = currentItem?.type;
    if (!type || seriesId == null || !seriesHasData(processedSeries, type, seriesId) || isSeriesHidden(processedSeries, type, seriesId)) {
      const previousSeries = getPreviousNonEmptySeries(processedSeries, compatibleSeriesTypes, type, seriesId);
      if (previousSeries === null) {
        return null;
      }
      type = previousSeries.type;
      seriesId = previousSeries.seriesId;
    }
    const maxLength = useCurrentSeriesMaxLength ? processedSeries[type]?.series[seriesId]?.data.length ?? 0 : getMaxSeriesLength(processedSeries, compatibleSeriesTypes);
    let dataIndex = currentItem?.dataIndex == null ? maxLength - 1 : currentItem.dataIndex - 1;
    if (allowCycles) {
      dataIndex = (maxLength + dataIndex) % maxLength;
    } else {
      dataIndex = Math.max(0, dataIndex);
    }
    const visibleDataIndex = findVisibleDataIndex({
      processedSeries,
      type,
      seriesId,
      startIndex: dataIndex,
      dataLength: maxLength,
      direction: -1,
      allowCycles
    });
    if (visibleDataIndex === null) {
      return null;
    }
    return {
      type: type,
      seriesId,
      dataIndex: visibleDataIndex
    };
  };
}
export function createGetNextSeriesFocusedItem(
/**
 * The set of series types compatible with this navigation action.
 */
compatibleSeriesTypes) {
  return function getNextSeriesFocusedItem(currentItem, state) {
    const processedSeries = selectorChartSeriesProcessed(state);
    let seriesId = currentItem?.seriesId;
    let type = currentItem?.type;
    const nextSeries = getNextNonEmptySeries(processedSeries, compatibleSeriesTypes, type, seriesId);
    if (nextSeries === null) {
      return null; // No series to move the focus to.
    }
    type = nextSeries.type;
    seriesId = nextSeries.seriesId;
    const data = processedSeries[type].series[seriesId].data;
    const startIndex = currentItem?.dataIndex == null ? 0 : Math.min(currentItem.dataIndex, data.length - 1);
    const visibleDataIndex = findVisibleDataIndex({
      processedSeries,
      type,
      seriesId,
      startIndex,
      dataLength: data.length,
      direction: 1,
      allowCycles: true
    });
    if (visibleDataIndex === null) {
      return null;
    }
    return {
      type: type,
      seriesId,
      dataIndex: visibleDataIndex
    };
  };
}
export function createGetPreviousSeriesFocusedItem(
/**
 * The set of series types compatible with this navigation action.
 */
compatibleSeriesTypes) {
  return function getPreviousSeriesFocusedItem(currentItem, state) {
    const processedSeries = selectorChartSeriesProcessed(state);
    let seriesId = currentItem?.seriesId;
    let type = currentItem?.type;
    const previousSeries = getPreviousNonEmptySeries(processedSeries, compatibleSeriesTypes, type, seriesId);
    if (previousSeries === null) {
      return null; // No series to move the focus to.
    }
    type = previousSeries.type;
    seriesId = previousSeries.seriesId;
    const data = processedSeries[type].series[seriesId].data;
    const startIndex = currentItem?.dataIndex == null ? data.length - 1 : Math.min(currentItem.dataIndex, data.length - 1);
    const visibleDataIndex = findVisibleDataIndex({
      processedSeries,
      type,
      seriesId,
      startIndex,
      dataLength: data.length,
      direction: -1,
      allowCycles: true
    });
    if (visibleDataIndex === null) {
      return null;
    }
    return {
      type: type,
      seriesId,
      dataIndex: visibleDataIndex
    };
  };
}
