// @mui/x-charts/internals/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  useSeries: true,
  useInteractionItemProps: true,
  useDrawingArea: true,
  useScatterChartProps: true,
  useScatterPlotData: true,
  scatterSeriesConfig: true,
  useLineChartProps: true,
  useAreaPlotData: true,
  useLinePlotData: true,
  processBarDataForPlot: true,
  useBarPlotData: true,
  useRadarChartProps: true,
  useSkipAnimation: true,
  useRegisterPointerInteractions: true,
  useGeoProjectionSelectors: true,
  useGeoProjectionTypes: true,
  getCartesianAxisTriggerTooltip: true,
  getCartesianAxisIndex: true,
  getRotationAxisIndex: true,
  createCommonKeyboardFocusHandler: true,
  getSeriesColorFn: true,
  resolveColorProcessor: true,
  processLineLikeSeries: true,
  getLineLikeTooltip: true,
  checkBarChartScaleErrors: true,
  getBandSize: true,
  EPSILON: true,
  useChartsTooltipUtilityClasses: true,
  getChartsTooltipUtilityClass: true,
  getAxisExtrema: true,
  evaluateCurveAtAngle: true,
  getLineItemAtPosition: true
};
Object.defineProperty(exports, "EPSILON", {
  enumerable: true,
  get: function () {
    return _epsilon.EPSILON;
  }
});
Object.defineProperty(exports, "checkBarChartScaleErrors", {
  enumerable: true,
  get: function () {
    return _checkBarChartScaleErrors.checkBarChartScaleErrors;
  }
});
Object.defineProperty(exports, "createCommonKeyboardFocusHandler", {
  enumerable: true,
  get: function () {
    return _createCommonKeyboardFocusHandler.createCommonKeyboardFocusHandler;
  }
});
Object.defineProperty(exports, "evaluateCurveAtAngle", {
  enumerable: true,
  get: function () {
    return _curveEvaluation.evaluateCurveAtAngle;
  }
});
Object.defineProperty(exports, "getAxisExtrema", {
  enumerable: true,
  get: function () {
    return _getAxisExtrema.getAxisExtrema;
  }
});
Object.defineProperty(exports, "getBandSize", {
  enumerable: true,
  get: function () {
    return _getBandSize.getBandSize;
  }
});
Object.defineProperty(exports, "getCartesianAxisIndex", {
  enumerable: true,
  get: function () {
    return _getAxisValue.getAxisIndex;
  }
});
Object.defineProperty(exports, "getCartesianAxisTriggerTooltip", {
  enumerable: true,
  get: function () {
    return _getAxisTriggerTooltip.getAxisTriggerTooltip;
  }
});
Object.defineProperty(exports, "getChartsTooltipUtilityClass", {
  enumerable: true,
  get: function () {
    return _chartsTooltipClasses.getChartsTooltipUtilityClass;
  }
});
Object.defineProperty(exports, "getLineItemAtPosition", {
  enumerable: true,
  get: function () {
    return _getItemAtPosition.default;
  }
});
Object.defineProperty(exports, "getLineLikeTooltip", {
  enumerable: true,
  get: function () {
    return _getLineLikeTooltip.getLineLikeTooltip;
  }
});
Object.defineProperty(exports, "getRotationAxisIndex", {
  enumerable: true,
  get: function () {
    return _getAxisIndex.getRotationAxisIndex;
  }
});
Object.defineProperty(exports, "getSeriesColorFn", {
  enumerable: true,
  get: function () {
    return _getSeriesColorFn.getSeriesColorFn;
  }
});
Object.defineProperty(exports, "processBarDataForPlot", {
  enumerable: true,
  get: function () {
    return _useBarPlotData.processBarDataForPlot;
  }
});
Object.defineProperty(exports, "processLineLikeSeries", {
  enumerable: true,
  get: function () {
    return _processLineLikeSeries.processLineLikeSeries;
  }
});
Object.defineProperty(exports, "resolveColorProcessor", {
  enumerable: true,
  get: function () {
    return _resolveColorProcessor.resolveColorProcessor;
  }
});
Object.defineProperty(exports, "scatterSeriesConfig", {
  enumerable: true,
  get: function () {
    return _seriesConfig.scatterSeriesConfig;
  }
});
Object.defineProperty(exports, "useAreaPlotData", {
  enumerable: true,
  get: function () {
    return _useAreaPlotData.useAreaPlotData;
  }
});
Object.defineProperty(exports, "useBarPlotData", {
  enumerable: true,
  get: function () {
    return _useBarPlotData.useBarPlotData;
  }
});
Object.defineProperty(exports, "useChartsTooltipUtilityClasses", {
  enumerable: true,
  get: function () {
    return _chartsTooltipClasses.useUtilityClasses;
  }
});
Object.defineProperty(exports, "useDrawingArea", {
  enumerable: true,
  get: function () {
    return _useDrawingArea.useDrawingArea;
  }
});
exports.useGeoProjectionTypes = exports.useGeoProjectionSelectors = void 0;
Object.defineProperty(exports, "useInteractionItemProps", {
  enumerable: true,
  get: function () {
    return _useInteractionItemProps.useInteractionItemProps;
  }
});
Object.defineProperty(exports, "useLineChartProps", {
  enumerable: true,
  get: function () {
    return _useLineChartProps.useLineChartProps;
  }
});
Object.defineProperty(exports, "useLinePlotData", {
  enumerable: true,
  get: function () {
    return _useLinePlotData.useLinePlotData;
  }
});
Object.defineProperty(exports, "useRadarChartProps", {
  enumerable: true,
  get: function () {
    return _useRadarChartProps.useRadarChartProps;
  }
});
Object.defineProperty(exports, "useRegisterPointerInteractions", {
  enumerable: true,
  get: function () {
    return _useRegisterPointerInteractions.useRegisterPointerInteractions;
  }
});
Object.defineProperty(exports, "useScatterChartProps", {
  enumerable: true,
  get: function () {
    return _useScatterChartProps.useScatterChartProps;
  }
});
Object.defineProperty(exports, "useScatterPlotData", {
  enumerable: true,
  get: function () {
    return _useScatterPlotData.useScatterPlotData;
  }
});
Object.defineProperty(exports, "useSeries", {
  enumerable: true,
  get: function () {
    return _useSeries.useSeries;
  }
});
Object.defineProperty(exports, "useSkipAnimation", {
  enumerable: true,
  get: function () {
    return _useSkipAnimation.useSkipAnimation;
  }
});
var _ChartsAxesGradients = require("./components/ChartsAxesGradients");
Object.keys(_ChartsAxesGradients).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ChartsAxesGradients[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ChartsAxesGradients[key];
    }
  });
});
var _ChartsLabelMark = require("../ChartsLabel/ChartsLabelMark");
Object.keys(_ChartsLabelMark).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ChartsLabelMark[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ChartsLabelMark[key];
    }
  });
});
var _NotRendered = require("./components/NotRendered");
Object.keys(_NotRendered).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _NotRendered[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _NotRendered[key];
    }
  });
});
var _BarLabelPlot = require("../BarChart/BarLabel/BarLabelPlot");
Object.keys(_BarLabelPlot).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _BarLabelPlot[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _BarLabelPlot[key];
    }
  });
});
var _BarClipPath = require("../BarChart/BarClipPath");
Object.keys(_BarClipPath).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _BarClipPath[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _BarClipPath[key];
    }
  });
});
var _useSeries = require("../hooks/useSeries");
var _useInteractionItemProps = require("../hooks/useInteractionItemProps");
var _useDrawingArea = require("../hooks/useDrawingArea");
var _useScatterChartProps = require("../ScatterChart/useScatterChartProps");
var _useScatterPlotData = require("../ScatterChart/useScatterPlotData");
var _seriesConfig = require("../ScatterChart/seriesConfig");
var _useLineChartProps = require("../LineChart/useLineChartProps");
var _useAreaPlotData = require("../LineChart/useAreaPlotData");
var _useLinePlotData = require("../LineChart/useLinePlotData");
var _useBarChartProps = require("../BarChart/useBarChartProps");
Object.keys(_useBarChartProps).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useBarChartProps[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useBarChartProps[key];
    }
  });
});
var _useBarPlotData = require("../BarChart/useBarPlotData");
var _useRadarChartProps = require("../RadarChart/useRadarChartProps");
var _useChartsContainerProps = require("../ChartsContainer/useChartsContainerProps");
Object.keys(_useChartsContainerProps).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartsContainerProps[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartsContainerProps[key];
    }
  });
});
var _useChartsDataProviderProps = require("../ChartsDataProvider/useChartsDataProviderProps");
Object.keys(_useChartsDataProviderProps).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartsDataProviderProps[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartsDataProviderProps[key];
    }
  });
});
var _useChartsRadialDataProviderProps = require("../ChartsRadialDataProvider/useChartsRadialDataProviderProps");
Object.keys(_useChartsRadialDataProviderProps).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartsRadialDataProviderProps[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartsRadialDataProviderProps[key];
    }
  });
});
var _seriesSelectorOfType = require("./seriesSelectorOfType");
Object.keys(_seriesSelectorOfType).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _seriesSelectorOfType[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _seriesSelectorOfType[key];
    }
  });
});
var _useSkipAnimation = require("../hooks/useSkipAnimation");
var _useRegisterPointerInteractions = require("./plugins/featurePlugins/shared/useRegisterPointerInteractions");
var _useChartId = require("./plugins/corePlugins/useChartId");
Object.keys(_useChartId).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartId[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartId[key];
    }
  });
});
var _useChartSeries = require("./plugins/corePlugins/useChartSeries");
Object.keys(_useChartSeries).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartSeries[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartSeries[key];
    }
  });
});
var _useChartDimensions = require("./plugins/corePlugins/useChartDimensions");
Object.keys(_useChartDimensions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartDimensions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartDimensions[key];
    }
  });
});
var _useChartInteractionListener = require("./plugins/corePlugins/useChartInteractionListener");
Object.keys(_useChartInteractionListener).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartInteractionListener[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartInteractionListener[key];
    }
  });
});
var _useChartSeriesConfig = require("./plugins/corePlugins/useChartSeriesConfig");
Object.keys(_useChartSeriesConfig).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartSeriesConfig[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartSeriesConfig[key];
    }
  });
});
var _useChartZAxis = require("./plugins/featurePlugins/useChartZAxis");
Object.keys(_useChartZAxis).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartZAxis[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartZAxis[key];
    }
  });
});
var _useChartCartesianAxis = require("./plugins/featurePlugins/useChartCartesianAxis");
Object.keys(_useChartCartesianAxis).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartCartesianAxis[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartCartesianAxis[key];
    }
  });
});
var _useChartPolarAxis = require("./plugins/featurePlugins/useChartPolarAxis");
Object.keys(_useChartPolarAxis).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartPolarAxis[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartPolarAxis[key];
    }
  });
});
var _useChartTooltip = require("./plugins/featurePlugins/useChartTooltip");
Object.keys(_useChartTooltip).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartTooltip[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartTooltip[key];
    }
  });
});
var _useChartInteraction = require("./plugins/featurePlugins/useChartInteraction");
Object.keys(_useChartInteraction).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartInteraction[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartInteraction[key];
    }
  });
});
var _useChartHighlight = require("./plugins/featurePlugins/useChartHighlight");
Object.keys(_useChartHighlight).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartHighlight[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartHighlight[key];
    }
  });
});
var _useChartVisibilityManager = require("./plugins/featurePlugins/useChartVisibilityManager");
Object.keys(_useChartVisibilityManager).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartVisibilityManager[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartVisibilityManager[key];
    }
  });
});
var _useChartKeyboardNavigation = require("./plugins/featurePlugins/useChartKeyboardNavigation");
Object.keys(_useChartKeyboardNavigation).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartKeyboardNavigation[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartKeyboardNavigation[key];
    }
  });
});
var _useChartClosestPoint = require("./plugins/featurePlugins/useChartClosestPoint");
Object.keys(_useChartClosestPoint).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartClosestPoint[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartClosestPoint[key];
    }
  });
});
var _useChartBrush = require("./plugins/featurePlugins/useChartBrush");
Object.keys(_useChartBrush).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartBrush[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartBrush[key];
    }
  });
});
var _useGeoProjectionSelectors = _interopRequireWildcard(require("./plugins/featurePlugins/useGeoProjection/useGeoProjection.selectors"));
exports.useGeoProjectionSelectors = _useGeoProjectionSelectors;
var _useGeoProjectionTypes = _interopRequireWildcard(require("./plugins/featurePlugins/useGeoProjection/useGeoProjection.types"));
exports.useGeoProjectionTypes = _useGeoProjectionTypes;
var _useChartItemClick = require("./plugins/featurePlugins/useChartItemClick");
Object.keys(_useChartItemClick).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useChartItemClick[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useChartItemClick[key];
    }
  });
});
var _useProgressiveRendering = require("./plugins/featurePlugins/useProgressiveRendering");
Object.keys(_useProgressiveRendering).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useProgressiveRendering[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useProgressiveRendering[key];
    }
  });
});
var _selectors = require("./plugins/utils/selectors");
Object.keys(_selectors).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _selectors[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _selectors[key];
    }
  });
});
var _getAxisTriggerTooltip = require("./plugins/featurePlugins/useChartCartesianAxis/getAxisTriggerTooltip");
var _getAxisValue = require("./plugins/featurePlugins/useChartCartesianAxis/getAxisValue");
var _getAxisIndex = require("./plugins/featurePlugins/useChartPolarAxis/getAxisIndex");
var _useCharts = require("./store/useCharts");
Object.keys(_useCharts).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useCharts[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useCharts[key];
    }
  });
});
var _useStore = require("./store/useStore");
Object.keys(_useStore).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useStore[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useStore[key];
    }
  });
});
var _BarChart = require("../BarChart/BarChart.plugins");
Object.keys(_BarChart).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _BarChart[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _BarChart[key];
    }
  });
});
var _LineChart = require("../LineChart/LineChart.plugins");
Object.keys(_LineChart).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _LineChart[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _LineChart[key];
    }
  });
});
var _ScatterChart = require("../ScatterChart/ScatterChart.plugins");
Object.keys(_ScatterChart).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ScatterChart[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ScatterChart[key];
    }
  });
});
var _RadarChart = require("../RadarChart/RadarChart.plugins");
Object.keys(_RadarChart).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _RadarChart[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _RadarChart[key];
    }
  });
});
var _PieChart = require("../PieChart/PieChart.plugins");
Object.keys(_PieChart).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _PieChart[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _PieChart[key];
    }
  });
});
var _configInit = require("./configInit");
Object.keys(_configInit).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _configInit[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _configInit[key];
    }
  });
});
var _clampAngle = require("./clampAngle");
Object.keys(_clampAngle).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _clampAngle[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _clampAngle[key];
    }
  });
});
var _getLabel = require("./getLabel");
Object.keys(_getLabel).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getLabel[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getLabel[key];
    }
  });
});
var _legendUtils = require("./legendUtils");
Object.keys(_legendUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _legendUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _legendUtils[key];
    }
  });
});
var _getChartPoint = require("./getChartPoint");
Object.keys(_getChartPoint).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getChartPoint[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getChartPoint[key];
    }
  });
});
var _coordinateTransformation = require("./plugins/featurePlugins/useChartPolarAxis/coordinateTransformation");
Object.keys(_coordinateTransformation).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _coordinateTransformation[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _coordinateTransformation[key];
    }
  });
});
var _isDefined = require("./isDefined");
Object.keys(_isDefined).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _isDefined[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _isDefined[key];
    }
  });
});
var _getScale = require("./getScale");
Object.keys(_getScale).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getScale[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getScale[key];
    }
  });
});
var _getAsNumber = require("./getAsNumber");
Object.keys(_getAsNumber).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getAsNumber[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getAsNumber[key];
    }
  });
});
var _stacking = require("./stacking");
Object.keys(_stacking).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _stacking[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _stacking[key];
    }
  });
});
var _getCurve = require("./getCurve");
Object.keys(_getCurve).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getCurve[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getCurve[key];
    }
  });
});
var _getSymbol = require("./getSymbol");
Object.keys(_getSymbol).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _getSymbol[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _getSymbol[key];
    }
  });
});
var _consumeSlots = require("./consumeSlots");
Object.keys(_consumeSlots).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _consumeSlots[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _consumeSlots[key];
    }
  });
});
var _consumeThemeProps = require("./consumeThemeProps");
Object.keys(_consumeThemeProps).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _consumeThemeProps[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _consumeThemeProps[key];
    }
  });
});
var _defaultizeMargin = require("./defaultizeMargin");
Object.keys(_defaultizeMargin).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _defaultizeMargin[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _defaultizeMargin[key];
    }
  });
});
var _colorScale = require("./colorScale");
Object.keys(_colorScale).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _colorScale[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _colorScale[key];
    }
  });
});
var _sizeScale = require("./sizeScale");
Object.keys(_sizeScale).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _sizeScale[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _sizeScale[key];
    }
  });
});
var _incompleteDatasetKeysError = require("./incompleteDatasetKeysError");
Object.keys(_incompleteDatasetKeysError).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _incompleteDatasetKeysError[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _incompleteDatasetKeysError[key];
    }
  });
});
var _ticks = require("./ticks");
Object.keys(_ticks).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ticks[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ticks[key];
    }
  });
});
var _dateHelpers = require("./dateHelpers");
Object.keys(_dateHelpers).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _dateHelpers[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _dateHelpers[key];
    }
  });
});
var _invertScale = require("./invertScale");
Object.keys(_invertScale).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _invertScale[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _invertScale[key];
    }
  });
});
var _scaleGuards = require("./scaleGuards");
Object.keys(_scaleGuards).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _scaleGuards[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _scaleGuards[key];
    }
  });
});
var _findMinMax = require("./findMinMax");
Object.keys(_findMinMax).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _findMinMax[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _findMinMax[key];
    }
  });
});
var _commonNextFocusItem = require("./commonNextFocusItem");
Object.keys(_commonNextFocusItem).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _commonNextFocusItem[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _commonNextFocusItem[key];
    }
  });
});
var _createCommonKeyboardFocusHandler = require("./createCommonKeyboardFocusHandler");
var _getSeriesColorFn = require("./getSeriesColorFn");
var _resolveColorProcessor = require("./resolveColorProcessor");
var _processLineLikeSeries = require("./processLineLikeSeries");
var _getLineLikeTooltip = require("./getLineLikeTooltip");
var _checkBarChartScaleErrors = require("../BarChart/checkBarChartScaleErrors");
var _getBandSize = require("./getBandSize");
var _defaultSeriesConfig = require("./plugins/utils/defaultSeriesConfig");
Object.keys(_defaultSeriesConfig).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _defaultSeriesConfig[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _defaultSeriesConfig[key];
    }
  });
});
var _epsilon = require("../utils/epsilon");
var _chartsTooltipClasses = require("../ChartsTooltip/chartsTooltipClasses");
var _getAxisExtrema = require("./plugins/featurePlugins/useChartCartesianAxis/getAxisExtrema");
var _ChartsProvider = require("../context/ChartsProvider");
Object.keys(_ChartsProvider).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ChartsProvider[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ChartsProvider[key];
    }
  });
});
var _ChartsSlotsContext = require("../context/ChartsSlotsContext");
Object.keys(_ChartsSlotsContext).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _ChartsSlotsContext[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _ChartsSlotsContext[key];
    }
  });
});
var _config = require("../models/seriesType/config");
Object.keys(_config).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _config[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _config[key];
    }
  });
});
var _common = require("../models/seriesType/common");
Object.keys(_common).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _common[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _common[key];
    }
  });
});
var _line = require("../models/seriesType/line");
Object.keys(_line).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _line[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _line[key];
    }
  });
});
var _bar = require("../models/seriesType/bar");
Object.keys(_bar).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _bar[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _bar[key];
    }
  });
});
var _zAxis = require("../models/z-axis");
Object.keys(_zAxis).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _zAxis[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _zAxis[key];
    }
  });
});
var _axis = require("../models/axis");
Object.keys(_axis).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _axis[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _axis[key];
    }
  });
});
var _composition = require("../models/seriesType/composition");
Object.keys(_composition).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _composition[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _composition[key];
    }
  });
});
var _models = require("./plugins/models");
Object.keys(_models).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _models[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _models[key];
    }
  });
});
var _material = require("./material");
Object.keys(_material).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _material[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _material[key];
    }
  });
});
var _createSvgIcon = require("./createSvgIcon");
Object.keys(_createSvgIcon).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _createSvgIcon[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _createSvgIcon[key];
    }
  });
});
var _constants = require("./constants");
Object.keys(_constants).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _constants[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _constants[key];
    }
  });
});
var _scales = require("./scales");
Object.keys(_scales).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _scales[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _scales[key];
    }
  });
});
var _identifierSerializer = require("./identifierSerializer");
Object.keys(_identifierSerializer).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _identifierSerializer[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _identifierSerializer[key];
    }
  });
});
var _identifierCleaner = require("./identifierCleaner");
Object.keys(_identifierCleaner).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _identifierCleaner[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _identifierCleaner[key];
    }
  });
});
var _curveEvaluation = require("../LineChart/seriesConfig/curveEvaluation");
var _getItemAtPosition = _interopRequireDefault(require("../LineChart/seriesConfig/getItemAtPosition"));
