// @mui/x-scheduler/event-calendar/EventCalendarRoot.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EventCalendarRoot = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _store = require("@base-ui/utils/store");
var _styles = require("@mui/material/styles");
var _Collapse = _interopRequireDefault(require("@mui/material/Collapse"));
var _Divider = _interopRequireDefault(require("@mui/material/Divider"));
var _useMergedRefs = require("@base-ui/utils/useMergedRefs");
var _eventCalendarSelectors = require("@mui/x-scheduler-internals/event-calendar-selectors");
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _responsiveTypography = require("../internals/constants/responsiveTypography");
var _ResponsiveTypographyContainer = require("../internals/components/ResponsiveTypographyContainer");
var _errorContainer = require("../internals/components/error-container");
var _WeekView = require("../week-view/WeekView");
var _agendaView = require("../agenda-view");
var _DayView = require("../day-view/DayView");
var _monthView = require("../month-view");
var _headerToolbar = require("./header-toolbar");
var _resourcesTree = require("./resources-tree");
var _miniCalendar = require("./mini-calendar");
var _EventCalendarStyledContext = require("./EventCalendarStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["children", "className"];
const EventCalendarRootStyled = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'Root'
})(({
  theme
}) => ({
  // Layout
  boxSizing: 'border-box',
  '*, *::before, *::after': {
    boxSizing: 'inherit'
  },
  position: 'relative',
  width: '100%',
  display: 'flex',
  flexDirection: 'column',
  gap: theme.spacing(1),
  height: '100%',
  minHeight: 0,
  overflow: 'hidden',
  fontFamily: theme.typography.fontFamily
}));
const EventCalendarSidePanel = (0, _styles.styled)('aside', {
  name: 'MuiEventCalendar',
  slot: 'SidePanel'
})(({
  theme
}) => ({
  minWidth: 250,
  width: 'fit-content',
  display: 'flex',
  flexDirection: 'column',
  gap: theme.spacing(2),
  border: '1px solid',
  borderColor: (theme.vars || theme).palette.divider,
  borderRadius: theme.shape.borderRadius,
  maxHeight: '100%',
  overflowY: 'hidden'
}));
const EventCalendarMainPanel = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MainPanel'
})(({
  theme
}) => ({
  display: 'flex',
  flexGrow: 1,
  gap: theme.spacing(1),
  minHeight: 0,
  '&[data-view="month"]': {
    maxHeight: '100%',
    overflow: 'hidden'
  }
}));
const EventCalendarContent = (0, _styles.styled)('section', {
  name: 'MuiEventCalendar',
  slot: 'Content'
})(() => (0, _extends2.default)({
  display: 'flex',
  flex: 1,
  overflow: 'auto',
  height: '100%',
  maxHeight: '100%'
}, _responsiveTypography.responsiveTypographyContainerQueries, {
  '&[data-side-panel-open="false"]': {
    gridColumn: '1 / -1'
  }
}));

/**
 * Internal component that renders the EventCalendar UI.
 * Used by both EventCalendar and EventCalendarPremium.
 * Expects the store to be provided via context.
 */
const EventCalendarRoot = exports.EventCalendarRoot = /*#__PURE__*/React.forwardRef(function EventCalendarRoot(props, forwardedRef) {
  const {
      children,
      className
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const {
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const view = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarViewSelectors.view);
  const isSidePanelOpen = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarPreferenceSelectors.isSidePanelOpen);
  let content;
  switch (view) {
    case 'week':
      content = /*#__PURE__*/(0, _jsxRuntime.jsx)(_WeekView.WeekView, {});
      break;
    case 'day':
      content = /*#__PURE__*/(0, _jsxRuntime.jsx)(_DayView.DayView, {});
      break;
    case 'month':
      content = /*#__PURE__*/(0, _jsxRuntime.jsx)(_monthView.MonthView, {});
      break;
    case 'agenda':
      content = /*#__PURE__*/(0, _jsxRuntime.jsx)(_agendaView.AgendaView, {});
      break;
    default:
      content = null;
  }
  const rootRef = React.useRef(null);
  const handleRootRef = (0, _useMergedRefs.useMergedRefs)(forwardedRef, rootRef);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(EventCalendarRootStyled, (0, _extends2.default)({
    className: (0, _clsx.default)(classes.root, className)
  }, other, {
    ref: handleRootRef,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_headerToolbar.HeaderToolbar, {}), /*#__PURE__*/(0, _jsxRuntime.jsxs)(EventCalendarMainPanel, {
      className: classes.mainPanel,
      "data-view": view,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_Collapse.default, {
        in: isSidePanelOpen,
        orientation: "horizontal",
        className: classes.sidePanelCollapse,
        children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(EventCalendarSidePanel, {
          className: classes.sidePanel,
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_miniCalendar.MiniCalendar, {}), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Divider.default, {
            className: classes.sidePanelDivider
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_resourcesTree.ResourcesTree, {})]
        })
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_ResponsiveTypographyContainer.ResponsiveTypographyContainer, {
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(EventCalendarContent, {
          className: classes.content,
          "data-view": view,
          "data-side-panel-open": isSidePanelOpen,
          "aria-label": localeText.calendarContentAriaLabel,
          children: content
        })
      })]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_errorContainer.ErrorContainer, {}), children]
  }));
});
if (process.env.NODE_ENV !== "production") EventCalendarRoot.displayName = "EventCalendarRoot";
