// @mui/x-scheduler/event-calendar/EventCalendar.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EventCalendar = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _styles = require("@mui/material/styles");
var _useEventCalendar = require("@mui/x-scheduler-internals/use-event-calendar");
var _useSchedulerStoreContext = require("@mui/x-scheduler-internals/use-scheduler-store-context");
var _internals = require("@mui/x-scheduler-internals/internals");
var _useId = require("@base-ui/utils/useId");
var _eventDialog = require("../internals/components/event-dialog");
var _SharedComponentsStyledContext = require("../internals/components/SharedComponentsStyledContext");
var _eventCalendarClasses = require("./eventCalendarClasses");
var _EventCalendarStyledContext = require("./EventCalendarStyledContext");
var _EventDialogStyledContext = require("../internals/components/event-dialog/EventDialogStyledContext");
var _defaultLocaleText = require("../internals/constants/defaultLocaleText");
var _EventCalendarRoot = require("./EventCalendarRoot");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className", "classes"],
  _excluded2 = ["localeText", "apiRef"];
const EventCalendar = exports.EventCalendar = /*#__PURE__*/React.forwardRef(function EventCalendar(inProps, forwardedRef) {
  const props = (0, _styles.useThemeProps)({
    props: inProps,
    name: 'MuiEventCalendar'
  });
  const _useExtractEventCalen = (0, _useEventCalendar.useExtractEventCalendarParameters)(props),
    {
      parameters,
      forwardedProps: {
        className,
        classes: classesProp
      }
    } = _useExtractEventCalen,
    forwardedProps = (0, _objectWithoutPropertiesLoose2.default)(_useExtractEventCalen.forwardedProps, _excluded);
  const store = (0, _useEventCalendar.useEventCalendar)(parameters);
  const classes = (0, _eventCalendarClasses.useEventCalendarUtilityClasses)(classesProp);
  const {
      localeText,
      apiRef
    } = forwardedProps,
    other = (0, _objectWithoutPropertiesLoose2.default)(forwardedProps, _excluded2);
  (0, _internals.useInitializeApiRef)(store, apiRef);
  const schedulerId = (0, _useId.useId)();
  const mergedLocaleText = React.useMemo(() => (0, _extends2.default)({}, _defaultLocaleText.EVENT_CALENDAR_DEFAULT_LOCALE_TEXT, localeText), [localeText]);
  const calendarStyledContextValue = React.useMemo(() => ({
    schedulerId,
    classes,
    localeText: mergedLocaleText
  }), [schedulerId, classes, mergedLocaleText]);
  const dialogStyledContextValue = React.useMemo(() => ({
    schedulerId,
    classes,
    localeText: mergedLocaleText
  }), [schedulerId, classes, mergedLocaleText]);
  const sharedComponentsStyledContextValue = React.useMemo(() => ({
    classes
  }), [classes]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_useSchedulerStoreContext.SchedulerStoreContext.Provider, {
    value: store,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_EventCalendarStyledContext.EventCalendarStyledContext.Provider, {
      value: calendarStyledContextValue,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_EventDialogStyledContext.EventDialogStyledContext.Provider, {
        value: dialogStyledContextValue,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_SharedComponentsStyledContext.SharedComponentsStyledContext.Provider, {
          value: sharedComponentsStyledContextValue,
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventDialog.EventDialogProvider, {
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_EventCalendarRoot.EventCalendarRoot, (0, _extends2.default)({
              className: className
            }, other, {
              ref: forwardedRef
            }))
          })
        })
      })
    })
  });
});
if (process.env.NODE_ENV !== "production") EventCalendar.displayName = "EventCalendar";
process.env.NODE_ENV !== "production" ? EventCalendar.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * The ref object that allows Event Calendar manipulation.
   * Can be instantiated with `useEventCalendarApiRef()`.
   */
  apiRef: _propTypes.default.shape({
    current: _propTypes.default.shape({
      setVisibleDate: _propTypes.default.func
    })
  }),
  /**
   * Whether the event can be dragged to change its start and end dates without changing the duration.
   * @default true
   */
  areEventsDraggable: _propTypes.default.bool,
  /**
   * Whether the event start or end can be dragged to change its duration without changing its other date.
   * If `true`, both start and end can be resized.
   * If `false`, the events are not resizable.
   * If `"start"`, only the start can be resized.
   * If `"end"`, only the end can be resized.
   * @default true
   */
  areEventsResizable: _propTypes.default.oneOfType([_propTypes.default.oneOf(['end', 'start']), _propTypes.default.bool]),
  /**
   * Whether events can be dragged from outside of the calendar and dropped into it.
   * @default false
   */
  canDragEventsFromTheOutside: _propTypes.default.bool,
  /**
   * Whether events can be dragged from inside of the calendar and dropped outside of it.
   * If true, when the mouse leaves the calendar, the event won't be rendered inside the calendar anymore.
   * If false, when the mouse leaves the calendar, the event will be rendered in its last valid position inside the calendar.
   * @default false
   */
  canDropEventsToTheOutside: _propTypes.default.bool,
  /**
   * Override or extend the styles applied to the component.
   */
  classes: _propTypes.default.object,
  /**
   * Data source for fetching events asynchronously.
   * When provided, events are fetched through the data source instead of the `events` prop.
   */
  dataSource: _propTypes.default.shape({
    getEvents: _propTypes.default.func.isRequired,
    persistEvents: _propTypes.default.func.isRequired
  }),
  /**
   * The locale object from `date-fns` used to format dates.
   * This affects day names, month names, week start day, and other locale-dependent formatting.
   * Import a locale from `date-fns/locale` and pass it to this prop.
   * @default enUS (English)
   */
  dateLocale: _propTypes.default.object,
  /**
   * The default preferences for the calendar.
   * To use controlled preferences, use the `preferences` prop.
   * @default { showWeekends: true, showWeekNumber: false, isSidePanelOpen: true, showEmptyDaysInAgenda: true, ampm: true }
   */
  defaultPreferences: _propTypes.default.shape({
    ampm: _propTypes.default.bool,
    isSidePanelOpen: _propTypes.default.bool,
    showEmptyDaysInAgenda: _propTypes.default.bool,
    showWeekends: _propTypes.default.bool,
    showWeekNumber: _propTypes.default.bool,
    weekStartsOn: _propTypes.default.oneOf([0, 1, 2, 3, 4, 5, 6])
  }),
  /**
   * The view initially displayed in the calendar.
   * To render a controlled calendar, use the `view` prop.
   * @default "week"
   */
  defaultView: _propTypes.default.oneOf(['agenda', 'day', 'month', 'week']),
  /**
   * The date initially used to determine the visible date range.
   * To render a controlled component, use the `visibleDate` prop.
   * @default today
   */
  defaultVisibleDate: _propTypes.default.instanceOf(Date),
  /**
   * The IDs of the resources initially visible.
   * To render a controlled scheduler, use the `visibleResources` prop.
   * @default {} - all resources are visible
   */
  defaultVisibleResources: _propTypes.default.object,
  /**
   * The timezone used to display events in the scheduler.
   *
   * Accepts any valid IANA timezone name
   * (for example "America/New_York", "Europe/Paris", "Asia/Tokyo"),
   * or keywords understood by the adapter, such as
   * "default" (use the adapter's default timezone),
   * "locale" (use the user's current locale timezone),
   * or "UTC".
   *
   * This timezone only affects rendering, events keep their original data timezone.
   * @default "default"
   */
  displayTimezone: _propTypes.default.string,
  /**
   * The color palette used for all events.
   * Can be overridden per resource using the `eventColor` property on the resource model.
   * Can be overridden per event using the `color` property on the event model.
   * @default "teal"
   */
  eventColor: _propTypes.default.oneOf(['amber', 'blue', 'green', 'grey', 'indigo', 'lime', 'orange', 'pink', 'purple', 'red', 'teal']),
  /**
   * Configures how events are created.
   * If `false`, event creation is disabled.
   * If `true`, event creation is enabled with default configuration.
   * If an object, event creation is enabled with the provided configuration.
   */
  eventCreation: _propTypes.default.oneOfType([_propTypes.default.shape({
    duration: _propTypes.default.number,
    interaction: _propTypes.default.oneOf(['click', 'double-click'])
  }), _propTypes.default.bool]),
  /**
   * The structure of the event model.
   * It defines how to read and write the properties of the event model.
   * If not provided, the event model is assumed to match the `CalendarEvent` interface.
   */
  eventModelStructure: _propTypes.default.object,
  /**
   * The events currently available in the calendar.
   * @default []
   */
  events: _propTypes.default.arrayOf(_propTypes.default.object),
  /**
   * Set the locale text of the Event Calendar.
   * You can find all the translation keys supported in [the source](https://github.com/mui/mui-x/blob/HEAD/packages/x-scheduler/src/models/translations.ts)
   * in the GitHub repository.
   */
  localeText: _propTypes.default.object,
  /**
   * Callback fired when some event of the calendar change.
   */
  onEventsChange: _propTypes.default.func,
  /**
   * Event handler called when the preferences change.
   */
  onPreferencesChange: _propTypes.default.func,
  /**
   * Event handler called when the view changes.
   */
  onViewChange: _propTypes.default.func,
  /**
   * Event handler called when the visible date changes.
   */
  onVisibleDateChange: _propTypes.default.func,
  /**
   * Event handler called when the visible resources change.
   */
  onVisibleResourcesChange: _propTypes.default.func,
  /**
   * Preferences currently displayed in the calendar.
   */
  preferences: _propTypes.default.shape({
    ampm: _propTypes.default.bool,
    isSidePanelOpen: _propTypes.default.bool,
    showEmptyDaysInAgenda: _propTypes.default.bool,
    showWeekends: _propTypes.default.bool,
    showWeekNumber: _propTypes.default.bool,
    weekStartsOn: _propTypes.default.oneOf([0, 1, 2, 3, 4, 5, 6])
  }),
  /**
   * Config of the preferences menu.
   * Defines which options are visible in the menu.
   * If `false`, the menu will be entirely hidden.
   * @default { toggleWeekendVisibility: true, toggleWeekNumberVisibility: true, toggleAmpm: true, toggleEmptyDaysInAgenda: true, toggleWeekStartsOn: false }
   */
  preferencesMenuConfig: _propTypes.default.oneOfType([_propTypes.default.oneOf([false]), _propTypes.default.shape({
    toggleAmpm: _propTypes.default.bool,
    toggleEmptyDaysInAgenda: _propTypes.default.bool,
    toggleWeekendVisibility: _propTypes.default.bool,
    toggleWeekNumberVisibility: _propTypes.default.bool,
    toggleWeekStartsOn: _propTypes.default.bool
  })]),
  /**
   * Whether the calendar is in read-only mode.
   * @default false
   */
  readOnly: _propTypes.default.bool,
  /**
   * The structure of the resource model.
   * It defines how to read and write the properties of the resource model.
   * If not provided, the resource model is assumed to match the `CalendarResource` interface.
   */
  resourceModelStructure: _propTypes.default.object,
  /**
   * The resources the events can be assigned to.
   */
  resources: _propTypes.default.arrayOf(_propTypes.default.object),
  /**
   * Whether each event must be assigned to a resource. When true, the resource cannot be cleared in the edit dialog and the form cannot be submitted without one.
   * @default false
   */
  shouldEventRequireResource: _propTypes.default.bool,
  /**
   * Whether the component should display the current time indicator.
   * @default true
   */
  showCurrentTimeIndicator: _propTypes.default.bool,
  /**
   * The system prop that allows defining system overrides as well as additional CSS styles.
   */
  sx: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object, _propTypes.default.bool])), _propTypes.default.func, _propTypes.default.object]),
  /**
   * The view currently displayed in the calendar.
   */
  view: _propTypes.default.oneOf(['agenda', 'day', 'month', 'week']),
  /**
   * The views available in the calendar.
   * @default ["day", "week", "month", "agenda"]
   */
  views: _propTypes.default.arrayOf(_propTypes.default.oneOf(['agenda', 'day', 'month', 'week']).isRequired),
  /**
   * The date currently used to determine the visible date range.
   */
  visibleDate: _propTypes.default.instanceOf(Date),
  /**
   * The IDs of the resources currently visible.
   * A resource is visible if it is not included in this object or if it is included with `true` value.
   */
  visibleResources: _propTypes.default.object
} : void 0;
