// @mui/x-charts/BarChart/BarElement.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BarElement = BarElement;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _useSlotProps = _interopRequireDefault(require("@mui/utils/useSlotProps"));
var _barClasses = require("./barClasses");
var _useItemHighlightState = require("../hooks/useItemHighlightState");
var _AnimatedBarElement = require("./AnimatedBarElement");
var _useIsItemFocused = require("../hooks/useIsItemFocused");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["seriesId", "dataIndex", "classes", "color", "slots", "slotProps", "style", "onClick", "skipAnimation", "layout", "x", "xOrigin", "y", "yOrigin", "width", "height", "hidden"];
function BarElement(props) {
  const {
      seriesId,
      dataIndex,
      classes: innerClasses,
      color,
      slots,
      slotProps,
      style,
      onClick,
      skipAnimation,
      layout,
      x,
      xOrigin,
      y,
      yOrigin,
      width,
      height,
      hidden
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const itemIdentifier = React.useMemo(() => ({
    type: 'bar',
    seriesId,
    dataIndex
  }), [seriesId, dataIndex]);
  const highlightState = (0, _useItemHighlightState.useItemHighlightState)(itemIdentifier);
  const isHighlighted = highlightState === 'highlighted';
  const isFaded = highlightState === 'faded';
  const isFocused = (0, _useIsItemFocused.useIsItemFocused)(itemIdentifier);
  const ownerState = {
    seriesId,
    dataIndex,
    classes: innerClasses,
    color,
    isFaded,
    isHighlighted,
    isFocused
  };
  const classes = (0, _barClasses.useUtilityClasses)(ownerState);
  const Bar = slots?.bar ?? _AnimatedBarElement.AnimatedBarElement;
  const barProps = (0, _useSlotProps.default)({
    elementType: Bar,
    externalSlotProps: slotProps?.bar,
    externalForwardedProps: other,
    additionalProps: {
      seriesId,
      dataIndex,
      color,
      x,
      xOrigin,
      y,
      yOrigin,
      width,
      height,
      style,
      onClick,
      cursor: onClick ? 'pointer' : undefined,
      stroke: 'none',
      fill: color,
      skipAnimation,
      layout,
      hidden
    },
    className: classes.element,
    ownerState
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(Bar, (0, _extends2.default)({}, barProps));
}
process.env.NODE_ENV !== "production" ? BarElement.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  classes: _propTypes.default.object,
  dataIndex: _propTypes.default.number.isRequired,
  hidden: _propTypes.default.bool,
  layout: _propTypes.default.oneOf(['horizontal', 'vertical']).isRequired,
  seriesId: _propTypes.default.string.isRequired,
  skipAnimation: _propTypes.default.bool.isRequired,
  /**
   * The props used for each component slot.
   * @default {}
   */
  slotProps: _propTypes.default.object,
  /**
   * Overridable component slots.
   * @default {}
   */
  slots: _propTypes.default.object,
  xOrigin: _propTypes.default.number.isRequired,
  yOrigin: _propTypes.default.number.isRequired
} : void 0;
