// @mui/x-charts/BarChart/BarPlot.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BarPlot = BarPlot;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _clsx = _interopRequireDefault(require("clsx"));
var _styles = require("@mui/material/styles");
var _hooks = require("../hooks");
var _BarLabelPlot = require("./BarLabel/BarLabelPlot");
var _useSkipAnimation = require("../hooks/useSkipAnimation");
var _useInternalIsZoomInteracting = require("../internals/plugins/featurePlugins/useChartCartesianAxis/useInternalIsZoomInteracting");
var _useBarPlotData = require("./useBarPlotData");
var _barClasses = require("./barClasses");
var _animation = require("../internals/animation/animation");
var _IndividualBarPlot = require("./IndividualBarPlot");
var _BatchBarPlot = require("./BatchBarPlot");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className", "skipAnimation", "onItemClick", "borderRadius", "renderer"];
const BarPlotRoot = (0, _styles.styled)('g', {
  name: 'MuiBarPlot',
  slot: 'Root'
})({
  [`& .${_barClasses.barClasses.element}`]: {
    transitionProperty: 'opacity, fill',
    transitionDuration: `${_animation.ANIMATION_DURATION_MS}ms`,
    transitionTimingFunction: _animation.ANIMATION_TIMING_FUNCTION
  }
});

/**
 * Demos:
 *
 * - [Bars](https://mui.com/x/react-charts/bars/)
 * - [Bar demonstration](https://mui.com/x/react-charts/bar-demo/)
 * - [Stacking](https://mui.com/x/react-charts/stacking/)
 *
 * API:
 *
 * - [BarPlot API](https://mui.com/x/api/charts/bar-plot/)
 */
function BarPlot(props) {
  const {
      className,
      skipAnimation: inSkipAnimation,
      onItemClick,
      borderRadius,
      renderer
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const isZoomInteracting = (0, _useInternalIsZoomInteracting.useInternalIsZoomInteracting)();
  const skipAnimation = (0, _useSkipAnimation.useSkipAnimation)(isZoomInteracting || inSkipAnimation);
  const batchSkipAnimation = (0, _useSkipAnimation.useSkipAnimation)(inSkipAnimation);
  const {
    xAxis: xAxes
  } = (0, _hooks.useXAxes)();
  const {
    yAxis: yAxes
  } = (0, _hooks.useYAxes)();
  const {
    completedData,
    masksData
  } = (0, _useBarPlotData.useBarPlotData)((0, _hooks.useDrawingArea)(), xAxes, yAxes);
  const classes = (0, _barClasses.useUtilityClasses)();
  const BarElementPlot = renderer === 'svg-batch' ? _BatchBarPlot.BatchBarPlot : _IndividualBarPlot.IndividualBarPlot;
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(BarPlotRoot, {
    className: (0, _clsx.default)(classes.root, className),
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(BarElementPlot, (0, _extends2.default)({
      completedData: completedData,
      masksData: masksData
      /* The batch renderer doesn't animate bars after the initial mount. Providing skipAnimation was causing an issue
       * where bars would animate again after a zoom interaction because skipAnimation would change from true to false. */,
      skipAnimation: renderer === 'svg-batch' ? batchSkipAnimation : skipAnimation,
      onItemClick: onItemClick,
      borderRadius: borderRadius
    }, other)), completedData.map(processedSeries => /*#__PURE__*/(0, _jsxRuntime.jsx)(_BarLabelPlot.BarLabelPlot, (0, _extends2.default)({
      className: classes.seriesLabels,
      processedSeries: processedSeries,
      skipAnimation: skipAnimation
    }, other), processedSeries.seriesId))]
  });
}
process.env.NODE_ENV !== "production" ? BarPlot.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * Defines the border radius of the bar element.
   */
  borderRadius: _propTypes.default.number,
  /**
   * A CSS class name applied to the root element.
   */
  className: _propTypes.default.string,
  /**
   * Callback fired when a bar item is clicked.
   * @param {MouseEvent | React.MouseEvent<SVGElement, MouseEvent>} event The event source of the callback.
   *        It is a native MouseEvent for `svg-batch` renderer and a React MouseEvent for `svg-single` renderer.
   * @param {BarItemIdentifier} barItemIdentifier The bar item identifier.
   */
  onItemClick: _propTypes.default.func,
  /**
   * The type of renderer to use for the bar plot.
   * - `svg-single`: Renders every bar in a `<rect />` element.
   * - `svg-batch`: Batch renders bars in `<path />` elements for better performance with large datasets, at the cost of some limitations.
   *                Read more: https://mui.com/x/react-charts/bars/#performance
   *
   * @default 'svg-single'
   */
  renderer: _propTypes.default.oneOf(['svg-batch', 'svg-single']),
  /**
   * If `true`, animations are skipped.
   * @default undefined
   */
  skipAnimation: _propTypes.default.bool,
  /**
   * The props used for each component slot.
   * @default {}
   */
  slotProps: _propTypes.default.object,
  /**
   * Overridable component slots.
   * @default {}
   */
  slots: _propTypes.default.object
} : void 0;
