// @mui/x-scheduler-internals/use-event-occurrences-with-day-grid-position/useEventOccurrencesWithDayGridPosition.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useEventOccurrencesWithDayGridPosition = useEventOccurrencesWithDayGridPosition;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _useAdapterContext = require("../use-adapter-context");
var _sortEventOccurrences = require("../sort-event-occurrences");
/**
 * Places event occurrences for a list of days, where if an event is rendered in a day, it fills the entire day cell (no notion of time).
 */
function useEventOccurrencesWithDayGridPosition(parameters) {
  const {
    days,
    occurrencesMap,
    shouldAddPosition
  } = parameters;
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  return React.useMemo(() => {
    const indexLookup = {};
    const dayListSize = days.length;
    const processedDays = days.map((day, dayIndex) => {
      indexLookup[day.key] = {
        occurrencesIndex: {},
        usedIndexes: new Set()
      };
      const needsPosition = [];
      const withoutPosition = [];

      // 1. Split occurrences into withPosition and withoutPosition
      for (const occurrence of occurrencesMap.get(day.key) ?? []) {
        const hasPosition = shouldAddPosition ? shouldAddPosition(occurrence) : true;
        if (hasPosition) {
          needsPosition.push(occurrence);
        } else {
          withoutPosition.push(occurrence);
        }
      }

      // 2. Sort the withPosition occurrences by start and end date
      const sortedNeedsPosition = (0, _sortEventOccurrences.sortEventOccurrences)(needsPosition);

      // 3. Assign position to each occurrence
      const withPosition = [];
      for (const occurrence of sortedNeedsPosition) {
        let position;
        const occurrenceIndexInPreviousDay = dayIndex === 0 ? null : indexLookup[days[dayIndex - 1].key].occurrencesIndex[occurrence.key];

        // If the event is present in the previous day, we keep the same index
        if (occurrenceIndexInPreviousDay != null) {
          position = {
            index: occurrenceIndexInPreviousDay,
            daySpan: 1,
            isInvisible: true
          };
        }
        // Otherwise, we find the smallest available index
        else {
          const usedIndexes = indexLookup[day.key].usedIndexes;
          let i = 1;
          while (usedIndexes.has(i)) {
            i += 1;
          }
          const durationInDays = adapter.differenceInDays(occurrence.displayTimezone.end.value, day.value) + 1;
          position = {
            index: i,
            daySpan: Math.min(durationInDays, dayListSize - dayIndex) // Don't go past the day list end
          };
        }
        indexLookup[day.key].occurrencesIndex[occurrence.key] = position.index;
        indexLookup[day.key].usedIndexes.add(position.index);
        withPosition.push((0, _extends2.default)({}, occurrence, {
          position
        }));
      }

      // Sort the occurrences by their index to make sure they are in the order they should be rendered in.
      withPosition.sort((a, b) => a.position.index - b.position.index);
      return (0, _extends2.default)({}, day, {
        withPosition,
        withoutPosition
      });
    });
    const usedIndexes = Object.values(indexLookup).flatMap(day => Array.from(day.usedIndexes));
    return {
      days: processedDays,
      maxIndex: usedIndexes.length === 0 ? 1 : Math.max(...usedIndexes)
    };
  }, [adapter, days, occurrencesMap, shouldAddPosition]);
}
