// @mui/x-scheduler/internals/components/create-modal/createModal.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createModal = createModal;
var _formatErrorMessage2 = _interopRequireDefault(require("@mui/x-internals/formatErrorMessage"));
var React = _interopRequireWildcard(require("react"));
var _useStableCallback = require("@base-ui/utils/useStableCallback");
var _EventManager = require("@mui/x-internals/EventManager");
var _jsxRuntime = require("react/jsx-runtime");
function createModal(config) {
  const Context = /*#__PURE__*/React.createContext(undefined);
  if (process.env.NODE_ENV !== "production") Context.displayName = "Context";
  function useModalContext() {
    const context = React.useContext(Context);
    if (!context) {
      throw new Error(process.env.NODE_ENV !== "production" ? `MUI X Scheduler: \`${config.contextName}\` is missing. Hook must be placed within its Provider.` : (0, _formatErrorMessage2.default)(206, config.contextName));
    }
    return context;
  }
  function Provider(props) {
    const {
      children,
      render,
      onOpen: onOpenProp,
      onClose: onCloseProp
    } = props;
    const anchorRef = React.useRef(null);
    const eventManager = React.useRef(new _EventManager.EventManager());
    const [state, setState] = React.useState({
      isOpen: false,
      data: null
    });
    const onOpen = (0, _useStableCallback.useStableCallback)((forwardedAnchorRef, data) => {
      anchorRef.current = forwardedAnchorRef?.current ?? null;
      setState({
        isOpen: true,
        data
      });
      onOpenProp?.(data);
      eventManager.current.emit('open', {
        data,
        anchorRef: anchorRef.current
      });
    });
    const onClose = (0, _useStableCallback.useStableCallback)(() => {
      onCloseProp?.();
      setState({
        isOpen: false,
        data: null
      });
      eventManager.current.emit('close');
    });
    const subscribeCloseHandler = React.useCallback(handler => {
      eventManager.current.on('close', handler);
      return () => {
        eventManager.current.removeListener('close', handler);
      };
    }, []);
    const contextValue = React.useMemo(() => ({
      onOpen,
      onClose,
      isOpen: state.isOpen || false,
      data: state.data,
      subscribeCloseHandler
    }), [onOpen, onClose, state.isOpen, state.data, subscribeCloseHandler]);
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)(Context.Provider, {
      value: contextValue,
      children: [children, state.data && anchorRef.current && render({
        isOpen: Boolean(state.isOpen),
        anchorRef,
        data: state.data,
        onClose
      })]
    });
  }
  const Trigger = /*#__PURE__*/React.forwardRef(function Trigger(props, ref) {
    const {
      data,
      onClick,
      children
    } = props;
    const {
      onOpen
    } = useModalContext();
    return /*#__PURE__*/React.cloneElement(children, {
      ref,
      onClick: event => {
        onClick?.(event);
        onOpen(ref, data);
      }
    });
  });
  if (process.env.NODE_ENV !== "production") Trigger.displayName = "Trigger";
  return {
    Context,
    useContext: useModalContext,
    Provider,
    Trigger
  };
}
