// @mui/x-charts/RadarChart/RadarAxis/useRadarAxis.mjs
import { warnOnce } from '@mui/x-internals/warning';
import { useRadiusAxes } from "../../hooks/useAxis.mjs";
import { useRotationScale } from "../../hooks/useScale.mjs";
import { useChartsContext } from "../../context/ChartsProvider/useChartsContext.mjs";
import { selectorChartPolarCenter } from "../../internals/plugins/featurePlugins/useChartPolarAxis/index.mjs";
import { isOrdinalScale } from "../../internals/scaleGuards.mjs";
import { degToRad } from "../../internals/degToRad.mjs";
import { clampAngle } from "../../internals/clampAngle.mjs";
import { rad2deg } from "../../internals/angleConversion.mjs";
/**
 * Returns an array with one item per metric with the different points to label.
 */
export function useRadarAxis(params) {
  const {
    metric,
    angle,
    divisions = 1
  } = params;
  const {
    instance,
    store
  } = useChartsContext();
  const rotationScale = useRotationScale();
  const {
    radiusAxis
  } = useRadiusAxes();
  const {
    cx,
    cy
  } = store.use(selectorChartPolarCenter);
  if (metric === undefined || !rotationScale || rotationScale.domain().length === 0) {
    return null;
  }
  const existingMetrics = rotationScale.domain();
  if (process.env.NODE_ENV !== 'production' && !existingMetrics.includes(metric)) {
    warnOnce([`MUI X Charts: Your radar axis tries to display values for the metric "${metric}" which does not exist.`, `Either add this metric to your radar, or pick one from the existing metrics: ${existingMetrics.join(', ')}`]);
  }
  const anglesWithDefault = angle !== undefined ? degToRad(angle) : rotationScale(metric) ?? 0;
  const radiusRatio = Array.from({
    length: divisions
  }, (_, index) => (index + 1) / divisions);
  const radiusScale = radiusAxis[metric].scale;
  const R = radiusScale.range()[1];
  if (isOrdinalScale(radiusScale)) {
    if (process.env.NODE_ENV !== 'production') {
      console.error('MUI X Charts: Radar chart does not support ordinal axes');
    }
    return null;
  }
  return {
    metric,
    angle: clampAngle(rad2deg(anglesWithDefault)),
    center: {
      x: cx,
      y: cy
    },
    labels: radiusRatio.map(ratio => {
      const radius = ratio * R;
      const [x, y] = instance.polar2svg(radius, anglesWithDefault);
      const value = radiusScale.invert(radius);
      const defaultTickLabel = value.toString();
      return {
        x,
        y,
        value,
        formattedValue: radiusAxis[metric].valueFormatter?.(radiusScale.invert(radius), {
          location: 'tick',
          scale: radiusScale,
          defaultTickLabel,
          tickNumber: divisions
        }) ?? defaultTickLabel
      };
    })
  };
}
