// @mui/x-scheduler-internals/calendar-grid/time-event/CalendarGridTimeEvent.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarGridTimeEvent = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _useStableCallback = require("@base-ui/utils/useStableCallback");
var _useId = require("@base-ui/utils/useId");
var _useButton = require("../../base-ui-copy/utils/useButton");
var _useRenderElement = require("../../base-ui-copy/utils/useRenderElement");
var _CalendarGridTimeEventCssVars = require("./CalendarGridTimeEventCssVars");
var _CalendarGridTimeColumnContext = require("../time-column/CalendarGridTimeColumnContext");
var _useDraggableEvent = require("../../internals/utils/useDraggableEvent");
var _useElementPositionInCollection = require("../../internals/utils/useElementPositionInCollection");
var _accessibilityUtils = require("../../internals/utils/accessibility-utils");
var _CalendarGridTimeEventContext = require("./CalendarGridTimeEventContext");
var _useAdapterContext = require("../../use-adapter-context");
var _useEventCalendarStoreContext = require("../../use-event-calendar-store-context");
var _schedulerSelectors = require("../../scheduler-selectors");
var _CalendarGridRootContext = require("../root/CalendarGridRootContext");
var _eventUtils = require("../../internals/utils/event-utils");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className", "render", "style", "start", "end", "eventId", "occurrenceKey", "renderDragPreview", "id", "isDraggable", "nativeButton"];
const CalendarGridTimeEvent = exports.CalendarGridTimeEvent = /*#__PURE__*/React.forwardRef(function CalendarGridTimeEvent(componentProps, forwardedRef) {
  const {
      // Rendering props

      // Internal props
      start,
      end,
      eventId,
      occurrenceKey,
      renderDragPreview,
      id: idProp,
      isDraggable = false,
      nativeButton = false
      // Props forwarded to the DOM element
    } = componentProps,
    elementProps = (0, _objectWithoutPropertiesLoose2.default)(componentProps, _excluded);

  // TODO: Expose a real `interactive` prop
  // to control whether the event should behave like a button
  const isInteractive = true;

  // Context hooks
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const {
    id: rootId
  } = (0, _CalendarGridRootContext.useCalendarGridRootContext)();
  const {
    start: columnStart,
    end: columnEnd,
    index: columnIndex,
    hasFocus: columnHasFocus,
    getCursorPositionInElementMs
  } = (0, _CalendarGridTimeColumnContext.useCalendarGridTimeColumnContext)();

  // Ref hooks
  const ref = React.useRef(null);

  // State hooks
  const id = (0, _useId.useId)(idProp);

  // Feature hooks
  const getSharedDragData = (0, _useStableCallback.useStableCallback)(input => {
    const offsetBeforeColumnStart = Math.max(adapter.getTime(columnStart) - start.timestamp, 0);
    const event = _schedulerSelectors.schedulerEventSelectors.processedEvent(store.state, eventId);
    const originalOccurrence = (0, _eventUtils.generateOccurrenceFromEvent)({
      event,
      eventId,
      occurrenceKey,
      start,
      end
    });
    const offsetInsideColumn = getCursorPositionInElementMs({
      input,
      elementRef: ref
    });
    return {
      eventId,
      occurrenceKey,
      originalOccurrence,
      start: start.value,
      end: end.value,
      initialCursorPositionInEventMs: offsetBeforeColumnStart + offsetInsideColumn
    };
  });
  const getDragData = (0, _useStableCallback.useStableCallback)(input => (0, _extends2.default)({}, getSharedDragData(input), {
    source: 'CalendarGridTimeEvent'
  }));
  const {
    state,
    preview,
    contextValue: draggableEventContextValue
  } = (0, _useDraggableEvent.useDraggableEvent)({
    ref,
    start,
    end,
    occurrenceKey,
    eventId,
    isDraggable,
    renderDragPreview,
    getDragData,
    collectionStart: columnStart,
    collectionEnd: columnEnd
  });
  const {
    getButtonProps,
    buttonRef
  } = (0, _useButton.useButton)({
    disabled: !isInteractive,
    native: nativeButton,
    tabIndex: columnHasFocus ? 0 : -1
  });
  const {
    position,
    duration
  } = (0, _useElementPositionInCollection.useElementPositionInCollection)({
    start,
    end,
    collectionStart: columnStart,
    collectionEnd: columnEnd
  });
  const columnHeaderId = (0, _accessibilityUtils.getCalendarGridHeaderCellId)(rootId, columnIndex);
  const contextValue = React.useMemo(() => (0, _extends2.default)({}, draggableEventContextValue, {
    getSharedDragData
  }), [draggableEventContextValue, getSharedDragData]);
  const element = (0, _useRenderElement.useRenderElement)('div', componentProps, {
    state,
    ref: [forwardedRef, buttonRef, ref],
    props: [elementProps, {
      id,
      'aria-labelledby': `${columnHeaderId} ${id}`,
      style: {
        [_CalendarGridTimeEventCssVars.CalendarGridTimeEventCssVars.yPosition]: `${position * 100}%`,
        [_CalendarGridTimeEventCssVars.CalendarGridTimeEventCssVars.height]: `${duration * 100}%`
      }
    }, getButtonProps]
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_CalendarGridTimeEventContext.CalendarGridTimeEventContext.Provider, {
    value: contextValue,
    children: [element, preview.element]
  });
});
if (process.env.NODE_ENV !== "production") CalendarGridTimeEvent.displayName = "CalendarGridTimeEvent";
