// @mui/x-scheduler/internals/components/event/event-item/EventItem.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EventItem = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _formatErrorMessage2 = _interopRequireDefault(require("@mui/x-internals/formatErrorMessage"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _styles = require("@mui/material/styles");
var _useId = require("@base-ui/utils/useId");
var _store = require("@base-ui/utils/store");
var _RepeatRounded = _interopRequireDefault(require("@mui/icons-material/RepeatRounded"));
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _button = require("@base-ui/react/button");
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _useFormatTime = require("../../../hooks/useFormatTime");
var _EventCalendarStyledContext = require("../../../../event-calendar/EventCalendarStyledContext");
var _tokens = require("../../../utils/tokens");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["occurrence", "ariaLabelledBy", "id", "variant", "className", "onClick"];
const EventItemCard = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'EventItemCard'
})(({
  theme
}) => ({
  padding: 0,
  borderRadius: theme.shape.borderRadius,
  '&:hover': {
    backgroundColor: (theme.vars || theme).palette.action.hover
  },
  '&:focus-visible': {
    outline: '2px solid var(--event-surface-accent)',
    outlineOffset: 1
  },
  '&[data-variant="compact"], &[data-variant="regular"]': {
    containerType: 'inline-size',
    cursor: 'pointer',
    height: 'fit-content'
  },
  '&[data-variant="filled"]': {
    backgroundColor: 'var(--event-surface-bold)',
    color: 'var(--event-on-surface-bold)',
    '&:hover': {
      backgroundColor: 'var(--event-surface-bold-hover)'
    },
    '&[data-editing]': {
      backgroundColor: 'var(--event-surface-selected)',
      color: 'var(--event-on-surface-selected)',
      '&:hover': {
        backgroundColor: 'var(--event-surface-selected-hover)'
      }
    }
  },
  '&[data-variant="regular"]': {
    cursor: 'pointer'
  },
  '&[data-editing]': {
    backgroundColor: 'var(--event-surface-selected)',
    '&:hover': {
      backgroundColor: 'var(--event-surface-selected-hover)'
    }
  },
  variants: (0, _tokens.getPaletteVariants)(theme)
}));
const EventItemCardWrapper = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'EventItemCardWrapper'
})(({
  theme
}) => ({
  display: 'flex',
  alignItems: 'center',
  gap: theme.spacing(1),
  padding: theme.spacing(0.5, 1),
  '&[data-variant="compact"], &[data-variant="filled"]': {
    padding: `0 ${theme.spacing(0.5)}`
  }
}));
const EventItemTitle = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'EventItemTitle'
})(({
  theme
}) => ({
  margin: 0,
  color: (theme.vars || theme).palette.text.primary,
  fontWeight: theme.typography.fontWeightMedium,
  fontSize: 'var(--EventCalendar-fontSize-eventTitle, 0.75rem)',
  lineHeight: 1.43,
  '[data-editing] &': {
    color: 'var(--event-on-surface-selected)'
  }
}));
const EventItemTime = (0, _styles.styled)('time', {
  name: 'MuiEventCalendar',
  slot: 'EventItemTime'
})(({
  theme
}) => ({
  display: 'inline-block',
  color: (theme.vars || theme).palette.text.secondary,
  fontWeight: theme.typography.fontWeightRegular,
  fontSize: theme.typography.caption.fontSize,
  lineHeight: 1.43,
  whiteSpace: 'nowrap',
  width: 150,
  '&[data-compact]': {
    width: 'fit-content',
    marginInlineEnd: theme.spacing(0.5)
  },
  '[data-editing] &': {
    color: 'var(--event-on-surface-selected)'
  }
}));
const EventItemRecurringIcon = (0, _styles.styled)(_RepeatRounded.default, {
  name: 'MuiEventCalendar',
  slot: 'EventItemRecurringIcon'
})(({
  theme
}) => ({
  color: (theme.vars || theme).palette.text.primary,
  '[data-editing] &': {
    color: 'var(--event-on-surface-selected)'
  }
}));
const ResourceLegendColor = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'ResourceLegendColor'
})({
  width: 8,
  height: 8,
  borderRadius: '50%',
  flexShrink: 0,
  backgroundColor: 'var(--event-main)',
  '[data-editing] &': {
    backgroundColor: 'var(--event-on-surface-selected)'
  }
});
const EventItemCardContent = (0, _styles.styled)('p', {
  name: 'MuiEventCalendar',
  slot: 'EventItemCardContent'
})({
  margin: 0,
  height: 20,
  lineHeight: '20px'
});
const EventItemLinesClamp = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'EventItemLinesClamp'
})({
  display: '-webkit-box',
  WebkitLineClamp: 'var(--number-of-lines)',
  WebkitBoxOrient: 'vertical',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  wordBreak: 'break-word',
  overflowWrap: 'break-word'
});

/**
 * Component used to display an event occurrence, without any positioning capabilities
 * Used in <AgendaView /> and in the event popover of <MonthView /> to display the list of events for a specific day.
 */
const EventItem = exports.EventItem = /*#__PURE__*/React.forwardRef(function EventItem(props, forwardedRef) {
  const {
      occurrence,
      ariaLabelledBy,
      id: idProp,
      variant = 'regular',
      className,
      onClick
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);

  // Context hooks
  const {
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const isEditing = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.isEditedEvent, occurrence.id);

  // State hooks
  const id = (0, _useId.useId)(idProp);

  // Selector hooks
  const resource = (0, _store.useStore)(store, _schedulerSelectors.schedulerResourceSelectors.processedResource, occurrence.resource);
  const color = (0, _store.useStore)(store, _schedulerSelectors.schedulerEventSelectors.color, occurrence.id);
  const isRecurring = (0, _store.useStore)(store, _schedulerSelectors.schedulerEventSelectors.isRecurring, occurrence.id);
  const formatTime = (0, _useFormatTime.useFormatTime)();
  const content = React.useMemo(() => {
    switch (variant) {
      case 'compact':
        return /*#__PURE__*/(0, _jsxRuntime.jsxs)(React.Fragment, {
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(ResourceLegendColor, {
            className: classes.resourceLegendColor,
            role: "img",
            "aria-label": resource?.title ? localeText.resourceAriaLabel(resource.title) : localeText.noResourceAriaLabel
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemLinesClamp, {
            className: classes.eventItemLinesClamp,
            style: {
              '--number-of-lines': 1
            },
            children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(EventItemCardContent, {
              className: classes.eventItemCardContent,
              children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemTime, {
                className: classes.eventItemTime,
                "data-compact": true,
                children: /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                  children: formatTime(occurrence.displayTimezone.start.value)
                })
              }), /*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemTitle, {
                className: classes.eventItemTitle,
                children: occurrence.title
              })]
            })
          }), isRecurring && /*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemRecurringIcon, {
            className: classes.eventItemRecurringIcon,
            "aria-hidden": "true",
            fontSize: "small"
          })]
        });
      case 'filled':
        return /*#__PURE__*/(0, _jsxRuntime.jsxs)(React.Fragment, {
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemLinesClamp, {
            className: classes.eventItemLinesClamp,
            style: {
              '--number-of-lines': 1
            },
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemTitle, {
              className: classes.eventItemTitle,
              children: occurrence.title
            })
          }), isRecurring && /*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemRecurringIcon, {
            className: classes.eventItemRecurringIcon,
            "aria-hidden": "true",
            fontSize: "small"
          })]
        });
      case 'regular':
        return /*#__PURE__*/(0, _jsxRuntime.jsxs)(React.Fragment, {
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(ResourceLegendColor, {
            className: classes.resourceLegendColor,
            role: "img",
            "aria-label": resource?.title ? localeText.resourceAriaLabel(resource.title) : localeText.noResourceAriaLabel
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemLinesClamp, {
            className: classes.eventItemLinesClamp,
            style: {
              '--number-of-lines': 1
            },
            children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(EventItemCardContent, {
              className: classes.eventItemCardContent,
              children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(MultiDayDateLabel, {
                occurrence: occurrence,
                formatTime: formatTime
              }), /*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemTitle, {
                className: classes.eventItemTitle,
                children: occurrence.title
              })]
            })
          }), isRecurring && /*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemRecurringIcon, {
            className: classes.eventItemRecurringIcon,
            "aria-hidden": "true",
            fontSize: "small"
          })]
        });
      default:
        throw new Error(process.env.NODE_ENV !== "production" ? 'MUI X Scheduler: Unsupported variant provided to EventItem component. ' + 'The EventItem component only supports specific variant values. ' + 'Check the component documentation for supported variants.' : (0, _formatErrorMessage2.default)(209));
    }
  }, [variant, resource?.title, localeText, formatTime, occurrence, isRecurring, classes]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_button.Button, {
    nativeButton: false,
    onClick: onClick,
    render: /*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemCard, (0, _extends2.default)({
      ref: forwardedRef,
      id: id,
      "data-variant": variant,
      "data-palette": color,
      "data-editing": isEditing || undefined,
      "aria-labelledby": `${ariaLabelledBy} ${id}`
    }, other, {
      className: (0, _clsx.default)(className, classes.eventItemCard, occurrence.className)
    })),
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemCardWrapper, {
      className: classes.eventItemCardWrapper,
      "data-variant": variant,
      children: content
    })
  });
});
if (process.env.NODE_ENV !== "production") EventItem.displayName = "EventItem";
function MultiDayDateLabel(props) {
  const {
    occurrence,
    formatTime
  } = props;
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const {
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  if (!adapter.isSameDay(occurrence.displayTimezone.start.value, occurrence.displayTimezone.end.value)) {
    const format = `${adapter.formats.dayOfMonth} ${adapter.formats.month3Letters}`;
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemTime, {
      className: classes.eventItemTime,
      as: "span",
      children: localeText.eventItemMultiDayLabel(adapter.formatByString(occurrence.displayTimezone.end.value, format))
    });
  }
  if (occurrence.allDay) {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(EventItemTime, {
      className: classes.eventItemTime,
      as: "span",
      children: localeText.allDay
    });
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(EventItemTime, {
    className: classes.eventItemTime,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
      children: formatTime(occurrence.displayTimezone.start.value)
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("span", {
      children: [" - ", formatTime(occurrence.displayTimezone.end.value)]
    })]
  });
}
