// @mui/x-scheduler-internals/event-calendar-selectors/eventCalendarPreferenceSelectors.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.eventCalendarPreferenceSelectors = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _store = require("@base-ui/utils/store");
var _EventCalendarStore = require("../use-event-calendar/EventCalendarStore");
const allPreferencesSelector = (0, _store.createSelectorMemoized)(state => state.preferences, preferences => (0, _extends2.default)({}, _EventCalendarStore.DEFAULT_EVENT_CALENDAR_PREFERENCES, preferences));
const eventCalendarPreferenceSelectors = exports.eventCalendarPreferenceSelectors = {
  all: allPreferencesSelector,
  menuConfig: (0, _store.createSelector)(state => state.preferencesMenuConfig),
  ampm: (0, _store.createSelector)(allPreferencesSelector, preferences => preferences.ampm),
  showWeekends: (0, _store.createSelector)(allPreferencesSelector, preferences => preferences.showWeekends),
  showWeekNumber: (0, _store.createSelector)(allPreferencesSelector, preferences => preferences.showWeekNumber),
  showEmptyDaysInAgenda: (0, _store.createSelector)(allPreferencesSelector, preferences => preferences.showEmptyDaysInAgenda),
  isSidePanelOpen: (0, _store.createSelector)(allPreferencesSelector, preferences => preferences.isSidePanelOpen),
  weekStartsOn: (0, _store.createSelector)(allPreferencesSelector, preferences => preferences.weekStartsOn)
};
