// @mui/x-scheduler-internals/event-calendar-selectors/eventCalendarAgendaSelectors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.eventCalendarAgendaSelectors = void 0;
var _store = require("@base-ui/utils/store");
var _schedulerSelectors = require("../scheduler-selectors");
var _eventCalendarPreferenceSelectors = require("./eventCalendarPreferenceSelectors");
var _useEventOccurrencesGroupedByDay = require("../use-event-occurrences-grouped-by-day");
var _constants = require("../constants");
var _getDayList = require("../get-day-list");
const eventCalendarAgendaSelectors = exports.eventCalendarAgendaSelectors = {
  visibleDays: (0, _store.createSelectorMemoized)(state => state.adapter, _schedulerSelectors.schedulerOtherSelectors.visibleDate, _schedulerSelectors.schedulerOtherSelectors.displayTimezone, _eventCalendarPreferenceSelectors.eventCalendarPreferenceSelectors.showWeekends, _eventCalendarPreferenceSelectors.eventCalendarPreferenceSelectors.showEmptyDaysInAgenda, _schedulerSelectors.schedulerEventSelectors.processedEventList, _schedulerSelectors.schedulerResourceSelectors.visibleMap, _schedulerSelectors.schedulerOtherSelectors.recurringEventsPlugin, (adapter, visibleDate, displayTimezone, showWeekends, showEmptyDaysInAgenda, events, visibleResources, recurringEventsPlugin) => {
    const amount = _constants.AGENDA_VIEW_DAYS_AMOUNT;

    // 1) First chunk of days
    let accumulatedDays = (0, _getDayList.getDayList)({
      adapter,
      start: visibleDate,
      end: adapter.addDays(visibleDate, amount - 1),
      excludeWeekends: !showWeekends
    });

    // Compute occurrences for the current accumulated range
    let occurrenceMap = (0, _useEventOccurrencesGroupedByDay.innerGetEventOccurrencesGroupedByDay)({
      adapter,
      days: accumulatedDays,
      events,
      visibleResources,
      displayTimezone,
      recurringEventsPlugin
    });
    const hasEvents = day => (occurrenceMap.get(day.key)?.length ?? 0) > 0;

    // 2) If we show empty days, just return the amount days
    if (showEmptyDaysInAgenda) {
      return accumulatedDays;
    }

    // 3) If we hide empty days, keep extending forward in blocks until we fill `amount` days with events
    let daysWithEvents = accumulatedDays.filter(hasEvents).slice(0, amount);
    while (daysWithEvents.length < amount) {
      // Stop if the calendar span already reaches the horizon
      const first = accumulatedDays[0]?.value;
      const last = accumulatedDays[accumulatedDays.length - 1]?.value;
      if (first && last) {
        const spanDays = adapter.differenceInDays(adapter.startOfDay(last), adapter.startOfDay(first)) + 1;

        // Hard stop to avoid scanning too far into the future
        if (spanDays >= _constants.AGENDA_MAX_HORIZON_DAYS) {
          break;
        }
      }

      // Extend forward by one more chunk and recompute occurrences over the accumulated range
      const nextStart = adapter.addDays(last ?? visibleDate, 1);
      const more = (0, _getDayList.getDayList)({
        adapter,
        start: nextStart,
        end: adapter.addDays(nextStart, amount),
        excludeWeekends: !showWeekends
      });
      accumulatedDays = accumulatedDays.concat(more);
      occurrenceMap = (0, _useEventOccurrencesGroupedByDay.innerGetEventOccurrencesGroupedByDay)({
        adapter,
        days: accumulatedDays,
        events,
        visibleResources,
        displayTimezone,
        recurringEventsPlugin
      });
      daysWithEvents = accumulatedDays.filter(hasEvents).slice(0, amount);
    }
    return daysWithEvents;
  })
};
