// @mui/x-scheduler/internals/components/event-dialog/EventDialogHeader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = EventDialogHeader;
var React = _interopRequireWildcard(require("react"));
var _styles = require("@mui/material/styles");
var _IconButton = _interopRequireDefault(require("@mui/material/IconButton"));
var _CloseRounded = _interopRequireDefault(require("@mui/icons-material/CloseRounded"));
var _EventDialogStyledContext = require("./EventDialogStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const EventDialogHeaderRoot = (0, _styles.styled)('header', {
  name: 'MuiEventDialog',
  slot: 'Header'
})(({
  theme
}) => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'space-between',
  gap: theme.spacing(1.5),
  padding: theme.spacing(3),
  cursor: 'move'
}));
function EventDialogHeader(props) {
  const {
    children,
    onClose,
    dragHandlerRef
  } = props;
  const {
    classes,
    localeText
  } = (0, _EventDialogStyledContext.useEventDialogStyledContext)();
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(EventDialogHeaderRoot, {
    ref: dragHandlerRef,
    className: classes.eventDialogHeader,
    children: [children, /*#__PURE__*/(0, _jsxRuntime.jsx)(_IconButton.default, {
      className: classes.eventDialogCloseButton,
      "aria-label": localeText.closeButtonAriaLabel,
      onClick: onClose,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_CloseRounded.default, {})
    })]
  });
}
