// @mui/x-scheduler/internals/components/event-dialog/eventDialogClasses.mjs
export const eventDialogClassKeys = ['eventDialog', 'eventDialogCloseButton', 'eventDialogHeader', 'eventDialogReadonlyContent', 'eventDialogActions', 'eventDialogDateTimeContainer', 'eventDialogTitle', 'eventDialogResourceContainer', 'eventDialogResourceLegendContainer', 'eventDialogResourceLegendColor', 'eventDialogResourceTitle', 'eventDialogPaper', 'eventDialogForm', 'eventDialogFormActions', 'eventDialogContent', 'eventDialogTabPanel', 'eventDialogTabContent', 'eventDialogDateTimeFieldsContainer', 'eventDialogDateTimeFieldsRow', 'eventDialogResourceMenuItem', 'eventDialogResourceMenuListSubheader', 'eventDialogResourceMenuColorDot', 'eventDialogResourceMenuColorToggleGroup', 'eventDialogResourceMenuColorToggle', 'eventDialogRecurrenceLabelContainer', 'eventDialogRecurrenceSelectorContainer', 'eventDialogInlineRow', 'eventDialogSectionHeaderTitle', 'eventDialogSectionFieldset', 'eventDialogRadioButtonLabel', 'eventDialogRepeatSectionLabel', 'eventDialogEndsRadioGroup', 'eventDialogRepeatSectionFieldset', 'eventDialogRepeatSectionContent', 'eventDialogRecurrenceSelectorToggleGroup', 'eventDialogWeekDaySelectorCheckbox', 'eventDialogFrequencySelect', 'eventDialogSmallNumberField', 'eventDialogTabsContainer', 'eventDialogTab', 'eventDialogFormDivider', 'eventDialogDeleteButton', 'eventDialogSaveButton', 'eventDialogCloseAction', 'eventDialogDateTimeIcon', 'eventDialogDateTimeLabel', 'eventDialogRecurrenceIcon', 'eventDialogRecurrenceLabel', 'eventDialogDescriptionLabel'];

// Create a slots object for reuse in useUtilityClasses (avoids duplication in EventCalendar.tsx and EventTimelinePremium.tsx)
export const eventDialogSlots = {
  eventDialog: ['eventDialog'],
  eventDialogCloseButton: ['eventDialogCloseButton'],
  eventDialogHeader: ['eventDialogHeader'],
  eventDialogReadonlyContent: ['eventDialogReadonlyContent'],
  eventDialogActions: ['eventDialogActions'],
  eventDialogDateTimeContainer: ['eventDialogDateTimeContainer'],
  eventDialogTitle: ['eventDialogTitle'],
  eventDialogResourceContainer: ['eventDialogResourceContainer'],
  eventDialogResourceLegendContainer: ['eventDialogResourceLegendContainer'],
  eventDialogResourceLegendColor: ['eventDialogResourceLegendColor'],
  eventDialogResourceTitle: ['eventDialogResourceTitle'],
  eventDialogPaper: ['eventDialogPaper'],
  eventDialogForm: ['eventDialogForm'],
  eventDialogFormActions: ['eventDialogFormActions'],
  eventDialogContent: ['eventDialogContent'],
  eventDialogTabPanel: ['eventDialogTabPanel'],
  eventDialogTabContent: ['eventDialogTabContent'],
  eventDialogDateTimeFieldsContainer: ['eventDialogDateTimeFieldsContainer'],
  eventDialogDateTimeFieldsRow: ['eventDialogDateTimeFieldsRow'],
  eventDialogResourceMenuItem: ['eventDialogResourceMenuItem'],
  eventDialogResourceMenuListSubheader: ['eventDialogResourceMenuListSubheader'],
  eventDialogResourceMenuColorDot: ['eventDialogResourceMenuColorDot'],
  eventDialogResourceMenuColorToggleGroup: ['eventDialogResourceMenuColorToggleGroup'],
  eventDialogResourceMenuColorToggle: ['eventDialogResourceMenuColorToggle'],
  eventDialogRecurrenceLabelContainer: ['eventDialogRecurrenceLabelContainer'],
  eventDialogRecurrenceSelectorContainer: ['eventDialogRecurrenceSelectorContainer'],
  eventDialogInlineRow: ['eventDialogInlineRow'],
  eventDialogSectionHeaderTitle: ['eventDialogSectionHeaderTitle'],
  eventDialogSectionFieldset: ['eventDialogSectionFieldset'],
  eventDialogRadioButtonLabel: ['eventDialogRadioButtonLabel'],
  eventDialogRepeatSectionLabel: ['eventDialogRepeatSectionLabel'],
  eventDialogEndsRadioGroup: ['eventDialogEndsRadioGroup'],
  eventDialogRepeatSectionFieldset: ['eventDialogRepeatSectionFieldset'],
  eventDialogRepeatSectionContent: ['eventDialogRepeatSectionContent'],
  eventDialogRecurrenceSelectorToggleGroup: ['eventDialogRecurrenceSelectorToggleGroup'],
  eventDialogWeekDaySelectorCheckbox: ['eventDialogWeekDaySelectorCheckbox'],
  eventDialogFrequencySelect: ['eventDialogFrequencySelect'],
  eventDialogSmallNumberField: ['eventDialogSmallNumberField'],
  eventDialogTabsContainer: ['eventDialogTabsContainer'],
  eventDialogTab: ['eventDialogTab'],
  eventDialogFormDivider: ['eventDialogFormDivider'],
  eventDialogDeleteButton: ['eventDialogDeleteButton'],
  eventDialogSaveButton: ['eventDialogSaveButton'],
  eventDialogCloseAction: ['eventDialogCloseAction'],
  eventDialogDateTimeIcon: ['eventDialogDateTimeIcon'],
  eventDialogDateTimeLabel: ['eventDialogDateTimeLabel'],
  eventDialogRecurrenceIcon: ['eventDialogRecurrenceIcon'],
  eventDialogRecurrenceLabel: ['eventDialogRecurrenceLabel'],
  eventDialogDescriptionLabel: ['eventDialogDescriptionLabel']
};
