// @mui/x-scheduler/internals/components/event-dialog/EventDialog.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EventDialogContext = exports.EventDialogContent = void 0;
exports.EventDialogProvider = EventDialogProvider;
exports.EventDialogTrigger = EventDialogTrigger;
exports.useEventDialogContext = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _store = require("@base-ui/utils/store");
var _empty = require("@base-ui/utils/empty");
var _Paper = _interopRequireDefault(require("@mui/material/Paper"));
var _Dialog = _interopRequireWildcard(require("@mui/material/Dialog"));
var _Backdrop = require("@mui/material/Backdrop");
var _styles = require("@mui/material/styles");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _useSchedulerStoreContext = require("@mui/x-scheduler-internals/use-scheduler-store-context");
var _useDraggableDialog = require("@mui/x-scheduler-internals/use-draggable-dialog");
var _createModal = require("../create-modal");
var _FormContent = require("./FormContent");
var _dialogUtils = require("../../utils/dialog-utils");
var _ReadonlyContent = _interopRequireDefault(require("./ReadonlyContent"));
var _EventDialogStyledContext = require("./EventDialogStyledContext");
var _EventDialogOptionalRenderersContext = require("./EventDialogOptionalRenderersContext");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["anchorRef", "dragHandlerRef", "className"],
  _excluded2 = ["style", "anchorRef", "occurrence", "onClose", "open"],
  _excluded3 = ["children", "optionalRenderers"],
  _excluded4 = ["occurrence"];
const EventDialogRoot = (0, _styles.styled)(_Dialog.default, {
  name: 'MuiEventDialog',
  slot: 'Root'
})({
  [`& .${_Backdrop.backdropClasses.root}`]: {
    backgroundColor: 'transparent'
  },
  [`& .${_Dialog.dialogClasses.container}`]: {
    width: '100%',
    justifyContent: 'unset',
    alignItems: 'unset'
  },
  [`& .${_Dialog.dialogClasses.paper}`]: {
    margin: 0
  }
});
const EventDialogPaper = (0, _styles.styled)(_Paper.default, {
  name: 'MuiEventDialog',
  slot: 'Paper'
})(({
  theme
}) => ({
  borderWidth: 0,
  borderTopWidth: 1,
  height: 'fit-content',
  overflow: 'hidden',
  '&[data-dragging]': {
    outline: `1px solid ${(theme.vars || theme).palette.primary.light}`
  }
}));
// 1. Setup the Draggable Paper Logic
const PaperComponent = function PaperComponent(props) {
  const nodeRef = React.useRef(null);
  const mutateStyle = React.useCallback(style => {
    if (nodeRef.current) {
      nodeRef.current.style.transform = style;
    }
  }, [nodeRef]);
  const {
      anchorRef,
      dragHandlerRef,
      className
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const resetDrag = (0, _useDraggableDialog.useDraggableDialog)(nodeRef, dragHandlerRef, mutateStyle);
  const updatePosition = React.useCallback((shouldResetDrag = false) => {
    const position = (0, _dialogUtils.calculatePosition)(anchorRef.current, nodeRef.current, 'left');
    if (position && nodeRef.current) {
      nodeRef.current.style.top = `${position.top}px`;
      nodeRef.current.style.left = `${position.left}px`;
      if (shouldResetDrag) {
        // Reset transform when position is recalculated
        resetDrag();
      }
    }
  }, [anchorRef, resetDrag, nodeRef]);
  React.useLayoutEffect(() => {
    updatePosition(true);
  }, [updatePosition]);
  React.useEffect(() => {
    const handleResize = () => {
      updatePosition(false);
      resetDrag();
    };
    window.addEventListener('resize', handleResize);
    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, [updatePosition, resetDrag]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(EventDialogPaper, (0, _extends2.default)({}, other, {
    ref: nodeRef,
    className: className
  }));
};
if (process.env.NODE_ENV !== "production") PaperComponent.displayName = "PaperComponent";
const EventDialog = (0, _createModal.createModal)({
  contextName: 'EventDialogContext'
});
const EventDialogContext = exports.EventDialogContext = EventDialog.Context;
const useEventDialogContext = exports.useEventDialogContext = EventDialog.useContext;
const EventDialogContent = exports.EventDialogContent = /*#__PURE__*/React.forwardRef(function EventDialogContent(inProps, forwardedRef) {
  // eslint-disable-next-line mui/material-ui-name-matches-component-name
  const props = (0, _styles.useThemeProps)({
    props: inProps,
    name: 'MuiEventDialog'
  });
  const {
      anchorRef,
      occurrence,
      onClose,
      open
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded2);
  // Context hooks
  const store = (0, _useSchedulerStoreContext.useSchedulerStoreContext)();
  const {
    schedulerId,
    classes
  } = (0, _EventDialogStyledContext.useEventDialogStyledContext)();

  // Selector hooks
  const isEventReadOnly = (0, _store.useStore)(store, _schedulerSelectors.schedulerEventSelectors.isReadOnly, occurrence.id);

  // Ref hooks
  const dragHandlerRef = React.useRef(null);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(EventDialogRoot, (0, _extends2.default)({
    ref: forwardedRef,
    open: open,
    onClose: onClose,
    PaperComponent: PaperComponent,
    "aria-labelledby": `${schedulerId}-event-dialog-title`,
    "aria-modal": "false",
    className: classes.eventDialog,
    slotProps: {
      paper: {
        className: classes.eventDialogPaper,
        anchorRef,
        dragHandlerRef
      }
    }
  }, other, {
    children: isEventReadOnly ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_ReadonlyContent.default, {
      occurrence: occurrence,
      onClose: onClose,
      dragHandlerRef: dragHandlerRef
    }) : /*#__PURE__*/(0, _jsxRuntime.jsx)(_FormContent.FormContent, {
      occurrence: occurrence,
      onClose: onClose,
      dragHandlerRef: dragHandlerRef
    })
  }));
});
if (process.env.NODE_ENV !== "production") EventDialogContent.displayName = "EventDialogContent";
function EventDialogProvider(props) {
  const {
      children,
      optionalRenderers
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded3);
  const store = (0, _useSchedulerStoreContext.useSchedulerStoreContext)();
  const isRecurringScopeDialogOpen = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.isRecurringScopeDialogOpen);
  const showRecurrence = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.areRecurringEventsAvailable);
  const RecurringScopeDialogRenderer = optionalRenderers?.recurringScopeDialog;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_EventDialogOptionalRenderersContext.EventDialogOptionalRenderersContext.Provider, {
    value: optionalRenderers ?? _empty.EMPTY_OBJECT,
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(EventDialog.Provider, {
      render: ({
        isOpen,
        anchorRef,
        data: occurrence,
        onClose
      }) => /*#__PURE__*/(0, _jsxRuntime.jsx)(EventDialogContent, (0, _extends2.default)({
        open: isOpen,
        anchorRef: anchorRef,
        occurrence: occurrence,
        onClose: onClose
      }, other)),
      onOpen: occurrence => {
        store.setEditedEventId(occurrence.id);
      },
      onClose: () => {
        store.setEditedEventId(null);
        store.setOccurrencePlaceholder(null);
      },
      children: [children, showRecurrence && isRecurringScopeDialogOpen && RecurringScopeDialogRenderer && /*#__PURE__*/(0, _jsxRuntime.jsx)(RecurringScopeDialogRenderer, {})]
    })
  });
}
function EventDialogTrigger(props) {
  const {
      occurrence
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded4);
  const ref = React.useRef(null);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(EventDialog.Trigger, (0, _extends2.default)({
    ref: ref,
    data: occurrence
  }, other));
}
