// @mui/x-scheduler/internals/components/event-dialog/FormContent.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormContent = FormContent;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _store = require("@base-ui/utils/store");
var _styles = require("@mui/material/styles");
var _Button = _interopRequireDefault(require("@mui/material/Button"));
var _DialogActions = _interopRequireDefault(require("@mui/material/DialogActions"));
var _DialogContent = _interopRequireDefault(require("@mui/material/DialogContent"));
var _Divider = _interopRequireDefault(require("@mui/material/Divider"));
var _InputBase = require("@mui/material/InputBase");
var _TextField = _interopRequireDefault(require("@mui/material/TextField"));
var _Tabs = _interopRequireDefault(require("@mui/material/Tabs"));
var _Tab = _interopRequireDefault(require("@mui/material/Tab"));
var _useSchedulerStoreContext = require("@mui/x-scheduler-internals/use-scheduler-store-context");
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _EventDialogStyledContext = require("./EventDialogStyledContext");
var _EventDialogOptionalRenderersContext = require("./EventDialogOptionalRenderersContext");
var _utils = require("./utils");
var _EventDialogHeader = _interopRequireDefault(require("./EventDialogHeader"));
var _GeneralTab = require("./GeneralTab");
var _jsxRuntime = require("react/jsx-runtime");
const FormActions = (0, _styles.styled)(_DialogActions.default, {
  name: 'MuiEventDialog',
  slot: 'FormActions'
})(({
  theme
}) => ({
  padding: theme.spacing(3),
  gap: theme.spacing(2)
}));
const DialogContent = (0, _styles.styled)(_DialogContent.default, {
  name: 'MuiEventDialog',
  slot: 'DialogContent'
})({
  cursor: 'default',
  userSelect: 'text',
  padding: 0,
  minWidth: 360,
  width: 450,
  display: 'flex',
  flexDirection: 'column',
  overflow: 'hidden'
});
const EventDialogTitleTextField = (0, _styles.styled)(_TextField.default, {
  name: 'MuiEventDialog',
  slot: 'TitleTextField'
})(({
  theme
}) => ({
  flex: 1,
  [`& .${_InputBase.inputBaseClasses.root}`]: {
    fontSize: theme.typography.h6.fontSize,
    lineHeight: theme.typography.h6.lineHeight,
    fontWeight: theme.typography.h6.fontWeight
  }
}));
const EventDialogForm = (0, _styles.styled)('form', {
  name: 'MuiEventDialog',
  slot: 'Form'
})({
  display: 'flex',
  flexDirection: 'column',
  flex: 1,
  minHeight: 0
});
const EventDialogTabsContainer = (0, _styles.styled)('div', {
  name: 'MuiEventDialog',
  slot: 'TabsContainer'
})(({
  theme
}) => ({
  borderBottom: `1px solid ${(theme.vars || theme).palette.divider}`
}));
const EventDialogTabs = (0, _styles.styled)(_Tabs.default, {
  name: 'MuiEventDialog',
  slot: 'Tabs'
})(({
  theme
}) => ({
  padding: theme.spacing(0, 3)
}));
function FormContent(props) {
  const {
    occurrence,
    onClose,
    dragHandlerRef
  } = props;

  // Context hooks
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const {
    schedulerId,
    classes,
    localeText
  } = (0, _EventDialogStyledContext.useEventDialogStyledContext)();
  const store = (0, _useSchedulerStoreContext.useSchedulerStoreContext)();

  // Selector hooks
  const isPropertyReadOnly = (0, _store.useStore)(store, _schedulerSelectors.schedulerEventSelectors.isPropertyReadOnly, occurrence.id);
  const rawPlaceholder = (0, _store.useStore)(store, _schedulerSelectors.schedulerOccurrencePlaceholderSelectors.value);
  const recurringEventsPlugin = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.recurringEventsPlugin);
  const displayTimezone = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.displayTimezone);
  const showRecurrence = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.areRecurringEventsAvailable);
  const shouldEventRequireResource = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.shouldEventRequireResource);

  // Optional renderer hooks
  const {
    recurrenceTab: RecurrenceTabRenderer
  } = (0, _EventDialogOptionalRenderersContext.useEventDialogOptionalRenderers)();
  const recurrencePresets = (0, _store.useStore)(store, _schedulerSelectors.schedulerRecurringEventSelectors.presets, occurrence.displayTimezone.start);
  const defaultRecurrencePresetKey = (0, _store.useStore)(store, _schedulerSelectors.schedulerRecurringEventSelectors.defaultPresetKey, occurrence.displayTimezone.rrule, occurrence.displayTimezone.start);
  const titleInputRef = React.useCallback(input => input?.focus(), []);

  // State hooks
  const [tabValue, setTabValue] = React.useState('general');
  const [errors, setErrors] = React.useState({});
  const [controlled, setControlled] = React.useState(() => {
    const fmtDate = d => adapter.formatByString(d.value, 'yyyy-MM-dd');
    const fmtTime = d => adapter.formatByString(d.value, 'HH:mm');
    const base = occurrence.displayTimezone.rrule;
    return {
      startDate: fmtDate(occurrence.displayTimezone.start),
      endDate: fmtDate(occurrence.displayTimezone.end),
      startTime: fmtTime(occurrence.displayTimezone.start),
      endTime: fmtTime(occurrence.displayTimezone.end),
      resourceId: occurrence.resource ?? null,
      allDay: !!occurrence.allDay,
      color: (0, _utils.hasProp)(occurrence, 'color') ? occurrence.color : null,
      recurrenceSelection: defaultRecurrencePresetKey,
      rruleDraft: (0, _extends2.default)({
        freq: base?.freq ?? 'WEEKLY',
        interval: base?.interval ?? 1,
        byDay: base?.byDay ?? [],
        byMonthDay: base?.byMonthDay ?? []
      }, base?.count ? {
        count: base.count
      } : {}, base?.until ? {
        until: base.until
      } : {})
    };
  });
  const handleSubmit = async event => {
    event.preventDefault();
    const {
      start,
      end
    } = (0, _utils.computeRange)(adapter, controlled, displayTimezone);
    const form = new FormData(event.currentTarget);
    setErrors({});
    const err = (0, _utils.validateRange)(adapter, start, end, controlled.allDay);
    if (err) {
      setErrors({
        [err.field]: localeText.startDateAfterEndDateError
      });
      return;
    }
    if (shouldEventRequireResource && controlled.resourceId === null) {
      setErrors({
        resource: localeText.requiredResourceError
      });
      return;
    }
    const metaChanges = {
      title: form.get('title').trim(),
      description: form.get('description').trim(),
      allDay: controlled.allDay,
      resource: controlled.resourceId === null ? undefined : controlled.resourceId,
      color: controlled.color === null ? undefined : controlled.color
    };
    let rruleToSubmit;
    if (!showRecurrence || !recurrencePresets) {
      rruleToSubmit = undefined;
    } else if (controlled.recurrenceSelection === null) {
      rruleToSubmit = undefined;
    } else if (controlled.recurrenceSelection === 'custom') {
      rruleToSubmit = controlled.rruleDraft;
    } else {
      rruleToSubmit = recurrencePresets[controlled.recurrenceSelection];
    }
    if (rawPlaceholder?.type === 'creation') {
      store.createEvent((0, _extends2.default)({}, metaChanges, {
        start,
        end,
        rrule: rruleToSubmit
      }));
    } else if (showRecurrence && recurringEventsPlugin && occurrence.displayTimezone.rrule) {
      const recurrenceModified = !_schedulerSelectors.schedulerRecurringEventSelectors.isSameRRule(store.state, occurrence.displayTimezone.rrule, rruleToSubmit);
      const changes = (0, _extends2.default)({}, metaChanges, {
        id: occurrence.id,
        start,
        end
      }, recurrenceModified ? {
        rrule: rruleToSubmit
      } : {});
      await store.updateRecurringEvent({
        occurrenceStart: occurrence.displayTimezone.start.value,
        changes,
        onSubmit: onClose
      });

      // don't close the dialog
      return;
    } else {
      store.updateEvent((0, _extends2.default)({
        id: occurrence.id
      }, metaChanges, {
        start,
        end,
        rrule: rruleToSubmit
      }));
    }
    onClose();
  };
  const handleDelete = () => {
    if (showRecurrence && recurringEventsPlugin && occurrence.displayTimezone.rrule) {
      store.deleteRecurringEvent({
        occurrenceStart: occurrence.displayTimezone.start.value,
        eventId: occurrence.id,
        onSubmit: onClose
      });

      // don't close the dialog
      return;
    }
    store.deleteEvent(occurrence.id);
    onClose();
  };
  const handleTabChange = (event, newValue) => {
    setTabValue(newValue);
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(DialogContent, {
    className: classes.eventDialogContent,
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(EventDialogForm, {
      onSubmit: handleSubmit,
      className: classes.eventDialogForm,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(_EventDialogHeader.default, {
        onClose: onClose,
        dragHandlerRef: dragHandlerRef,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
          id: `${schedulerId}-event-dialog-title`,
          style: {
            position: 'absolute',
            width: 0,
            height: 0,
            overflow: 'hidden'
          },
          children: occurrence.title
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(EventDialogTitleTextField, {
          name: "title",
          defaultValue: occurrence.title,
          required: true,
          inputRef: titleInputRef,
          slotProps: {
            input: {
              readOnly: isPropertyReadOnly('title'),
              'aria-label': localeText.eventTitleAriaLabel
            },
            formHelperText: {
              role: 'alert'
            }
          },
          error: !!errors.title,
          helperText: errors.title,
          fullWidth: true,
          size: "small"
        })]
      }), showRecurrence && RecurrenceTabRenderer && /*#__PURE__*/(0, _jsxRuntime.jsx)(EventDialogTabsContainer, {
        className: classes.eventDialogTabsContainer,
        children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(EventDialogTabs, {
          value: tabValue,
          onChange: handleTabChange,
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_Tab.default, {
            id: `${schedulerId}-general-tab`,
            className: classes.eventDialogTab,
            label: localeText.generalTabLabel,
            value: "general"
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Tab.default, {
            id: `${schedulerId}-recurrence-tab`,
            className: classes.eventDialogTab,
            label: localeText.recurrenceTabLabel,
            value: "recurrence"
          })]
        })
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_GeneralTab.GeneralTab, {
        occurrence: occurrence,
        errors: errors,
        setErrors: setErrors,
        controlled: controlled,
        setControlled: setControlled,
        value: showRecurrence && RecurrenceTabRenderer ? tabValue : 'general'
      }), showRecurrence && RecurrenceTabRenderer && /*#__PURE__*/(0, _jsxRuntime.jsx)(RecurrenceTabRenderer, {
        occurrence: occurrence,
        controlled: controlled,
        setControlled: setControlled,
        tabValue: tabValue
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Divider.default, {
        className: classes.eventDialogFormDivider
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(FormActions, {
        className: classes.eventDialogFormActions,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_Button.default, {
          className: classes.eventDialogDeleteButton,
          color: "error",
          type: "button",
          onClick: handleDelete,
          children: localeText.deleteEvent
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Button.default, {
          className: classes.eventDialogSaveButton,
          variant: "contained",
          type: "submit",
          children: localeText.saveChanges
        })]
      })]
    })
  });
}
