// @mui/x-scheduler/internals/components/event-dialog/GeneralTab.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GeneralTab = GeneralTab;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _store = require("@base-ui/utils/store");
var _styles = require("@mui/material/styles");
var _TextField = _interopRequireDefault(require("@mui/material/TextField"));
var _Switch = _interopRequireDefault(require("@mui/material/Switch"));
var _FormControlLabel = _interopRequireWildcard(require("@mui/material/FormControlLabel"));
var _Divider = _interopRequireDefault(require("@mui/material/Divider"));
var _useSchedulerStoreContext = require("@mui/x-scheduler-internals/use-scheduler-store-context");
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _EventDialogStyledContext = require("./EventDialogStyledContext");
var _utils = require("./utils");
var _ResourceAndColorSection = _interopRequireDefault(require("./ResourceAndColorSection"));
var _EventDialogTabPanel = require("./EventDialogTabPanel");
var _jsxRuntime = require("react/jsx-runtime");
const SectionFieldset = (0, _styles.styled)('fieldset', {
  name: 'MuiEventDialog',
  slot: 'SectionFieldset'
})(({
  theme
}) => ({
  display: 'flex',
  flexDirection: 'column',
  gap: theme.spacing(2),
  border: 0,
  margin: 0,
  padding: 0,
  minInlineSize: 'min-content'
}));
const SectionHeaderTitle = (0, _styles.styled)('legend', {
  name: 'MuiEventDialog',
  slot: 'SectionHeaderTitle'
})(({
  theme
}) => (0, _extends2.default)({}, theme.typography.subtitle2, {
  padding: 0,
  marginBlockEnd: theme.spacing(2),
  textTransform: 'uppercase',
  color: (theme.vars || theme).palette.text.secondary
}));
const DateTimeFieldsContainer = (0, _styles.styled)('div', {
  name: 'MuiEventDialog',
  slot: 'DateTimeFieldsContainer'
})(({
  theme
}) => ({
  display: 'flex',
  flexDirection: 'column',
  gap: theme.spacing(3)
}));
const DateTimeFieldsRow = (0, _styles.styled)('div', {
  name: 'MuiEventDialog',
  slot: 'DateTimeFieldsRow'
})(({
  theme
}) => ({
  display: 'flex',
  gap: theme.spacing(1),
  '& > :first-of-type': {
    flex: '1.5 1 0%'
  },
  '& > :nth-of-type(2)': {
    flex: '1 1 0%'
  }
}));
const AllDayFormControlLabel = (0, _styles.styled)(_FormControlLabel.default, {
  name: 'MuiEventDialog',
  slot: 'AllDayFormControlLabel'
})({
  width: '100%',
  justifyContent: 'space-between',
  [`&.${_FormControlLabel.formControlLabelClasses.root}`]: {
    marginLeft: 0
  }
});
function GeneralTab(props) {
  const {
    occurrence,
    errors,
    setErrors,
    controlled,
    setControlled,
    value
  } = props;

  // Context hooks
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const {
    schedulerId,
    classes,
    localeText
  } = (0, _EventDialogStyledContext.useEventDialogStyledContext)();
  const store = (0, _useSchedulerStoreContext.useSchedulerStoreContext)();

  // Selector hooks
  const displayTimezone = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.displayTimezone);
  const shouldEventRequireResource = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.shouldEventRequireResource);
  const isPropertyReadOnly = (0, _store.useStore)(store, _schedulerSelectors.schedulerEventSelectors.isPropertyReadOnly, occurrence.id);
  const rawPlaceholder = (0, _store.useStore)(store, _schedulerSelectors.schedulerOccurrencePlaceholderSelectors.value);
  function pushPlaceholder(next) {
    if (rawPlaceholder?.type !== 'creation') {
      return;
    }
    const {
      start,
      end,
      surfaceType
    } = (0, _utils.computeRange)(adapter, next, displayTimezone);
    const surfaceTypeToUse = rawPlaceholder.lockSurfaceType ? rawPlaceholder.surfaceType : surfaceType;
    store.setOccurrencePlaceholder({
      type: 'creation',
      surfaceType: surfaceTypeToUse,
      resourceId: next.resourceId,
      start,
      end,
      lockSurfaceType: rawPlaceholder.lockSurfaceType
    });
  }
  const createHandleChangeDateOrTimeField = field => event => {
    const inputValue = event.currentTarget.value;
    setErrors({});
    const newState = (0, _extends2.default)({}, controlled, {
      [field]: inputValue
    });
    pushPlaceholder(newState);
    setControlled(newState);
  };
  const handleToggleAllDay = checked => {
    const newState = (0, _extends2.default)({}, controlled, {
      allDay: checked
    });
    pushPlaceholder(newState);
    setControlled(newState);
  };
  const handleResourceChange = newResource => {
    const nextErrors = (0, _extends2.default)({}, errors);
    delete nextErrors.resource;
    setErrors(nextErrors);
    const newState = (0, _extends2.default)({}, controlled, {
      resourceId: newResource
    });
    pushPlaceholder(newState);
    setControlled(newState);
  };
  const handleColorChange = newColor => {
    const newState = (0, _extends2.default)({}, controlled, {
      color: newColor
    });
    pushPlaceholder(newState);
    setControlled(newState);
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_EventDialogTabPanel.EventDialogTabPanel, {
    role: "tabpanel",
    id: `${schedulerId}-general-tabpanel`,
    "aria-labelledby": `${schedulerId}-general-tab`,
    className: classes.eventDialogTabPanel,
    hidden: value !== 'general',
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(_EventDialogTabPanel.EventDialogTabContent, {
      className: classes.eventDialogTabContent,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(SectionFieldset, {
        className: classes.eventDialogSectionFieldset,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(SectionHeaderTitle, {
          className: classes.eventDialogSectionHeaderTitle,
          children: localeText.dateTimeSectionLabel
        }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(DateTimeFieldsContainer, {
          className: classes.eventDialogDateTimeFieldsContainer,
          children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(DateTimeFieldsRow, {
            className: classes.eventDialogDateTimeFieldsRow,
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_TextField.default, {
              name: "startDate",
              label: localeText.startDateLabel,
              type: "date",
              value: controlled.startDate,
              onChange: createHandleChangeDateOrTimeField('startDate'),
              required: true,
              slotProps: {
                inputLabel: {
                  shrink: true
                },
                input: {
                  readOnly: isPropertyReadOnly('start')
                },
                formHelperText: {
                  role: 'alert'
                }
              },
              error: !!errors.startDate,
              helperText: errors.startDate,
              size: "small"
            }), !controlled.allDay && /*#__PURE__*/(0, _jsxRuntime.jsx)(_TextField.default, {
              name: "startTime",
              label: localeText.startTimeLabel,
              type: "time",
              value: controlled.startTime,
              onChange: createHandleChangeDateOrTimeField('startTime'),
              required: true,
              slotProps: {
                inputLabel: {
                  shrink: true
                },
                input: {
                  readOnly: isPropertyReadOnly('start')
                }
              },
              size: "small"
            })]
          }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(DateTimeFieldsRow, {
            className: classes.eventDialogDateTimeFieldsRow,
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_TextField.default, {
              name: "endDate",
              label: localeText.endDateLabel,
              type: "date",
              value: controlled.endDate,
              onChange: createHandleChangeDateOrTimeField('endDate'),
              required: true,
              slotProps: {
                inputLabel: {
                  shrink: true
                },
                input: {
                  readOnly: isPropertyReadOnly('end')
                }
              },
              size: "small"
            }), !controlled.allDay && /*#__PURE__*/(0, _jsxRuntime.jsx)(_TextField.default, {
              name: "endTime",
              label: localeText.endTimeLabel,
              type: "time",
              value: controlled.endTime,
              onChange: createHandleChangeDateOrTimeField('endTime'),
              required: true,
              slotProps: {
                inputLabel: {
                  shrink: true
                },
                input: {
                  readOnly: isPropertyReadOnly('end')
                }
              },
              size: "small"
            })]
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(AllDayFormControlLabel, {
            control: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Switch.default, {
              id: `${schedulerId}-enable-all-day-switch`,
              checked: controlled.allDay,
              onChange: event => handleToggleAllDay(event.target.checked),
              disabled: isPropertyReadOnly('allDay')
            }),
            label: localeText.allDayLabel,
            labelPlacement: "start"
          })]
        })]
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Divider.default, {}), /*#__PURE__*/(0, _jsxRuntime.jsxs)(SectionFieldset, {
        className: classes.eventDialogSectionFieldset,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(SectionHeaderTitle, {
          className: classes.eventDialogSectionHeaderTitle,
          children: localeText.resourceColorSectionLabel
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_ResourceAndColorSection.default, {
          readOnly: isPropertyReadOnly('resource'),
          resourceId: controlled.resourceId,
          onResourceChange: handleResourceChange,
          onColorChange: handleColorChange,
          color: controlled.color,
          error: shouldEventRequireResource && typeof errors.resource === 'string' ? errors.resource : undefined
        })]
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Divider.default, {}), /*#__PURE__*/(0, _jsxRuntime.jsx)(_TextField.default, {
        name: "description",
        label: localeText.descriptionLabel,
        defaultValue: (0, _utils.hasProp)(occurrence, 'description') ? occurrence.description : '',
        multiline: true,
        rows: 5,
        fullWidth: true,
        slotProps: {
          input: {
            readOnly: isPropertyReadOnly('description')
          }
        }
      })]
    })
  });
}
