// @mui/x-scheduler/internals/components/event-dialog/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  EventDialogContent: true,
  EventDialogProvider: true,
  EventDialogTrigger: true,
  useEventDialogContext: true,
  getEndsSelectionFromRRule: true,
  getWeekdayToken: true,
  EventDialogTabPanel: true,
  EventDialogTabContent: true,
  EventDialogOptionalRenderersContext: true
};
Object.defineProperty(exports, "EventDialogContent", {
  enumerable: true,
  get: function () {
    return _EventDialog.EventDialogContent;
  }
});
Object.defineProperty(exports, "EventDialogOptionalRenderersContext", {
  enumerable: true,
  get: function () {
    return _EventDialogOptionalRenderersContext.EventDialogOptionalRenderersContext;
  }
});
Object.defineProperty(exports, "EventDialogProvider", {
  enumerable: true,
  get: function () {
    return _EventDialog.EventDialogProvider;
  }
});
Object.defineProperty(exports, "EventDialogTabContent", {
  enumerable: true,
  get: function () {
    return _EventDialogTabPanel.EventDialogTabContent;
  }
});
Object.defineProperty(exports, "EventDialogTabPanel", {
  enumerable: true,
  get: function () {
    return _EventDialogTabPanel.EventDialogTabPanel;
  }
});
Object.defineProperty(exports, "EventDialogTrigger", {
  enumerable: true,
  get: function () {
    return _EventDialog.EventDialogTrigger;
  }
});
Object.defineProperty(exports, "getEndsSelectionFromRRule", {
  enumerable: true,
  get: function () {
    return _utils.getEndsSelectionFromRRule;
  }
});
Object.defineProperty(exports, "getWeekdayToken", {
  enumerable: true,
  get: function () {
    return _utils.getWeekdayToken;
  }
});
Object.defineProperty(exports, "useEventDialogContext", {
  enumerable: true,
  get: function () {
    return _EventDialog.useEventDialogContext;
  }
});
var _EventDialog = require("./EventDialog");
var _eventDialogClasses = require("./eventDialogClasses");
Object.keys(_eventDialogClasses).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _eventDialogClasses[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _eventDialogClasses[key];
    }
  });
});
var _EventDialogStyledContext = require("./EventDialogStyledContext");
Object.keys(_EventDialogStyledContext).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _EventDialogStyledContext[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _EventDialogStyledContext[key];
    }
  });
});
var _utils = require("./utils");
var _EventDialogTabPanel = require("./EventDialogTabPanel");
var _EventDialogOptionalRenderersContext = require("./EventDialogOptionalRenderersContext");
