// @mui/x-scheduler/internals/components/event-dialog/ReadonlyContent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ReadonlyContent;
var React = _interopRequireWildcard(require("react"));
var _CalendarMonthRounded = _interopRequireDefault(require("@mui/icons-material/CalendarMonthRounded"));
var _RepeatRounded = _interopRequireDefault(require("@mui/icons-material/RepeatRounded"));
var _store = require("@base-ui/utils/store");
var _styles = require("@mui/material/styles");
var _Button = _interopRequireDefault(require("@mui/material/Button"));
var _Typography = _interopRequireDefault(require("@mui/material/Typography"));
var _useSchedulerStoreContext = require("@mui/x-scheduler-internals/use-scheduler-store-context");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _EventDialogHeader = _interopRequireDefault(require("./EventDialogHeader"));
var _EventDialogStyledContext = require("./EventDialogStyledContext");
var _utils = require("./utils");
var _useFormatTime = require("../../hooks/useFormatTime");
var _tokens = require("../../utils/tokens");
var _jsxRuntime = require("react/jsx-runtime");
const ReadonlyContentDragContainer = (0, _styles.styled)('section', {
  name: 'MuiEventDialog',
  slot: 'ReadonlyContentDragContainer'
})({
  cursor: 'move'
});
const ReadonlyContentRoot = (0, _styles.styled)('div', {
  name: 'MuiEventDialog',
  slot: 'ReadonlyContent'
})(({
  theme
}) => ({
  display: 'flex',
  padding: theme.spacing(0, 3),
  flexDirection: 'column',
  gap: theme.spacing(2)
}));
const RecurrenceLabelContainer = (0, _styles.styled)('div', {
  name: 'MuiEventDialog',
  slot: 'RecurrenceLabelContainer'
})(({
  theme
}) => ({
  display: 'flex',
  gap: theme.spacing(1)
}));
const EventDialogActions = (0, _styles.styled)('div', {
  name: 'MuiEventDialog',
  slot: 'Actions'
})(({
  theme
}) => ({
  display: 'flex',
  justifyContent: 'space-between',
  padding: theme.spacing(3)
}));
const EventDialogDateTimeContainer = (0, _styles.styled)('div', {
  name: 'MuiEventDialog',
  slot: 'DateTimeContainer'
})(({
  theme
}) => ({
  display: 'flex',
  alignItems: 'center',
  gap: theme.spacing(1)
}));
const EventDialogTitle = (0, _styles.styled)(_Typography.default, {
  name: 'MuiEventDialog',
  slot: 'Title'
})({
  margin: 0,
  color: 'var(--event-on-surface-subtle-primary)'
});
const EventDialogResourceContainer = (0, _styles.styled)('div', {
  name: 'MuiEventDialog',
  slot: 'ResourceContainer'
})(({
  theme
}) => ({
  display: 'flex',
  alignItems: 'center',
  gap: theme.spacing(1)
}));
const EventDialogResourceLegendContainer = (0, _styles.styled)('div', {
  name: 'MuiEventDialog',
  slot: 'ResourceLegendContainer'
})(({
  theme
}) => ({
  display: 'flex',
  alignItems: 'center',
  gap: theme.spacing(1)
}));
const ResourceLegendColorDot = (0, _styles.styled)('span', {
  name: 'MuiEventDialog',
  slot: 'ResourceLegendColor'
})(({
  theme
}) => ({
  width: 14,
  height: 14,
  margin: theme.spacing(0, 0.2),
  borderRadius: 2,
  flexShrink: 0,
  backgroundColor: 'var(--event-main)',
  variants: (0, _tokens.getPaletteVariants)(theme)
}));
const EventDialogResourceTitle = (0, _styles.styled)(_Typography.default, {
  name: 'MuiEventDialog',
  slot: 'ResourceTitle'
})(({
  theme
}) => ({
  margin: 0,
  fontSize: theme.typography.body2.fontSize,
  color: 'var(--event-on-surface-subtle-secondary)',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap'
}));
function ReadonlyContent(props) {
  const {
    occurrence,
    onClose,
    dragHandlerRef
  } = props;

  // Context hooks
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const {
    schedulerId,
    classes,
    localeText
  } = (0, _EventDialogStyledContext.useEventDialogStyledContext)();
  const store = (0, _useSchedulerStoreContext.useSchedulerStoreContext)();

  // Selector hooks
  const color = (0, _store.useStore)(store, _schedulerSelectors.schedulerEventSelectors.color, occurrence.id);
  const resource = (0, _store.useStore)(store, _schedulerSelectors.schedulerResourceSelectors.processedResource, occurrence.resource);
  const defaultRecurrenceKey = (0, _store.useStore)(store, _schedulerSelectors.schedulerRecurringEventSelectors.defaultPresetKey, occurrence.displayTimezone.rrule, occurrence.displayTimezone.start);
  const showRecurrence = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.areRecurringEventsAvailable);

  // Feature hook
  const formatTime = (0, _useFormatTime.useFormatTime)();
  const recurrenceLabel = (0, _utils.getRecurrenceLabel)(adapter, occurrence.displayTimezone.start, defaultRecurrenceKey, localeText);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(ReadonlyContentDragContainer, {
    ref: dragHandlerRef,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_EventDialogHeader.default, {
      onClose: onClose,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(EventDialogTitle, {
        variant: "h6",
        id: `${schedulerId}-event-dialog-title`,
        className: classes.eventDialogTitle,
        children: occurrence.title
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(ReadonlyContentRoot, {
      className: classes.eventDialogReadonlyContent,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(EventDialogResourceContainer, {
        className: classes.eventDialogResourceContainer,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(EventDialogResourceLegendContainer, {
          className: classes.eventDialogResourceLegendContainer,
          children: [resource?.eventColor && resource.eventColor !== color && /*#__PURE__*/(0, _jsxRuntime.jsx)(ResourceLegendColorDot, {
            className: classes.eventDialogResourceLegendColor,
            "data-palette": resource.eventColor
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(ResourceLegendColorDot, {
            className: classes.eventDialogResourceLegendColor,
            "data-palette": color
          })]
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(EventDialogResourceTitle, {
          className: classes.eventDialogResourceTitle,
          children: resource?.title || localeText.noResourceAriaLabel
        })]
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(EventDialogDateTimeContainer, {
        className: classes.eventDialogDateTimeContainer,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_CalendarMonthRounded.default, {
          className: classes.eventDialogDateTimeIcon,
          fontSize: "small"
        }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(_Typography.default, {
          className: classes.eventDialogDateTimeLabel,
          variant: "body2",
          component: "p",
          noWrap: true,
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("time", {
            dateTime: adapter.format(occurrence.displayTimezone.start.value, 'localizedNumericDate'),
            children: /*#__PURE__*/(0, _jsxRuntime.jsxs)("span", {
              children: [adapter.format(occurrence.displayTimezone.start.value, 'localizedDateWithFullMonthAndWeekDay'), ",", ' ']
            })
          }), occurrence.allDay ? /*#__PURE__*/(0, _jsxRuntime.jsxs)("span", {
            children: [" ", localeText.allDayLabel]
          }) : /*#__PURE__*/(0, _jsxRuntime.jsxs)("time", {
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
              children: formatTime(occurrence.displayTimezone.start.value)
            }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("span", {
              children: [" - ", formatTime(occurrence.displayTimezone.end.value)]
            })]
          })]
        })]
      }), showRecurrence && defaultRecurrenceKey != null && /*#__PURE__*/(0, _jsxRuntime.jsxs)(RecurrenceLabelContainer, {
        className: classes.eventDialogRecurrenceLabelContainer,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_RepeatRounded.default, {
          className: classes.eventDialogRecurrenceIcon,
          fontSize: "small"
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Typography.default, {
          className: classes.eventDialogRecurrenceLabel,
          variant: "body2",
          component: "em",
          sx: {
            color: 'text.secondary'
          },
          children: recurrenceLabel
        })]
      }), (0, _utils.hasProp)(occurrence, 'description') && !!occurrence.description ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_Typography.default, {
        className: classes.eventDialogDescriptionLabel,
        variant: "body2",
        children: occurrence.description
      }) : null]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(EventDialogActions, {
      className: classes.eventDialogActions,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Button.default, {
        className: classes.eventDialogCloseAction,
        variant: "contained",
        type: "button",
        onClick: onClose,
        children: localeText.closeButtonLabel
      })
    })]
  });
}
