// @mui/x-scheduler/internals/components/event-dialog/ResourceAndColorSection.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ResourceAndColorSection;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _Check = _interopRequireDefault(require("@mui/icons-material/Check"));
var _styles = require("@mui/material/styles");
var _Select = _interopRequireDefault(require("@mui/material/Select"));
var _MenuItem = _interopRequireDefault(require("@mui/material/MenuItem"));
var _Divider = _interopRequireDefault(require("@mui/material/Divider"));
var _ListSubheader = _interopRequireDefault(require("@mui/material/ListSubheader"));
var _InputAdornment = _interopRequireDefault(require("@mui/material/InputAdornment"));
var _InputLabel = _interopRequireDefault(require("@mui/material/InputLabel"));
var _FormControl = _interopRequireDefault(require("@mui/material/FormControl"));
var _FormHelperText = _interopRequireDefault(require("@mui/material/FormHelperText"));
var _ListItemIcon = _interopRequireDefault(require("@mui/material/ListItemIcon"));
var _ListItemText = _interopRequireDefault(require("@mui/material/ListItemText"));
var _toggle = require("@base-ui/react/toggle");
var _toggleGroup = require("@base-ui/react/toggle-group");
var _constants = require("@mui/x-scheduler-internals/constants");
var _useSchedulerStoreContext = require("@mui/x-scheduler-internals/use-scheduler-store-context");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _store = require("@base-ui/utils/store");
var _tokens = require("../../utils/tokens");
var _EventDialogStyledContext = require("./EventDialogStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const NO_RESOURCE_VALUE = '';
const ResourceMenuItem = (0, _styles.styled)(_MenuItem.default, {
  name: 'MuiEventDialog',
  slot: 'ResourceMenuItem'
})(({
  theme
}) => ({
  paddingLeft: `calc(${theme.spacing(2)} + var(--resource-indent) * ${theme.spacing(2)})`
}));
const ResourceMenuListSubheader = (0, _styles.styled)(_ListSubheader.default, {
  name: 'MuiEventDialog',
  slot: 'ResourceMenuListSubheader'
})({});
const ResourceMenuColorDot = (0, _styles.styled)('span', {
  name: 'MuiEventDialog',
  slot: 'ResourceMenuColorDot'
})(({
  theme
}) => ({
  width: 14,
  height: 14,
  borderRadius: '2px',
  flexShrink: 0,
  backgroundColor: 'var(--event-main)',
  variants: (0, _tokens.getPaletteVariants)(theme),
  [`&[data-no-resource="true"]`]: {
    backgroundColor: 'var(--event-surface-subtle)',
    border: '1.2px dashed var(--event-main)'
  }
}));
const ResourceMenuColorToggleGroup = (0, _styles.styled)(_toggleGroup.ToggleGroup, {
  name: 'MuiEventDialog',
  slot: 'ResourceMenuColorToggleGroup'
})(({
  theme
}) => ({
  display: 'flex',
  flexWrap: 'wrap',
  gap: theme.spacing(1),
  borderRadius: theme.shape.borderRadius
}));
const ResourceMenuColorToggle = (0, _styles.styled)(_toggle.Toggle, {
  name: 'MuiEventDialog',
  slot: 'ResourceMenuColorToggle'
})(({
  theme
}) => ({
  width: 24,
  height: 24,
  borderRadius: theme.shape.borderRadius,
  border: 'none',
  cursor: 'pointer',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  backgroundColor: 'var(--event-main)',
  color: 'white',
  '&:disabled': {
    cursor: 'not-allowed',
    opacity: 0.5
  },
  variants: (0, _tokens.getPaletteVariants)(theme)
}));
function ResourceSelectAdornment(props) {
  const {
    resource
  } = props;
  const store = (0, _useSchedulerStoreContext.useSchedulerStoreContext)();
  const {
    classes
  } = (0, _EventDialogStyledContext.useEventDialogStyledContext)();
  const resourceColor = (0, _store.useStore)(store, _schedulerSelectors.schedulerResourceSelectors.defaultEventColor, resource?.value);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(ResourceMenuColorDot, {
    className: classes.eventDialogResourceMenuColorDot,
    "data-palette": resourceColor,
    "data-no-resource": Boolean(resource?.value === null)
  });
}
function ResourceAndColorSection(props) {
  const {
    readOnly,
    resourceId,
    onResourceChange,
    onColorChange,
    color,
    error
  } = props;

  // Context hooks
  const {
    schedulerId,
    classes,
    localeText
  } = (0, _EventDialogStyledContext.useEventDialogStyledContext)();
  const store = (0, _useSchedulerStoreContext.useSchedulerStoreContext)();

  // Selector hooks
  const resources = (0, _store.useStore)(store, _schedulerSelectors.schedulerResourceSelectors.processedResourceFlatList);
  const resourceDepthLookup = (0, _store.useStore)(store, _schedulerSelectors.schedulerResourceSelectors.resourceDepthLookup);
  const childrenIdLookup = (0, _store.useStore)(store, _schedulerSelectors.schedulerResourceSelectors.childrenIdLookup);
  const eventDefaultColor = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.defaultEventColor);
  const shouldEventRequireResource = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.shouldEventRequireResource);
  const resourcesOptions = React.useMemo(() => {
    const hasNesting = resources.some(resource => (childrenIdLookup.get(resource.id)?.length ?? 0) > 0);
    const firstTopLevelIndex = resources.findIndex(resource => (resourceDepthLookup.get(resource.id) ?? 0) === 0);
    return [
    // The no-resource option must stay in the rendered options list so MUI Select's
    // `value=""` keeps matching a MenuItem when an event has no resource yet — otherwise
    // MUI logs an "out-of-range value" warning. It's hidden from the menu when
    // `shouldEventRequireResource` is `true` so the user can't pick it.
    {
      label: localeText.labelNoResource,
      value: null,
      eventColor: eventDefaultColor,
      isGroupRoot: false,
      indentLevel: 0,
      showDivider: false,
      hidden: shouldEventRequireResource
    }, ...resources.map((resource, index) => {
      const depth = resourceDepthLookup.get(resource.id) ?? 0;
      const hasChildren = (childrenIdLookup.get(resource.id)?.length ?? 0) > 0;
      const isTopLevel = depth === 0;
      // Skip the divider above the first top-level group when nothing precedes it visually
      // (the no-resource option is hidden).
      const isFirstTopLevel = index === firstTopLevelIndex;
      const showDivider = hasNesting && isTopLevel && (!isFirstTopLevel || !shouldEventRequireResource);
      return {
        label: resource.title,
        value: resource.id,
        eventColor: resource.eventColor ?? eventDefaultColor,
        isGroupRoot: isTopLevel && hasChildren,
        indentLevel: Math.max(0, depth - 1),
        showDivider
      };
    })];
  }, [resources, resourceDepthLookup, childrenIdLookup, localeText.labelNoResource, eventDefaultColor, shouldEventRequireResource]);
  const resource = React.useMemo(() => resourcesOptions.find(option => resourceId ? option.value === resourceId : option.value === null) || null, [resourcesOptions, resourceId]);
  const handleChange = event => {
    const value = event.target.value;
    onResourceChange(value === NO_RESOURCE_VALUE ? null : value);
  };
  const errorId = `${schedulerId}-resource-error`;
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(React.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(_FormControl.default, {
      size: "small",
      fullWidth: true,
      error: !!error,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_InputLabel.default, {
        id: `${schedulerId}-resource-select-label`,
        children: localeText.resourceLabel
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Select.default, {
        labelId: `${schedulerId}-resource-select-label`,
        label: localeText.resourceLabel,
        value: resourceId ?? NO_RESOURCE_VALUE,
        displayEmpty: true,
        onChange: handleChange,
        readOnly: readOnly,
        "aria-describedby": error ? errorId : undefined,
        startAdornment: /*#__PURE__*/(0, _jsxRuntime.jsx)(_InputAdornment.default, {
          position: "start",
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(ResourceSelectAdornment, {
            resource: resource
          })
        }),
        renderValue: () => {
          if (resource) {
            return resource.label;
          }
          // `resourceId == null` means the resource is unset, not invalid.
          if (resourceId == null) {
            return localeText.labelNoResource;
          }
          return localeText.labelInvalidResource;
        },
        children: resourcesOptions.flatMap(resourceOption => {
          const items = [];
          if (resourceOption.showDivider) {
            items.push(/*#__PURE__*/(0, _jsxRuntime.jsx)(_Divider.default, {}, `divider-${resourceOption.value}`));
          }
          if (resourceOption.isGroupRoot) {
            items.push(/*#__PURE__*/(0, _jsxRuntime.jsx)(ResourceMenuListSubheader, {
              className: classes.eventDialogResourceMenuListSubheader,
              children: resourceOption.label.toUpperCase()
            }, `header-${resourceOption.value}`));
          }
          items.push(/*#__PURE__*/(0, _jsxRuntime.jsxs)(ResourceMenuItem, {
            value: resourceOption.value ?? NO_RESOURCE_VALUE,
            "aria-label": resourceOption.label,
            className: classes.eventDialogResourceMenuItem,
            style: (0, _extends2.default)({
              '--resource-indent': resourceOption.indentLevel
            }, resourceOption.hidden && {
              display: 'none'
            }),
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_ListItemIcon.default, {
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(ResourceMenuColorDot, {
                className: classes.eventDialogResourceMenuColorDot,
                "data-palette": resourceOption.eventColor,
                "data-no-resource": Boolean(resourceOption.value === null)
              })
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_ListItemText.default, {
              children: resourceOption.label
            })]
          }, resourceOption.value ?? NO_RESOURCE_VALUE));
          return items;
        })
      }), error && /*#__PURE__*/(0, _jsxRuntime.jsx)(_FormHelperText.default, {
        id: errorId,
        role: "alert",
        children: error
      })]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(ResourceMenuColorToggleGroup, {
      value: color ? [color] : [],
      onValueChange: values => {
        const next = values[values.length - 1];
        onColorChange(next ?? null);
      },
      "aria-label": localeText.colorPickerLabel,
      disabled: readOnly,
      className: classes.eventDialogResourceMenuColorToggleGroup,
      children: _constants.EVENT_COLORS.map(colorOption => /*#__PURE__*/(0, _jsxRuntime.jsx)(ResourceMenuColorToggle, {
        value: colorOption,
        "aria-label": localeText.selectColorAriaLabel(colorOption),
        "data-palette": colorOption,
        className: classes.eventDialogResourceMenuColorToggle,
        children: color === colorOption && /*#__PURE__*/(0, _jsxRuntime.jsx)(_Check.default, {
          fontSize: "small"
        })
      }, colorOption))
    })]
  });
}
