// @mui/x-data-grid/hooks/features/virtualization/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  EMPTY_RENDER_CONTEXT: true,
  gridVirtualizationSelector: true,
  gridVirtualizationEnabledSelector: true,
  gridVirtualizationColumnEnabledSelector: true,
  gridVirtualizationRowEnabledSelector: true,
  gridRenderContextSelector: true,
  gridRenderContextColumnsSelector: true
};
Object.defineProperty(exports, "EMPTY_RENDER_CONTEXT", {
  enumerable: true,
  get: function () {
    return _xVirtualizer.EMPTY_RENDER_CONTEXT;
  }
});
Object.defineProperty(exports, "gridRenderContextColumnsSelector", {
  enumerable: true,
  get: function () {
    return _gridVirtualizationSelectors.gridRenderContextColumnsSelector;
  }
});
Object.defineProperty(exports, "gridRenderContextSelector", {
  enumerable: true,
  get: function () {
    return _gridVirtualizationSelectors.gridRenderContextSelector;
  }
});
Object.defineProperty(exports, "gridVirtualizationColumnEnabledSelector", {
  enumerable: true,
  get: function () {
    return _gridVirtualizationSelectors.gridVirtualizationColumnEnabledSelector;
  }
});
Object.defineProperty(exports, "gridVirtualizationEnabledSelector", {
  enumerable: true,
  get: function () {
    return _gridVirtualizationSelectors.gridVirtualizationEnabledSelector;
  }
});
Object.defineProperty(exports, "gridVirtualizationRowEnabledSelector", {
  enumerable: true,
  get: function () {
    return _gridVirtualizationSelectors.gridVirtualizationRowEnabledSelector;
  }
});
Object.defineProperty(exports, "gridVirtualizationSelector", {
  enumerable: true,
  get: function () {
    return _gridVirtualizationSelectors.gridVirtualizationSelector;
  }
});
var _xVirtualizer = require("@mui/x-virtualizer");
var _useGridVirtualization = require("./useGridVirtualization");
Object.keys(_useGridVirtualization).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useGridVirtualization[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _useGridVirtualization[key];
    }
  });
});
var _gridVirtualizationSelectors = require("./gridVirtualizationSelectors");
