// @mui/x-scheduler/month-view/month-view-row/MonthViewCell.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MonthViewCell = void 0;
var React = _interopRequireWildcard(require("react"));
var _styles = require("@mui/material/styles");
var _store = require("@base-ui/utils/store");
var _Button = _interopRequireDefault(require("@mui/material/Button"));
var _useMergedRefs = require("@base-ui/utils/useMergedRefs");
var _useAdapter = require("@mui/x-scheduler-internals/use-adapter");
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _calendarGrid = require("@mui/x-scheduler-internals/calendar-grid");
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _eventCalendarSelectors = require("@mui/x-scheduler-internals/event-calendar-selectors");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _DayGridEvent = require("../../internals/components/event/day-grid-event/DayGridEvent");
var _MoreEventsPopover = require("../../internals/components/more-events-popover/MoreEventsPopover");
var _dateUtils = require("../../internals/utils/date-utils");
var _eventUtils = require("../../internals/utils/event-utils");
var _eventDialog = require("../../internals/components/event-dialog");
var _EventDialog = require("../../internals/components/event-dialog/EventDialog");
var _EventCalendarStyledContext = require("../../event-calendar/EventCalendarStyledContext");
var _eventCalendarClasses = require("../../event-calendar/eventCalendarClasses");
var _eventSkeleton = require("../../internals/components/event-skeleton");
var _tokens = require("../../internals/utils/tokens");
var _jsxRuntime = require("react/jsx-runtime");
const MonthViewCellRoot = (0, _styles.styled)(_calendarGrid.CalendarGrid.DayCell, {
  name: 'MuiEventCalendar',
  slot: 'MonthViewCell'
})(({
  theme
}) => ({
  display: 'grid',
  gridTemplateRows: 'repeat(var(--row-count), minmax(auto, 18px))',
  gap: theme.spacing(0.5),
  padding: theme.spacing(0.5),
  fontSize: theme.typography.body2.fontSize,
  lineHeight: '18px',
  borderInlineStart: `1px solid transparent`,
  '&:not(:first-of-type)': {
    borderInlineStartColor: (theme.vars || theme).palette.divider
  },
  '&[data-weekend]': {
    backgroundColor: (theme.vars || theme).palette.action.hover,
    color: (theme.vars || theme).palette.text.primary
  },
  '&[data-current]': {
    backgroundColor: theme.alpha((theme.vars || theme).palette.primary.light, 0.05)
  },
  [`&[data-current] .${_eventCalendarClasses.eventCalendarClasses.monthViewCellNumber}`]: {
    backgroundColor: (theme.vars || theme).palette.primary.main,
    color: (theme.vars || theme).palette.primary.contrastText
  },
  '&[data-other-month]': {
    color: (theme.vars || theme).palette.text.disabled
  },
  '&:focus-visible': {
    outline: 'none',
    backgroundColor: (0, _tokens.getCellFocusBackground)(theme)
  },
  // Today button states
  [`&[data-current] > .${_eventCalendarClasses.eventCalendarClasses.monthViewCellNumberButton} > .${_eventCalendarClasses.eventCalendarClasses.monthViewCellNumber}`]: {
    backgroundColor: 'transparent'
  },
  [`&[data-current] > .${_eventCalendarClasses.eventCalendarClasses.monthViewCellNumberButton}:hover > .${_eventCalendarClasses.eventCalendarClasses.monthViewCellNumber}`]: {
    backgroundColor: 'transparent'
  },
  [`&[data-current] > .${_eventCalendarClasses.eventCalendarClasses.monthViewCellNumberButton}:active > .${_eventCalendarClasses.eventCalendarClasses.monthViewCellNumber}`]: {
    backgroundColor: 'transparent'
  },
  [`&[data-current] > .${_eventCalendarClasses.eventCalendarClasses.monthViewCellNumberButton}`]: {
    backgroundColor: (theme.vars || theme).palette.primary.main,
    color: (theme.vars || theme).palette.primary.contrastText,
    marginTop: 1
  },
  [`&[data-current] > .${_eventCalendarClasses.eventCalendarClasses.monthViewCellNumberButton}:hover`]: {
    backgroundColor: (theme.vars || theme).palette.primary.dark
  },
  [`&[data-current] > .${_eventCalendarClasses.eventCalendarClasses.monthViewCellNumberButton}:active`]: {
    backgroundColor: (theme.vars || theme).palette.primary.dark
  }
}));
const MonthViewCellNumber = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'MonthViewCellNumber'
})(({
  theme
}) => ({
  gridRow: 1,
  justifySelf: 'end',
  alignSelf: 'flex-end',
  padding: theme.spacing(0, 0.5),
  borderRadius: theme.shape.borderRadius
}));
const MonthViewCellNumberButton = (0, _styles.styled)('button', {
  name: 'MuiEventCalendar',
  slot: 'MonthViewCellNumberButton'
})(({
  theme
}) => ({
  gridRow: 1,
  justifySelf: 'end',
  width: 'fit-content',
  alignSelf: 'flex-end',
  borderRadius: theme.shape.borderRadius,
  border: 'none',
  background: 'none',
  padding: 0,
  cursor: 'pointer',
  font: 'inherit',
  color: 'inherit',
  '&:hover': {
    backgroundColor: (theme.vars || theme).palette.action.hover
  },
  '&:active': {
    backgroundColor: (theme.vars || theme).palette.action.selected
  },
  '&:focus-visible': {
    backgroundColor: (0, _tokens.getCellFocusBackground)(theme),
    outline: `2px solid ${(theme.vars || theme).palette.primary.main}`,
    outlineOffset: 2
  },
  '&:focus-visible:hover': {
    backgroundColor: (theme.vars || theme).palette.action.hover
  }
}));
const MonthViewCellEvents = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MonthViewCellEvents'
})(({
  theme
}) => ({
  position: 'relative',
  display: 'grid',
  gap: theme.spacing(0.5)
}));
const MonthViewMoreEvents = (0, _styles.styled)(_Button.default, {
  name: 'MuiEventCalendar',
  slot: 'MonthViewMoreEvents'
})(({
  theme
}) => ({
  margin: 0,
  color: (theme.vars || theme).palette.text.secondary,
  fontSize: theme.typography.caption.fontSize,
  lineHeight: '18px',
  paddingInlineStart: theme.spacing(0.5),
  justifyContent: 'flex-start',
  textTransform: 'none'
}));
const MonthViewPlaceholderEventContainer = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MonthViewPlaceholderContainer'
})(({
  theme
}) => ({
  position: 'absolute',
  top: 0,
  left: 0,
  right: 0,
  display: 'grid',
  gridTemplateColumns: '1fr',
  gridTemplateRows: 'repeat(var(--row-count), minmax(auto, 18px))',
  gap: theme.spacing(0.5)
}));
const MonthViewCell = exports.MonthViewCell = /*#__PURE__*/React.forwardRef(function MonthViewCell(props, ref) {
  const {
    day,
    row,
    maxEvents,
    colIndex,
    ariaLabelledBy
  } = props;

  // Context hooks
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const {
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const {
    onOpen: startEditing
  } = (0, _EventDialog.useEventDialogContext)();

  // Selector hooks
  const hasDayView = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarViewSelectors.hasDayView);
  const visibleDate = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.visibleDate);
  const isCreatingAnEvent = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarOccurrencePlaceholderSelectors.isCreatingInDayCell, day.value);
  const isToday = (0, _store.useStore)(store, _schedulerSelectors.schedulerNowSelectors.isCurrentDay, day.value);
  const isLoading = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.isLoading);
  const placeholder = _calendarGrid.CalendarGrid.usePlaceholderInDay(day.value, row, maxEvents);

  // Ref hooks
  const cellRef = React.useRef(null);
  const handleRef = (0, _useMergedRefs.useMergedRefs)(ref, cellRef);
  const isCurrentMonth = adapter.isSameMonth(day.value, visibleDate);
  const isFirstDayOfMonth = adapter.isSameDay(day.value, adapter.startOfMonth(day.value));
  const visibleOccurrences = day.withPosition.length > maxEvents ? day.withPosition.slice(0, maxEvents - 1) : day.withPosition;
  const hiddenCount = day.withPosition.length - visibleOccurrences.length;
  const cellNumberContent = /*#__PURE__*/(0, _jsxRuntime.jsx)(MonthViewCellNumber, {
    className: classes.monthViewCellNumber,
    children: isFirstDayOfMonth ? (0, _dateUtils.formatMonthAndDayOfMonth)(day.value, adapter) : adapter.format(day.value, 'dayOfMonth')
  });

  // Day number header + max events
  const rowCount = 1 + maxEvents;
  React.useEffect(() => {
    if (!isCreatingAnEvent || !placeholder || !cellRef.current) {
      return;
    }
    startEditing(cellRef, placeholder);
  }, [isCreatingAnEvent, placeholder, startEditing]);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(MonthViewCellRoot, {
    className: classes.monthViewCell,
    ref: handleRef,
    value: day.value,
    "aria-colindex": colIndex,
    "aria-labelledby": ariaLabelledBy,
    "data-current": isToday || undefined,
    "data-other-month": !isCurrentMonth || undefined,
    "data-weekend": (0, _useAdapter.isWeekend)(adapter, day.value) || undefined,
    lockSurfaceType: true,
    style: {
      '--row-count': rowCount
    },
    children: [hasDayView ? /*#__PURE__*/(0, _jsxRuntime.jsx)(MonthViewCellNumberButton, {
      type: "button",
      className: classes.monthViewCellNumberButton,
      onClick: event => store.switchToDay(day.value, event),
      tabIndex: -1,
      children: cellNumberContent
    }) : cellNumberContent, /*#__PURE__*/(0, _jsxRuntime.jsxs)(MonthViewCellEvents, {
      className: classes.monthViewCellEvents,
      children: [isLoading && /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventSkeleton.EventSkeleton, {
        "data-variant": "day-grid"
      }), !isLoading && visibleOccurrences.map(occurrence => {
        if (occurrence.position.isInvisible) {
          return /*#__PURE__*/(0, _jsxRuntime.jsx)(_DayGridEvent.DayGridEvent, {
            occurrence: occurrence,
            variant: "invisible"
          }, occurrence.key);
        }
        return /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventDialog.EventDialogTrigger, {
          occurrence: occurrence,
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_DayGridEvent.DayGridEvent, {
            occurrence: occurrence,
            variant: (0, _eventUtils.isOccurrenceAllDayOrMultipleDay)(occurrence, adapter) ? 'filled' : 'compact'
          })
        }, occurrence.key);
      }), hiddenCount > 0 && /*#__PURE__*/(0, _jsxRuntime.jsx)(_MoreEventsPopover.MoreEventsPopoverTrigger, {
        occurrences: day.withPosition,
        day: day,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(MonthViewMoreEvents, {
          size: "small",
          "aria-label": localeText.hiddenEvents(hiddenCount),
          className: classes.monthViewMoreEvents,
          children: localeText.hiddenEvents(hiddenCount)
        })
      }), placeholder != null && /*#__PURE__*/(0, _jsxRuntime.jsx)(MonthViewPlaceholderEventContainer, {
        className: classes.monthViewPlaceholderContainer,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_DayGridEvent.DayGridEvent, {
          occurrence: placeholder,
          variant: "placeholder"
        })
      })]
    })]
  }, day.key);
});
if (process.env.NODE_ENV !== "production") MonthViewCell.displayName = "MonthViewCell";
