// @mui/x-scheduler/month-view/month-view-row/MonthViewWeekRow.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = MonthViewWeekRow;
var React = _interopRequireWildcard(require("react"));
var _styles = require("@mui/material/styles");
var _store = require("@base-ui/utils/store");
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _calendarGrid = require("@mui/x-scheduler-internals/calendar-grid");
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _useEventOccurrencesWithDayGridPosition = require("@mui/x-scheduler-internals/use-event-occurrences-with-day-grid-position");
var _eventCalendarSelectors = require("@mui/x-scheduler-internals/event-calendar-selectors");
var _internals = require("@mui/x-scheduler-internals/internals");
var _MonthViewCell = require("./MonthViewCell");
var _EventCalendarStyledContext = require("../../event-calendar/EventCalendarStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const FIXED_CELL_WIDTH = 28;
const MonthViewRow = (0, _styles.styled)(_calendarGrid.CalendarGrid.DayRow, {
  name: 'MuiEventCalendar',
  slot: 'MonthViewRow'
})(({
  theme
}) => ({
  display: 'grid',
  gridTemplateColumns: 'repeat(auto-fit, minmax(0, 1fr))',
  '&[data-show-week-number]': {
    gridTemplateColumns: `${FIXED_CELL_WIDTH}px repeat(auto-fit, minmax(0, 1fr))`
  },
  '&:not(:last-child)': {
    borderBlockEnd: `1px solid ${(theme.vars || theme).palette.divider}`
  }
}));
const MonthViewWeekNumberCell = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MonthViewWeekNumberCell'
})(({
  theme
}) => ({
  padding: theme.spacing(1, 0),
  textAlign: 'center',
  fontSize: theme.typography.caption.fontSize,
  lineHeight: '18px',
  color: (theme.vars || theme).palette.text.secondary
}));
function MonthViewWeekRow(props) {
  const {
    rowIndex,
    maxEvents,
    days,
    occurrencesMap,
    firstDayRef
  } = props;
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const showWeekNumber = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarPreferenceSelectors.showWeekNumber);
  const weekStartsOn = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarPreferenceSelectors.weekStartsOn);
  const {
    schedulerId,
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const occurrences = (0, _useEventOccurrencesWithDayGridPosition.useEventOccurrencesWithDayGridPosition)({
    days,
    occurrencesMap
  });
  const weekNumber = (0, _internals.getWeekNumber)(adapter, days[0].value, weekStartsOn);
  const {
    start,
    end
  } = React.useMemo(() => ({
    start: days[0].value,
    end: adapter.endOfDay(days[days.length - 1].value)
  }), [adapter, days]);
  const weekNumberId = showWeekNumber ? `${schedulerId}-MonthViewWeekNumber-${weekNumber}` : undefined;
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(MonthViewRow, {
    className: classes.monthViewRow,
    start: start,
    end: end,
    rowIndex: rowIndex,
    "data-show-week-number": showWeekNumber || undefined,
    children: [showWeekNumber && /*#__PURE__*/(0, _jsxRuntime.jsx)(MonthViewWeekNumberCell, {
      className: classes.monthViewWeekNumberCell,
      id: weekNumberId,
      "aria-label": localeText.weekNumberAriaLabel(weekNumber),
      "aria-hidden": "true",
      children: weekNumber
    }), occurrences.days.map((day, dayIdx) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_MonthViewCell.MonthViewCell, {
      ref: dayIdx === 0 ? firstDayRef : undefined,
      day: day,
      maxEvents: maxEvents,
      row: occurrences,
      colIndex: dayIdx + 1,
      ariaLabelledBy: weekNumberId
    }, day.key))]
  });
}
