// @mui/x-charts/ScatterChart/seriesConfig/index.mjs
import { getExtremumX, getExtremumY } from "./extremums.mjs";
import seriesProcessor from "./seriesProcessor.mjs";
import getColor from "./getColor.mjs";
import legendGetter from "./legend.mjs";
import tooltipGetter from "./tooltip.mjs";
import getSeriesWithDefaultValues from "./getSeriesWithDefaultValues.mjs";
import tooltipItemPositionGetter from "./tooltipPosition.mjs";
import keyboardFocusHandler from "./keyboardFocusHandler.mjs";
import { identifierSerializerSeriesIdDataIndex } from "../../internals/identifierSerializer.mjs";
import { identifierCleanerSeriesIdDataIndex } from "../../internals/identifierCleaner.mjs";
import { createIsHighlighted, createIsFaded } from "../../internals/plugins/featurePlugins/useChartHighlight/index.mjs";
import descriptionGetter from "./descriptionGetter.mjs";
export const scatterSeriesConfig = {
  seriesProcessor,
  colorProcessor: getColor,
  legendGetter,
  tooltipGetter,
  tooltipItemPositionGetter,
  xExtremumGetter: getExtremumX,
  yExtremumGetter: getExtremumY,
  getSeriesWithDefaultValues,
  keyboardFocusHandler,
  identifierSerializer: identifierSerializerSeriesIdDataIndex,
  identifierCleaner: identifierCleanerSeriesIdDataIndex,
  descriptionGetter,
  isHighlightedCreator: createIsHighlighted,
  isFadedCreator: createIsFaded
};
