// @mui/x-charts/ScatterChart/seriesConfig/seriesProcessor.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _incompleteDatasetKeysError = require("../../internals/incompleteDatasetKeysError");
const seriesProcessor = ({
  series,
  seriesOrder
}, dataset, isItemVisible) => {
  const completeSeries = Object.fromEntries(Object.entries(series).map(([seriesId, seriesData]) => {
    const datasetKeys = seriesData?.datasetKeys;
    const missingKeys = ['x', 'y'].filter(key => typeof datasetKeys?.[key] !== 'string');
    if (seriesData?.datasetKeys && missingKeys.length > 0) {
      (0, _incompleteDatasetKeysError.incompleteDatasetKeysError)('Scatter', seriesId, missingKeys);
    }
    let data;
    if (seriesData.valueGetter) {
      data = dataset?.map(seriesData.valueGetter) ?? [];
    } else if (datasetKeys) {
      data = dataset?.map(d => {
        const x = d[datasetKeys.x];
        const y = d[datasetKeys.y];
        const rep = {
          x,
          y
        };
        if (datasetKeys.colorValue !== undefined) {
          rep.colorValue = d[datasetKeys.colorValue];
        }
        if (datasetKeys.sizeValue !== undefined) {
          rep.sizeValue = d[datasetKeys.sizeValue];
        }
        if (datasetKeys.z !== undefined) {
          rep.z = d[datasetKeys.z];
        }
        if (datasetKeys.id !== undefined) {
          rep.id = d[datasetKeys.id];
        }
        return rep;
      }) ?? [];
    } else {
      data = seriesData.data ?? [];
    }
    return [seriesId, (0, _extends2.default)({
      labelMarkType: 'circle',
      markerSize: 4
    }, seriesData, {
      preview: (0, _extends2.default)({
        markerSize: 1
      }, seriesData?.preview),
      data,
      hidden: !isItemVisible?.({
        type: 'scatter',
        seriesId
      }),
      valueFormatter: seriesData.valueFormatter ?? (v => v && `(${v.x}, ${v.y})`)
    })];
  }));
  return {
    series: completeSeries,
    seriesOrder
  };
};
var _default = exports.default = seriesProcessor;
