// @mui/x-charts/ChartsReferenceLine/ChartsReferenceLine.mjs
'use client';

import _extends from "@babel/runtime/helpers/esm/extends";
import _formatErrorMessage from "@mui/x-internals/formatErrorMessage";
import PropTypes from 'prop-types';
import { ChartsXReferenceLine } from "./ChartsXReferenceLine.mjs";
import { ChartsYReferenceLine } from "./ChartsYReferenceLine.mjs";
import { jsx as _jsx } from "react/jsx-runtime";
function ChartsReferenceLine(props) {
  const {
    x,
    y
  } = props;
  if (x !== undefined && y !== undefined) {
    throw new Error(process.env.NODE_ENV !== "production" ? 'MUI X Charts: The ChartsReferenceLine cannot have both `x` and `y` props set. ' + 'A reference line can only be vertical (x) or horizontal (y), not both. ' + 'Remove either the `x` or `y` prop.' : _formatErrorMessage(4));
  }
  if (x === undefined && y === undefined) {
    throw new Error(process.env.NODE_ENV !== "production" ? 'MUI X Charts: The ChartsReferenceLine requires either an `x` or `y` prop. ' + 'Without a value, the reference line cannot be positioned on the chart. ' + 'Provide either an `x` value for a vertical line or a `y` value for a horizontal line.' : _formatErrorMessage(5));
  }
  if (x !== undefined) {
    return /*#__PURE__*/_jsx(ChartsXReferenceLine, _extends({}, props));
  }
  return /*#__PURE__*/_jsx(ChartsYReferenceLine, _extends({}, props));
}
process.env.NODE_ENV !== "production" ? ChartsReferenceLine.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * The id of the axis used for the reference value.
   * @default The `id` of the first defined axis.
   */
  axisId: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /**
   * Override or extend the styles applied to the component.
   */
  classes: PropTypes.object,
  /**
   * The label to display along the reference line.
   */
  label: PropTypes.string,
  /**
   * The alignment if the label is in the chart drawing area.
   * @default 'middle'
   */
  labelAlign: PropTypes.oneOf(['end', 'middle', 'start']),
  /**
   * The style applied to the label.
   */
  labelStyle: PropTypes.object,
  /**
   * The style applied to the line.
   */
  lineStyle: PropTypes.object,
  /**
   * Additional space around the label in px.
   * Can be a number or an object `{ x, y }` to distinguish space with the reference line and space with axes.
   * @default { x: 0, y: 5 } on a horizontal line and { x: 5, y: 0 } on a vertical line.
   */
  spacing: PropTypes.oneOfType([PropTypes.number, PropTypes.shape({
    x: PropTypes.number,
    y: PropTypes.number
  })]),
  /**
   * The x value associated with the reference line.
   * If defined the reference line will be vertical.
   */
  x: PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.number, PropTypes.string]),
  /**
   * The y value associated with the reference line.
   * If defined the reference line will be horizontal.
   */
  y: PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.number, PropTypes.string])
} : void 0;
export { ChartsReferenceLine };
