// @mui/x-scheduler-internals/calendar-grid/current-time-indicator/CalendarGridCurrentTimeIndicator.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarGridCurrentTimeIndicator = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _store = require("@base-ui/utils/store");
var _useRenderElement = require("../../base-ui-copy/utils/useRenderElement");
var _useAdapterContext = require("../../use-adapter-context");
var _CalendarGridTimeColumnContext = require("../time-column/CalendarGridTimeColumnContext");
var _useElementPositionInCollection = require("../../internals/utils/useElementPositionInCollection");
var _CalendarGridCurrentTimeIndicatorCssVars = require("./CalendarGridCurrentTimeIndicatorCssVars");
var _dateUtils = require("../../internals/utils/date-utils");
var _useEventCalendarStoreContext = require("../../use-event-calendar-store-context");
var _schedulerSelectors = require("../../scheduler-selectors");
var _processDate = require("../../process-date");
const _excluded = ["className", "render", "style"];
const CalendarGridCurrentTimeIndicator = exports.CalendarGridCurrentTimeIndicator = /*#__PURE__*/React.forwardRef(function CalendarGridCurrentTimeIndicator(componentProps, forwardedRef) {
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const elementProps = (0, _objectWithoutPropertiesLoose2.default)(componentProps, _excluded);
  const {
    start: columnStart,
    end: columnEnd
  } = (0, _CalendarGridTimeColumnContext.useCalendarGridTimeColumnContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const now = (0, _store.useStore)(store, _schedulerSelectors.schedulerNowSelectors.nowUpdatedEveryMinute);
  const nowForColumn = React.useMemo(() => (0, _processDate.processDate)((0, _dateUtils.mergeDateAndTime)(adapter, columnStart, now), adapter), [adapter, columnStart, now]);
  const endForCalc = React.useMemo(() => (0, _processDate.processDate)(adapter.addMinutes(nowForColumn.value, 1), adapter), [adapter, nowForColumn]);
  const {
    position
  } = (0, _useElementPositionInCollection.useElementPositionInCollection)({
    start: nowForColumn,
    end: endForCalc,
    collectionStart: columnStart,
    collectionEnd: columnEnd
  });
  const isOutOfRange = adapter.isBefore(nowForColumn.value, columnStart) || adapter.isAfter(nowForColumn.value, columnEnd);
  return (0, _useRenderElement.useRenderElement)('div', componentProps, {
    ref: [forwardedRef],
    props: [elementProps, {
      style: {
        [_CalendarGridCurrentTimeIndicatorCssVars.CalendarGridCurrentTimeIndicatorCssVars.yPosition]: `${position * 100}%`
      }
    }],
    enabled: !isOutOfRange
  });
});
if (process.env.NODE_ENV !== "production") CalendarGridCurrentTimeIndicator.displayName = "CalendarGridCurrentTimeIndicator";
