// @mui/x-scheduler/internals/components/event-skeleton/EventSkeleton.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EventSkeleton = EventSkeleton;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _clsx = _interopRequireDefault(require("clsx"));
var _Skeleton = _interopRequireDefault(require("@mui/material/Skeleton"));
var _styles = require("@mui/material/styles");
var _SharedComponentsStyledContext = require("../SharedComponentsStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const EventSkeletonRoot = (0, _styles.styled)(_Skeleton.default, {
  name: 'MuiEventSkeleton',
  slot: 'Root'
})(({
  theme
}) => {
  // Typing the styles per variant makes adding a new `EventSkeletonVariant`
  // without a matching style block a compile error.
  const variantStyles = {
    'time-column': {
      position: 'absolute',
      left: 0,
      right: 12,
      top: 0,
      height: '100%',
      borderRadius: 0
    },
    'day-grid': {
      height: 18
    },
    agenda: {
      height: 28
    },
    'timeline-row': {
      height: `calc(${theme.typography.body2.lineHeight}em + ${theme.spacing(1.125)})`,
      width: '100%'
    }
  };
  return (0, _extends2.default)({
    opacity: 0.5
  }, Object.fromEntries(Object.entries(variantStyles).map(([variant, style]) => [`&[data-variant="${variant}"]`, style])));
});
function EventSkeleton(props) {
  const {
    classes
  } = (0, _SharedComponentsStyledContext.useSharedComponentsStyledContext)();
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(EventSkeletonRoot, (0, _extends2.default)({
    variant: "rounded",
    animation: "wave",
    "aria-hidden": true
  }, props, {
    className: (0, _clsx.default)(classes.eventSkeleton, props.className)
  }));
}
