// @mui/x-scheduler/event-calendar/resources-tree/ResourcesTree.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ResourcesTree = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _styles = require("@mui/material/styles");
var _Checkbox = require("@mui/material/Checkbox");
var _Typography = _interopRequireDefault(require("@mui/material/Typography"));
var _store = require("@base-ui/utils/store");
var _useStableCallback = require("@base-ui/utils/useStableCallback");
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _RichTreeView = require("@mui/x-tree-view/RichTreeView");
var _TreeItem = require("@mui/x-tree-view/TreeItem");
var _hooks = require("@mui/x-tree-view/hooks");
var _clsx = _interopRequireDefault(require("clsx"));
var _tokens = require("../../internals/utils/tokens");
var _EventCalendarStyledContext = require("../EventCalendarStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["itemId", "label"];
const ResourcesTreeRoot = (0, _styles.styled)('section', {
  name: 'MuiEventCalendar',
  slot: 'ResourcesTree'
})(({
  theme
}) => ({
  display: 'flex',
  flexDirection: 'column',
  padding: theme.spacing(1),
  overflowY: 'auto',
  scrollbarWidth: 'thin',
  [`&[data-flat] .${_TreeItem.treeItemClasses.iconContainer}`]: {
    display: 'none'
  }
}));
const ResourcesTreeLabel = (0, _styles.styled)(_Typography.default, {
  name: 'MuiEventCalendar',
  slot: 'ResourcesTreeLabel'
})(({
  theme
}) => (0, _extends2.default)({}, theme.typography.subtitle2, {
  fontWeight: theme.typography.fontWeightMedium,
  paddingLeft: theme.spacing(1.25),
  paddingBottom: theme.spacing(1)
}));
const ResourcesTreeItemRoot = (0, _styles.styled)(_TreeItem.TreeItem, {
  name: 'MuiEventCalendar',
  slot: 'ResourcesTreeItem'
})(({
  theme
}) => ({
  [`& .${_TreeItem.treeItemClasses.content}[data-selected]`]: {
    backgroundColor: 'transparent',
    '&:hover': {
      backgroundColor: (theme.vars || theme).palette.action.hover,
      '@media (hover: none)': {
        backgroundColor: 'transparent'
      }
    },
    '&[data-focused]': {
      backgroundColor: (theme.vars || theme).palette.action.focus
    }
  },
  [`& .${_Checkbox.checkboxClasses.root}`]: {
    color: 'var(--event-main)',
    [`&.${_Checkbox.checkboxClasses.checked}`]: {
      color: 'var(--event-main)'
    }
  },
  variants: (0, _tokens.getPaletteVariants)(theme)
}));
const ResourcesTreeItemLabel = (0, _styles.styled)(_Typography.default, {
  name: 'MuiEventCalendar',
  slot: 'ResourcesTreeItemLabel'
})(({
  theme
}) => ({
  fontSize: theme.typography.body2.fontSize
}));
const getItemLabel = item => item.title;
function ResourcesTreeItem(props) {
  const {
      itemId,
      label
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const {
    classes
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const eventColor = (0, _store.useStore)(store, _schedulerSelectors.schedulerResourceSelectors.defaultEventColor, itemId);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(ResourcesTreeItemRoot, (0, _extends2.default)({
    itemId: itemId,
    label: /*#__PURE__*/(0, _jsxRuntime.jsx)(ResourcesTreeItemLabel, {
      className: classes.resourcesTreeItemLabel,
      children: label
    }),
    "data-palette": eventColor,
    className: classes.resourcesTreeItem,
    slotProps: {
      checkbox: {
        className: classes.resourcesTreeItemCheckbox,
        size: 'small',
        slotProps: {
          input: {
            tabIndex: -1
          }
        }
      }
    }
  }, other));
}
const ResourcesTree = exports.ResourcesTree = /*#__PURE__*/React.forwardRef(function ResourcesTree(props, forwardedRef) {
  const {
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const apiRef = (0, _hooks.useRichTreeViewApiRef)();
  const headingId = React.useId();
  const resources = (0, _store.useStore)(store, _schedulerSelectors.schedulerResourceSelectors.processedResourceList);
  const childrenLookup = (0, _store.useStore)(store, _schedulerSelectors.schedulerResourceSelectors.processedResourceChildrenLookup);
  const flatList = (0, _store.useStore)(store, _schedulerSelectors.schedulerResourceSelectors.processedResourceFlatList);
  const visibleMap = (0, _store.useStore)(store, _schedulerSelectors.schedulerResourceSelectors.visibleMap);
  const getItemChildren = React.useCallback(item => childrenLookup.get(item.id) ?? [], [childrenLookup]);
  const selectedItems = React.useMemo(() => {
    // A resource is "fully selected" only if itself and every descendant are visible.
    const fullyVisible = new Set();
    for (let i = flatList.length - 1; i >= 0; i -= 1) {
      const resource = flatList[i];
      if (visibleMap[resource.id] === false) {
        continue;
      }
      const children = childrenLookup.get(resource.id);
      if (!children || children.every(child => fullyVisible.has(child.id))) {
        fullyVisible.add(resource.id);
      }
    }
    return flatList.filter(r => fullyVisible.has(r.id)).map(r => r.id);
  }, [flatList, visibleMap, childrenLookup]);
  const defaultExpandedItems = React.useMemo(() => Array.from(childrenLookup.keys()), [childrenLookup]);
  const handleSelectedItemsChange = (0, _useStableCallback.useStableCallback)((event, newSelectedIds) => {
    const newSelected = new Set(newSelectedIds);
    const newVisibleResources = {};
    // Walk in reverse (leaves before parents) so a parent is marked visible when any descendant is selected, even if the parent itself was removed.
    for (let i = flatList.length - 1; i >= 0; i -= 1) {
      const resource = flatList[i];
      if (newSelected.has(resource.id)) {
        newVisibleResources[resource.id] = true;
        continue;
      }
      const children = childrenLookup.get(resource.id);
      const hasVisibleDescendant = children?.some(child => newVisibleResources[child.id] === true);
      newVisibleResources[resource.id] = hasVisibleDescendant ?? false;
    }
    store.setVisibleResources(newVisibleResources, event?.nativeEvent);
  });
  const handleItemClick = (0, _useStableCallback.useStableCallback)((event, itemId) => {
    // Clicks already handled by the checkbox or the expand-collapse icon must not double-toggle.
    const target = event.target;
    if (target.closest(`.${_TreeItem.treeItemClasses.checkbox}`) || target.closest(`.${_TreeItem.treeItemClasses.iconContainer}`)) {
      return;
    }
    apiRef.current?.setItemSelection({
      event,
      itemId,
      keepExistingSelection: true
    });
  });
  if (resources.length === 0) {
    return null;
  }
  const isFlat = childrenLookup.size === 0;
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(ResourcesTreeRoot, (0, _extends2.default)({
    ref: forwardedRef
  }, props, {
    className: (0, _clsx.default)(props.className, classes.resourcesTree),
    "data-flat": isFlat || undefined,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(ResourcesTreeLabel, {
      id: headingId,
      className: classes.resourcesTreeLabel,
      children: localeText.resourcesLabel
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_RichTreeView.RichTreeView, {
      items: resources,
      apiRef: apiRef,
      getItemChildren: getItemChildren,
      getItemLabel: getItemLabel,
      multiSelect: true,
      checkboxSelection: true,
      selectionPropagation: {
        parents: true,
        descendants: true
      },
      expansionTrigger: "iconContainer",
      selectedItems: selectedItems,
      defaultExpandedItems: defaultExpandedItems,
      onSelectedItemsChange: handleSelectedItemsChange,
      onItemClick: handleItemClick,
      slots: {
        item: ResourcesTreeItem
      },
      "aria-labelledby": headingId
    })]
  }));
});
if (process.env.NODE_ENV !== "production") ResourcesTree.displayName = "ResourcesTree";
