// @mui/x-scheduler-internals/calendar-grid/day-event/CalendarGridDayEvent.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarGridDayEvent = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _useStableCallback = require("@base-ui/utils/useStableCallback");
var _useStore = require("@base-ui/utils/store/useStore");
var _useId = require("@base-ui/utils/useId");
var _useButton = require("../../base-ui-copy/utils/useButton");
var _useRenderElement = require("../../base-ui-copy/utils/useRenderElement");
var _useDraggableEvent = require("../../internals/utils/useDraggableEvent");
var _useAdapterContext = require("../../use-adapter-context");
var _CalendarGridDayRowContext = require("../day-row/CalendarGridDayRowContext");
var _schedulerSelectors = require("../../scheduler-selectors");
var _accessibilityUtils = require("../../internals/utils/accessibility-utils");
var _CalendarGridDayEventContext = require("./CalendarGridDayEventContext");
var _useEventCalendarStoreContext = require("../../use-event-calendar-store-context");
var _CalendarGridDayCellContext = require("../day-cell/CalendarGridDayCellContext");
var _CalendarGridRootContext = require("../root/CalendarGridRootContext");
var _eventUtils = require("../../internals/utils/event-utils");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className", "render", "style", "start", "end", "eventId", "occurrenceKey", "renderDragPreview", "id", "isDraggable", "nativeButton"];
const overflowStateAttributesMapping = {
  startingBeforeEdge: value => value ? {
    'data-starting-before-edge': ''
  } : null,
  endingAfterEdge: value => value ? {
    'data-ending-after-edge': ''
  } : null
};
const CalendarGridDayEvent = exports.CalendarGridDayEvent = /*#__PURE__*/React.forwardRef(function CalendarGridDayEvent(componentProps, forwardedRef) {
  const {
      // Rendering props

      // Internal props
      start,
      end,
      eventId,
      occurrenceKey,
      renderDragPreview,
      id: idProp,
      isDraggable = false,
      nativeButton = false
      // Props forwarded to the DOM element
    } = componentProps,
    elementProps = (0, _objectWithoutPropertiesLoose2.default)(componentProps, _excluded);

  // TODO: Expose a real `interactive` prop
  // to control whether the event should behave like a button
  const isInteractive = true;

  // Context hooks
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const {
    id: rootId
  } = (0, _CalendarGridRootContext.useCalendarGridRootContext)();
  const {
    start: rowStart,
    end: rowEnd
  } = (0, _CalendarGridDayRowContext.useCalendarGridDayRowContext)();
  const {
    index: cellIndex,
    hasFocus: cellHasFocus
  } = (0, _CalendarGridDayCellContext.useCalendarGridDayCellContext)();

  // Ref hooks
  const ref = React.useRef(null);

  // Selector hooks
  const hasPlaceholder = (0, _useStore.useStore)(store, _schedulerSelectors.schedulerOccurrencePlaceholderSelectors.isDefined);

  // State hooks
  const id = (0, _useId.useId)(idProp);

  // Feature hooks
  const getDraggedDay = (0, _useStableCallback.useStableCallback)(input => {
    if (!ref.current) {
      return start.value;
    }
    const eventStartInRow = adapter.isBefore(start.value, rowStart) ? rowStart : start.value;
    const eventEndInRow = adapter.isAfter(end.value, rowEnd) ? rowEnd : end.value;
    const eventDayLengthInRow = adapter.differenceInDays(eventEndInRow, eventStartInRow) + 1;
    const clientX = input.clientX;
    const elementPosition = ref.current.getBoundingClientRect();
    const positionX = (clientX - elementPosition.x) / ref.current.offsetWidth;
    return adapter.startOfDay(adapter.addDays(eventStartInRow, Math.ceil(positionX * eventDayLengthInRow) - 1));
  });
  const firstEventOfSeries = _schedulerSelectors.schedulerEventSelectors.processedEvent(store.state, eventId);
  const originalOccurrence = (0, _eventUtils.generateOccurrenceFromEvent)({
    event: firstEventOfSeries,
    eventId,
    occurrenceKey,
    start,
    end
  });
  const getSharedDragData = (0, _useStableCallback.useStableCallback)(() => ({
    eventId,
    occurrenceKey,
    originalOccurrence,
    start: start.value,
    end: end.value
  }));
  const getDragData = (0, _useStableCallback.useStableCallback)(input => (0, _extends2.default)({}, getSharedDragData(input), {
    source: 'CalendarGridDayEvent',
    draggedDay: getDraggedDay(input)
  }));
  const {
    state,
    preview,
    contextValue: draggableEventContextValue
  } = (0, _useDraggableEvent.useDraggableEvent)({
    ref,
    start,
    end,
    occurrenceKey,
    eventId,
    isDraggable,
    renderDragPreview,
    getDragData,
    collectionStart: rowStart,
    collectionEnd: rowEnd
  });
  const startingBeforeEdge = draggableEventContextValue.doesEventStartBeforeCollectionStart;
  const endingAfterEdge = draggableEventContextValue.doesEventEndAfterCollectionEnd;
  const mergedState = (0, _extends2.default)({}, state, {
    startingBeforeEdge,
    endingAfterEdge
  });
  const {
    getButtonProps,
    buttonRef
  } = (0, _useButton.useButton)({
    disabled: !isInteractive,
    native: nativeButton,
    tabIndex: cellHasFocus ? 0 : -1
  });

  // Rendering hooks

  const columnHeaderId = (0, _accessibilityUtils.getCalendarGridHeaderCellId)(rootId, cellIndex);
  const contextValue = React.useMemo(() => (0, _extends2.default)({}, draggableEventContextValue, {
    getSharedDragData
  }), [draggableEventContextValue, getSharedDragData]);
  const element = (0, _useRenderElement.useRenderElement)('div', componentProps, {
    state: mergedState,
    ref: [forwardedRef, buttonRef, ref],
    props: [elementProps, {
      id,
      'aria-labelledby': `${columnHeaderId} ${id}`,
      style: hasPlaceholder ? {
        pointerEvents: 'none'
      } : undefined
    }, getButtonProps],
    stateAttributesMapping: overflowStateAttributesMapping
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_CalendarGridDayEventContext.CalendarGridDayEventContext.Provider, {
    value: contextValue,
    children: [element, preview.element]
  });
});
if (process.env.NODE_ENV !== "production") CalendarGridDayEvent.displayName = "CalendarGridDayEvent";
