// @mui/x-scheduler-internals/calendar-grid/use-placeholder-in-day/useCalendarGridPlaceholderInDay.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCalendarGridPlaceholderInDay = useCalendarGridPlaceholderInDay;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _useStore = require("@base-ui/utils/store/useStore");
var _schedulerSelectors = require("../../scheduler-selectors");
var _useEventCalendarStoreContext = require("../../use-event-calendar-store-context");
var _CalendarGridDayRowContext = require("../day-row/CalendarGridDayRowContext");
var _useAdapterContext = require("../../use-adapter-context");
var _eventCalendarSelectors = require("../../event-calendar-selectors");
var _processDate = require("../../process-date");
var _dragUtils = require("../../internals/utils/drag-utils");
function useCalendarGridPlaceholderInDay(day, row, maxEvents) {
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const {
    start: rowStart,
    end: rowEnd
  } = (0, _CalendarGridDayRowContext.useCalendarGridDayRowContext)();
  const rawPlaceholder = (0, _useStore.useStore)(store, _eventCalendarSelectors.eventCalendarOccurrencePlaceholderSelectors.placeholderInDayCell, day, rowStart);
  const originalEventId = (0, _dragUtils.isInternalDragOrResizePlaceholder)(rawPlaceholder) ? rawPlaceholder.eventId : null;
  const originalEvent = (0, _useStore.useStore)(store, _schedulerSelectors.schedulerEventSelectors.processedEvent, originalEventId);
  return React.useMemo(() => {
    if (!rawPlaceholder) {
      return null;
    }
    const sharedProperties = {
      key: 'occurrence-placeholder',
      id: originalEventId ?? 'occurrence-placeholder',
      title: originalEvent ? originalEvent.title : ''
    };

    // Creation mode
    if (rawPlaceholder.type === 'creation') {
      const startProcessed = (0, _processDate.processDate)(day, adapter);
      const endProcessed = (0, _processDate.processDate)(adapter.isAfter(rawPlaceholder.end, rowEnd) ? rowEnd : rawPlaceholder.end, adapter);
      const timezone = adapter.getTimezone(day);
      return (0, _extends2.default)({}, sharedProperties, {
        title: '',
        allDay: true,
        displayTimezone: {
          start: startProcessed,
          end: endProcessed,
          timezone
        },
        position: {
          index: 1,
          daySpan: adapter.differenceInDays(rawPlaceholder.end, day) + 1
        }
      });
    }
    if (rawPlaceholder.type === 'external-drag') {
      const startProcessed = (0, _processDate.processDate)(rawPlaceholder.start, adapter);
      const endProcessed = (0, _processDate.processDate)(rawPlaceholder.end, adapter);
      const timezone = adapter.getTimezone(rawPlaceholder.start);
      return (0, _extends2.default)({}, sharedProperties, {
        title: rawPlaceholder.eventData.title ?? '',
        displayTimezone: {
          start: startProcessed,
          end: endProcessed,
          timezone
        },
        position: {
          index: 1,
          daySpan: adapter.differenceInDays(rawPlaceholder.end, day) + 1
        }
      });
    }
    let positionIndex = 1;
    const targetDay = row.days.find(rowDay => adapter.isSameDay(rowDay.value, day));
    if (targetDay) {
      const usedIndexes = new Set(targetDay.withPosition.filter(occ => occ.key !== rawPlaceholder.occurrenceKey).map(occ => occ.position.index));
      while (usedIndexes.has(positionIndex)) {
        positionIndex += 1;
      }
    }

    // If the position exceeds the available event rows, clamp it so the
    // placeholder renders on top of an existing event instead of overflowing.
    if (maxEvents != null && positionIndex > maxEvents) {
      positionIndex = maxEvents;
    }
    return (0, _extends2.default)({}, sharedProperties, {
      start: (0, _processDate.processDate)(rawPlaceholder.start, adapter),
      end: (0, _processDate.processDate)(rawPlaceholder.end, adapter),
      displayTimezone: (0, _extends2.default)({}, originalEvent.displayTimezone),
      position: {
        index: positionIndex,
        daySpan: adapter.differenceInDays(rawPlaceholder.end, day) + 1
      }
    });
  }, [adapter, day, maxEvents, originalEvent, originalEventId, rawPlaceholder, row.days, rowEnd]);
}
