// @mui/x-data-grid/components/GridScrollbarFillerCell.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridScrollbarFillerCell = GridScrollbarFillerCell;
var _clsx = _interopRequireDefault(require("clsx"));
var _constants = require("../constants");
var _useGridPrivateApiContext = require("../hooks/utils/useGridPrivateApiContext");
var _usePinnedScrollOffset = require("../hooks/utils/usePinnedScrollOffset");
var _constants2 = require("../internals/constants");
var _jsxRuntime = require("react/jsx-runtime");
const classes = {
  root: _constants.gridClasses.scrollbarFiller,
  pinnedRight: _constants.gridClasses['scrollbarFiller--pinnedRight']
};
function GridScrollbarFillerCell({
  pinnedRight
}) {
  const apiRef = (0, _useGridPrivateApiContext.useGridPrivateApiContext)();
  const pinnedScrollOffset = (0, _usePinnedScrollOffset.usePinnedScrollOffset)(apiRef, pinnedRight ? _constants2.PinnedColumnPosition.RIGHT : undefined);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    role: "none",
    className: (0, _clsx.default)(classes.root, pinnedRight && classes.pinnedRight),
    style: {
      right: pinnedScrollOffset
    }
  });
}
