// @mui/x-scheduler-internals/calendar-grid/header-cell/CalendarGridHeaderCell.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarGridHeaderCell = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _store = require("@base-ui/utils/store");
var _useRenderElement = require("../../base-ui-copy/utils/useRenderElement");
var _useCompositeListItem = require("../../base-ui-copy/composite/list/useCompositeListItem");
var _CompositeListContext = require("../../base-ui-copy/composite/list/CompositeListContext");
var _useAdapterContext = require("../../use-adapter-context");
var _useEventCalendarStoreContext = require("../../use-event-calendar-store-context");
var _accessibilityUtils = require("../../internals/utils/accessibility-utils");
var _getNavigationTarget = require("../../internals/utils/getNavigationTarget");
var _CalendarGridRootContext = require("../root/CalendarGridRootContext");
var _schedulerSelectors = require("../../scheduler-selectors");
const _excluded = ["className", "render", "style", "date", "skipDataCurrent", "ariaLabelFormat"];
const selectorIsCurrentDate = (0, _store.createSelector)((state, date, skipDataCurrent) => !skipDataCurrent && _schedulerSelectors.schedulerNowSelectors.isCurrentDay(state, date));
const CalendarGridHeaderCell = exports.CalendarGridHeaderCell = /*#__PURE__*/React.forwardRef(function CalendarGridHeaderCell(componentProps, forwardedRef) {
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const {
      // Rendering props

      // Internal props
      date,
      skipDataCurrent,
      ariaLabelFormat = adapter.formats.weekday
      // Props forwarded to the DOM element
    } = componentProps,
    elementProps = (0, _objectWithoutPropertiesLoose2.default)(componentProps, _excluded);
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const {
    id: rootId,
    focusedCell,
    setFocusedCell,
    rowTypes,
    rowsPerType
  } = (0, _CalendarGridRootContext.useCalendarGridRootContext)();
  const isCurrentDay = (0, _store.useStore)(store, selectorIsCurrentDate, date.value, skipDataCurrent);
  const {
    ref: listItemRef,
    index
  } = (0, _useCompositeListItem.useCompositeListItem)();
  const {
    elementsRef
  } = (0, _CompositeListContext.useCompositeListContext)();
  const id = (0, _accessibilityUtils.getCalendarGridHeaderCellId)(rootId, index);
  const cellRef = React.useRef(null);
  const hasFocus = focusedCell?.rowType === 'header' && focusedCell?.rowIndex === 0 && focusedCell?.columnIndex === index;

  // Apply DOM focus when this cell becomes the focused cell
  React.useEffect(() => {
    if (hasFocus && cellRef.current && !cellRef.current.contains(document.activeElement)) {
      cellRef.current.focus({
        preventScroll: true
      });
    }
  }, [hasFocus]);
  const handleKeyDown = event => {
    const target = (0, _getNavigationTarget.getNavigationTarget)(event.key, 'header', 0, index, {
      columnCount: elementsRef.current.length,
      rowTypes,
      rowsPerType
    });
    if (target) {
      event.preventDefault();
      setFocusedCell(target);
      return;
    }
    if (event.key === 'Enter' && event.target === event.currentTarget) {
      const button = event.currentTarget.querySelector('button');
      if (button) {
        button.click();
      }
    }
  };
  const handleFocus = event => {
    if (event.target === event.currentTarget) {
      setFocusedCell({
        rowType: 'header',
        rowIndex: 0,
        columnIndex: index
      });
    }
  };
  const state = React.useMemo(() => ({
    current: isCurrentDay
  }), [isCurrentDay]);
  const keyboardProps = {
    tabIndex: 0,
    onKeyDown: handleKeyDown,
    onFocus: handleFocus
  };
  return (0, _useRenderElement.useRenderElement)('div', componentProps, {
    state,
    ref: [forwardedRef, listItemRef, cellRef],
    props: [{
      role: 'columnheader',
      id,
      'aria-label': `${adapter.formatByString(date.value, ariaLabelFormat)}`,
      'aria-colindex': index + 1
    }, keyboardProps, elementProps]
  });
});
if (process.env.NODE_ENV !== "production") CalendarGridHeaderCell.displayName = "CalendarGridHeaderCell";
