// @mui/x-scheduler/event-calendar/mini-calendar/MiniCalendar.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MiniCalendar = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _store = require("@base-ui/utils/store");
var _styles = require("@mui/material/styles");
var _IconButton = _interopRequireDefault(require("@mui/material/IconButton"));
var _ChevronLeft = _interopRequireDefault(require("@mui/icons-material/ChevronLeft"));
var _ChevronRight = _interopRequireDefault(require("@mui/icons-material/ChevronRight"));
var _useAdapter = require("@mui/x-scheduler-internals/use-adapter");
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _eventCalendarSelectors = require("@mui/x-scheduler-internals/event-calendar-selectors");
var _getDayList = require("@mui/x-scheduler-internals/get-day-list");
var _internals = require("@mui/x-scheduler-internals/internals");
var _EventCalendarStyledContext = require("../EventCalendarStyledContext");
var _dateUtils = require("../../internals/utils/date-utils");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className"];
const MiniCalendarRoot = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MiniCalendar'
})(({
  theme
}) => ({
  userSelect: 'none',
  padding: theme.spacing(1)
}));
const MiniCalendarHeader = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MiniCalendarHeader'
})(({
  theme
}) => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'space-between',
  marginBottom: theme.spacing(1)
}));
const MiniCalendarNavigation = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MiniCalendarNavigation'
})(({
  theme
}) => ({
  display: 'flex',
  alignItems: 'center',
  gap: theme.spacing(0.5)
}));
const MiniCalendarMonthLabel = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'MiniCalendarMonthLabel'
})(({
  theme
}) => (0, _extends2.default)({}, theme.typography.subtitle2, {
  fontWeight: theme.typography.fontWeightMedium,
  paddingLeft: theme.spacing(1.25)
}));
const MiniCalendarWeekdayHeader = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MiniCalendarWeekdayHeader'
})(({
  theme
}) => ({
  display: 'grid',
  gridTemplateColumns: 'repeat(7, 1fr)',
  textAlign: 'center',
  marginBottom: theme.spacing(0.5)
}));
const MiniCalendarWeekdayCell = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'MiniCalendarWeekdayCell'
})(({
  theme
}) => ({
  fontSize: theme.typography.caption.fontSize,
  color: (theme.vars || theme).palette.text.secondary,
  padding: theme.spacing(0.5, 0),
  '&[data-weekend]': {
    color: (theme.vars || theme).palette.error.main
  }
}));
const MiniCalendarGrid = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MiniCalendarGrid'
})({
  display: 'flex',
  flexDirection: 'column'
});
const MiniCalendarWeekRow = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MiniCalendarWeekRow'
})({
  display: 'grid',
  gridTemplateColumns: 'repeat(7, 1fr)'
});
const MiniCalendarDayCell = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MiniCalendarDayCell'
})({
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  aspectRatio: '1'
});
const MiniCalendarDayButton = (0, _styles.styled)('button', {
  name: 'MuiEventCalendar',
  slot: 'MiniCalendarDayButton'
})(({
  theme
}) => ({
  width: 28,
  height: 28,
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  border: 'none',
  borderRadius: '50%',
  background: 'none',
  cursor: 'pointer',
  fontSize: theme.typography.caption.fontSize,
  color: (theme.vars || theme).palette.text.primary,
  padding: 0,
  '&:hover': {
    backgroundColor: theme.alpha((theme.vars || theme).palette.primary.main, 0.08)
  },
  '&:focus-visible': {
    outline: `2px solid ${(theme.vars || theme).palette.primary.main}`,
    outlineOffset: 2
  },
  '&[data-other-month]': {
    color: (theme.vars || theme).palette.text.disabled
  },
  '&[data-today]:not([data-active])': {
    fontWeight: theme.typography.fontWeightBold,
    color: (theme.vars || theme).palette.primary.main,
    backgroundColor: theme.alpha((theme.vars || theme).palette.primary.main, 0.15)
  },
  '&[data-active]': {
    backgroundColor: (theme.vars || theme).palette.primary.main,
    color: (theme.vars || theme).palette.primary.contrastText,
    '&:hover': {
      backgroundColor: (theme.vars || theme).palette.primary.dark
    }
  },
  '&[data-today][data-active]': {
    fontWeight: theme.typography.fontWeightBold
  }
}));

/**
 * A compact month calendar for navigating to specific days.
 */
const MiniCalendar = exports.MiniCalendar = /*#__PURE__*/React.forwardRef(function MiniCalendar(props, forwardedRef) {
  const {
      className
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const {
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const visibleDate = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.visibleDate);
  const now = (0, _store.useStore)(store, _schedulerSelectors.schedulerNowSelectors.nowUpdatedEveryMinute);
  const weekStartsOn = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarPreferenceSelectors.weekStartsOn);
  const [displayedMonth, setDisplayedMonth] = React.useState(() => adapter.startOfMonth(visibleDate));

  // Sync displayed month when visibleDate changes externally
  React.useEffect(() => {
    setDisplayedMonth(adapter.startOfMonth(visibleDate));
  }, [adapter, visibleDate]);

  // Generate days for the displayed month (including padding days from adjacent months)
  // Always show 6 weeks (42 days) for consistent height
  const days = React.useMemo(() => {
    const monthStart = adapter.startOfMonth(displayedMonth);
    const gridStart = (0, _internals.getStartOfWeek)(adapter, monthStart, weekStartsOn);
    // 6 weeks = 42 days, so end is 41 days after start
    const gridEnd = adapter.addDays(gridStart, 41);
    return (0, _getDayList.getDayList)({
      adapter,
      start: gridStart,
      end: gridEnd
    });
  }, [adapter, displayedMonth, weekStartsOn]);

  // Group days into weeks
  const weeks = React.useMemo(() => {
    const result = [];
    for (let i = 0; i < days.length; i += 7) {
      result.push(days.slice(i, i + 7));
    }
    return result;
  }, [days]);

  // Get weekday headers from first week
  const weekdays = weeks[0] ?? [];
  const monthYearLabel = React.useMemo(() => (0, _dateUtils.formatMonthFullLetterAndYear)(displayedMonth, adapter), [adapter, displayedMonth]);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(MiniCalendarRoot, (0, _extends2.default)({
    ref: forwardedRef,
    role: "grid",
    "aria-label": localeText.miniCalendarLabel,
    "aria-rowcount": 1 + weeks.length,
    "aria-colcount": 7,
    className: (0, _clsx.default)(classes.miniCalendar, className)
  }, other, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(MiniCalendarHeader, {
      className: classes.miniCalendarHeader,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(MiniCalendarMonthLabel, {
        className: classes.miniCalendarMonthLabel,
        children: monthYearLabel
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(MiniCalendarNavigation, {
        className: classes.miniCalendarNavigation,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_IconButton.default, {
          className: classes.miniCalendarPreviousButton,
          size: "small",
          "aria-label": localeText.miniCalendarGoToPreviousMonth,
          onClick: () => setDisplayedMonth(prev => adapter.addMonths(prev, -1)),
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_ChevronLeft.default, {
            fontSize: "small"
          })
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_IconButton.default, {
          className: classes.miniCalendarNextButton,
          size: "small",
          "aria-label": localeText.miniCalendarGoToNextMonth,
          onClick: () => setDisplayedMonth(prev => adapter.addMonths(prev, 1)),
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_ChevronRight.default, {
            fontSize: "small"
          })
        })]
      })]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(MiniCalendarWeekdayHeader, {
      role: "row",
      "aria-rowindex": 1,
      className: classes.miniCalendarWeekdayHeader,
      children: weekdays.map((day, dayIndex) => /*#__PURE__*/(0, _jsxRuntime.jsx)(MiniCalendarWeekdayCell, {
        role: "columnheader",
        "aria-label": adapter.formatByString(day.value, adapter.formats.weekday),
        "aria-colindex": dayIndex + 1,
        className: classes.miniCalendarWeekdayCell,
        "data-weekend": (0, _useAdapter.isWeekend)(adapter, day.value) || undefined,
        children: adapter.formatByString(day.value, adapter.formats.weekday1Letter)
      }, day.key))
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(MiniCalendarGrid, {
      role: "rowgroup",
      className: classes.miniCalendarGrid,
      children: weeks.map((week, weekIndex) => /*#__PURE__*/(0, _jsxRuntime.jsx)(MiniCalendarWeekRow, {
        role: "row",
        "aria-rowindex": weekIndex + 2,
        className: classes.miniCalendarWeekRow,
        children: week.map((day, dayIndex) => {
          const isToday = adapter.isSameDay(day.value, now);
          const isActive = adapter.isSameDay(day.value, visibleDate);
          const isOtherMonth = !adapter.isSameMonth(day.value, displayedMonth);

          // Create a full date label for accessibility
          const fullDateLabel = adapter.formatByString(day.value, adapter.formats.localizedDateWithFullMonthAndWeekDay);
          return /*#__PURE__*/(0, _jsxRuntime.jsx)(MiniCalendarDayCell, {
            role: "gridcell",
            "aria-colindex": dayIndex + 1,
            className: classes.miniCalendarDayCell,
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(MiniCalendarDayButton, {
              type: "button",
              className: classes.miniCalendarDayButton,
              "data-other-month": isOtherMonth || undefined,
              "data-today": isToday || undefined,
              "data-active": isActive || undefined,
              "aria-label": fullDateLabel,
              "aria-current": isToday ? 'date' : undefined,
              "aria-selected": isActive,
              tabIndex: isActive ? 0 : -1,
              onClick: event => store.goToDate(day.value, event),
              children: adapter.formatByString(day.value, adapter.formats.dayOfMonth)
            })
          }, day.key);
        })
      }, weekIndex))
    })]
  }));
});
if (process.env.NODE_ENV !== "production") MiniCalendar.displayName = "MiniCalendar";
