// @mui/x-charts/ChartsSvgLayer/ChartsSvgLayer.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartsSvgLayer = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _styles = require("@mui/material/styles");
var _propTypes = _interopRequireDefault(require("prop-types"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _ChartsAxesGradients = require("../internals/components/ChartsAxesGradients");
var _ChartsProvider = require("../context/ChartsProvider");
var _useChartDimensions = require("../internals/plugins/corePlugins/useChartDimensions/useChartDimensions.selectors");
var _chartsSvgLayerClasses = require("./chartsSvgLayerClasses");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["children", "className"];
const ChartsSvgLayerStyles = (0, _styles.styled)('svg', {
  name: 'MuiChartsSvgLayer',
  slot: 'Root'
})({
  width: '100%',
  height: '100%',
  position: 'absolute',
  inset: 0
});

/**
 * A layer that provides the drawing area SVG the chart elements.
 * Must be wrapped in a `<ChartsLayerContainer>`.
 *
 * Demos:
 *
 * - [Composition](https://mui.com/x/api/charts/composition/)
 *
 * API:
 *
 * - [ChartsSvgLayer API](https://mui.com/x/api/charts/charts-svg-layer/)
 */
const ChartsSvgLayer = exports.ChartsSvgLayer = /*#__PURE__*/React.forwardRef(function ChartsSvgLayer(inProps, ref) {
  const {
    store
  } = (0, _ChartsProvider.useChartsContext)();
  const svgWidth = store.use(_useChartDimensions.selectorChartSvgWidth);
  const svgHeight = store.use(_useChartDimensions.selectorChartSvgHeight);
  const themeProps = (0, _styles.useThemeProps)({
    props: inProps,
    name: 'MuiChartsSvgLayer'
  });
  const {
      children,
      className
    } = themeProps,
    other = (0, _objectWithoutPropertiesLoose2.default)(themeProps, _excluded);
  const classes = (0, _chartsSvgLayerClasses.useUtilityClasses)();
  const hasIntrinsicSize = svgHeight > 0 && svgWidth > 0;
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(ChartsSvgLayerStyles, (0, _extends2.default)({
    viewBox: `0 0 ${svgWidth} ${svgHeight}`,
    className: (0, _clsx.default)(classes.root, className)
  }, other, {
    ref: ref,
    "aria-hidden": true,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_ChartsAxesGradients.ChartsAxesGradients, {}), hasIntrinsicSize && children]
  }));
});
if (process.env.NODE_ENV !== "production") ChartsSvgLayer.displayName = "ChartsSvgLayer";
process.env.NODE_ENV !== "production" ? ChartsSvgLayer.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  children: _propTypes.default.node,
  className: _propTypes.default.string,
  sx: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object, _propTypes.default.bool])), _propTypes.default.func, _propTypes.default.object])
} : void 0;
