// @mui/x-charts/PieChart/PieArcLabel.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PieArcLabel = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _clsx = _interopRequireDefault(require("clsx"));
var _styles = require("@mui/material/styles");
var _animation = require("../internals/animation/animation");
var _useAnimatePieArcLabel = require("../hooks/animation/useAnimatePieArcLabel");
var _pieClasses = require("./pieClasses");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["seriesId", "classes", "color", "startAngle", "endAngle", "paddingAngle", "arcLabelRadius", "cornerRadius", "formattedArcLabel", "isHighlighted", "isFaded", "skipAnimation", "hidden", "className"];
const PieArcLabelRoot = (0, _styles.styled)('text', {
  name: 'MuiPieArcLabel',
  slot: 'Root'
})(({
  theme
}) => ({
  fill: (theme.vars || theme).palette.text.primary,
  textAnchor: 'middle',
  dominantBaseline: 'middle',
  pointerEvents: 'none',
  animationName: 'animate-opacity',
  animationDuration: '0s',
  animationTimingFunction: _animation.ANIMATION_TIMING_FUNCTION,
  transitionDuration: `${_animation.ANIMATION_DURATION_MS}ms`,
  transitionProperty: 'opacity',
  transitionTimingFunction: _animation.ANIMATION_TIMING_FUNCTION,
  [`&.${_pieClasses.pieClasses.animate}`]: {
    animationDuration: `${_animation.ANIMATION_DURATION_MS}ms`
  },
  '@keyframes animate-opacity': {
    from: {
      opacity: 0
    }
  }
}));
const PieArcLabel = exports.PieArcLabel = /*#__PURE__*/React.forwardRef(function PieArcLabel(props, ref) {
  const {
      seriesId,
      classes: innerClasses,
      color,
      startAngle,
      endAngle,
      paddingAngle,
      arcLabelRadius,
      cornerRadius,
      formattedArcLabel,
      isHighlighted,
      isFaded,
      skipAnimation,
      hidden,
      className
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const ownerState = {
    seriesId,
    classes: innerClasses,
    color,
    isFaded,
    isHighlighted,
    skipAnimation
  };
  const classes = (0, _pieClasses.useUtilityClasses)(ownerState);
  const animatedProps = (0, _useAnimatePieArcLabel.useAnimatePieArcLabel)({
    cornerRadius,
    startAngle,
    endAngle,
    arcLabelRadius,
    paddingAngle,
    skipAnimation,
    ref
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(PieArcLabelRoot, (0, _extends2.default)({
    className: (0, _clsx.default)(classes.arcLabel, className),
    "data-highlighted": isHighlighted || undefined,
    "data-faded": isFaded || undefined
  }, other, animatedProps, {
    opacity: hidden ? 0 : 1,
    children: formattedArcLabel
  }));
});
if (process.env.NODE_ENV !== "production") PieArcLabel.displayName = "PieArcLabel";
process.env.NODE_ENV !== "production" ? PieArcLabel.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  arcLabelRadius: _propTypes.default.number.isRequired,
  classes: _propTypes.default.object,
  color: _propTypes.default.string.isRequired,
  cornerRadius: _propTypes.default.number.isRequired,
  endAngle: _propTypes.default.number.isRequired,
  formattedArcLabel: _propTypes.default.string,
  hidden: _propTypes.default.bool,
  isFaded: _propTypes.default.bool.isRequired,
  isHighlighted: _propTypes.default.bool.isRequired,
  paddingAngle: _propTypes.default.number.isRequired,
  seriesId: _propTypes.default.string.isRequired,
  skipAnimation: _propTypes.default.bool.isRequired,
  startAngle: _propTypes.default.number.isRequired
} : void 0;
