// @mui/x-charts/PieChart/PieArc.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PieArc = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _clsx = _interopRequireDefault(require("clsx"));
var _styles = require("@mui/material/styles");
var _hooks = require("../hooks");
var _animation = require("../internals/animation/animation");
var _useInteractionItemProps = require("../hooks/useInteractionItemProps");
var _pieClasses = require("./pieClasses");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className", "classes", "color", "dataIndex", "seriesId", "isFaded", "isHighlighted", "isFocused", "onClick", "cornerRadius", "startAngle", "endAngle", "innerRadius", "outerRadius", "paddingAngle", "skipAnimation", "stroke", "skipInteraction"];
const PieArcRoot = (0, _styles.styled)('path', {
  name: 'MuiPieArc',
  slot: 'Root'
})({
  transitionProperty: 'opacity, fill, filter',
  transitionDuration: `${_animation.ANIMATION_DURATION_MS}ms`,
  transitionTimingFunction: _animation.ANIMATION_TIMING_FUNCTION
});
const PieArc = exports.PieArc = /*#__PURE__*/React.forwardRef(function PieArc(props, ref) {
  const {
      className,
      classes: innerClasses,
      color,
      dataIndex,
      seriesId,
      isFaded,
      isHighlighted,
      isFocused,
      onClick,
      cornerRadius,
      startAngle,
      endAngle,
      innerRadius,
      outerRadius,
      paddingAngle,
      skipAnimation,
      stroke: strokeProp
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const theme = (0, _styles.useTheme)();
  const stroke = strokeProp ?? (theme.vars || theme).palette.background.paper;
  const ownerState = {
    seriesId,
    dataIndex,
    classes: innerClasses,
    color,
    isFaded,
    isHighlighted,
    isFocused
  };
  const classes = (0, _pieClasses.useUtilityClasses)(ownerState);
  const interactionProps = (0, _useInteractionItemProps.useInteractionItemProps)({
    type: 'pie',
    seriesId,
    dataIndex
  });
  const animatedProps = (0, _hooks.useAnimatePieArc)({
    cornerRadius,
    startAngle,
    endAngle,
    innerRadius,
    outerRadius,
    paddingAngle,
    skipAnimation,
    ref
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(PieArcRoot, (0, _extends2.default)({
    onClick: onClick,
    cursor: onClick ? 'pointer' : 'unset',
    ownerState: ownerState,
    className: (0, _clsx.default)(classes.arc, className),
    fill: color,
    opacity: isFaded ? 0.3 : 1,
    filter: isHighlighted ? 'brightness(120%)' : 'none',
    stroke: stroke,
    strokeWidth: 1,
    strokeLinejoin: "round",
    "data-highlighted": isHighlighted || undefined,
    "data-faded": isFaded || undefined,
    "data-index": dataIndex
  }, other, interactionProps, animatedProps));
});
if (process.env.NODE_ENV !== "production") PieArc.displayName = "PieArc";
process.env.NODE_ENV !== "production" ? PieArc.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  classes: _propTypes.default.object,
  cornerRadius: _propTypes.default.number.isRequired,
  dataIndex: _propTypes.default.number.isRequired,
  endAngle: _propTypes.default.number.isRequired,
  innerRadius: _propTypes.default.number.isRequired,
  isFaded: _propTypes.default.bool.isRequired,
  isFocused: _propTypes.default.bool.isRequired,
  isHighlighted: _propTypes.default.bool.isRequired,
  outerRadius: _propTypes.default.number.isRequired,
  paddingAngle: _propTypes.default.number.isRequired,
  seriesId: _propTypes.default.string.isRequired,
  /**
   * If `true`, the animation is disabled.
   */
  skipAnimation: _propTypes.default.bool,
  /**
   * If `true`, the default event handlers are disabled.
   * Those are used, for example, to display a tooltip or highlight the arc on hover.
   */
  skipInteraction: _propTypes.default.bool,
  startAngle: _propTypes.default.number.isRequired
} : void 0;
