// @mui/x-charts/PieChart/PiePlot.mjs
'use client';

import clsx from 'clsx';
import PropTypes from 'prop-types';
import { PieArcPlot } from "./PieArcPlot.mjs";
import { PieArcLabelPlot } from "./PieArcLabelPlot.mjs";
import { usePieSeriesContext, usePieSeriesLayout } from "../hooks/usePieSeries.mjs";
import { useSkipAnimation } from "../hooks/useSkipAnimation.mjs";
import { useUtilityClasses } from "./pieClasses.mjs";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/**
 * Demos:
 *
 * - [Pie](https://mui.com/x/react-charts/pie/)
 * - [Pie demonstration](https://mui.com/x/react-charts/pie-demo/)
 *
 * API:
 *
 * - [PiePlot API](https://mui.com/x/api/charts/pie-plot/)
 */
function PiePlot(props) {
  const {
    skipAnimation: inSkipAnimation,
    slots,
    slotProps,
    onItemClick,
    className
  } = props;
  const seriesData = usePieSeriesContext();
  const seriesLayout = usePieSeriesLayout();
  const skipAnimation = useSkipAnimation(inSkipAnimation);
  const classes = useUtilityClasses();
  if (seriesData === undefined) {
    return null;
  }
  const {
    series,
    seriesOrder
  } = seriesData;
  return /*#__PURE__*/_jsxs("g", {
    className: clsx(classes.root, className),
    children: [seriesOrder.map(seriesId => {
      const {
        cornerRadius,
        paddingAngle,
        data,
        highlighted,
        faded
      } = series[seriesId];
      return /*#__PURE__*/_jsx("g", {
        className: classes.series,
        transform: `translate(${seriesLayout[seriesId].center.x}, ${seriesLayout[seriesId].center.y})`,
        "data-series": seriesId,
        children: /*#__PURE__*/_jsx(PieArcPlot, {
          innerRadius: seriesLayout[seriesId].radius.inner,
          outerRadius: seriesLayout[seriesId].radius.outer,
          cornerRadius: cornerRadius,
          paddingAngle: paddingAngle,
          seriesId: seriesId,
          data: data,
          skipAnimation: skipAnimation,
          highlighted: highlighted,
          faded: faded,
          onItemClick: onItemClick,
          slots: slots,
          slotProps: slotProps
        })
      }, seriesId);
    }), seriesOrder.map(seriesId => {
      const {
        cornerRadius,
        paddingAngle,
        arcLabel,
        arcLabelMinAngle,
        data
      } = series[seriesId];
      return /*#__PURE__*/_jsx("g", {
        className: classes.seriesLabels,
        transform: `translate(${seriesLayout[seriesId].center.x}, ${seriesLayout[seriesId].center.y})`,
        "data-series": seriesId,
        children: /*#__PURE__*/_jsx(PieArcLabelPlot, {
          innerRadius: seriesLayout[seriesId].radius.inner,
          outerRadius: seriesLayout[seriesId].radius.outer,
          arcLabelRadius: seriesLayout[seriesId].radius.label,
          cornerRadius: cornerRadius,
          paddingAngle: paddingAngle,
          seriesId: seriesId,
          data: data,
          skipAnimation: skipAnimation,
          arcLabel: arcLabel,
          arcLabelMinAngle: arcLabelMinAngle,
          slots: slots,
          slotProps: slotProps
        })
      }, seriesId);
    })]
  });
}
process.env.NODE_ENV !== "production" ? PiePlot.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  className: PropTypes.string,
  /**
   * Callback fired when a pie item is clicked.
   * @param {React.MouseEvent<SVGPathElement, MouseEvent>} event The event source of the callback.
   * @param {PieItemIdentifier} pieItemIdentifier The pie item identifier.
   * @param {DefaultizedPieValueType} item The pie item.
   */
  onItemClick: PropTypes.func,
  /**
   * If `true`, animations are skipped.
   * @default false
   */
  skipAnimation: PropTypes.bool,
  /**
   * The props used for each component slot.
   * @default {}
   */
  slotProps: PropTypes.object,
  /**
   * Overridable component slots.
   * @default {}
   */
  slots: PropTypes.object
} : void 0;
export { PiePlot };
