// @mui/x-charts/ChartsLayerContainer/ChartsLayerContainer.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartsLayerContainer = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _clsx = _interopRequireDefault(require("clsx"));
var _warning = require("@mui/x-internals/warning");
var _styles = require("@mui/material/styles");
var _useForkRef = _interopRequireDefault(require("@mui/utils/useForkRef"));
var _useId = _interopRequireDefault(require("@mui/utils/useId"));
var _chartsSurfaceClasses = require("../ChartsSurface/chartsSurfaceClasses");
var _useChartDimensions = require("../internals/plugins/corePlugins/useChartDimensions");
var _useChartKeyboardNavigation = require("../internals/plugins/featurePlugins/useChartKeyboardNavigation");
var _ChartsProvider = require("../context/ChartsProvider");
var _hooks = require("../hooks");
var _useRegisterPointerInteractions = require("../internals/plugins/featurePlugins/shared/useRegisterPointerInteractions");
var _ChartsSurface = require("../ChartsSurface");
var _ChartsAccessibilityProxy = require("../internals/components/ChartsAccessibilityProxy");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["children", "title", "desc", "className"]; // eslint-disable-next-line import/no-cycle
const ChartsLayerContainerDiv = (0, _styles.styled)('div', {
  name: 'MuiChartsLayerContainer',
  slot: 'Root'
})(({
  ownerState
}) => ({
  width: ownerState.width ?? '100%',
  height: ownerState.height ?? '100%',
  display: 'flex',
  position: 'relative',
  flexDirection: 'column',
  alignItems: 'center',
  justifyContent: 'center',
  touchAction: 'pan-y',
  userSelect: 'none',
  gridArea: 'chart',
  '&:focus': {
    outline: 'none' // By default, don't show focus outline
  }
}));
/**
 * A component that contains the chart layers, such as `<ChartsSvgLayer>`, and `<ChartsWebGLLayer>`.
 * It is responsible for positioning itself and providing the dimensions and interaction context to its children layers.
 */
const ChartsLayerContainer = exports.ChartsLayerContainer = /*#__PURE__*/React.forwardRef(function ChartsLayerContainer(inProps, ref) {
  const {
    store,
    instance
  } = (0, _ChartsProvider.useChartsContext)();
  const propsWidth = store.use(_useChartDimensions.selectorChartPropsWidth);
  const propsHeight = store.use(_useChartDimensions.selectorChartPropsHeight);
  const isKeyboardNavigationEnabled = store.use(_useChartKeyboardNavigation.selectorChartsIsKeyboardNavigationEnabled);
  (0, _useRegisterPointerInteractions.useRegisterPointerInteractions)();
  const themeProps = (0, _styles.useThemeProps)({
    props: inProps,
    name: 'MuiChartsLayerContainer'
  });
  const {
      children,
      title,
      desc,
      className
    } = themeProps,
    other = (0, _objectWithoutPropertiesLoose2.default)(themeProps, _excluded);
  const classes = (0, _chartsSurfaceClasses.useUtilityClasses)();
  const chartsLayerContainerRef = (0, _hooks.useChartsLayerContainerRef)();
  const handleRef = (0, _useForkRef.default)(chartsLayerContainerRef, ref);
  const descId = (0, _useId.default)();
  if (process.env.NODE_ENV !== 'production') {
    React.Children.forEach(children, child => {
      if (typeof child === 'object' && child != null && 'type' in child && child.type === _ChartsSurface.ChartsSurface) {
        (0, _warning.warnOnce)('MUI X Charts: ChartsSurface should not be used inside ChartsLayerContainer. Render a ChartsSvgLayer instead.', 'error');
      }
    });
  }
  return (
    /*#__PURE__*/
    // `role="none"` is an alias for `role="presentation"`, but aria-query treats them differently.
    // See https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/issues/1090
    // eslint-disable-next-line jsx-a11y/role-supports-aria-props
    (0, _jsxRuntime.jsxs)(ChartsLayerContainerDiv, (0, _extends2.default)({
      ref: handleRef,
      ownerState: {
        width: propsWidth,
        height: propsHeight
      },
      role: "none",
      "aria-label": title,
      "aria-describedby": desc ? descId : undefined,
      className: (0, _clsx.default)(classes.root, className)
    }, other, {
      onPointerEnter: event => {
        other.onPointerEnter?.(event);
        instance.handlePointerEnter?.(event);
      },
      onPointerLeave: event => {
        other.onPointerLeave?.(event);
        instance.handlePointerLeave?.(event);
      },
      onClick: event => {
        other.onClick?.(event);
        instance.handleClick?.(event);
      },
      children: [isKeyboardNavigationEnabled && /*#__PURE__*/(0, _jsxRuntime.jsx)(_ChartsAccessibilityProxy.ChartsAccessibilityProxy, {}), desc && /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
        id: descId,
        style: {
          display: 'none'
        },
        children: desc
      }), children]
    }))
  );
});
if (process.env.NODE_ENV !== "production") ChartsLayerContainer.displayName = "ChartsLayerContainer";
process.env.NODE_ENV !== "production" ? ChartsLayerContainer.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  /**
   * The description of the chart.
   * Used to provide an accessible description for the chart.
   */
  desc: _propTypes.default.string,
  sx: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object, _propTypes.default.bool])), _propTypes.default.func, _propTypes.default.object]),
  /**
   * The title of the chart.
   * Used to provide an accessible label for the chart.
   */
  title: _propTypes.default.string
} : void 0;
