// @mui/x-charts/internals/plugins/featurePlugins/useGeoProjection/useGeoProjection.selectors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.selectorGeoTooltipPosition = exports.selectorChartRawScale = exports.selectorChartRawProjection = exports.selectorChartProjectionFactory = exports.selectorChartProjection = exports.selectorChartGeoProjectionState = exports.selectorChartGeoPath = exports.selectorChartGeoFeatureKey = exports.selectorChartGeoFeatureIndexesByName = exports.selectorChartGeoData = void 0;
var _store = require("@mui/x-internals/store");
var _d3Geo = require("@mui/x-charts-vendor/d3-geo");
var _useChartDimensions = require("../../corePlugins/useChartDimensions/useChartDimensions.selectors");
const isConicProjection = projection => {
  return 'parallels' in projection && typeof projection.parallels === 'function';
};
const selectorChartGeoProjectionState = state => state.geoProjection;
exports.selectorChartGeoProjectionState = selectorChartGeoProjectionState;
const selectorChartGeoData = exports.selectorChartGeoData = (0, _store.createSelector)(selectorChartGeoProjectionState, geoProjection => geoProjection?.geoData ?? null);
const selectorChartGeoFeatureKey = exports.selectorChartGeoFeatureKey = (0, _store.createSelector)(selectorChartGeoProjectionState, geoProjection => geoProjection?.geoFeatureKey ?? 'name');
const selectorChartRawProjection = exports.selectorChartRawProjection = (0, _store.createSelector)(selectorChartGeoProjectionState, geoProjection => geoProjection?.projection ?? null);
const selectorChartProjectionFactory = exports.selectorChartProjectionFactory = (0, _store.createSelector)(selectorChartGeoProjectionState, geoProjection => geoProjection?.factories ?? null);
const selectorChartRawScale = exports.selectorChartRawScale = (0, _store.createSelector)(selectorChartGeoProjectionState, geoProjection => geoProjection?.scale ?? null);
const selectorChartRotate = (0, _store.createSelectorMemoized)(selectorChartGeoProjectionState, geoProjection => geoProjection?.rotate ?? null);
const selectorChartTranslate = (0, _store.createSelectorMemoized)(selectorChartGeoProjectionState, geoProjection => geoProjection?.translate ?? null);
const selectorChartParallels = (0, _store.createSelectorMemoized)(selectorChartGeoProjectionState, selectorChartRotate, (geoProjection, rotate) => geoProjection?.parallels ?? (rotate ? [rotate[1] - 15, rotate[1] + 15] : [30, 30]));
/**
 * Map a feature's `properties.name` to its index in `geoData.features`,
 * for fast lookup by name when joining series rows to features.
 *
 * Features without a string `properties.name` are skipped; on duplicates,
 * the first occurrence wins.
 */
const selectorChartGeoFeatureIndexesByName = exports.selectorChartGeoFeatureIndexesByName = (0, _store.createSelectorMemoized)(selectorChartGeoData, selectorChartGeoFeatureKey, (geoData, geoFeatureKey) => {
  const map = new Map();
  if (!geoData) {
    return map;
  }
  geoData.features.forEach((feature, index) => {
    const name = typeof geoFeatureKey === 'function' ? geoFeatureKey(feature) : feature.properties?.[geoFeatureKey];
    if (typeof name !== 'string') {
      return;
    }
    if (map.has(name)) {
      map.get(name).push(index);
      return;
    }
    map.set(name, [index]);
  });
  return map;
});

/**
 * Resolves the raw `projection` input into a ready-to-use `GeoProjection` instance
 * fitted to the chart's drawing area.
 *
 * - String inputs (e.g. `'mercator'`) are mapped to the matching d3-geo factory.
 * - `GeoProjection` instances are used as-is, then fitted.
 * - Returns `null` when no projection is registered or the name is unknown.
 */
const selectorChartProjection = exports.selectorChartProjection = (0, _store.createSelectorMemoized)(selectorChartRawProjection, selectorChartProjectionFactory, selectorChartGeoData, selectorChartParallels, selectorChartRotate, selectorChartTranslate, selectorChartRawScale, _useChartDimensions.selectorChartDrawingArea, (projectionInput, projectionFactory, geoData, parallels, rotate, translate, scale, drawingArea) => {
  if (!projectionInput || !projectionFactory) {
    return null;
  }
  let projection;
  if (typeof projectionInput === 'string') {
    const factory = projectionFactory[projectionInput];
    if (!factory) {
      if (process.env.NODE_ENV !== 'production') {
        console.error(`MUI X Charts: Unknown projection name '${projectionInput}'. ` + `Expected one of: ${Object.keys(projectionFactory).join(', ')}.`);
      }
      return null;
    }
    projection = factory();
    if (isConicProjection(projection)) {
      projection.parallels(parallels);
    }
  } else {
    projection = projectionInput;
  }
  if (geoData) {
    if (isConicProjection(projection)) {
      if (rotate) {
        projection.rotate?.(rotate);
      }
      if (!scale) {
        const [[x0, y0], [x1, y1]] = (0, _d3Geo.geoPath)(projection).bounds(geoData);
        const currentScale = projection.scale();
        const fitScale = Math.min(currentScale * (drawingArea.width / (x1 - x0)), currentScale * (drawingArea.height / (y1 - y0)));
        projection.scale(fitScale);
      } else {
        projection.scale(scale);
      }
      return projection;
    }
    if (rotate) {
      projection.rotate?.(rotate);
    }
    if (scale) {
      projection.scale(scale);
      projection.clipExtent?.([[drawingArea.left, drawingArea.top], [drawingArea.left + drawingArea.width, drawingArea.top + drawingArea.height]]);
    } else {
      projection.fitExtent?.([[drawingArea.left, drawingArea.top], [drawingArea.left + drawingArea.width, drawingArea.top + drawingArea.height]], geoData);
    }
    if (translate) {
      projection.translate(translate);
    }
  }
  return projection;
});

/**
 * Resolves the raw `projection` input into a ready-to-use `GeoPath` instance
 * fitted to the chart's drawing area.
 */
const selectorChartGeoPath = exports.selectorChartGeoPath = (0, _store.createSelectorMemoized)(selectorChartProjection, projection => {
  if (!projection) {
    return null;
  }
  return (0, _d3Geo.geoPath)(projection);
});
const selectorGeoTooltipPosition = exports.selectorGeoTooltipPosition = (0, _store.createSelectorMemoized)(selectorChartGeoData, selectorChartProjection, selectorChartGeoFeatureIndexesByName, (geoData, projection, featureIndexesByName) => {
  return {
    geoData,
    projection,
    featureIndexesByName
  };
});
