// @mui/x-charts/RadarChart/radarClasses.mjs
import generateUtilityClass from '@mui/utils/generateUtilityClass';
import composeClasses from '@mui/utils/composeClasses';
import generateUtilityClasses from '@mui/utils/generateUtilityClasses';
export function getRadarUtilityClass(slot) {
  return generateUtilityClass('MuiRadarChart', slot);
}
export const radarClasses = generateUtilityClasses('MuiRadarChart', ['axisRoot', 'axisLine', 'axisLabel', 'gridRadial', 'gridDivider', 'gridStripe', 'seriesRoot', 'seriesArea', 'seriesMark', 'axisHighlightRoot', 'axisHighlightLine', 'axisHighlightDot']);
export const useUtilityClasses = classes => {
  const slots = {
    axisRoot: ['axisRoot'],
    axisLine: ['axisLine'],
    axisLabel: ['axisLabel'],
    gridRadial: ['gridRadial'],
    gridDivider: ['gridDivider'],
    gridStripe: ['gridStripe'],
    seriesRoot: ['seriesRoot'],
    seriesArea: ['seriesArea'],
    seriesMark: ['seriesMark'],
    axisHighlightRoot: ['axisHighlightRoot'],
    axisHighlightLine: ['axisHighlightLine'],
    axisHighlightDot: ['axisHighlightDot']
  };
  return composeClasses(slots, getRadarUtilityClass, classes);
};
