// @mui/x-charts/RadarChart/radarClasses.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRadarUtilityClass = getRadarUtilityClass;
exports.useUtilityClasses = exports.radarClasses = void 0;
var _generateUtilityClass = _interopRequireDefault(require("@mui/utils/generateUtilityClass"));
var _composeClasses = _interopRequireDefault(require("@mui/utils/composeClasses"));
var _generateUtilityClasses = _interopRequireDefault(require("@mui/utils/generateUtilityClasses"));
function getRadarUtilityClass(slot) {
  return (0, _generateUtilityClass.default)('MuiRadarChart', slot);
}
const radarClasses = exports.radarClasses = (0, _generateUtilityClasses.default)('MuiRadarChart', ['axisRoot', 'axisLine', 'axisLabel', 'gridRadial', 'gridDivider', 'gridStripe', 'seriesRoot', 'seriesArea', 'seriesMark', 'axisHighlightRoot', 'axisHighlightLine', 'axisHighlightDot']);
const useUtilityClasses = classes => {
  const slots = {
    axisRoot: ['axisRoot'],
    axisLine: ['axisLine'],
    axisLabel: ['axisLabel'],
    gridRadial: ['gridRadial'],
    gridDivider: ['gridDivider'],
    gridStripe: ['gridStripe'],
    seriesRoot: ['seriesRoot'],
    seriesArea: ['seriesArea'],
    seriesMark: ['seriesMark'],
    axisHighlightRoot: ['axisHighlightRoot'],
    axisHighlightLine: ['axisHighlightLine'],
    axisHighlightDot: ['axisHighlightDot']
  };
  return (0, _composeClasses.default)(slots, getRadarUtilityClass, classes);
};
exports.useUtilityClasses = useUtilityClasses;
