// @mui/x-charts/ChartsRadialAxisHighlight/ChartsRadialAxisHighlight.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartsRadialAxisHighlight = ChartsRadialAxisHighlight;
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _composeClasses = _interopRequireDefault(require("@mui/utils/composeClasses"));
var _chartsRadialAxisHighlightClasses = require("./chartsRadialAxisHighlightClasses");
var _ChartsRotationAxisHighlight = _interopRequireDefault(require("./ChartsRotationAxisHighlight"));
var _ChartsRadiusAxisHighlight = _interopRequireDefault(require("./ChartsRadiusAxisHighlight"));
var _jsxRuntime = require("react/jsx-runtime");
const useUtilityClasses = () => {
  const slots = {
    root: ['root']
  };
  return (0, _composeClasses.default)(slots, _chartsRadialAxisHighlightClasses.getRadialAxisHighlightUtilityClass);
};

/**
 * Demos:
 *
 * - [Radial Lines](https://mui.com/x/react-charts/radial-lines/)
 *
 * API:
 *
 * - [ChartsRadialAxisHighlight API](https://mui.com/x/api/charts/charts-radial-axis-highlight/)
 */
function ChartsRadialAxisHighlight(props) {
  const {
    rotation: rotationHighlight,
    radius: radiusHighlight
  } = props;
  const classes = useUtilityClasses();
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(React.Fragment, {
    children: [rotationHighlight && rotationHighlight !== 'none' && /*#__PURE__*/(0, _jsxRuntime.jsx)(_ChartsRotationAxisHighlight.default, {
      type: rotationHighlight,
      classes: classes
    }), radiusHighlight && radiusHighlight !== 'none' && /*#__PURE__*/(0, _jsxRuntime.jsx)(_ChartsRadiusAxisHighlight.default, {
      type: radiusHighlight,
      classes: classes
    })]
  });
}
process.env.NODE_ENV !== "production" ? ChartsRadialAxisHighlight.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  radius: _propTypes.default.oneOf(['band', 'line', 'none']),
  rotation: _propTypes.default.oneOf(['band', 'line', 'none'])
} : void 0;
