// @mui/x-charts/ChartsRadialAxisHighlight/ChartsRadialAxisHighlightPath.mjs
'use client';

import _extends from "@babel/runtime/helpers/esm/extends";
import { styled } from '@mui/material/styles';
export const ChartsRadialAxisHighlightPath = styled('path', {
  name: 'MuiChartsRadialAxisHighlight',
  slot: 'Root'
})(({
  theme
}) => ({
  pointerEvents: 'none',
  variants: [{
    props: {
      axisHighlight: 'band'
    },
    style: _extends({
      fill: 'white',
      fillOpacity: 0.1
    }, theme.applyStyles('light', {
      fill: 'gray'
    }))
  }, {
    props: {
      axisHighlight: 'line'
    },
    style: _extends({
      fill: 'none',
      strokeDasharray: '5 2',
      stroke: '#ffffff'
    }, theme.applyStyles('light', {
      stroke: '#000000'
    }))
  }]
}));
export const ChartsRadialAxisHighlightCircle = styled('circle', {
  name: 'MuiChartsRadialAxisHighlight',
  slot: 'Root'
})(({
  theme
}) => ({
  pointerEvents: 'none',
  variants: [{
    props: {
      axisHighlight: 'band'
    },
    style: _extends({
      fill: 'white',
      fillOpacity: 0.1
    }, theme.applyStyles('light', {
      fill: 'gray'
    }))
  }, {
    props: {
      axisHighlight: 'line'
    },
    style: _extends({
      fill: 'none',
      strokeDasharray: '5 2',
      stroke: '#ffffff'
    }, theme.applyStyles('light', {
      stroke: '#000000'
    }))
  }]
}));
