// @mui/x-charts/ChartsRadiusAxis/ChartsRadiusAxis.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartsRadiusAxis = ChartsRadiusAxis;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _clsx = _interopRequireDefault(require("clsx"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _styles = require("@mui/material/styles");
var _useTicks = require("../hooks/useTicks");
var _useAxis = require("../hooks/useAxis");
var _ChartsProvider = require("../context/ChartsProvider");
var _useChartPolarAxis = require("../internals/plugins/featurePlugins/useChartPolarAxis");
var _chartsRadiusAxisClasses = require("./chartsRadiusAxisClasses");
var _createGetLabelTextAnchors = require("./createGetLabelTextAnchors");
var _getLabelTransform = require("./getLabelTransform");
var _jsxRuntime = require("react/jsx-runtime");
/* Gap between a tick and its label. */
const TICK_LABEL_GAP = 3;
const getLabelTextAnchors = (0, _createGetLabelTextAnchors.createGetLabelTextAnchors)(_getLabelTransform.getLabelTransform);

/**
 * Get the angle to use to display the radius axis.
 * @param position The position props
 * @param rotationAxis The default rotation axis
 * @returns the angle in radians to use.
 */
function getAxisAngleInRadians(position, rotationAxis) {
  if (position === 'start') {
    return rotationAxis?.scale.range()[0] ?? 0;
  }
  if (position === 'end') {
    return rotationAxis?.scale.range()[1] ?? 0;
  }
  if (typeof position === 'number') {
    return position * Math.PI / 180;
  }
  return 0;
}
function ChartsRadiusAxis(props) {
  const radiusAxis = (0, _useAxis.useRadiusAxis)(props.axisId);
  const settings = (0, _extends2.default)({}, radiusAxis, props);
  const {
    position = 'start',
    disableLine,
    disableTicks,
    disableTickLabel,
    tickLabelPosition: tickLabelPositionProp = 'auto',
    tickPosition = position === 'start' ? 'before' : 'after',
    tickSize = 6,
    className,
    classes: classesProp
  } = settings;
  let tickLabelPosition = tickLabelPositionProp;
  if (tickLabelPosition === 'auto') {
    tickLabelPosition = position === 'start' ? 'before' : 'after';
  }
  const isCentered = tickLabelPosition === 'center';
  const classes = (0, _chartsRadiusAxisClasses.useUtilityClasses)({
    classes: classesProp,
    isCentered
  });
  const theme = (0, _styles.useTheme)();
  const {
    store
  } = (0, _ChartsProvider.useChartsContext)();
  const {
    cx,
    cy
  } = store.use(_useChartPolarAxis.selectorChartPolarCenter);
  const rotationAxis = (0, _useAxis.useRotationAxis)();
  const ticks = (0, _useTicks.useTicks)({
    scale: radiusAxis.scale,
    tickNumber: radiusAxis?.tickNumber ?? 5,
    tickInterval: radiusAxis?.tickInterval,
    tickSpacing: radiusAxis?.tickSpacing,
    valueFormatter: radiusAxis?.valueFormatter,
    direction: 'radius'
  });
  if (!radiusAxis || settings.position === 'none') {
    return null;
  }
  const angle = getAxisAngleInRadians(position, rotationAxis);
  // Convert "0 = up" convention to SVG math angle (0 = right, clockwise y-down).
  const dx = Math.sin(angle);
  const dy = -Math.cos(angle);
  // Perpendicular offset for tick marks (rotated 90° clockwise).
  const px = -dy;
  const py = dx;
  const [innerRadius, outerRadius] = radiusAxis.scale.range();
  const stroke = (theme.vars ?? theme).palette.text.primary;
  const tickDx = (tickPosition === 'after' ? 1 : -1) * px * tickSize;
  const tickDy = (tickPosition === 'after' ? 1 : -1) * py * tickSize;
  const tickLabelGap = isCentered ? 0 : TICK_LABEL_GAP;
  const tickLabelGapDx = (tickLabelPosition === 'after' ? 1 : -1) * px * tickLabelGap;
  const tickLabelGapDy = (tickLabelPosition === 'after' ? 1 : -1) * py * tickLabelGap;
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("g", {
    className: (0, _clsx.default)(classes.root, className),
    children: [!disableLine && /*#__PURE__*/(0, _jsxRuntime.jsx)("line", {
      x1: cx + dx * innerRadius,
      y1: cy + dy * innerRadius,
      x2: cx + dx * outerRadius,
      y2: cy + dy * outerRadius,
      stroke: stroke,
      className: classes.line
    }), ticks.map(({
      offset: radius,
      labelOffset,
      formattedValue
    }, index) => {
      if (!formattedValue) {
        return null;
      }
      const tickX = cx + dx * radius;
      const tickY = cy + dy * radius;

      // Compute the label position.
      let labelX = cx + dx * (radius + labelOffset);
      let labelY = cy + dy * (radius + labelOffset);
      if (tickLabelGap !== 0) {
        labelX += tickLabelGapDx;
        labelY += tickLabelGapDy;
      }
      if (!isCentered && tickLabelPosition === tickPosition && !disableTicks) {
        // Add the size of the tick if they are in the same direction.
        labelX += tickDx;
        labelY += tickDy;
      }
      return /*#__PURE__*/(0, _jsxRuntime.jsxs)("g", {
        className: classes.tickContainer,
        children: [!disableTicks && /*#__PURE__*/(0, _jsxRuntime.jsx)("line", {
          x1: tickX,
          y1: tickY,
          x2: tickX + tickDx,
          y2: tickY + tickDy,
          stroke: stroke,
          className: classes.tick
        }), !disableTickLabel && /*#__PURE__*/(0, _jsxRuntime.jsx)("text", (0, _extends2.default)({
          x: labelX,
          y: labelY,
          fill: stroke,
          fontSize: 12,
          className: classes.tickLabel,
          pointerEvents: "none"
        }, getLabelTextAnchors(dx, dy, tickLabelPosition), {
          children: formattedValue
        }))]
      }, index);
    })]
  });
}
process.env.NODE_ENV !== "production" ? ChartsRadiusAxis.propTypes /* remove-proptypes */ = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  axis: _propTypes.default.oneOf(['radius']),
  /**
   * The id of the axis to render.
   * If undefined, it will be the first defined axis.
   */
  axisId: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  /**
   * Override or extend the styles applied to the component.
   */
  classes: _propTypes.default.object,
  /**
   * A CSS class name applied to the root element.
   */
  className: _propTypes.default.string,
  /**
   * If true, the axis line is disabled.
   * @default false
   */
  disableLine: _propTypes.default.bool,
  /**
   * If true, the tick labels are not rendered.
   * @default false
   */
  disableTickLabel: _propTypes.default.bool,
  /**
   * If true, the ticks are disabled.
   * @default false
   */
  disableTicks: _propTypes.default.bool,
  /**
   * The label of the axis.
   */
  label: _propTypes.default.string,
  /**
   * The style applied to the axis label.
   */
  labelStyle: _propTypes.default.object,
  /**
   * The maximal radius.
   * Can be a number (in pixels), a pixel string (for example `'80px'`), or a percentage string
   * (for example `'80%'`) relative to the available radius (half the smallest side of the drawing area).
   * @default '100%'
   */
  maxRadius: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  /**
   * The minimal radius.
   * Can be a number (in pixels), a pixel string (for example `'20px'`), or a percentage string
   * (for example `'20%'`) relative to the available radius (half the smallest side of the drawing area).
   * @default 0
   */
  minRadius: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  /**
   * The position of the axis in polar coordinates.
   * It can be 'start', 'end', or a specific angle in degrees.
   * @default 'start'
   */
  position: _propTypes.default.oneOfType([_propTypes.default.oneOf(['end', 'none', 'start']), _propTypes.default.number]),
  sx: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object, _propTypes.default.bool])), _propTypes.default.func, _propTypes.default.object]),
  /**
   * Defines which ticks are displayed.
   * Its value can be:
   * - 'auto' In such case the ticks are computed based on axis scale and other parameters.
   * - a filtering function of the form `(value, index) => boolean` which is available only if the axis has "point" scale.
   * - an array containing the values where ticks should be displayed.
   * @see See {@link https://mui.com/x/react-charts/axis/#fixed-tick-positions}
   * @default 'auto'
   */
  tickInterval: _propTypes.default.oneOfType([_propTypes.default.oneOf(['auto']), _propTypes.default.array, _propTypes.default.func]),
  /**
   * Defines which ticks get its label displayed. Its value can be:
   * - 'auto' In such case, labels are displayed if they do not overlap with the previous one.
   * - a filtering function of the form (value, index) => boolean. Warning: the index is tick index, not data ones.
   * @default 'auto'
   */
  tickLabelInterval: _propTypes.default.oneOfType([_propTypes.default.oneOf(['auto']), _propTypes.default.func]),
  /**
   * The placement of ticks label. Can be the middle of the band, or the tick position.
   * Only used if scale is 'band'.
   * @default 'middle'
   */
  tickLabelPlacement: _propTypes.default.oneOf(['middle', 'tick']),
  /**
   * Set the position of the tick labels relative to the axis line.
   * The before/after is defined based on clockwise direction.
   * Using `'auto'` sets it to `'before'` if position is `'start'` and `'after'` otherwise.
   * @default 'auto'
   */
  tickLabelPosition: _propTypes.default.oneOf(['after', 'auto', 'before', 'center']),
  /**
   * The style applied to ticks text.
   */
  tickLabelStyle: _propTypes.default.object,
  /**
   * Maximal step between two ticks.
   * When using time data, the value is assumed to be in ms.
   * Not supported by categorical axis (band, points).
   */
  tickMaxStep: _propTypes.default.number,
  /**
   * Minimal step between two ticks.
   * When using time data, the value is assumed to be in ms.
   * Not supported by categorical axis (band, points).
   */
  tickMinStep: _propTypes.default.number,
  /**
   * The number of ticks. This number is not guaranteed.
   * Not supported by categorical axis (band, points).
   */
  tickNumber: _propTypes.default.number,
  /**
   * The placement of ticks in regard to the band interval.
   * Only used if scale is 'band'.
   * @default 'extremities'
   */
  tickPlacement: _propTypes.default.oneOf(['end', 'extremities', 'middle', 'start']),
  /**
   * Set the position of the tick relative to the axis line.
   * The before/after is defined based on clockwise direction.
   * @default position === 'start' ? 'before' : 'after'
   */
  tickPosition: _propTypes.default.oneOf(['after', 'before']),
  /**
   * The size of the ticks.
   * @default 6
   */
  tickSize: _propTypes.default.number,
  /**
   * The minimum space between ticks when using an ordinal scale. It defines the minimum distance in pixels between two ticks.
   * @default 0
   */
  tickSpacing: _propTypes.default.number
} : void 0;
