// @mui/x-charts/LineChart/seriesConfig/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.lineSeriesConfig = void 0;
var _extremums = require("./extremums");
var _seriesProcessor = _interopRequireDefault(require("./seriesProcessor"));
var _getColor = _interopRequireDefault(require("./getColor"));
var _legend = _interopRequireDefault(require("./legend"));
var _tooltip = _interopRequireWildcard(require("./tooltip"));
var _getSeriesWithDefaultValues = _interopRequireDefault(require("./getSeriesWithDefaultValues"));
var _tooltipPosition = _interopRequireDefault(require("./tooltipPosition"));
var _getItemAtPosition = _interopRequireDefault(require("./getItemAtPosition"));
var _keyboardFocusHandler = _interopRequireDefault(require("./keyboardFocusHandler"));
var _identifierSerializer = require("../../internals/identifierSerializer");
var _identifierCleaner = require("../../internals/identifierCleaner");
var _useChartHighlight = require("../../internals/plugins/featurePlugins/useChartHighlight");
var _descriptionGetter = _interopRequireDefault(require("./descriptionGetter"));
const lineSeriesConfig = exports.lineSeriesConfig = {
  colorProcessor: _getColor.default,
  seriesProcessor: _seriesProcessor.default,
  legendGetter: _legend.default,
  tooltipGetter: _tooltip.default,
  tooltipItemPositionGetter: _tooltipPosition.default,
  axisTooltipGetter: _tooltip.axisTooltipGetter,
  xExtremumGetter: _extremums.getExtremumX,
  yExtremumGetter: _extremums.getExtremumY,
  getSeriesWithDefaultValues: _getSeriesWithDefaultValues.default,
  getItemAtPosition: _getItemAtPosition.default,
  keyboardFocusHandler: _keyboardFocusHandler.default,
  identifierSerializer: _identifierSerializer.identifierSerializerSeriesIdDataIndex,
  identifierCleaner: _identifierCleaner.identifierCleanerSeriesIdDataIndex,
  descriptionGetter: _descriptionGetter.default,
  isHighlightedCreator: _useChartHighlight.createIsHighlighted,
  isFadedCreator: _useChartHighlight.createIsFaded
};
