// @mui/x-charts/LineChart/seriesConfig/index.mjs
import { getExtremumX, getExtremumY } from "./extremums.mjs";
import seriesProcessor from "./seriesProcessor.mjs";
import getColor from "./getColor.mjs";
import legendGetter from "./legend.mjs";
import tooltipGetter, { axisTooltipGetter } from "./tooltip.mjs";
import getSeriesWithDefaultValues from "./getSeriesWithDefaultValues.mjs";
import tooltipItemPositionGetter from "./tooltipPosition.mjs";
import getItemAtPosition from "./getItemAtPosition.mjs";
import keyboardFocusHandler from "./keyboardFocusHandler.mjs";
import { identifierSerializerSeriesIdDataIndex } from "../../internals/identifierSerializer.mjs";
import { identifierCleanerSeriesIdDataIndex } from "../../internals/identifierCleaner.mjs";
import { createIsHighlighted, createIsFaded } from "../../internals/plugins/featurePlugins/useChartHighlight/index.mjs";
import descriptionGetter from "./descriptionGetter.mjs";
export const lineSeriesConfig = {
  colorProcessor: getColor,
  seriesProcessor,
  legendGetter,
  tooltipGetter,
  tooltipItemPositionGetter,
  axisTooltipGetter,
  xExtremumGetter: getExtremumX,
  yExtremumGetter: getExtremumY,
  getSeriesWithDefaultValues,
  getItemAtPosition,
  keyboardFocusHandler,
  identifierSerializer: identifierSerializerSeriesIdDataIndex,
  identifierCleaner: identifierCleanerSeriesIdDataIndex,
  descriptionGetter,
  isHighlightedCreator: createIsHighlighted,
  isFadedCreator: createIsFaded
};
