// @mui/x-scheduler/internals/components/more-events-popover/MoreEventsPopover.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MoreEventsPopoverContext = void 0;
exports.MoreEventsPopoverProvider = MoreEventsPopoverProvider;
exports.MoreEventsPopoverTrigger = MoreEventsPopoverTrigger;
exports.default = MoreEventsPopoverContent;
exports.useMoreEventsPopoverContext = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _styles = require("@mui/material/styles");
var _Popover = _interopRequireDefault(require("@mui/material/Popover"));
var _Typography = _interopRequireDefault(require("@mui/material/Typography"));
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _EventItem = require("../event/event-item/EventItem");
var _createModal = require("../create-modal");
var _eventUtils = require("../../utils/event-utils");
var _dateUtils = require("../../utils/date-utils");
var _eventDialog = require("../event-dialog");
var _EventCalendarStyledContext = require("../../../event-calendar/EventCalendarStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["occurrences", "day"];
const MoreEventsPopoverHeader = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MoreEventsPopoverHeader'
})(({
  theme
}) => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'space-between',
  padding: theme.spacing(1),
  borderBottom: `1px solid ${(theme.vars || theme).palette.divider}`
}));
const MoreEventsPopoverTitle = (0, _styles.styled)(_Typography.default, {
  name: 'MuiEventCalendar',
  slot: 'MoreEventsPopoverTitle'
})(({
  theme
}) => ({
  fontSize: theme.typography.body2.fontSize,
  fontWeight: theme.typography.fontWeightMedium,
  color: (theme.vars || theme).palette.text.primary,
  lineHeight: 1.5,
  margin: 0
}));
const MoreEventsPopoverBody = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MoreEventsPopoverBody'
})(({
  theme
}) => ({
  padding: theme.spacing(1),
  display: 'flex',
  flexDirection: 'column',
  gap: theme.spacing(1),
  width: 'fit-content',
  minWidth: 200
}));
const MoreEventsPopover = (0, _createModal.createModal)({
  contextName: 'MoreEventsPopoverContext'
});
const MoreEventsPopoverContext = exports.MoreEventsPopoverContext = MoreEventsPopover.Context;
const useMoreEventsPopoverContext = exports.useMoreEventsPopoverContext = MoreEventsPopover.useContext;
function MoreEventsPopoverContent(props) {
  const {
    open,
    anchor,
    occurrences,
    day,
    onClose
  } = props;

  // Context hooks
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const {
    schedulerId,
    classes
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const {
    subscribeCloseHandler
  } = (0, _eventDialog.useEventDialogContext)();
  React.useEffect(() => {
    subscribeCloseHandler(() => {
      onClose();
    });
  }, [subscribeCloseHandler, onClose]);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_Popover.default, {
    className: classes.moreEventsPopover,
    open: open,
    anchorEl: anchor,
    onClose: onClose,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(MoreEventsPopoverHeader, {
      className: classes.moreEventsPopoverHeader,
      id: `${schedulerId}-PopoverHeader-${day.key}`,
      "aria-label": `${(0, _dateUtils.formatWeekDayMonthAndDayOfMonth)(day.value, adapter)}`,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(MoreEventsPopoverTitle, {
        className: classes.moreEventsPopoverTitle,
        children: (0, _dateUtils.formatWeekDayMonthAndDayOfMonth)(day.value, adapter)
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(MoreEventsPopoverBody, {
      className: classes.moreEventsPopoverBody,
      children: occurrences.map(occurrence => /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventDialog.EventDialogTrigger, {
        occurrence: occurrence,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_EventItem.EventItem, {
          variant: (0, _eventUtils.isOccurrenceAllDayOrMultipleDay)(occurrence, adapter) ? 'filled' : 'compact',
          occurrence: occurrence,
          date: day,
          ariaLabelledBy: `${schedulerId}-PopoverHeader-${day.key}`
        })
      }, occurrence.key))
    })]
  });
}
function MoreEventsPopoverProvider(props) {
  const {
    children
  } = props;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(MoreEventsPopover.Provider, {
    render: ({
      isOpen,
      anchorRef,
      data,
      onClose
    }) => /*#__PURE__*/(0, _jsxRuntime.jsx)(MoreEventsPopoverContent, {
      open: isOpen,
      anchor: anchorRef.current,
      occurrences: data.occurrences,
      count: data.count,
      day: data.day,
      onClose: onClose
    }),
    children: children
  });
}
function MoreEventsPopoverTrigger(props) {
  const {
      occurrences,
      day
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const ref = React.useRef(null);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(MoreEventsPopover.Trigger, (0, _extends2.default)({
    ref: ref,
    data: {
      occurrences,
      count: occurrences.length,
      day
    }
  }, other));
}
