// @mui/x-scheduler-internals/calendar-grid/day-row/CalendarGridDayRow.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarGridDayRow = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _useRenderElement = require("../../base-ui-copy/utils/useRenderElement");
var _CompositeList = require("../../base-ui-copy/composite/list/CompositeList");
var _CalendarGridDayRowContext = require("./CalendarGridDayRowContext");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className", "render", "style", "start", "end", "rowIndex"];
const CalendarGridDayRow = exports.CalendarGridDayRow = /*#__PURE__*/React.forwardRef(function CalendarGridDayRow(componentProps, forwardedRef) {
  const {
      // Rendering props

      // Internal props
      start,
      end,
      rowIndex = 0
      // Props forwarded to the DOM element
    } = componentProps,
    elementProps = (0, _objectWithoutPropertiesLoose2.default)(componentProps, _excluded);
  const cellsRefs = React.useRef([]);
  const contextValue = React.useMemo(() => ({
    start,
    end,
    rowIndex
  }), [start, end, rowIndex]);
  const element = (0, _useRenderElement.useRenderElement)('div', componentProps, {
    ref: [forwardedRef],
    props: [{
      role: 'row',
      'aria-rowindex': rowIndex + 2
    }, elementProps]
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_CompositeList.CompositeList, {
    elementsRef: cellsRefs,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_CalendarGridDayRowContext.CalendarGridDayRowContext.Provider, {
      value: contextValue,
      children: element
    })
  });
});
if (process.env.NODE_ENV !== "production") CalendarGridDayRow.displayName = "CalendarGridDayRow";
