// @mui/x-scheduler/internals/components/error-container/ErrorContainer.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ErrorContainer = ErrorContainer;
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _styles = require("@mui/material/styles");
var _store = require("@base-ui/utils/store");
var _Alert = _interopRequireDefault(require("@mui/material/Alert"));
var _Typography = _interopRequireDefault(require("@mui/material/Typography"));
var _useSchedulerStoreContext = require("@mui/x-scheduler-internals/use-scheduler-store-context");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _SharedComponentsStyledContext = require("../SharedComponentsStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const ErrorContainerRoot = (0, _styles.styled)('div', {
  name: 'MuiEventErrorContainer',
  slot: 'Root'
})({
  position: 'absolute',
  zIndex: 1300,
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'flex-end',
  gap: 16,
  bottom: 16,
  right: 16
});
const ErrorAlert = (0, _styles.styled)(_Alert.default, {
  name: 'MuiEventErrorContainer',
  slot: 'Alert'
})({
  maxWidth: 400
});
const ErrorMessage = (0, _styles.styled)(_Typography.default, {
  name: 'MuiEventErrorContainer',
  slot: 'Message'
})({
  wordBreak: 'break-word',
  display: '-webkit-box',
  WebkitLineClamp: 3,
  WebkitBoxOrient: 'vertical',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  overflowWrap: 'break-word'
});
function ErrorContainer(props) {
  const {
    className
  } = props;
  const store = (0, _useSchedulerStoreContext.useSchedulerStoreContext)();
  const {
    classes
  } = (0, _SharedComponentsStyledContext.useSharedComponentsStyledContext)();
  const errors = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.errors);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(ErrorContainerRoot, {
    className: (0, _clsx.default)(classes.errorContainer, className),
    children: errors.map(({
      error,
      key
    }) => /*#__PURE__*/(0, _jsxRuntime.jsx)(ErrorAlert, {
      className: classes.errorAlert,
      severity: "error",
      onClose: () => store.dismissError(key),
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(ErrorMessage, {
        className: classes.errorMessage,
        variant: "body2",
        children: error.message
      })
    }, key))
  });
}
