// @mui/x-scheduler-internals/internals/utils/useDraggableEvent.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDraggableEvent = useDraggableEvent;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _adapter = require("@atlaskit/pragmatic-drag-and-drop/element/adapter");
var _disableNativeDragPreview = require("@atlaskit/pragmatic-drag-and-drop/element/disable-native-drag-preview");
var _store = require("@base-ui/utils/store");
var _useSchedulerStoreContext = require("../../use-scheduler-store-context");
var _schedulerSelectors = require("../../scheduler-selectors");
var _useDragPreview = require("./useDragPreview");
var _useEvent = require("./useEvent");
var _useAdapterContext = require("../../use-adapter-context");
function useDraggableEvent(parameters) {
  const {
    ref,
    start,
    end,
    occurrenceKey,
    eventId,
    renderDragPreview,
    getDragData,
    collectionStart,
    collectionEnd,
    isDraggable = false
  } = parameters;

  // Context hooks
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const store = (0, _useSchedulerStoreContext.useSchedulerStoreContext)();

  // Selector hooks
  const placeholderAction = (0, _store.useStore)(store, _schedulerSelectors.schedulerOccurrencePlaceholderSelectors.actionForOccurrence, occurrenceKey);
  const event = (0, _store.useStore)(store, _schedulerSelectors.schedulerEventSelectors.processedEvent, eventId);

  // Feature hooks
  const {
    state: eventState
  } = (0, _useEvent.useEvent)({
    start,
    end,
    eventId
  });
  const preview = (0, _useDragPreview.useDragPreview)({
    type: 'internal-event',
    data: event,
    renderDragPreview,
    showPreviewOnDragStart: false
  });
  const state = (0, _extends2.default)({}, eventState, {
    dragging: placeholderAction === 'internal-drag',
    resizing: placeholderAction === 'internal-resize'
  });
  React.useEffect(() => {
    if (!isDraggable || !ref.current) {
      return;
    }

    // eslint-disable-next-line consistent-return
    return (0, _adapter.draggable)({
      element: ref.current,
      getInitialData: ({
        input
      }) => getDragData(input),
      onGenerateDragPreview: ({
        nativeSetDragImage
      }) => {
        (0, _disableNativeDragPreview.disableNativeDragPreview)({
          nativeSetDragImage
        });
      },
      onDragStart: ({
        location
      }) => {
        preview.actions.onDragStart(location);
      },
      onDrag: ({
        location
      }) => {
        preview.actions.onDrag(location);
      },
      onDrop: () => {
        store.setOccurrencePlaceholder(null);
        preview.actions.onDrop();
      }
    });
  }, [ref, getDragData, isDraggable, store, preview.actions]);
  const contextValue = React.useMemo(() => ({
    doesEventStartBeforeCollectionStart: adapter.isBefore(start.value, collectionStart),
    doesEventEndAfterCollectionEnd: adapter.isAfter(end.value, collectionEnd)
  }), [adapter, start, end, collectionStart, collectionEnd]);
  return {
    state,
    preview,
    contextValue
  };
}
