// @mui/x-scheduler-internals/internals/utils/useDragPreview.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDragPreview = useDragPreview;
var React = _interopRequireWildcard(require("react"));
var _useSchedulerStoreContext = require("../../use-scheduler-store-context");
var _schedulerSelectors = require("../../scheduler-selectors");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * Returns the drag preview to render when the dragged event is not over a valid drop target.
 */
function useDragPreview(parameters) {
  const {
    renderDragPreview,
    showPreviewOnDragStart,
    data,
    type
  } = parameters;
  const store = (0, _useSchedulerStoreContext.useSchedulerStoreContext)(true);
  const [state, setState] = React.useState({
    isDragging: false,
    dragPosition: null
  });
  const element = React.useMemo(() => {
    if (state.dragPosition == null) {
      return null;
    }
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      style: {
        position: 'fixed',
        top: state.dragPosition.clientY,
        left: state.dragPosition.clientX,
        pointerEvents: 'none',
        zIndex: 9999
      },
      children: renderDragPreview({
        data,
        type
      })
    });
  }, [state.dragPosition, renderDragPreview, data, type]);
  const actions = React.useMemo(() => ({
    onDragStart: location => {
      setState({
        isDragging: true,
        dragPosition: showPreviewOnDragStart ? location.current.input : null
      });
    },
    onDrag: location => {
      let shouldShowPreview = true;
      if (store && type === 'internal-event' && !_schedulerSelectors.schedulerEventSelectors.canDropEventsToTheOutside(store.state)) {
        shouldShowPreview = false;
      } else if (location.current.dropTargets.some(el => el.data.isSchedulerDropTarget)) {
        shouldShowPreview = false;
      }
      setState({
        isDragging: true,
        dragPosition: shouldShowPreview ? location.current.input : null
      });
    },
    onDrop: () => {
      setState({
        isDragging: false,
        dragPosition: null
      });
    }
  }), [store, showPreviewOnDragStart, type]);
  return {
    element,
    state,
    actions
  };
}
