// @mui/x-scheduler-internals/internals/utils/useEventResizeHandler.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useEventResizeHandler = useEventResizeHandler;
var React = _interopRequireWildcard(require("react"));
var _adapter = require("@atlaskit/pragmatic-drag-and-drop/element/adapter");
var _disableNativeDragPreview = require("@atlaskit/pragmatic-drag-and-drop/element/disable-native-drag-preview");
function useEventResizeHandler(parameters) {
  const {
    ref,
    side,
    getDragData,
    contextValue: {
      doesEventStartBeforeCollectionStart,
      doesEventEndAfterCollectionEnd
    }
  } = parameters;
  const enabled = side === 'start' && !doesEventStartBeforeCollectionStart || side === 'end' && !doesEventEndAfterCollectionEnd;
  const state = React.useMemo(() => ({
    start: side === 'start',
    end: side === 'end'
  }), [side]);
  React.useEffect(() => {
    if (!ref.current || !enabled) {
      return undefined;
    }
    return (0, _adapter.draggable)({
      element: ref.current,
      getInitialData: ({
        input
      }) => getDragData(input),
      onGenerateDragPreview: ({
        nativeSetDragImage
      }) => {
        (0, _disableNativeDragPreview.disableNativeDragPreview)({
          nativeSetDragImage
        });
      }
    });
  }, [ref, enabled, side, getDragData]);
  return {
    state,
    enabled
  };
}
