// @base-ui/react/scroll-area/content/ScrollAreaContentDataAttributes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScrollAreaContentDataAttributes = void 0;
let ScrollAreaContentDataAttributes = exports.ScrollAreaContentDataAttributes = /*#__PURE__*/function (ScrollAreaContentDataAttributes) {
  /**
   * Present when the user scrolls inside the scroll area.
   */
  ScrollAreaContentDataAttributes["scrolling"] = "data-scrolling";
  /**
   * Present when the scroll area content is wider than the viewport.
   */
  ScrollAreaContentDataAttributes["hasOverflowX"] = "data-has-overflow-x";
  /**
   * Present when the scroll area content is taller than the viewport.
   */
  ScrollAreaContentDataAttributes["hasOverflowY"] = "data-has-overflow-y";
  /**
   * Present when there is overflow on the horizontal start side.
   */
  ScrollAreaContentDataAttributes["overflowXStart"] = "data-overflow-x-start";
  /**
   * Present when there is overflow on the horizontal end side.
   */
  ScrollAreaContentDataAttributes["overflowXEnd"] = "data-overflow-x-end";
  /**
   * Present when there is overflow on the vertical start side.
   */
  ScrollAreaContentDataAttributes["overflowYStart"] = "data-overflow-y-start";
  /**
   * Present when there is overflow on the vertical end side.
   */
  ScrollAreaContentDataAttributes["overflowYEnd"] = "data-overflow-y-end";
  return ScrollAreaContentDataAttributes;
}({});
