// @mui/x-charts/ChartsRadialGrid/styledComponents.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridRoot = exports.GridPath = exports.GridLine = exports.GridCircle = void 0;
var _styles = require("@mui/material/styles");
var _chartsRadialGridClasses = require("./chartsRadialGridClasses");
const GridRoot = exports.GridRoot = (0, _styles.styled)('g', {
  name: 'MuiChartsRadialGrid',
  slot: 'Root',
  overridesResolver: (props, styles) => [{
    [`& .${_chartsRadialGridClasses.chartsRadialGridClasses.rotationLine}`]: styles.rotationLine
  }, {
    [`& .${_chartsRadialGridClasses.chartsRadialGridClasses.radiusLine}`]: styles.radiusLine
  }, styles.root]
})({});
const GridLine = exports.GridLine = (0, _styles.styled)('line', {
  slot: 'internal',
  shouldForwardProp: undefined
})(({
  theme
}) => ({
  stroke: (theme.vars || theme).palette.divider,
  shapeRendering: 'crispEdges',
  strokeWidth: 1,
  fill: 'none'
}));
const GridPath = exports.GridPath = (0, _styles.styled)('path', {
  slot: 'internal',
  shouldForwardProp: undefined
})(({
  theme
}) => ({
  stroke: (theme.vars || theme).palette.divider,
  shapeRendering: 'crispEdges',
  strokeWidth: 1,
  fill: 'none'
}));
const GridCircle = exports.GridCircle = (0, _styles.styled)('circle', {
  slot: 'internal',
  shouldForwardProp: undefined
})(({
  theme
}) => ({
  stroke: (theme.vars || theme).palette.divider,
  shapeRendering: 'crispEdges',
  strokeWidth: 1,
  fill: 'none'
}));
