// @mui/x-scheduler-internals/standalone-event/StandaloneEvent.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StandaloneEvent = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _adapter = require("@atlaskit/pragmatic-drag-and-drop/element/adapter");
var _disableNativeDragPreview = require("@atlaskit/pragmatic-drag-and-drop/element/disable-native-drag-preview");
var _useStableCallback = require("@base-ui/utils/useStableCallback");
var _useButton = require("../base-ui-copy/utils/useButton");
var _useRenderElement = require("../base-ui-copy/utils/useRenderElement");
var _useDragPreview = require("../internals/utils/useDragPreview");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className", "render", "style", "data", "onEventDrop", "renderDragPreview"];
const StandaloneEvent = exports.StandaloneEvent = /*#__PURE__*/React.forwardRef(function StandaloneEvent(componentProps, forwardedRef) {
  const {
      // Rendering props

      // Internal props
      data,
      onEventDrop,
      renderDragPreview
      // Props forwarded to the DOM element
    } = componentProps,
    elementProps = (0, _objectWithoutPropertiesLoose2.default)(componentProps, _excluded);

  // TODO: Expose a real `interactive` prop
  // to control whether the event should behave like a button
  const isInteractive = true;
  const ref = React.useRef(null);
  const {
    getButtonProps,
    buttonRef
  } = (0, _useButton.useButton)({
    disabled: !isInteractive,
    native: false
  });
  const preview = (0, _useDragPreview.useDragPreview)({
    type: 'standalone-event',
    data,
    renderDragPreview,
    showPreviewOnDragStart: true
  });
  const state = React.useMemo(() => ({
    dragging: preview.state.isDragging
  }), [preview.state.isDragging]);
  const getDragData = (0, _useStableCallback.useStableCallback)(() => ({
    source: 'StandaloneEvent',
    eventData: data,
    onEventDrop,
    eventId: data.id,
    occurrenceKey: `external-${data.id}`
  }));
  React.useEffect(() => {
    return (0, _adapter.draggable)({
      element: ref.current,
      getInitialData: getDragData,
      onGenerateDragPreview: ({
        nativeSetDragImage
      }) => {
        (0, _disableNativeDragPreview.disableNativeDragPreview)({
          nativeSetDragImage
        });
      },
      onDragStart: ({
        location
      }) => {
        preview.actions.onDragStart(location);
      },
      onDrag: ({
        location
      }) => {
        preview.actions.onDrag(location);
      },
      onDrop: () => {
        preview.actions.onDrop();
      }
    });
  }, [getDragData, preview.actions]);
  const element = (0, _useRenderElement.useRenderElement)('div', componentProps, {
    state,
    ref: [forwardedRef, buttonRef, ref],
    props: [elementProps, getButtonProps]
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(React.Fragment, {
    children: [element, preview.element]
  });
});
if (process.env.NODE_ENV !== "production") StandaloneEvent.displayName = "StandaloneEvent";
