// @mui/x-scheduler/internals/components/event-drag-preview/EventDragPreview.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EventDragPreview = EventDragPreview;
var React = _interopRequireWildcard(require("react"));
var _styles = require("@mui/material/styles");
var _store = require("@base-ui/utils/store");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _useSchedulerStoreContext = require("@mui/x-scheduler-internals/use-scheduler-store-context");
var _tokens = require("../../utils/tokens");
var _EventCalendarStyledContext = require("../../../event-calendar/EventCalendarStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const EventDragPreviewRoot = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'EventDragPreview'
})(({
  theme
}) => ({
  padding: theme.spacing(0.5, 1),
  borderRadius: theme.spacing(0.5),
  fontSize: theme.typography.body2.fontSize,
  backgroundColor: 'var(--event-surface-bold)',
  color: 'var(--event-on-surface-bold)',
  variants: (0, _tokens.getPaletteVariants)(theme)
}));
const fakeStore = {
  subscribe: () => {},
  getSnapshot: () => ({})
};
function EventDragPreview(props) {
  const store = (0, _useSchedulerStoreContext.useSchedulerStoreContext)(true);
  const styledContext = React.useContext(_EventCalendarStyledContext.EventCalendarStyledContext);
  const color = (0, _store.useStore)(store ?? fakeStore, store ? _schedulerSelectors.schedulerEventSelectors.color : () => 'teal', props.data.id);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(EventDragPreviewRoot, {
    className: styledContext?.classes.eventDragPreview,
    "data-palette": color,
    children: props.data.title
  });
}
