// @mui/x-scheduler-internals/calendar-grid/day-cell/CalendarGridDayCell.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarGridDayCell = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _useRenderElement = require("../../base-ui-copy/utils/useRenderElement");
var _useCompositeListItem = require("../../base-ui-copy/composite/list/useCompositeListItem");
var _CompositeListContext = require("../../base-ui-copy/composite/list/CompositeListContext");
var _useAdapterContext = require("../../use-adapter-context");
var _useEventCreation = require("../../internals/utils/useEventCreation");
var _accessibilityUtils = require("../../internals/utils/accessibility-utils");
var _useKeyboardEventCreation = require("../../internals/utils/useKeyboardEventCreation");
var _getNavigationTarget = require("../../internals/utils/getNavigationTarget");
var _CalendarGridRootContext = require("../root/CalendarGridRootContext");
var _CalendarGridDayRowContext = require("../day-row/CalendarGridDayRowContext");
var _useDayCellDropTarget = require("./useDayCellDropTarget");
var _CalendarGridDayCellContext = require("./CalendarGridDayCellContext");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className", "render", "style", "value", "addPropertiesToDroppedEvent", "lockSurfaceType"],
  _excluded2 = ["aria-labelledby"];
const CalendarGridDayCell = exports.CalendarGridDayCell = /*#__PURE__*/React.forwardRef(function CalendarGridDayCell(componentProps, forwardedRef) {
  const {
      // Rendering props

      // Internal props
      value,
      addPropertiesToDroppedEvent,
      lockSurfaceType
      // Props forwarded to the DOM element
    } = componentProps,
    elementProps = (0, _objectWithoutPropertiesLoose2.default)(componentProps, _excluded);
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const {
    id: rootId,
    focusedCell,
    setFocusedCell,
    rowTypes,
    rowsPerType
  } = (0, _CalendarGridRootContext.useCalendarGridRootContext)();
  const {
    rowIndex
  } = (0, _CalendarGridDayRowContext.useCalendarGridDayRowContext)();
  const {
    ref: listItemRef,
    index
  } = (0, _useCompositeListItem.useCompositeListItem)();
  const {
    elementsRef
  } = (0, _CompositeListContext.useCompositeListContext)();
  const dropTargetRef = (0, _useDayCellDropTarget.useDayCellDropTarget)({
    value,
    addPropertiesToDroppedEvent
  });
  const columnHeaderId = (0, _accessibilityUtils.getCalendarGridHeaderCellId)(rootId, index);
  const cellRef = React.useRef(null);
  const hasFocus = focusedCell?.rowType === 'day-grid' && focusedCell?.rowIndex === rowIndex && focusedCell?.columnIndex === index;
  const getPlaceholderFields = () => ({
    surfaceType: 'day-grid',
    start: adapter.startOfDay(value),
    end: adapter.endOfDay(value),
    lockSurfaceType,
    resourceId: null
  });
  const eventCreationProps = (0, _useEventCreation.useEventCreation)(getPlaceholderFields);
  const triggerKeyboardCreation = (0, _useKeyboardEventCreation.useKeyboardEventCreation)(getPlaceholderFields);

  // Apply DOM focus when this cell becomes the focused cell
  React.useEffect(() => {
    if (hasFocus && cellRef.current && !cellRef.current.contains(document.activeElement)) {
      cellRef.current.focus({
        preventScroll: true
      });
    }
  }, [hasFocus]);
  const handleKeyDown = event => {
    const target = (0, _getNavigationTarget.getNavigationTarget)(event.key, 'day-grid', rowIndex, index, {
      columnCount: elementsRef.current.length,
      rowTypes,
      rowsPerType
    });
    if (target) {
      event.preventDefault();
      setFocusedCell(target);
      return;
    }
    if (event.key === 'Enter' && event.target === event.currentTarget && triggerKeyboardCreation) {
      event.preventDefault();
      triggerKeyboardCreation();
    }
  };
  const handleFocus = event => {
    if (event.target === event.currentTarget) {
      setFocusedCell({
        rowType: 'day-grid',
        rowIndex,
        columnIndex: index
      });
    }
  };
  const contextValue = React.useMemo(() => ({
    index,
    hasFocus
  }), [index, hasFocus]);

  // Associate this cell with its column header, matching the pattern used by DayEvent and TimeEvent.
  // Any additional aria-labelledby passed by the styled layer (e.g., an "All day" row header) is appended.
  const {
      'aria-labelledby': elementAriaLabelledBy
    } = elementProps,
    restElementProps = (0, _objectWithoutPropertiesLoose2.default)(elementProps, _excluded2);
  const ariaLabelledBy = [columnHeaderId, elementAriaLabelledBy].filter(Boolean).join(' ');
  const keyboardProps = {
    // All cells are always tabbable so Tab flows through: cell → events → next cell → events.
    // Arrow keys navigate programmatically via setFocusedCell, independent of tabIndex.
    tabIndex: 0,
    onKeyDown: handleKeyDown,
    onFocus: handleFocus
  };
  const element = (0, _useRenderElement.useRenderElement)('div', componentProps, {
    ref: [forwardedRef, dropTargetRef, listItemRef, cellRef],
    props: [{
      role: 'gridcell',
      'aria-labelledby': ariaLabelledBy || undefined,
      'aria-colindex': index + 1
    }, keyboardProps, eventCreationProps, restElementProps]
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_CalendarGridDayCellContext.CalendarGridDayCellContext.Provider, {
    value: contextValue,
    children: element
  });
});
if (process.env.NODE_ENV !== "production") CalendarGridDayCell.displayName = "CalendarGridDayCell";
