// @mui/x-charts/ChartsDataProvider/useChartsDataProviderProps.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useChartsDataProviderProps = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _styles = require("@mui/material/styles");
var _defaultSeriesConfig = require("../internals/plugins/utils/defaultSeriesConfig");
var _allPlugins = require("../internals/plugins/allPlugins");
const _excluded = ["children", "localeText", "plugins", "slots", "slotProps", "seriesConfig"];
const useChartsDataProviderProps = inProps => {
  // eslint-disable-next-line mui/material-ui-name-matches-component-name
  const props = (0, _styles.useThemeProps)({
    props: inProps,
    name: 'MuiChartsDataProvider'
  });
  const {
      children,
      localeText,
      plugins = _allPlugins.DEFAULT_PLUGINS,
      slots,
      slotProps,
      seriesConfig = _defaultSeriesConfig.defaultSeriesConfig
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const theme = (0, _styles.useTheme)();
  const chartProviderProps = {
    plugins: plugins,
    pluginParams: (0, _extends2.default)({
      theme: theme.palette.mode,
      seriesConfig
    }, other)
  };
  return {
    children,
    localeText,
    chartProviderProps,
    slots,
    slotProps
  };
};
exports.useChartsDataProviderProps = useChartsDataProviderProps;
