// @mui/x-charts/ChartsDataProvider/useChartsDataProviderProps.mjs
'use client';

import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
const _excluded = ["children", "localeText", "plugins", "slots", "slotProps", "seriesConfig"];
import { useTheme, useThemeProps } from '@mui/material/styles';
import { defaultSeriesConfig } from "../internals/plugins/utils/defaultSeriesConfig.mjs";
import { DEFAULT_PLUGINS } from "../internals/plugins/allPlugins.mjs";
export const useChartsDataProviderProps = inProps => {
  // eslint-disable-next-line mui/material-ui-name-matches-component-name
  const props = useThemeProps({
    props: inProps,
    name: 'MuiChartsDataProvider'
  });
  const {
      children,
      localeText,
      plugins = DEFAULT_PLUGINS,
      slots,
      slotProps,
      seriesConfig = defaultSeriesConfig
    } = props,
    other = _objectWithoutPropertiesLoose(props, _excluded);
  const theme = useTheme();
  const chartProviderProps = {
    plugins: plugins,
    pluginParams: _extends({
      theme: theme.palette.mode,
      seriesConfig
    }, other)
  };
  return {
    children,
    localeText,
    chartProviderProps,
    slots,
    slotProps
  };
};
