// @mui/x-scheduler-internals/calendar-grid/time-scrollable-content/CalendarGridTimeScrollableContent.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarGridTimeScrollableContent = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _element = require("@atlaskit/pragmatic-drag-and-drop-auto-scroll/element");
var _useRenderElement = require("../../base-ui-copy/utils/useRenderElement");
var _CompositeList = require("../../base-ui-copy/composite/list/CompositeList");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className", "render", "style"];
const CalendarGridTimeScrollableContent = exports.CalendarGridTimeScrollableContent = /*#__PURE__*/React.forwardRef(function CalendarGridScrollableContent(componentProps, forwardedRef) {
  const elementProps = (0, _objectWithoutPropertiesLoose2.default)(componentProps, _excluded);
  const ref = React.useRef(null);
  const columnsRefs = React.useRef([]);
  React.useEffect(() => {
    // TODO: Try to add the behavior back in the test
    // For now, it causes the following error in JSDOM:
    // "Auto scrolling has been attached to an element that appears not to be scrollable"
    if (!ref.current || process.env.NODE_ENV === 'test') {
      return undefined;
    }
    return (0, _element.autoScrollForElements)({
      element: ref.current
    });
  }, []);
  const element = (0, _useRenderElement.useRenderElement)('div', componentProps, {
    ref: [forwardedRef, ref],
    props: [elementProps]
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_CompositeList.CompositeList, {
    elementsRef: columnsRefs,
    children: element
  });
});
if (process.env.NODE_ENV !== "production") CalendarGridTimeScrollableContent.displayName = "CalendarGridTimeScrollableContent";
