// @mui/x-scheduler-internals/calendar-grid/time-event-placeholder/CalendarGridTimeEventPlaceholder.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarGridTimeEventPlaceholder = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _useRenderElement = require("../../base-ui-copy/utils/useRenderElement");
var _CalendarGridTimeEventPlaceholderCssVars = require("./CalendarGridTimeEventPlaceholderCssVars");
var _CalendarGridTimeColumnContext = require("../time-column/CalendarGridTimeColumnContext");
var _useEvent = require("../../internals/utils/useEvent");
var _useElementPositionInCollection = require("../../internals/utils/useElementPositionInCollection");
const _excluded = ["className", "render", "style", "start", "end"];
const CalendarGridTimeEventPlaceholder = exports.CalendarGridTimeEventPlaceholder = /*#__PURE__*/React.forwardRef(function CalendarGridTimeEventPlaceholder(componentProps, forwardedRef) {
  const {
      // Rendering props

      // Internal props
      start,
      end
      // Props forwarded to the DOM element
    } = componentProps,
    elementProps = (0, _objectWithoutPropertiesLoose2.default)(componentProps, _excluded);
  const {
    start: columnStart,
    end: columnEnd
  } = (0, _CalendarGridTimeColumnContext.useCalendarGridTimeColumnContext)();
  const {
    position,
    duration
  } = (0, _useElementPositionInCollection.useElementPositionInCollection)({
    start,
    end,
    collectionStart: columnStart,
    collectionEnd: columnEnd
  });
  const {
    state
  } = (0, _useEvent.useEvent)({
    start,
    end
  });
  return (0, _useRenderElement.useRenderElement)('div', componentProps, {
    state,
    ref: [forwardedRef],
    props: [elementProps, {
      style: {
        [_CalendarGridTimeEventPlaceholderCssVars.CalendarGridTimeEventPlaceholderCssVars.yPosition]: `${position * 100}%`,
        [_CalendarGridTimeEventPlaceholderCssVars.CalendarGridTimeEventPlaceholderCssVars.height]: `${duration * 100}%`
      }
    }]
  });
});
if (process.env.NODE_ENV !== "production") CalendarGridTimeEventPlaceholder.displayName = "CalendarGridTimeEventPlaceholder";
