// @mui/x-scheduler-internals/calendar-grid/root/CalendarGridRoot.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarGridRoot = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _useId = require("@base-ui/utils/useId");
var _useRenderElement = require("../../base-ui-copy/utils/useRenderElement");
var _getNavigationTarget = require("../../internals/utils/getNavigationTarget");
var _CalendarGridRootContext = require("./CalendarGridRootContext");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className", "render", "style", "id", "rowTypes", "rowsPerType"];
const DEFAULT_ROWS_PER_TYPE = {};
const CalendarGridRoot = exports.CalendarGridRoot = /*#__PURE__*/React.forwardRef(function CalendarGridRoot(componentProps, forwardedRef) {
  const {
      // Rendering props

      // Internal props
      id: idProp,
      rowTypes = _getNavigationTarget.DEFAULT_ROW_TYPES,
      rowsPerType: rowsPerTypeProp = DEFAULT_ROWS_PER_TYPE
      // Props forwarded to the DOM element
    } = componentProps,
    elementProps = (0, _objectWithoutPropertiesLoose2.default)(componentProps, _excluded);
  const id = (0, _useId.useId)(idProp);
  const rowsPerType = rowsPerTypeProp;
  const rootRef = React.useRef(null);
  const [focusedCell, setFocusedCellState] = React.useState(null);
  const setFocusedCell = React.useCallback(coordinates => {
    setFocusedCellState(prev => prev?.rowType === coordinates.rowType && prev.rowIndex === coordinates.rowIndex && prev.columnIndex === coordinates.columnIndex ? prev : coordinates);
  }, []);
  const handleBlur = React.useCallback(event => {
    if (!rootRef.current?.contains(event.relatedTarget)) {
      setFocusedCellState(null);
    }
  }, []);
  const contextValue = React.useMemo(() => ({
    id,
    focusedCell,
    setFocusedCell,
    rowTypes,
    rowsPerType
  }), [id, focusedCell, setFocusedCell, rowTypes, rowsPerType]);
  const element = (0, _useRenderElement.useRenderElement)('div', componentProps, {
    ref: [forwardedRef, rootRef],
    props: [elementProps, {
      role: 'grid',
      id,
      onBlur: handleBlur
    }]
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_CalendarGridRootContext.CalendarGridRootContext.Provider, {
    value: contextValue,
    children: element
  });
});
if (process.env.NODE_ENV !== "production") CalendarGridRoot.displayName = "CalendarGridRoot";
