// @mui/x-charts/internals/plugins/featurePlugins/useChartCartesianAxis/defaultizeZoom.mjs
import _extends from "@babel/runtime/helpers/esm/extends";
import { DEFAULT_ZOOM_SLIDER_PREVIEW_SIZE, DEFAULT_ZOOM_SLIDER_SHOW_TOOLTIP, DEFAULT_ZOOM_SLIDER_SIZE } from "../../../constants.mjs";
/**
 * Ordinal Y axes (band/point) render with `domain[0]` at the top of the chart
 * (see the `range.reverse()` applied in `selectorChartYScales`), which is the
 * opposite of the cartesian convention assumed by the zoom/pan math.
 *
 * Returns the `reverse` flag the zoom handlers should use so that pan/wheel
 * gestures follow the visual direction of the axis.
 */
export const getEffectiveZoomReverse = (axisDirection, scaleType, reverse) => {
  const resolvedReverse = reverse ?? false;
  const isOrdinal = scaleType === 'band' || scaleType === 'point';
  if (axisDirection === 'y' && isOrdinal) {
    return !resolvedReverse;
  }
  return resolvedReverse;
};
export const defaultZoomOptions = {
  minStart: 0,
  maxEnd: 100,
  step: 5,
  minSpan: 10,
  maxSpan: 100,
  panning: true,
  filterMode: 'keep',
  reverse: false,
  slider: {
    enabled: false,
    preview: false,
    size: DEFAULT_ZOOM_SLIDER_SIZE,
    showTooltip: DEFAULT_ZOOM_SLIDER_SHOW_TOOLTIP
  }
};
export const defaultizeZoom = (zoom, axisId, axisDirection, reverse) => {
  if (!zoom) {
    return undefined;
  }
  if (zoom === true) {
    return _extends({
      axisId,
      axisDirection
    }, defaultZoomOptions, {
      reverse: reverse ?? false
    });
  }
  return _extends({
    axisId,
    axisDirection
  }, defaultZoomOptions, {
    reverse: reverse ?? false
  }, zoom, {
    slider: _extends({}, defaultZoomOptions.slider, {
      size: zoom.slider?.preview ?? defaultZoomOptions.slider.preview ? DEFAULT_ZOOM_SLIDER_PREVIEW_SIZE : DEFAULT_ZOOM_SLIDER_SIZE
    }, zoom.slider)
  });
};
