// @mui/x-charts/internals/plugins/featurePlugins/useChartCartesianAxis/defaultizeZoom.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getEffectiveZoomReverse = exports.defaultizeZoom = exports.defaultZoomOptions = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _constants = require("../../../constants");
/**
 * Ordinal Y axes (band/point) render with `domain[0]` at the top of the chart
 * (see the `range.reverse()` applied in `selectorChartYScales`), which is the
 * opposite of the cartesian convention assumed by the zoom/pan math.
 *
 * Returns the `reverse` flag the zoom handlers should use so that pan/wheel
 * gestures follow the visual direction of the axis.
 */
const getEffectiveZoomReverse = (axisDirection, scaleType, reverse) => {
  const resolvedReverse = reverse ?? false;
  const isOrdinal = scaleType === 'band' || scaleType === 'point';
  if (axisDirection === 'y' && isOrdinal) {
    return !resolvedReverse;
  }
  return resolvedReverse;
};
exports.getEffectiveZoomReverse = getEffectiveZoomReverse;
const defaultZoomOptions = exports.defaultZoomOptions = {
  minStart: 0,
  maxEnd: 100,
  step: 5,
  minSpan: 10,
  maxSpan: 100,
  panning: true,
  filterMode: 'keep',
  reverse: false,
  slider: {
    enabled: false,
    preview: false,
    size: _constants.DEFAULT_ZOOM_SLIDER_SIZE,
    showTooltip: _constants.DEFAULT_ZOOM_SLIDER_SHOW_TOOLTIP
  }
};
const defaultizeZoom = (zoom, axisId, axisDirection, reverse) => {
  if (!zoom) {
    return undefined;
  }
  if (zoom === true) {
    return (0, _extends2.default)({
      axisId,
      axisDirection
    }, defaultZoomOptions, {
      reverse: reverse ?? false
    });
  }
  return (0, _extends2.default)({
    axisId,
    axisDirection
  }, defaultZoomOptions, {
    reverse: reverse ?? false
  }, zoom, {
    slider: (0, _extends2.default)({}, defaultZoomOptions.slider, {
      size: zoom.slider?.preview ?? defaultZoomOptions.slider.preview ? _constants.DEFAULT_ZOOM_SLIDER_PREVIEW_SIZE : _constants.DEFAULT_ZOOM_SLIDER_SIZE
    }, zoom.slider)
  });
};
exports.defaultizeZoom = defaultizeZoom;
