// @mui/x-scheduler/event-calendar/header-toolbar/preferences-menu/PreferencesMenu.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreferencesMenu = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _styles = require("@mui/material/styles");
var _useMergedRefs = require("@base-ui/utils/useMergedRefs");
var _store = require("@base-ui/utils/store");
var _Check = _interopRequireDefault(require("@mui/icons-material/Check"));
var _SettingsOutlined = _interopRequireDefault(require("@mui/icons-material/SettingsOutlined"));
var _IconButton = _interopRequireDefault(require("@mui/material/IconButton"));
var _Menu = _interopRequireDefault(require("@mui/material/Menu"));
var _MenuItem = _interopRequireDefault(require("@mui/material/MenuItem"));
var _ListItemIcon = _interopRequireDefault(require("@mui/material/ListItemIcon"));
var _ListItemText = _interopRequireDefault(require("@mui/material/ListItemText"));
var _Divider = _interopRequireDefault(require("@mui/material/Divider"));
var _ListSubheader = _interopRequireDefault(require("@mui/material/ListSubheader"));
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _eventCalendarSelectors = require("@mui/x-scheduler-internals/event-calendar-selectors");
var _clsx = _interopRequireDefault(require("clsx"));
var _EventCalendarStyledContext = require("../../EventCalendarStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const PreferencesMenuRoot = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'PreferencesMenu'
})({});
const PreferencesMenuListItemIcon = (0, _styles.styled)(_ListItemIcon.default, {
  name: 'MuiEventCalendar',
  slot: 'PreferencesMenuListItemIcon'
})({
  justifyContent: 'flex-end',
  '&[data-checked="false"]': {
    visibility: 'hidden'
  }
});
const PreferencesMenuListSubheader = (0, _styles.styled)(_ListSubheader.default, {
  name: 'MuiEventCalendar',
  slot: 'PreferencesMenuListSubheader'
})(({
  theme
}) => (0, _extends2.default)({}, theme.typography.button, {
  color: (theme.vars || theme).palette.text.disabled,
  paddingBottom: theme.spacing(1),
  backgroundColor: 'transparent'
}));
const PreferencesMenu = exports.PreferencesMenu = /*#__PURE__*/React.forwardRef(function PreferencesMenu(props, forwardedRef) {
  // Context hooks
  const {
    schedulerId,
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();

  // Ref hooks
  const containerRef = React.useRef(null);
  const handleRef = (0, _useMergedRefs.useMergedRefs)(forwardedRef, containerRef);

  // Selector hooks
  const currentView = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarViewSelectors.view);
  const preferences = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarPreferenceSelectors.all);
  const preferencesMenuConfig = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarPreferenceSelectors.menuConfig);

  // State hooks (must come before any early returns)
  const [anchorEl, setAnchorEl] = React.useState(null);
  const open = Boolean(anchorEl);
  const handleToggle = (key, checked, event) => {
    store.setPreferences({
      [key]: checked
    }, event);
  };
  const handleTimeFormatChange = (value, event) => {
    store.setPreferences({
      ampm: value === '12'
    }, event);
  };
  const handleWeekStartsOnChange = (value, event) => {
    store.setPreferences({
      weekStartsOn: value
    }, event);
  };
  const handleClick = event => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };

  // Early return if config is false
  if (preferencesMenuConfig === false) {
    return null;
  }
  const preferenceOptions = [{
    configKey: 'toggleWeekendVisibility',
    preferenceKey: 'showWeekends',
    label: localeText.showWeekends
  }, {
    configKey: 'toggleWeekNumberVisibility',
    preferenceKey: 'showWeekNumber',
    label: localeText.showWeekNumber
  }];
  const viewSpecificPreferenceOptions = [{
    configKey: 'toggleEmptyDaysInAgenda',
    preferenceKey: 'showEmptyDaysInAgenda',
    view: 'agenda',
    label: localeText.showEmptyDaysInAgenda
  }];
  const visibleOptions = preferenceOptions.filter(option => preferencesMenuConfig?.[option.configKey] !== false);
  const visibleViewSpecificOptions = viewSpecificPreferenceOptions.filter(option => preferencesMenuConfig?.[option.configKey] !== false && option.view === currentView);
  const showSpecificOptions = visibleViewSpecificOptions.length > 0;
  const showTimeFormatSubmenu = preferencesMenuConfig?.toggleAmpm !== false;
  const showWeekStartsOnSubmenu = preferencesMenuConfig?.toggleWeekStartsOn !== false;
  const showDividerBeforeTimeFormat = showTimeFormatSubmenu && visibleOptions.length > 0;
  const showDividerBeforeWeekStartsOn = showWeekStartsOnSubmenu && (visibleOptions.length > 0 || showTimeFormatSubmenu);
  const showDividerBeforeSpecificOptions = showSpecificOptions && (visibleOptions.length > 0 || showTimeFormatSubmenu || showWeekStartsOnSubmenu);

  // Early return if no menu items to show
  if (!showTimeFormatSubmenu && !showWeekStartsOnSubmenu && visibleOptions.length === 0 && !showSpecificOptions) {
    return null;
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(PreferencesMenuRoot, (0, _extends2.default)({
    ref: handleRef
  }, props, {
    className: (0, _clsx.default)(props.className, classes.preferencesMenu),
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_IconButton.default, {
      className: classes.preferencesMenuButton,
      "aria-label": localeText.preferencesMenu,
      onClick: handleClick,
      "aria-controls": open ? `${schedulerId}-preferences-menu` : undefined,
      "aria-haspopup": "true",
      "aria-expanded": open ? 'true' : undefined,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_SettingsOutlined.default, {})
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(_Menu.default, {
      className: classes.preferencesMenuList,
      id: `${schedulerId}-preferences-menu`,
      anchorEl: anchorEl,
      open: open,
      onClose: handleClose,
      container: containerRef.current,
      anchorOrigin: {
        vertical: 'bottom',
        horizontal: 'right'
      },
      transformOrigin: {
        vertical: 'top',
        horizontal: 'right'
      },
      slotProps: {
        list: {
          'aria-label': localeText.preferencesMenu
        },
        paper: {
          sx: {
            minWidth: 220
          }
        }
      },
      children: [visibleOptions.map(option => /*#__PURE__*/(0, _jsxRuntime.jsxs)(_MenuItem.default, {
        className: classes.preferencesMenuItem,
        role: "menuitemcheckbox",
        "aria-checked": !!preferences[option.preferenceKey],
        onClick: event => {
          handleToggle(option.preferenceKey, !preferences[option.preferenceKey], event.nativeEvent);
        },
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_ListItemText.default, {
          className: classes.preferencesMenuListItemText,
          children: option.label
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(PreferencesMenuListItemIcon, {
          className: classes.preferencesMenuListItemIcon,
          "data-checked": !!preferences[option.preferenceKey],
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Check.default, {
            fontSize: "small"
          })
        })]
      }, option.configKey)), showDividerBeforeTimeFormat && /*#__PURE__*/(0, _jsxRuntime.jsx)(_Divider.default, {
        className: classes.preferencesMenuDivider
      }), showTimeFormatSubmenu && /*#__PURE__*/(0, _jsxRuntime.jsx)(PreferencesMenuListSubheader, {
        className: classes.preferencesMenuListSubheader,
        children: localeText.timeFormat
      }), showTimeFormatSubmenu && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_MenuItem.default, {
        className: classes.preferencesMenuItem,
        role: "menuitemradio",
        "aria-checked": !!preferences.ampm,
        onClick: event => {
          handleTimeFormatChange('12', event.nativeEvent);
        },
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_ListItemText.default, {
          className: classes.preferencesMenuListItemText,
          children: localeText.amPm12h
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(PreferencesMenuListItemIcon, {
          className: classes.preferencesMenuListItemIcon,
          "data-checked": !!preferences.ampm,
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Check.default, {
            fontSize: "small"
          })
        })]
      }), showTimeFormatSubmenu && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_MenuItem.default, {
        className: classes.preferencesMenuItem,
        role: "menuitemradio",
        "aria-checked": !preferences.ampm,
        onClick: event => {
          handleTimeFormatChange('24', event.nativeEvent);
        },
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_ListItemText.default, {
          className: classes.preferencesMenuListItemText,
          children: localeText.hour24h
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(PreferencesMenuListItemIcon, {
          className: classes.preferencesMenuListItemIcon,
          "data-checked": !preferences.ampm,
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Check.default, {
            fontSize: "small"
          })
        })]
      }), showDividerBeforeWeekStartsOn && /*#__PURE__*/(0, _jsxRuntime.jsx)(_Divider.default, {
        className: classes.preferencesMenuDivider
      }), showWeekStartsOnSubmenu && /*#__PURE__*/(0, _jsxRuntime.jsx)(PreferencesMenuListSubheader, {
        className: classes.preferencesMenuListSubheader,
        children: localeText.startWeekOn
      }), showWeekStartsOnSubmenu && [{
        value: 0,
        label: localeText.weekdaySunday
      }, {
        value: 1,
        label: localeText.weekdayMonday
      }, {
        value: 6,
        label: localeText.weekdaySaturday
      }].map(({
        value,
        label
      }) => /*#__PURE__*/(0, _jsxRuntime.jsxs)(_MenuItem.default, {
        className: classes.preferencesMenuItem,
        role: "menuitemradio",
        "aria-checked": preferences.weekStartsOn === value,
        onClick: event => {
          handleWeekStartsOnChange(value, event.nativeEvent);
        },
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_ListItemText.default, {
          className: classes.preferencesMenuListItemText,
          children: label
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(PreferencesMenuListItemIcon, {
          className: classes.preferencesMenuListItemIcon,
          "data-checked": preferences.weekStartsOn === value,
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Check.default, {
            fontSize: "small"
          })
        })]
      }, value)), showDividerBeforeSpecificOptions && /*#__PURE__*/(0, _jsxRuntime.jsx)(_Divider.default, {
        className: classes.preferencesMenuDivider
      }), showSpecificOptions && /*#__PURE__*/(0, _jsxRuntime.jsx)(PreferencesMenuListSubheader, {
        className: classes.preferencesMenuListSubheader,
        children: localeText.viewSpecificOptions(currentView)
      }), showSpecificOptions && visibleViewSpecificOptions.map(option => /*#__PURE__*/(0, _jsxRuntime.jsxs)(_MenuItem.default, {
        className: classes.preferencesMenuItem,
        role: "menuitemcheckbox",
        "aria-checked": !!preferences[option.preferenceKey],
        onClick: event => {
          handleToggle(option.preferenceKey, !preferences[option.preferenceKey], event.nativeEvent);
        },
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_ListItemText.default, {
          className: classes.preferencesMenuListItemText,
          children: option.label
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(PreferencesMenuListItemIcon, {
          className: classes.preferencesMenuListItemIcon,
          "data-checked": !!preferences[option.preferenceKey],
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Check.default, {
            fontSize: "small"
          })
        })]
      }, option.configKey))]
    })]
  }));
});
if (process.env.NODE_ENV !== "production") PreferencesMenu.displayName = "PreferencesMenu";
