// @mui/x-scheduler/internals/utils/day-time-grid-view-config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDayTimeGridViewConfig = createDayTimeGridViewConfig;
var _store = require("@base-ui/utils/store");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _eventCalendarSelectors = require("@mui/x-scheduler-internals/event-calendar-selectors");
var _getDayList = require("@mui/x-scheduler-internals/get-day-list");
var _internals = require("@mui/x-scheduler-internals/internals");
var _processDate = require("@mui/x-scheduler-internals/process-date");
const DAYS_IN_WEEK = 7;

/**
 * Number of consecutive days a day-time-grid based view can render. Restricted to the
 * values the views actually use so the `=== 7` (week-aligned) branch below cannot be
 * reached with an out-of-domain count (`0`, negative, or fractional) that would silently
 * produce an empty or garbage day list.
 */

/**
 * Builds an `EventCalendarViewConfig` for a day-time-grid based view.
 *
 * When `dayCount === 7` the config is week-aligned: it snaps the visible range to
 * `startOfWeek` and honors the `showWeekends` and `weekStartsOn` preferences. For any
 * other value, the config shows `dayCount` consecutive days starting from the current
 * visible date.
 */
function createDayTimeGridViewConfig(dayCount) {
  if (dayCount === DAYS_IN_WEEK) {
    return {
      siblingVisibleDateGetter: ({
        state,
        delta
      }) => state.adapter.addWeeks((0, _internals.getStartOfWeek)(state.adapter, _schedulerSelectors.schedulerOtherSelectors.visibleDate(state), _eventCalendarSelectors.eventCalendarPreferenceSelectors.weekStartsOn(state)), delta),
      visibleDaysSelector: (0, _store.createSelectorMemoized)(state => state.adapter, _schedulerSelectors.schedulerOtherSelectors.visibleDate, _eventCalendarSelectors.eventCalendarPreferenceSelectors.showWeekends, _eventCalendarSelectors.eventCalendarPreferenceSelectors.weekStartsOn, (adapter, visibleDate, showWeekends, weekStartsOn) => (0, _getDayList.getDayList)({
        adapter,
        start: (0, _internals.getStartOfWeek)(adapter, visibleDate, weekStartsOn),
        end: (0, _internals.getEndOfWeek)(adapter, visibleDate, weekStartsOn),
        excludeWeekends: !showWeekends
      }))
    };
  }
  return {
    siblingVisibleDateGetter: ({
      state,
      delta
    }) => state.adapter.addDays(_schedulerSelectors.schedulerOtherSelectors.visibleDate(state), delta * dayCount),
    visibleDaysSelector: (0, _store.createSelectorMemoized)(state => state.adapter, _schedulerSelectors.schedulerOtherSelectors.visibleDate, (adapter, visibleDate) => Array.from({
      length: dayCount
    }, (_, index) => (0, _processDate.processDate)(adapter.addDays(visibleDate, index), adapter)))
  };
}
