// @mui/x-scheduler/internals/utils/test-utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.changeTo12HoursFormat = changeTo12HoursFormat;
exports.changeTo24HoursFormat = changeTo24HoursFormat;
exports.findPreferencesMenu = findPreferencesMenu;
exports.getPreferencesMenu = getPreferencesMenu;
exports.openPreferencesMenu = openPreferencesMenu;
exports.toggleShowEmptyDaysInAgenda = toggleShowEmptyDaysInAgenda;
exports.toggleShowWeekNumber = toggleShowWeekNumber;
exports.toggleShowWeekends = toggleShowWeekends;
var _internalTestUtils = require("@mui/internal-test-utils");
function getPreferencesMenu() {
  return _internalTestUtils.screen.queryByRole('button', {
    name: /settings/i
  });
}
async function findPreferencesMenu() {
  return _internalTestUtils.screen.findByRole('button', {
    name: /settings/i
  });
}
async function openPreferencesMenu(user) {
  const button = await findPreferencesMenu();
  await user.click(button);
  // MUI Menu doesn't easily expose aria-label, just wait for the menu to appear
  await _internalTestUtils.screen.findByRole('menu');
}
async function toggleShowWeekends(user) {
  const menuItem = await _internalTestUtils.screen.findByRole('menuitemcheckbox', {
    name: /show weekends/i
  });
  await user.click(menuItem);
}
async function toggleShowWeekNumber(user) {
  const menuItem = await _internalTestUtils.screen.findByRole('menuitemcheckbox', {
    name: /show week number/i
  });
  await user.click(menuItem);
}
async function toggleShowEmptyDaysInAgenda(user) {
  const menuItem = await _internalTestUtils.screen.findByRole('menuitemcheckbox', {
    name: /show empty days/i
  });
  await user.click(menuItem);
}
async function changeTo24HoursFormat(user) {
  const h24Option = await _internalTestUtils.screen.findByRole('menuitemradio', {
    name: /24-hour \(13:00\)/i
  });
  await user.click(h24Option);
}
async function changeTo12HoursFormat(user) {
  const h12Option = await _internalTestUtils.screen.findByRole('menuitemradio', {
    name: /12-hour \(1:00PM\)/i
  });
  await user.click(h12Option);
}
