// @mui/x-scheduler/agenda-view/AgendaView.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AgendaView = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _store = require("@base-ui/utils/store");
var _styles = require("@mui/material/styles");
var _useMergedRefs = require("@base-ui/utils/useMergedRefs");
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _useEventCalendarView = require("@mui/x-scheduler-internals/use-event-calendar-view");
var _sortEventOccurrences = require("@mui/x-scheduler-internals/sort-event-occurrences");
var _eventCalendarSelectors = require("@mui/x-scheduler-internals/event-calendar-selectors");
var _useEventOccurrencesGroupedByDay = require("@mui/x-scheduler-internals/use-event-occurrences-grouped-by-day");
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _constants = require("@mui/x-scheduler-internals/constants");
var _internals = require("@mui/x-scheduler-internals/internals");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _clsx = _interopRequireDefault(require("clsx"));
var _EventItem = require("../internals/components/event/event-item/EventItem");
var _eventSkeleton = require("../internals/components/event-skeleton");
var _EventCalendarStyledContext = require("../event-calendar/EventCalendarStyledContext");
var _eventDialog = require("../internals/components/event-dialog");
var _jsxRuntime = require("react/jsx-runtime");
const AgendaViewRoot = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'AgendaView'
})(({
  theme
}) => ({
  width: '100%',
  maxHeight: '100%',
  border: `1px solid ${(theme.vars || theme).palette.divider}`,
  borderRadius: theme.shape.borderRadius,
  overflowY: 'auto',
  position: 'relative'
}));
const AgendaViewRow = (0, _styles.styled)('section', {
  name: 'MuiEventCalendar',
  slot: 'AgendaViewRow'
})(({
  theme
}) => ({
  display: 'grid',
  gridTemplateColumns: '190px 1fr',
  '&:not(:last-child)': {
    borderBottom: `1px solid ${(theme.vars || theme).palette.divider}`
  }
}));
const DayHeaderCell = (0, _styles.styled)('header', {
  name: 'MuiEventCalendar',
  slot: 'AgendaViewDayHeaderCell'
})(({
  theme
}) => ({
  display: 'flex',
  alignItems: 'flex-start',
  borderRight: `1px solid ${(theme.vars || theme).palette.divider}`,
  padding: theme.spacing(2),
  gap: theme.spacing(0.5),
  '&[data-current]': {
    backgroundColor: theme.alpha((theme.vars || theme).palette.primary.light, 0.05),
    color: (theme.vars || theme).palette.primary.main
  }
}));
const DayNumberCell = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'AgendaViewDayNumberCell'
})(({
  theme
}) => ({
  fontSize: 'var(--EventCalendar-fontSize-agendaDayNumber, 1.5rem)',
  fontWeight: theme.typography.fontWeightMedium,
  lineHeight: 1,
  minWidth: '4ch',
  textAlign: 'center',
  color: (theme.vars || theme).palette.text.primary,
  '&[data-current]': {
    color: (theme.vars || theme).palette.primary.main
  }
}));
const WeekDayCell = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'AgendaViewWeekDayCell'
})(({
  theme
}) => ({
  display: 'flex',
  flexDirection: 'column',
  gap: theme.spacing(1)
}));
const AgendaWeekDayNameLabel = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'AgendaViewWeekDayNameLabel'
})(({
  theme
}) => ({
  fontSize: theme.typography.body2.fontSize,
  lineHeight: 1,
  display: '-webkit-box',
  WebkitLineClamp: 'var(--number-of-lines)',
  WebkitBoxOrient: 'vertical',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  wordBreak: 'break-word',
  overflowWrap: 'break-word'
}));
const AgendaYearAndMonthLabel = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'AgendaViewYearAndMonthLabel'
})(({
  theme
}) => ({
  fontSize: theme.typography.caption.fontSize,
  lineHeight: 1,
  color: (theme.vars || theme).palette.text.secondary,
  display: '-webkit-box',
  WebkitLineClamp: 'var(--number-of-lines)',
  WebkitBoxOrient: 'vertical',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  wordBreak: 'break-word',
  overflowWrap: 'break-word',
  '[data-current] &': {
    color: 'inherit'
  }
}));
const EventsList = (0, _styles.styled)('ul', {
  name: 'MuiEventCalendar',
  slot: 'AgendaViewEventsList'
})(({
  theme
}) => ({
  display: 'flex',
  flexDirection: 'column',
  padding: theme.spacing(1),
  margin: 0,
  listStyle: 'none',
  flexGrow: 1
}));
const AgendaViewWeekNumberLabel = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'AgendaViewWeekNumberLabel'
})(({
  theme
}) => ({
  fontSize: '0.875rem',
  fontWeight: theme.typography.fontWeightRegular,
  lineHeight: 1,
  color: (theme.vars || theme).palette.text.primary
}));
const AgendaViewWeekNumberRow = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'AgendaViewWeekNumberRow'
})(({
  theme
}) => (0, _extends2.default)({
  padding: theme.spacing(0.5, 2),
  backgroundColor: (theme.vars || theme).palette.grey[200],
  borderBottom: `1px solid ${(theme.vars || theme).palette.divider}`
}, theme.applyStyles('dark', {
  backgroundColor: (theme.vars || theme).palette.grey[900]
})));
const AGENDA_VIEW_CONFIG = {
  siblingVisibleDateGetter: ({
    state,
    delta
  }) => state.adapter.addDays(_schedulerSelectors.schedulerOtherSelectors.visibleDate(state), _constants.AGENDA_VIEW_DAYS_AMOUNT * delta),
  visibleDaysSelector: _eventCalendarSelectors.eventCalendarAgendaSelectors.visibleDays
};

/**
 * An Agenda View to use inside the Event Calendar.
 */
const AgendaView = exports.AgendaView = /*#__PURE__*/React.memo(/*#__PURE__*/React.forwardRef(function AgendaView(props, forwardedRef) {
  // Context hooks
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const {
    schedulerId,
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();

  // Ref hooks
  const containerRef = React.useRef(null);
  const handleRef = (0, _useMergedRefs.useMergedRefs)(forwardedRef, containerRef);

  // Selector hooks
  const now = (0, _store.useStore)(store, _schedulerSelectors.schedulerNowSelectors.nowUpdatedEveryMinute);
  const showWeekNumber = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarPreferenceSelectors.showWeekNumber);
  const weekStartsOn = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarPreferenceSelectors.weekStartsOn);

  // Feature hooks
  const {
    days
  } = (0, _useEventCalendarView.useEventCalendarView)(AGENDA_VIEW_CONFIG);
  const occurrencesMap = (0, _useEventOccurrencesGroupedByDay.useEventOccurrencesGroupedByDay)({
    days
  });

  // Selector hooks
  const isLoading = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.isLoading);
  const daysWithOccurrences = React.useMemo(() => days.map((date, index) => {
    const occurrences = (0, _sortEventOccurrences.sortEventOccurrences)(occurrencesMap.get(date.key) || []);
    const weekNumber = (0, _internals.getWeekNumber)(adapter, date.value, weekStartsOn);
    const isFirstDayOfWeek = index === 0 || !adapter.isSameDay((0, _internals.getStartOfWeek)(adapter, date.value, weekStartsOn), (0, _internals.getStartOfWeek)(adapter, days[index - 1].value, weekStartsOn));
    return {
      date,
      occurrences,
      weekNumber,
      isFirstDayOfWeek
    };
  }), [adapter, days, occurrencesMap, weekStartsOn]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(AgendaViewRoot, (0, _extends2.default)({}, props, {
    ref: handleRef,
    className: (0, _clsx.default)(props.className, classes.agendaView),
    children: daysWithOccurrences.map(({
      date,
      occurrences,
      isFirstDayOfWeek,
      weekNumber
    }) => /*#__PURE__*/(0, _jsxRuntime.jsxs)(React.Fragment, {
      children: [showWeekNumber && isFirstDayOfWeek && /*#__PURE__*/(0, _jsxRuntime.jsx)(AgendaViewWeekNumberRow, {
        className: classes.agendaViewWeekNumberRow,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(AgendaViewWeekNumberLabel, {
          className: classes.agendaViewWeekNumberLabel,
          "aria-label": localeText.weekNumberAriaLabel(weekNumber),
          children: `${localeText.week} ${weekNumber}`
        })
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(AgendaViewRow, {
        className: classes.agendaViewRow,
        id: `${schedulerId}-AgendaViewRow-${date.key}`,
        "aria-labelledby": `${schedulerId}-DayHeaderCell-${date.key}`,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(DayHeaderCell, {
          className: classes.agendaViewDayHeaderCell,
          id: `${schedulerId}-DayHeaderCell-${date.key}`,
          "aria-label": `${adapter.format(date.value, 'weekday')} ${adapter.format(date.value, 'dayOfMonth')}`,
          "data-current": adapter.isSameDay(date.value, now) ? '' : undefined,
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(DayNumberCell, {
            className: classes.agendaViewDayNumberCell,
            "data-current": adapter.isSameDay(date.value, now) ? '' : undefined,
            children: adapter.format(date.value, 'dayOfMonth')
          }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(WeekDayCell, {
            className: classes.agendaViewWeekDayCell,
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(AgendaWeekDayNameLabel, {
              className: classes.agendaViewWeekDayNameLabel,
              style: {
                '--number-of-lines': 1
              },
              children: adapter.format(date.value, 'weekday')
            }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(AgendaYearAndMonthLabel, {
              className: classes.agendaViewYearAndMonthLabel,
              style: {
                '--number-of-lines': 1
              },
              children: [adapter.format(date.value, 'monthFullLetter'), ",", ' ', adapter.format(date.value, 'yearPadded')]
            })]
          })]
        }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(EventsList, {
          className: classes.agendaViewEventsList,
          children: [isLoading && /*#__PURE__*/(0, _jsxRuntime.jsx)("li", {
            className: classes.agendaViewEventListItem,
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventSkeleton.EventSkeleton, {
              "data-variant": "agenda"
            })
          }), !isLoading && occurrences.map(occurrence => /*#__PURE__*/(0, _jsxRuntime.jsx)("li", {
            className: classes.agendaViewEventListItem,
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventDialog.EventDialogTrigger, {
              occurrence: occurrence,
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_EventItem.EventItem, {
                occurrence: occurrence,
                date: date,
                variant: "regular",
                ariaLabelledBy: `${schedulerId}-DayHeaderCell-${date.key}`
              })
            })
          }, occurrence.key))]
        })]
      })]
    }, date.key))
  }));
}));
if (process.env.NODE_ENV !== "production") AgendaView.displayName = "AgendaView";
