// @mui/x-scheduler-internals/calendar-grid/header-row/CalendarGridHeaderRow.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarGridHeaderRow = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _useRenderElement = require("../../base-ui-copy/utils/useRenderElement");
var _CompositeList = require("../../base-ui-copy/composite/list/CompositeList");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className", "render", "style"];
const CalendarGridHeaderRow = exports.CalendarGridHeaderRow = /*#__PURE__*/React.forwardRef(function CalendarGridHeaderRow(componentProps, forwardedRef) {
  const elementProps = (0, _objectWithoutPropertiesLoose2.default)(componentProps, _excluded);
  const cellsRefs = React.useRef([]);
  const element = (0, _useRenderElement.useRenderElement)('div', componentProps, {
    ref: [forwardedRef],
    props: [{
      role: 'row',
      'aria-rowindex': 1
    }, elementProps]
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_CompositeList.CompositeList, {
    elementsRef: cellsRefs,
    children: element
  });
});
if (process.env.NODE_ENV !== "production") CalendarGridHeaderRow.displayName = "CalendarGridHeaderRow";
