// @mui/x-scheduler-internals/scheduler-selectors/schedulerEventSelectors.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schedulerEventSelectors = void 0;
var _formatErrorMessage2 = _interopRequireDefault(require("@mui/x-internals/formatErrorMessage"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _store = require("@base-ui/utils/store");
var _schedulerResourceSelectors = require("./schedulerResourceSelectors");
var _constants = require("../constants");
const processedEventSelector = (0, _store.createSelector)(state => state.processedEventLookup, (processedEventLookup, eventId) => eventId == null ? null : processedEventLookup.get(eventId));
const isEventReadOnlySelector = (0, _store.createSelector)((state, eventId) => {
  const processedEvent = processedEventSelector(state, eventId);
  if (!processedEvent) {
    return false;
  }
  return resolveEventProperty({
    state,
    resourceId: processedEvent.resource,
    valueInEvent: processedEvent.modelInBuiltInFormat?.readOnly,
    getValueInResource: r => r.areEventsReadOnly,
    valueInState: state.readOnly ?? false
  });
});
const schedulerEventSelectors = exports.schedulerEventSelectors = {
  creationConfig: (0, _store.createSelectorMemoized)(state => state.readOnly, state => state.eventCreation, (isSchedulerReadOnly, creationConfig) => {
    if (isSchedulerReadOnly) {
      return false;
    }
    if (creationConfig === false) {
      return false;
    }
    if (creationConfig === true) {
      return _constants.DEFAULT_EVENT_CREATION_CONFIG;
    }
    return (0, _extends2.default)({}, _constants.DEFAULT_EVENT_CREATION_CONFIG, creationConfig);
  }),
  /**
   * Gets the default duration (in minutes) for newly created events.
   * This can be used when you need the value event on read-only calendar.
   */
  defaultEventDuration: (0, _store.createSelector)(state => state.eventCreation, eventCreation => {
    if (typeof eventCreation === 'boolean') {
      return _constants.DEFAULT_EVENT_CREATION_CONFIG.duration;
    }
    return eventCreation?.duration ?? _constants.DEFAULT_EVENT_CREATION_CONFIG.duration;
  }),
  processedEvent: processedEventSelector,
  processedEventRequired: (0, _store.createSelector)(processedEventSelector, (event, eventId) => {
    if (!event) {
      throw new Error(process.env.NODE_ENV !== "production" ? `MUI X Scheduler: Event with id="${eventId}" was not found. ` + 'The requested event does not exist in the scheduler state. ' + 'Verify the event id is correct and the event has been added.' : (0, _formatErrorMessage2.default)(210, eventId));
    }
    return event;
  }),
  isReadOnly: isEventReadOnlySelector,
  color: (0, _store.createSelector)((state, eventId) => {
    const event = processedEventSelector(state, eventId);
    if (!event) {
      return state.eventColor;
    }
    return resolveEventProperty({
      state,
      resourceId: event.resource,
      valueInEvent: event.color,
      getValueInResource: r => r.eventColor,
      valueInState: state.eventColor
    });
  }),
  isPropertyReadOnly: (0, _store.createSelectorMemoized)(isEventReadOnlySelector, state => state.eventModelStructure, (isEventReadOnly, eventModelStructure, _eventId) => {
    if (isEventReadOnly) {
      return () => true;
    }
    return property => {
      if (eventModelStructure?.[property] && !eventModelStructure?.[property].setter) {
        return true;
      }
      return false;
    };
  }),
  processedEventList: (0, _store.createSelectorMemoized)(state => state.eventIdList, state => state.processedEventLookup, (eventIds, processedEventLookup) => eventIds.map(id => processedEventLookup.get(id))),
  idList: (0, _store.createSelector)(state => state.eventIdList),
  modelList: (0, _store.createSelector)(state => state.eventModelList),
  modelLookup: (0, _store.createSelector)(state => state.eventModelLookup),
  canDragEventsFromTheOutside: (0, _store.createSelector)(state => state.canDragEventsFromTheOutside && !state.readOnly),
  canDropEventsToTheOutside: (0, _store.createSelector)(state => state.canDropEventsToTheOutside && !state.readOnly),
  isDraggable: (0, _store.createSelector)((state, eventId) => {
    if (isEventReadOnlySelector(state, eventId)) {
      return false;
    }
    const eventModelStructure = state.eventModelStructure;
    if (eventModelStructure?.start && !eventModelStructure?.start.setter) {
      return false;
    }
    if (eventModelStructure?.end && !eventModelStructure?.end.setter) {
      return false;
    }
    const processedEvent = processedEventSelector(state, eventId);
    if (!processedEvent) {
      return false;
    }
    return resolveEventProperty({
      state,
      resourceId: processedEvent.resource,
      valueInEvent: processedEvent.draggable,
      getValueInResource: r => r.areEventsDraggable,
      valueInState: state.areEventsDraggable
    });
  }),
  isResizable: (0, _store.createSelector)((state, eventId, side) => {
    if (isEventReadOnlySelector(state, eventId)) {
      return false;
    }
    const eventModelStructure = state.eventModelStructure;
    if (side === 'start' && eventModelStructure?.start && !eventModelStructure?.start.setter) {
      return false;
    }
    if (side === 'end' && eventModelStructure?.end && !eventModelStructure?.end.setter) {
      return false;
    }
    const processedEvent = processedEventSelector(state, eventId);
    if (!processedEvent) {
      return false;
    }
    return resolveEventProperty({
      state,
      resourceId: processedEvent.resource,
      valueInEvent: getIsResizableFromProperty(processedEvent.resizable, side) ?? undefined,
      getValueInResource: r => getIsResizableFromProperty(r.areEventsResizable, side) ?? undefined,
      valueInState: getIsResizableFromProperty(state.areEventsResizable, side) ?? false
    });
  }),
  isRecurring: (0, _store.createSelector)(processedEventSelector, state => state.recurringEventsPlugin, (event, recurringEventsPlugin, _eventId) => recurringEventsPlugin != null && Boolean(event?.dataTimezone.rrule))
};
function getIsResizableFromProperty(propertyValue, side) {
  if (propertyValue === undefined) {
    return null;
  }
  if (propertyValue === true) {
    return true;
  }
  if (propertyValue === false) {
    return false;
  }
  if (propertyValue === side) {
    return true;
  }

  // If the property is a specific side (e.g., 'start' or 'end') but doesn't match the current side,
  // return false because the property explicitly restricts resizing to a specific side.
  return false;
}

/**
 * Resolves an event property by checking (in order of priority):
 * 1. The event itself (`valueInEvent`)
 * 2. The resource hierarchy, child → parent → … (`getValueInResource`)
 * 3. The component-level state (`valueInState`)
 */
function resolveEventProperty(parameters) {
  const {
    state,
    resourceId,
    valueInEvent,
    getValueInResource,
    valueInState
  } = parameters;
  if (valueInEvent !== undefined) {
    return valueInEvent;
  }
  return (0, _schedulerResourceSelectors.resolveResourceProperty)(state, resourceId, getValueInResource, valueInState);
}
