// @mui/x-scheduler-internals/scheduler-selectors/schedulerOtherSelectors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schedulerOtherSelectors = void 0;
var _store = require("@base-ui/utils/store");
// Warning: Only add selectors here that do not belong to any specific feature.
const schedulerOtherSelectors = exports.schedulerOtherSelectors = {
  /**
   * Returns `true` if the event with the given ID is the currently active event.
   */
  isEditedEvent: (0, _store.createSelector)(state => state.editedEventId, (editedEventId, eventId) => editedEventId === eventId),
  visibleDate: (0, _store.createSelectorMemoized)(state => state.adapter, state => state.visibleDate, state => state.displayTimezone, (adapter, visibleDate, timezone) => adapter.setTimezone(visibleDate, timezone)),
  isRecurringScopeDialogOpen: (0, _store.createSelector)(state => state.pendingRecurringEventOperation != null),
  /**
   * The default event color used when no color is specified on the event or its resource.
   */
  defaultEventColor: (0, _store.createSelector)(state => state.eventColor),
  displayTimezone: (0, _store.createSelector)(state => state.displayTimezone),
  /**
   * Whether each event must be assigned to a resource. When true, the resource cannot be cleared in the edit dialog and the form cannot be submitted without one.
   */
  shouldEventRequireResource: (0, _store.createSelector)(state => state.shouldEventRequireResource),
  recurringEventsPlugin: (0, _store.createSelector)(state => state.recurringEventsPlugin),
  areRecurringEventsAvailable: (0, _store.createSelector)(state => state.recurringEventsPlugin != null),
  isLoading: (0, _store.createSelector)(state => state.isLoading),
  errors: (0, _store.createSelector)(state => state.errors)
};
