// @mui/x-scheduler-internals/scheduler-selectors/schedulerOccurrenceSelectors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.schedulerOccurrenceSelectors = void 0;
var _store = require("@base-ui/utils/store");
var _empty = require("@base-ui/utils/empty");
var _schedulerEventSelectors = require("./schedulerEventSelectors");
var _schedulerResourceSelectors = require("./schedulerResourceSelectors");
var _eventUtils = require("../internals/utils/event-utils");
var _schedulerOtherSelectors = require("./schedulerOtherSelectors");
const occurrencesGroupedByResourceListSelector = (0, _store.createSelectorMemoized)(state => state.adapter, _schedulerEventSelectors.schedulerEventSelectors.processedEventList, _schedulerResourceSelectors.schedulerResourceSelectors.visibleMap, _schedulerResourceSelectors.schedulerResourceSelectors.processedResourceList, _schedulerResourceSelectors.schedulerResourceSelectors.processedResourceChildrenLookup, _schedulerOtherSelectors.schedulerOtherSelectors.displayTimezone, state => state.recurringEventsPlugin, (adapter, events, visibleResources, resources, resourcesChildrenMap, displayTimezone, recurringEventsPlugin, start, end) => {
  const occurrencesGroupedByResource = new Map();
  const occurrences = (0, _eventUtils.getOccurrencesFromEvents)({
    adapter,
    start,
    end,
    events,
    visibleResources,
    displayTimezone,
    recurringEventsPlugin
  });
  for (const occurrence of occurrences) {
    const resourceId = occurrence.resource;
    if (resourceId) {
      if (!occurrencesGroupedByResource.has(resourceId)) {
        occurrencesGroupedByResource.set(resourceId, []);
      }
      occurrencesGroupedByResource.get(resourceId).push(occurrence);
    }
  }
  const processResources = innerResources => {
    const sortedResources = innerResources.toSorted((a, b) => a.title.localeCompare(b.title));
    const result = [];
    for (const resource of sortedResources) {
      if (visibleResources[resource.id] === false) {
        continue;
      }
      result.push({
        resource,
        occurrences: occurrencesGroupedByResource.get(resource.id) ?? []
      });
      const children = resourcesChildrenMap.get(resource.id) ?? [];
      if (children.length > 0) {
        result.push(...processResources(children));
      }
    }
    return result;
  };
  return processResources(resources);
});
const occurrencesGroupedByResourceMapSelector = (0, _store.createSelectorMemoized)(occurrencesGroupedByResourceListSelector, (groupedByResourceList, _start, _end) => {
  const map = new Map();
  for (const {
    resource,
    occurrences
  } of groupedByResourceList) {
    map.set(resource.id, occurrences);
  }
  return map;
});
const schedulerOccurrenceSelectors = exports.schedulerOccurrenceSelectors = {
  isStarted: (0, _store.createSelector)(state => state.adapter, state => state.nowUpdatedEveryMinute, (adapter, now, start) => {
    return adapter.isBefore(start.value, now) || adapter.isEqual(start.value, now);
  }),
  isEnded: (0, _store.createSelector)(state => state.adapter, state => state.nowUpdatedEveryMinute, (adapter, now, end) => {
    return adapter.isBefore(end.value, now);
  }),
  groupedByResourceList: occurrencesGroupedByResourceListSelector,
  resourceOccurrences: (0, _store.createSelector)(occurrencesGroupedByResourceMapSelector, (map, _start, end, resourceId) => map.get(resourceId) ?? _empty.EMPTY_ARRAY)
};
