// @mui/x-charts/ScatterChart/async/ScatterAsyncBatch.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScatterAsyncBatch = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _useSlotProps2 = _interopRequireDefault(require("@mui/utils/useSlotProps"));
var _useInteractionItemProps = require("../../hooks/useInteractionItemProps");
var _useStore = require("../../internals/store/useStore");
var _useItemHighlightStateGetter = require("../../hooks/useItemHighlightStateGetter");
var _useChartClosestPoint = require("../../internals/plugins/featurePlugins/useChartClosestPoint");
var _ScatterMarker = require("../ScatterMarker");
var _scatterClasses = require("../scatterClasses");
var _ChartsProvider = require("../../context/ChartsProvider");
var _scatterRenderData = require("./scatterRenderData.selectors");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["ownerState"];
/**
 * @ignore - internal component.
 */
function ScatterAsyncBatchComponent(props) {
  const {
    series,
    colorGetter,
    onItemClick,
    slots,
    slotProps,
    start,
    step,
    revealed,
    isInteracting,
    classes: inClasses
  } = props;
  const classes = (0, _scatterClasses.useUtilityClasses)({
    classes: inClasses
  });
  const {
    instance
  } = (0, _ChartsProvider.useChartsContext)();
  const store = (0, _useStore.useStore)();
  const isVoronoiEnabled = store.use(_useChartClosestPoint.selectorChartsIsVoronoiEnabled);
  const skipInteractionHandlers = isVoronoiEnabled;
  const getHighlightState = (0, _useItemHighlightStateGetter.useItemHighlightStateGetter)();
  const renderData = store.use(_scatterRenderData.selectorScatterSeriesRenderData, series.id);
  const Marker = slots?.marker ?? _ScatterMarker.ScatterMarker;
  const _useSlotProps = (0, _useSlotProps2.default)({
      elementType: Marker,
      externalSlotProps: slotProps?.marker,
      additionalProps: {
        seriesId: series.id,
        size: series.markerSize
      },
      ownerState: {}
    }),
    markerProps = (0, _objectWithoutPropertiesLoose2.default)(_useSlotProps, _excluded);
  if (renderData === undefined || !revealed) {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)("g", {
      "data-series": series.id,
      className: classes.series
    });
  }
  const {
    coords,
    count
  } = renderData;
  const markers = [];
  const safeStep = Math.max(1, step);
  for (let dataIndex = start; dataIndex < count; dataIndex += safeStep) {
    // Skip off-screen points (kept in-array to keep batches stable across pan).
    if (coords[dataIndex * 3 + 2] === 0) {
      continue;
    }
    const x = coords[dataIndex * 3];
    const y = coords[dataIndex * 3 + 1];
    const dataPoint = {
      x,
      y,
      dataIndex,
      seriesId: series.id,
      type: 'scatter'
    };
    const highlightState = isInteracting ? 'none' : getHighlightState(dataPoint);
    const isItemHighlighted = highlightState === 'highlighted';
    const isItemFaded = highlightState === 'faded';
    markers.push(/*#__PURE__*/(0, _jsxRuntime.jsx)(Marker, (0, _extends2.default)({
      className: (0, _clsx.default)(classes.marker, markerProps.className),
      dataIndex: dataIndex,
      color: colorGetter(dataIndex),
      isHighlighted: isItemHighlighted,
      isFaded: isItemFaded,
      x: x,
      y: y,
      onClick: onItemClick && (event => onItemClick(event, {
        type: 'scatter',
        seriesId: series.id,
        dataIndex
      })),
      "data-highlighted": isItemHighlighted || undefined,
      "data-faded": isItemFaded || undefined
    }, skipInteractionHandlers || isInteracting ? undefined : (0, _useInteractionItemProps.getInteractionItemProps)(instance, dataPoint), markerProps), dataIndex));
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("g", {
    "data-series": series.id,
    className: classes.series,
    children: markers
  });
}

// Memoized so a reveal tick only re-renders the batch whose `revealed` changed.
const ScatterAsyncBatch = exports.ScatterAsyncBatch = /*#__PURE__*/React.memo(ScatterAsyncBatchComponent);
if (process.env.NODE_ENV !== "production") ScatterAsyncBatch.displayName = "ScatterAsyncBatch";
