// @mui/x-charts/ScatterChart/async/ScatterAsyncBatch.mjs
'use client';

import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
const _excluded = ["ownerState"];
import * as React from 'react';
import clsx from 'clsx';
import useSlotProps from '@mui/utils/useSlotProps';
import { getInteractionItemProps } from "../../hooks/useInteractionItemProps.mjs";
import { useStore } from "../../internals/store/useStore.mjs";
import { useItemHighlightStateGetter } from "../../hooks/useItemHighlightStateGetter.mjs";
import { selectorChartsIsVoronoiEnabled } from "../../internals/plugins/featurePlugins/useChartClosestPoint/index.mjs";
import { ScatterMarker } from "../ScatterMarker.mjs";
import { useUtilityClasses } from "../scatterClasses.mjs";
import { useChartsContext } from "../../context/ChartsProvider/index.mjs";
import { selectorScatterSeriesRenderData } from "./scatterRenderData.selectors.mjs";
import { jsx as _jsx } from "react/jsx-runtime";
/**
 * @ignore - internal component.
 */
function ScatterAsyncBatchComponent(props) {
  const {
    series,
    colorGetter,
    onItemClick,
    slots,
    slotProps,
    start,
    step,
    revealed,
    isInteracting,
    classes: inClasses
  } = props;
  const classes = useUtilityClasses({
    classes: inClasses
  });
  const {
    instance
  } = useChartsContext();
  const store = useStore();
  const isVoronoiEnabled = store.use(selectorChartsIsVoronoiEnabled);
  const skipInteractionHandlers = isVoronoiEnabled;
  const getHighlightState = useItemHighlightStateGetter();
  const renderData = store.use(selectorScatterSeriesRenderData, series.id);
  const Marker = slots?.marker ?? ScatterMarker;
  const _useSlotProps = useSlotProps({
      elementType: Marker,
      externalSlotProps: slotProps?.marker,
      additionalProps: {
        seriesId: series.id,
        size: series.markerSize
      },
      ownerState: {}
    }),
    markerProps = _objectWithoutPropertiesLoose(_useSlotProps, _excluded);
  if (renderData === undefined || !revealed) {
    return /*#__PURE__*/_jsx("g", {
      "data-series": series.id,
      className: classes.series
    });
  }
  const {
    coords,
    count
  } = renderData;
  const markers = [];
  const safeStep = Math.max(1, step);
  for (let dataIndex = start; dataIndex < count; dataIndex += safeStep) {
    // Skip off-screen points (kept in-array to keep batches stable across pan).
    if (coords[dataIndex * 3 + 2] === 0) {
      continue;
    }
    const x = coords[dataIndex * 3];
    const y = coords[dataIndex * 3 + 1];
    const dataPoint = {
      x,
      y,
      dataIndex,
      seriesId: series.id,
      type: 'scatter'
    };
    const highlightState = isInteracting ? 'none' : getHighlightState(dataPoint);
    const isItemHighlighted = highlightState === 'highlighted';
    const isItemFaded = highlightState === 'faded';
    markers.push(/*#__PURE__*/_jsx(Marker, _extends({
      className: clsx(classes.marker, markerProps.className),
      dataIndex: dataIndex,
      color: colorGetter(dataIndex),
      isHighlighted: isItemHighlighted,
      isFaded: isItemFaded,
      x: x,
      y: y,
      onClick: onItemClick && (event => onItemClick(event, {
        type: 'scatter',
        seriesId: series.id,
        dataIndex
      })),
      "data-highlighted": isItemHighlighted || undefined,
      "data-faded": isItemFaded || undefined
    }, skipInteractionHandlers || isInteracting ? undefined : getInteractionItemProps(instance, dataPoint), markerProps), dataIndex));
  }
  return /*#__PURE__*/_jsx("g", {
    "data-series": series.id,
    className: classes.series,
    children: markers
  });
}

// Memoized so a reveal tick only re-renders the batch whose `revealed` changed.
const ScatterAsyncBatch = /*#__PURE__*/React.memo(ScatterAsyncBatchComponent);
if (process.env.NODE_ENV !== "production") ScatterAsyncBatch.displayName = "ScatterAsyncBatch";
export { ScatterAsyncBatch };
