// @mui/x-scheduler/month-view/MonthView.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MonthView = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _styles = require("@mui/material/styles");
var _useMergedRefs = require("@base-ui/utils/useMergedRefs");
var _store = require("@base-ui/utils/store");
var _useResizeObserver = require("@mui/x-internals/useResizeObserver");
var _getDayList = require("@mui/x-scheduler-internals/get-day-list");
var _internals = require("@mui/x-scheduler-internals/internals");
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _useEventCalendarView = require("@mui/x-scheduler-internals/use-event-calendar-view");
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _eventCalendarSelectors = require("@mui/x-scheduler-internals/event-calendar-selectors");
var _calendarGrid = require("@mui/x-scheduler-internals/calendar-grid");
var _useEventOccurrencesGroupedByDay = require("@mui/x-scheduler-internals/use-event-occurrences-grouped-by-day");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _clsx = _interopRequireDefault(require("clsx"));
var _MonthViewWeekRow = _interopRequireDefault(require("./month-view-row/MonthViewWeekRow"));
var _moreEventsPopover = require("../internals/components/more-events-popover");
var _EventCalendarStyledContext = require("../event-calendar/EventCalendarStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const FIXED_CELL_WIDTH = 28;
const MonthViewRoot = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MonthView'
})(({
  theme
}) => ({
  width: '100%',
  height: '100%',
  borderRadius: theme.shape.borderRadius,
  border: `1px solid ${(theme.vars || theme).palette.divider}`,
  overflow: 'hidden',
  display: 'flex',
  flexDirection: 'column',
  maxHeight: '100%',
  overflowY: 'auto'
}));
const MonthViewGrid = (0, _styles.styled)(_calendarGrid.CalendarGrid.Root, {
  name: 'MuiEventCalendar',
  slot: 'MonthViewGrid'
})({
  flex: 1,
  display: 'flex',
  flexDirection: 'column',
  minHeight: 0
});
const MonthViewHeader = (0, _styles.styled)(_calendarGrid.CalendarGrid.HeaderRow, {
  name: 'MuiEventCalendar',
  slot: 'MonthViewHeader'
})(({
  theme,
  ownerState
}) => ({
  display: 'grid',
  gridTemplateColumns: ownerState.showWeekNumber ? `${FIXED_CELL_WIDTH}px repeat(auto-fit, minmax(0, 1fr))` : 'repeat(auto-fit, minmax(0, 1fr))',
  borderBlockEnd: `1px solid ${(theme.vars || theme).palette.divider}`
}));
const MonthViewHeaderCell = (0, _styles.styled)(_calendarGrid.CalendarGrid.HeaderCell, {
  name: 'MuiEventCalendar',
  slot: 'MonthViewHeaderCell'
})(({
  theme
}) => ({
  padding: theme.spacing(1),
  textAlign: 'center',
  fontSize: theme.typography.body2.fontSize,
  lineHeight: '18px',
  '&:not(:first-of-type)': {
    borderInlineStart: `1px solid ${(theme.vars || theme).palette.divider}`
  },
  '&:focus-visible': {
    outline: 'none',
    borderRadius: theme.shape.borderRadius,
    boxShadow: `inset 0 0 0 2px ${(theme.vars || theme).palette.primary.main}`
  }
}));
const MonthViewWeekHeaderCell = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MonthViewWeekHeaderCell'
})(({
  theme
}) => ({
  padding: theme.spacing(1, 0),
  textAlign: 'center',
  fontSize: theme.typography.caption.fontSize,
  lineHeight: '18px',
  color: (theme.vars || theme).palette.text.secondary,
  fontStyle: 'italic'
}));
const MonthViewBody = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'MonthViewBody'
})({
  flex: 1,
  display: 'grid',
  gridAutoRows: 'minmax(0, 1fr)',
  position: 'relative',
  flexGrow: 1,
  overflow: 'hidden'
});
const MONTH_VIEW_ROW_TYPES = ['header', 'day-grid'];
const CELL_PADDING = 5; // theme.spacing(0.5) * 2
const DAY_NUMBER_HEADER_HEIGHT = 22; // event height (18px) + gap (4px)
const EVENT_HEIGHT = 18;
const EVENT_GAP = 4; // theme.spacing(0.5) = 4px

const MONTH_VIEW_CONFIG = {
  siblingVisibleDateGetter: ({
    state,
    delta
  }) => state.adapter.addMonths(state.adapter.startOfMonth(_schedulerSelectors.schedulerOtherSelectors.visibleDate(state)), delta),
  visibleDaysSelector: (0, _store.createSelectorMemoized)(state => state.adapter, _schedulerSelectors.schedulerOtherSelectors.visibleDate, _eventCalendarSelectors.eventCalendarPreferenceSelectors.showWeekends, _eventCalendarSelectors.eventCalendarPreferenceSelectors.weekStartsOn, (adapter, visibleDate, showWeekends, weekStartsOn) => (0, _getDayList.getDayList)({
    adapter,
    start: (0, _internals.getStartOfWeek)(adapter, adapter.startOfMonth(visibleDate), weekStartsOn),
    end: (0, _internals.getEndOfWeek)(adapter, adapter.endOfMonth(visibleDate), weekStartsOn),
    excludeWeekends: !showWeekends
  }))
};

/**
 * A Month View to use inside the Event Calendar.
 */
const MonthView = exports.MonthView = /*#__PURE__*/React.memo(/*#__PURE__*/React.forwardRef(function MonthView(props, forwardedRef) {
  // Context hooks
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const {
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();

  // Ref hooks
  const containerRef = React.useRef(null);
  const handleRef = (0, _useMergedRefs.useMergedRefs)(forwardedRef, containerRef);
  const cellRef = React.useRef(null);

  // Selector hooks
  const showWeekNumber = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarPreferenceSelectors.showWeekNumber);
  const showWeekends = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarPreferenceSelectors.showWeekends);

  // State hooks
  const [maxEvents, setMaxEvents] = React.useState(2);

  // Feature hooks
  const {
    days
  } = (0, _useEventCalendarView.useEventCalendarView)(MONTH_VIEW_CONFIG);
  const weeks = React.useMemo(() => {
    const chunkSize = showWeekends ? 7 : 5;
    const result = [];
    for (let i = 0; i < days.length; i += chunkSize) {
      result.push(days.slice(i, i + chunkSize));
    }
    return result;
  }, [days, showWeekends]);
  const monthViewRowsPerType = React.useMemo(() => ({
    'day-grid': weeks.length
  }), [weeks.length]);
  const occurrencesMap = (0, _useEventOccurrencesGroupedByDay.useEventOccurrencesGroupedByDay)({
    days
  });
  (0, _useResizeObserver.useResizeObserver)(cellRef, () => {
    const cellHeight = cellRef.current.clientHeight;
    const eventContainerHeight = cellHeight - CELL_PADDING - DAY_NUMBER_HEADER_HEIGHT;
    const maxEventsCount = Math.floor(eventContainerHeight / (EVENT_HEIGHT + EVENT_GAP));
    setMaxEvents(Math.max(1, maxEventsCount));
  }, true);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(MonthViewRoot, (0, _extends2.default)({}, props, {
    ref: handleRef,
    className: (0, _clsx.default)(props.className, classes.monthView),
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_moreEventsPopover.MoreEventsPopoverProvider, {
      children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(MonthViewGrid, {
        className: classes.monthViewGrid,
        rowTypes: MONTH_VIEW_ROW_TYPES,
        rowsPerType: monthViewRowsPerType,
        "aria-rowcount": 1 + weeks.length,
        "aria-colcount": weeks[0].length,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(MonthViewHeader, {
          className: classes.monthViewHeader,
          ownerState: {
            showWeekNumber
          },
          "aria-rowindex": 1,
          children: [showWeekNumber && /*#__PURE__*/(0, _jsxRuntime.jsx)(MonthViewWeekHeaderCell, {
            className: classes.monthViewWeekHeaderCell,
            "aria-hidden": "true",
            children: localeText.weekAbbreviation
          }), weeks[0].map((weekDay, dayIdx) => /*#__PURE__*/(0, _jsxRuntime.jsx)(MonthViewHeaderCell, {
            className: classes.monthViewHeaderCell,
            date: weekDay,
            "aria-colindex": dayIdx + 1,
            skipDataCurrent: true,
            children: adapter.formatByString(weekDay.value, 'ccc')
          }, weekDay.key))]
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(MonthViewBody, {
          className: classes.monthViewBody,
          children: weeks.map((week, weekIdx) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_MonthViewWeekRow.default, {
            rowIndex: weekIdx,
            maxEvents: maxEvents,
            days: week,
            occurrencesMap: occurrencesMap,
            firstDayRef: weekIdx === 0 ? cellRef : undefined
          }, weekIdx))
        })]
      })
    })
  }));
}));
if (process.env.NODE_ENV !== "production") MonthView.displayName = "MonthView";
