// @mui/x-scheduler/event-calendar/header-toolbar/view-switcher/ViewSwitcher.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ViewSwitcher = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _clsx = _interopRequireDefault(require("clsx"));
var _styles = require("@mui/material/styles");
var _ExpandMoreOutlined = _interopRequireDefault(require("@mui/icons-material/ExpandMoreOutlined"));
var _useMergedRefs = require("@base-ui/utils/useMergedRefs");
var _Button = _interopRequireDefault(require("@mui/material/Button"));
var _Menu = _interopRequireDefault(require("@mui/material/Menu"));
var _MenuItem = _interopRequireDefault(require("@mui/material/MenuItem"));
var _EventCalendarStyledContext = require("../../EventCalendarStyledContext");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["views", "onViewChange", "view", "className"];
const ViewSwitcherRoot = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'ViewSwitcher'
})({
  display: 'flex'
});
const ViewSwitcher = exports.ViewSwitcher = /*#__PURE__*/React.forwardRef(function ViewSwitcher(props, forwardedRef) {
  const {
      views,
      onViewChange,
      view,
      className
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const containerRef = React.useRef(null);
  const handleRef = (0, _useMergedRefs.useMergedRefs)(forwardedRef, containerRef);
  const {
    schedulerId,
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const [anchorEl, setAnchorEl] = React.useState(null);
  const open = Boolean(anchorEl);
  const handleMenuOpen = event => {
    setAnchorEl(event.currentTarget);
  };
  const handleMenuClose = () => {
    setAnchorEl(null);
  };
  const handleMenuItemClick = (event, newView) => {
    onViewChange(newView, event.nativeEvent);
    handleMenuClose();
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(ViewSwitcherRoot, (0, _extends2.default)({
    ref: handleRef
  }, other, {
    className: (0, _clsx.default)(className, classes.viewSwitcher),
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_Button.default, {
      className: classes.viewSwitcherButton,
      size: "medium",
      id: `${schedulerId}-view-switcher-button`,
      "aria-controls": open ? `${schedulerId}-view-switcher-menu` : undefined,
      "aria-haspopup": "true",
      "aria-expanded": open ? 'true' : undefined,
      onClick: handleMenuOpen,
      endIcon: /*#__PURE__*/(0, _jsxRuntime.jsx)(_ExpandMoreOutlined.default, {}),
      children: localeText[view]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Menu.default, {
      className: classes.viewSwitcherMenu,
      id: `${schedulerId}-view-switcher-menu`,
      anchorEl: anchorEl,
      open: open,
      onClose: handleMenuClose,
      slotProps: {
        list: {
          'aria-labelledby': `${schedulerId}-view-switcher-button`,
          role: 'listbox'
        }
      },
      children: views.map(viewItem => /*#__PURE__*/(0, _jsxRuntime.jsx)(_MenuItem.default, {
        selected: view === viewItem,
        "aria-selected": view === viewItem,
        onClick: event => handleMenuItemClick(event, viewItem),
        children: localeText[viewItem]
      }, viewItem))
    })]
  }));
});
if (process.env.NODE_ENV !== "production") ViewSwitcher.displayName = "ViewSwitcher";
