// @mui/x-scheduler/internals/components/day-time-grid/DayGridCell.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DayGridCell = DayGridCell;
var React = _interopRequireWildcard(require("react"));
var _styles = require("@mui/material/styles");
var _store = require("@base-ui/utils/store");
var _calendarGrid = require("@mui/x-scheduler-internals/calendar-grid");
var _useAdapter = require("@mui/x-scheduler-internals/use-adapter");
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _eventCalendarSelectors = require("@mui/x-scheduler-internals/event-calendar-selectors");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _event = require("../event");
var _eventDialog = require("../event-dialog");
var _EventDialog = require("../event-dialog/EventDialog");
var _eventSkeleton = require("../event-skeleton");
var _EventCalendarStyledContext = require("../../../event-calendar/EventCalendarStyledContext");
var _tokens = require("../../utils/tokens");
var _jsxRuntime = require("react/jsx-runtime");
const EVENT_HEIGHT = 22;
const DayTimeGridAllDayEventsCell = (0, _styles.styled)(_calendarGrid.CalendarGrid.DayCell, {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridAllDayEventsCell'
})(({
  theme
}) => ({
  flexGrow: 1,
  flexShrink: 0,
  flexBasis: 0,
  minWidth: 0,
  padding: theme.spacing(0.5),
  display: 'grid',
  gridTemplateRows: 'repeat(var(--row-count), minmax(auto, 18px))',
  gap: theme.spacing(0.5),
  lineHeight: '18px',
  minHeight: `calc(var(--row-count, 0) * ${EVENT_HEIGHT}px + ${theme.spacing(0.5)})`,
  '&[data-weekend]': {
    backgroundColor: (theme.vars || theme).palette.action.hover
  },
  '&:focus-visible': {
    outline: 'none',
    backgroundColor: (0, _tokens.getCellFocusBackground)(theme)
  }
}));
const DayTimeGridAllDayEventsCellEvents = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridAllDayEventsCellEvents'
})(({
  theme
}) => ({
  position: 'relative',
  display: 'grid',
  gap: theme.spacing(0.5)
}));
const DayTimeGridAllDayEventContainer = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridAllDayEventContainer'
})({
  display: 'contents'
});
function DayGridCell(props) {
  const {
    day,
    row,
    colIndex
  } = props;

  // Context hooks
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const {
    onOpen: startEditing
  } = (0, _EventDialog.useEventDialogContext)();
  const {
    schedulerId,
    classes
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();

  // Ref hooks
  const cellRef = React.useRef(null);

  // Selector hooks
  const isCreatingAnEvent = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarOccurrencePlaceholderSelectors.isCreatingInDayCell, day.value);
  const placeholder = _calendarGrid.CalendarGrid.usePlaceholderInDay(day.value, row);
  const isLoading = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.isLoading);
  const rowCount = Math.max(row.maxIndex, placeholder?.position.index ?? 0);
  React.useEffect(() => {
    if (!isCreatingAnEvent || !placeholder || !cellRef.current) {
      return;
    }
    startEditing(cellRef, placeholder);
  }, [isCreatingAnEvent, placeholder, startEditing]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridAllDayEventsCell, {
    className: classes.dayTimeGridAllDayEventsCell,
    ref: cellRef,
    value: day.value,
    addPropertiesToDroppedEvent: addPropertiesToDroppedEvent,
    style: {
      '--row-count': rowCount
    },
    "aria-labelledby": `${schedulerId}-DayTimeGridAllDayEventsHeaderCell`,
    "aria-colindex": colIndex,
    role: "gridcell",
    "data-weekend": (0, _useAdapter.isWeekend)(adapter, day.value) || undefined,
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(DayTimeGridAllDayEventsCellEvents, {
      className: classes.dayTimeGridAllDayEventsCellEvents,
      children: [isLoading && /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventSkeleton.EventSkeleton, {
        "data-variant": "day-grid"
      }), day.withPosition.map(occurrence => {
        if (occurrence.position.isInvisible) {
          return /*#__PURE__*/(0, _jsxRuntime.jsx)(_event.DayGridEvent, {
            occurrence: occurrence,
            variant: "invisible"
          }, occurrence.key);
        }
        return /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventDialog.EventDialogTrigger, {
          occurrence: occurrence,
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_event.DayGridEvent, {
            occurrence: occurrence,
            variant: "filled"
          })
        }, occurrence.key);
      }), placeholder != null && /*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridAllDayEventContainer, {
        className: classes.dayTimeGridAllDayEventContainer,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_event.DayGridEvent, {
          occurrence: placeholder,
          variant: "placeholder"
        })
      })]
    })
  });
}
/**
 * Makes sure any event dropped in the day cell is turned into an all-day event.
 */
function addPropertiesToDroppedEvent() {
  return {
    allDay: true
  };
}
