// @mui/x-scheduler/internals/components/day-time-grid/DayTimeGrid.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DayTimeGrid = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _styles = require("@mui/material/styles");
var _useMergedRefs = require("@base-ui/utils/useMergedRefs");
var _useIsoLayoutEffect = require("@base-ui/utils/useIsoLayoutEffect");
var _store = require("@base-ui/utils/store");
var _useResizeObserver = require("@mui/x-internals/useResizeObserver");
var _useEventOccurrencesGroupedByDay = require("@mui/x-scheduler-internals/use-event-occurrences-grouped-by-day");
var _useEventOccurrencesWithDayGridPosition = require("@mui/x-scheduler-internals/use-event-occurrences-with-day-grid-position");
var _eventCalendarSelectors = require("@mui/x-scheduler-internals/event-calendar-selectors");
var _useAdapter = require("@mui/x-scheduler-internals/use-adapter");
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _calendarGrid = require("@mui/x-scheduler-internals/calendar-grid");
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _clsx = _interopRequireDefault(require("clsx"));
var _TimeGridColumn = require("./TimeGridColumn");
var _DayGridCell = require("./DayGridCell");
var _useFormatTime = require("../../../internals/hooks/useFormatTime");
var _eventUtils = require("../../utils/event-utils");
var _EventCalendarStyledContext = require("../../../event-calendar/EventCalendarStyledContext");
var _eventCalendarClasses = require("../../../event-calendar/eventCalendarClasses");
var _responsiveTypography = require("../../constants/responsiveTypography");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["days", "className"];
const HOUR_HEIGHT = 46;
const HOURS_IN_DAY = 24;
const DayTimeGridContainer = (0, _styles.styled)(_calendarGrid.CalendarGrid.Root, {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridContainer'
})(({
  theme
}) => ({
  '--fixed-cell-width': 'var(--EventCalendar-size-fixedCellWidth)',
  '--hour-height': `${HOUR_HEIGHT}px`,
  width: '100%',
  height: '100%',
  display: 'flex',
  flexDirection: 'column',
  overflow: 'hidden',
  border: `1px solid ${(theme.vars || theme).palette.divider}`,
  borderRadius: theme.shape.borderRadius,
  maxHeight: '100%'
}));
const DayTimeGridRoot = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGrid'
})({
  display: 'flex',
  flexDirection: 'column',
  flex: 1,
  overflowY: 'auto'
});
const DayTimeGridHeader = (0, _styles.styled)(_calendarGrid.CalendarGrid.HeaderRow, {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridHeader'
})(({
  theme
}) => ({
  display: 'grid',
  gridTemplateColumns: 'minmax(var(--fixed-cell-width), auto) repeat(var(--column-count), minmax(0, 1fr)) fit-content(100%)',
  width: '100%',
  borderBlockEnd: `1px solid ${(theme.vars || theme).palette.divider}`
}));
const DayTimeGridAllDayEventsGrid = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridAllDayEventsGrid'
})(({
  theme
}) => ({
  display: 'grid',
  gridTemplateColumns: 'var(--fixed-cell-width) repeat(var(--column-count), 1fr) fit-content(100%)',
  width: '100%',
  borderBlockEnd: `1px solid ${(theme.vars || theme).palette.divider}`,
  [`&[data-has-scroll] .${_eventCalendarClasses.eventCalendarClasses.dayTimeGridScrollablePlaceholder}`]: {
    overflowY: 'scroll',
    scrollbarGutter: 'stable',
    height: '100%',
    gridColumn: '-1',
    /* Make scrollbar invisible by setting transparent colors */
    scrollbarColor: 'transparent transparent',
    background: 'transparent',
    /* Webkit browsers (Chrome, Safari, Edge) */
    '&::-webkit-scrollbar': {
      background: 'transparent'
    },
    '&::-webkit-scrollbar-thumb': {
      background: 'transparent'
    },
    '&::-webkit-scrollbar-track': {
      background: 'transparent'
    }
  },
  /* Weekend columns show visible scrollbar for visual distinction */
  [`&[data-weekend][data-has-scroll] .${_eventCalendarClasses.eventCalendarClasses.dayTimeGridScrollablePlaceholder}`]: {
    background: (theme.vars || theme).palette.action.hover,
    '&::-webkit-scrollbar': {
      background: (theme.vars || theme).palette.action.hover
    }
  }
}));
const DayTimeGridAllDayEventsRow = (0, _styles.styled)(_calendarGrid.CalendarGrid.DayRow, {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridAllDayEventsRow'
})(({
  theme
}) => ({
  position: 'relative',
  backgroundColor: 'transparent',
  display: 'grid',
  gridTemplateColumns: 'subgrid',
  gridColumn: '2 / -1',
  gridRow: 1,
  width: '100%',
  height: '100%',
  '& > *': {
    borderInlineStart: `1px solid ${(theme.vars || theme).palette.divider}`
  }
}));
const DayTimeGridAllDayEventsCell = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridAllDayEventsCell'
})(({
  theme
}) => ({
  padding: theme.spacing(0.5),
  backgroundColor: (theme.vars || theme).palette.background.paper
}));
const DayTimeGridAllDayEventsHeaderCell = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridAllDayEventsHeaderCell'
})(({
  theme
}) => ({
  gridColumn: '1',
  gridRow: '1',
  fontSize: 'var(--EventCalendar-fontSize-timeText, 0.75rem)',
  fontStyle: 'italic',
  padding: theme.spacing(1),
  textAlign: 'end',
  color: (theme.vars || theme).palette.text.secondary,
  overflowWrap: 'break-word',
  wordBreak: 'break-word',
  hyphens: 'auto'
}));
const DayTimeGridHeaderContent = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridHeaderContent'
})(({
  theme
}) => ({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
  gap: theme.spacing(1),
  padding: theme.spacing(1.25)
}));
const DayTimeGridHeaderCell = (0, _styles.styled)(_calendarGrid.CalendarGrid.HeaderCell, {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridHeaderCell'
})(({
  theme
}) => ({
  display: 'flex',
  '&:focus-visible': {
    outline: `2px solid ${(theme.vars || theme).palette.primary.main}`,
    outlineOffset: -2,
    borderRadius: theme.shape.borderRadius
  },
  '&:last-of-type': {
    gridColumn: 'span 2'
  },
  '&[data-has-scroll]': {
    display: 'grid',
    gridTemplateColumns: 'subgrid'
  },
  [`&[data-has-scroll] .${_eventCalendarClasses.eventCalendarClasses.dayTimeGridScrollablePlaceholder}`]: {
    overflowY: 'scroll',
    scrollbarGutter: 'stable',
    /* Make scrollbar invisible by setting transparent colors */
    scrollbarColor: 'transparent transparent',
    background: 'transparent',
    /* Webkit browsers (Chrome, Safari, Edge) */
    '&::-webkit-scrollbar': {
      background: 'transparent'
    },
    '&::-webkit-scrollbar-thumb': {
      background: 'transparent'
    },
    '&::-webkit-scrollbar-track': {
      background: 'transparent'
    }
  }
}));
const DayTimeGridHeaderButton = (0, _styles.styled)('button', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridHeaderButton'
})(({
  theme
}) => ({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
  gap: theme.spacing(1),
  background: 'none',
  border: 'none',
  cursor: 'pointer',
  font: 'inherit',
  color: 'inherit',
  padding: theme.spacing(0.25),
  flexGrow: 1,
  '&:focus-visible': {
    outline: `2px solid ${(theme.vars || theme).palette.primary.main}`,
    outlineOffset: -2,
    borderRadius: theme.shape.borderRadius
  }
}));
const DayTimeGridHeaderDayName = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridHeaderDayName'
})(({
  theme
}) => ({
  fontSize: theme.typography.body2.fontSize,
  color: (theme.vars || theme).palette.text.secondary,
  lineHeight: 1,
  '[data-current] &': {
    color: (theme.vars || theme).palette.primary.main
  }
}));
const DayTimeGridHeaderDayNumber = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridHeaderDayNumber'
})(({
  theme
}) => ({
  fontSize: 'var(--EventCalendar-fontSize-dayNumber, 1.5rem)',
  lineHeight: 1,
  width: 46,
  height: 46,
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  borderRadius: '50%',
  'button:hover &': {
    backgroundColor: (theme.vars || theme).palette.action.hover
  },
  '[data-current] &': {
    backgroundColor: (theme.vars || theme).palette.primary.main,
    color: (theme.vars || theme).palette.primary.contrastText
  },
  '[data-current] button:hover &': {
    backgroundColor: (theme.vars || theme).palette.primary.dark
  },
  [`@container ${_responsiveTypography.EVENT_CALENDAR_CONTAINER_NAME} (width < ${_responsiveTypography.RESPONSIVE_TYPOGRAPHY_BREAKPOINT_SM}px)`]: {
    width: 32,
    height: 32
  }
}));
const DayTimeGridBody = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridBody'
})({
  flex: 1,
  height: '100%'
});
const DayTimeGridScrollableContent = (0, _styles.styled)(_calendarGrid.CalendarGrid.TimeScrollableContent, {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridScrollableContent'
})({
  display: 'flex',
  height: `calc(var(--hour-height) * 24)`,
  position: 'relative',
  overflowY: 'auto',
  overflowX: 'clip'
});
const DayTimeGridTimeAxis = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridTimeAxis'
})({
  display: 'flex',
  flexDirection: 'column',
  flexShrink: 0,
  width: 'var(--fixed-cell-width)'
});
const DayTimeGridTimeAxisCell = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridTimeAxisCell'
})(({
  theme
}) => ({
  height: `calc(100% / ${HOURS_IN_DAY})`,
  lineHeight: `calc(100% / ${HOURS_IN_DAY})`,
  paddingInline: theme.spacing(1),
  textAlign: 'end',
  '&:not(:first-of-type)::after': {
    content: '""',
    position: 'absolute',
    left: 'var(--fixed-cell-width)',
    right: 0,
    borderBlockEnd: `1px solid ${(theme.vars || theme).palette.divider}`,
    top: 'calc(var(--hour) * var(--hour-height))',
    zIndex: 1
  }
}));
const DayTimeGridTimeAxisText = (0, _styles.styled)('time', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridTimeAxisText'
})(({
  theme
}) => ({
  fontSize: 'var(--EventCalendar-fontSize-timeText, 0.75rem)',
  lineHeight: 'calc(100% / 24)',
  color: (theme.vars || theme).palette.text.secondary,
  whiteSpace: 'nowrap'
}));
const DayTimeGridGrid = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridGrid'
})(({
  theme
}) => ({
  display: 'grid',
  gridTemplateColumns: 'repeat(auto-fit, minmax(0, 1fr))',
  width: '100%',
  position: 'relative',
  [`&[data-has-scroll] .${_eventCalendarClasses.eventCalendarClasses.dayTimeGridColumn}:last-of-type`]: {
    borderInlineEnd: `1px solid ${(theme.vars || theme).palette.divider}`
  }
}));
const DayTimeGrid = exports.DayTimeGrid = /*#__PURE__*/React.forwardRef(function DayTimeGrid(props, forwardedRef) {
  const {
      days,
      className
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);

  // Context hooks
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const {
    schedulerId,
    classes,
    localeText
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();

  // Ref hooks
  const bodyRef = React.useRef(null);
  const allDayHeaderWrapperRef = React.useRef(null);
  const containerRef = React.useRef(null);
  const handleRef = (0, _useMergedRefs.useMergedRefs)(forwardedRef, containerRef);

  // Selector hooks
  const hasDayView = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarViewSelectors.hasDayView);
  const now = (0, _store.useStore)(store, _schedulerSelectors.schedulerNowSelectors.nowUpdatedEveryMinute);
  const showCurrentTimeIndicator = (0, _store.useStore)(store, _schedulerSelectors.schedulerNowSelectors.showCurrentTimeIndicator);

  // Feature hooks
  const occurrencesMap = (0, _useEventOccurrencesGroupedByDay.useEventOccurrencesGroupedByDay)({
    days
  });
  const occurrences = (0, _useEventOccurrencesWithDayGridPosition.useEventOccurrencesWithDayGridPosition)({
    days,
    occurrencesMap,
    shouldAddPosition: React.useCallback(occurrence => (0, _eventUtils.isOccurrenceAllDayOrMultipleDay)(occurrence, adapter), [adapter])
  });
  const formatTime = (0, _useFormatTime.useFormatTime)();
  const [hasScroll, setHasScroll] = React.useState(false);
  const {
    start,
    end
  } = React.useMemo(() => ({
    start: days[0].value,
    end: adapter.endOfDay(days[days.length - 1].value)
  }), [adapter, days]);
  const isTodayInView = React.useMemo(() => adapter.isWithinRange(now, [adapter.startOfDay(days[0].value), adapter.endOfDay(days[days.length - 1].value)]), [adapter, days, now]);
  const updateHasScroll = React.useCallback(() => {
    const body = bodyRef.current;
    const allDayHeader = allDayHeaderWrapperRef.current;
    if (!body || !allDayHeader) {
      return;
    }
    setHasScroll(body.scrollHeight > body.clientHeight);
  }, []);
  (0, _useIsoLayoutEffect.useIsoLayoutEffect)(updateHasScroll, [occurrencesMap, updateHasScroll]);
  (0, _useResizeObserver.useResizeObserver)(bodyRef, updateHasScroll);
  const lastIsWeekend = (0, _useAdapter.isWeekend)(adapter, days[days.length - 1].value);
  const template = adapter.date('2020-01-01T00:00:00', 'default');
  const renderHeaderContent = day => /*#__PURE__*/(0, _jsxRuntime.jsxs)(DayTimeGridHeaderContent, {
    className: classes.dayTimeGridHeaderContent,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridHeaderDayName, {
      className: classes.dayTimeGridHeaderDayName,
      children: adapter.format(day.value, 'weekday3Letters')
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridHeaderDayNumber, {
      className: classes.dayTimeGridHeaderDayNumber,
      children: adapter.format(day.value, 'dayOfMonth')
    })]
  });
  if (process.env.NODE_ENV !== "production") renderHeaderContent.displayName = "renderHeaderContent";
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(DayTimeGridContainer, (0, _extends2.default)({
    ref: handleRef
  }, other, {
    className: (0, _clsx.default)(className, classes.dayTimeGridContainer),
    "aria-rowcount": 3,
    "aria-colcount": days.length,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(DayTimeGridHeader, {
      className: classes.dayTimeGridHeader,
      as: _calendarGrid.CalendarGrid.HeaderRow,
      "aria-rowindex": 1,
      style: {
        '--column-count': days.length
      },
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridAllDayEventsCell, {
        className: classes.dayTimeGridAllDayEventsCell,
        "aria-hidden": "true"
      }), days.map((day, index) => /*#__PURE__*/(0, _jsxRuntime.jsxs)(DayTimeGridHeaderCell, {
        className: classes.dayTimeGridHeaderCell,
        date: day,
        ariaLabelFormat: `${adapter.formats.weekday} ${adapter.formats.dayOfMonth}`,
        "aria-colindex": index + 1,
        "data-has-scroll": index === days.length - 1 && hasScroll || undefined,
        children: [hasDayView ? /*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridHeaderButton, {
          className: classes.dayTimeGridHeaderButton,
          type: "button",
          onClick: event => store.switchToDay(day.value, event),
          tabIndex: -1,
          children: renderHeaderContent(day)
        }) : renderHeaderContent(day), index === days.length - 1 && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: classes.dayTimeGridScrollablePlaceholder
        })]
      }, day.key))]
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(DayTimeGridAllDayEventsGrid, {
      className: classes.dayTimeGridAllDayEventsGrid,
      ref: allDayHeaderWrapperRef,
      "data-weekend": lastIsWeekend || undefined,
      "data-has-scroll": hasScroll || undefined,
      style: {
        '--column-count': days.length
      },
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridAllDayEventsHeaderCell, {
        className: classes.dayTimeGridAllDayEventsHeaderCell,
        id: `${schedulerId}-DayTimeGridAllDayEventsHeaderCell`,
        "aria-hidden": "true",
        children: localeText.allDay
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridAllDayEventsRow, {
        as: _calendarGrid.CalendarGrid.DayRow,
        start: start,
        end: end,
        className: classes.dayTimeGridAllDayEventsRow,
        children: occurrences.days.map((day, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_DayGridCell.DayGridCell, {
          day: day,
          row: occurrences,
          colIndex: index + 1
        }, day.key))
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        className: classes.dayTimeGridScrollablePlaceholder
      })]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridRoot, {
      className: classes.dayTimeGrid,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridBody, {
        className: classes.dayTimeGridBody,
        ref: bodyRef,
        children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(DayTimeGridScrollableContent, {
          className: classes.dayTimeGridScrollableContent,
          as: _calendarGrid.CalendarGrid.TimeScrollableContent,
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridTimeAxis, {
            className: classes.dayTimeGridTimeAxis,
            "aria-hidden": "true",
            children: Array.from({
              length: 24
            }, (_, hour) => /*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridTimeAxisCell, {
              className: classes.dayTimeGridTimeAxisCell,
              style: {
                '--hour': hour
              },
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridTimeAxisText, {
                className: classes.dayTimeGridTimeAxisText,
                as: "time",
                children: hour === 0 ? null : formatTime(adapter.setHours(template, hour))
              })
            }, hour))
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridGrid, {
            className: classes.dayTimeGridGrid,
            role: "row",
            "aria-rowindex": 3,
            "data-has-scroll": hasScroll || undefined,
            children: occurrences.days.map((day, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_TimeGridColumn.TimeGridColumn, {
              day: day,
              index: index,
              colIndex: index + 1,
              showCurrentTimeIndicator: showCurrentTimeIndicator && isTodayInView
            }, day.key))
          })]
        })
      })
    })]
  }));
});
if (process.env.NODE_ENV !== "production") DayTimeGrid.displayName = "DayTimeGrid";
