// @mui/x-scheduler/internals/components/day-time-grid/TimeGridColumn.js
"use strict";
'use client';

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TimeGridColumn = TimeGridColumn;
var React = _interopRequireWildcard(require("react"));
var _styles = require("@mui/material/styles");
var _store = require("@base-ui/utils/store");
var _calendarGrid = require("@mui/x-scheduler-internals/calendar-grid");
var _useEventCalendarStoreContext = require("@mui/x-scheduler-internals/use-event-calendar-store-context");
var _useAdapter = require("@mui/x-scheduler-internals/use-adapter");
var _useAdapterContext = require("@mui/x-scheduler-internals/use-adapter-context");
var _useEventOccurrencesWithTimelinePosition = require("@mui/x-scheduler-internals/use-event-occurrences-with-timeline-position");
var _eventCalendarSelectors = require("@mui/x-scheduler-internals/event-calendar-selectors");
var _schedulerSelectors = require("@mui/x-scheduler-internals/scheduler-selectors");
var _eventSkeleton = require("../event-skeleton");
var _EventDialog = require("../event-dialog/EventDialog");
var _EventCalendarStyledContext = require("../../../event-calendar/EventCalendarStyledContext");
var _tokens = require("../../utils/tokens");
var _DayTimeGridInternalRenderersContext = require("./DayTimeGridInternalRenderersContext");
var _jsxRuntime = require("react/jsx-runtime");
const DayTimeGridColumn = (0, _styles.styled)(_calendarGrid.CalendarGrid.TimeColumn, {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridColumn'
})(({
  theme
}) => ({
  borderInlineStart: `1px solid ${(theme.vars || theme).palette.divider}`,
  flexGrow: 1,
  flexShrink: 0,
  flexBasis: 0,
  minWidth: 0,
  position: 'relative',
  '&[data-weekend]': {
    backgroundColor: (theme.vars || theme).palette.action.hover
  },
  '&:focus-visible': {
    outline: 'none',
    backgroundColor: (0, _tokens.getCellFocusBackground)(theme)
  }
}));
const DayTimeGridColumnInteractiveLayer = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridColumnInteractiveLayer'
})({
  position: 'absolute',
  top: 0,
  bottom: 0,
  left: 0,
  right: 0
});
const DayTimeGridCurrentTimeIndicator = (0, _styles.styled)(_calendarGrid.CalendarGrid.CurrentTimeIndicator, {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridCurrentTimeIndicator'
})(({
  theme
}) => ({
  position: 'absolute',
  zIndex: 2,
  top: 'var(--y-position)',
  left: 0,
  right: -1,
  height: 0,
  borderTop: `2px solid ${(theme.vars || theme).palette.primary.main}`
}));
const DayTimeGridCurrentTimeIndicatorCircle = (0, _styles.styled)('span', {
  name: 'MuiEventCalendar',
  slot: 'DayTimeGridCurrentTimeIndicatorCircle'
})(({
  theme
}) => ({
  position: 'absolute',
  zIndex: 1,
  left: -5,
  top: -5,
  width: 8,
  height: 8,
  borderRadius: '50%',
  backgroundColor: (theme.vars || theme).palette.primary.main
}));
function TimeGridColumn(props) {
  const {
    day,
    showCurrentTimeIndicator,
    index,
    colIndex
  } = props;
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const {
    classes
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const start = React.useMemo(() => adapter.startOfDay(day.value), [adapter, day]);
  const end = React.useMemo(() => adapter.endOfDay(day.value), [adapter, day]);
  const {
    occurrences,
    maxIndex
  } = (0, _useEventOccurrencesWithTimelinePosition.useEventOccurrencesWithTimelinePosition)({
    occurrences: day.withoutPosition,
    maxSpan: Infinity
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridColumn, {
    className: classes.dayTimeGridColumn,
    start: start,
    end: end,
    addPropertiesToDroppedEvent: addPropertiesToDroppedEvent,
    "aria-colindex": colIndex,
    "data-weekend": (0, _useAdapter.isWeekend)(adapter, day.value) || undefined,
    style: {
      '--columns-count': maxIndex
    },
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(ColumnInteractiveLayer, {
      start: start,
      end: end,
      showCurrentTimeIndicator: showCurrentTimeIndicator,
      index: index,
      occurrences: occurrences,
      maxIndex: maxIndex
    })
  });
}
function ColumnInteractiveLayer({
  start,
  end,
  showCurrentTimeIndicator,
  index,
  occurrences,
  maxIndex
}) {
  // Context hooks
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const {
    onOpen: startEditing
  } = (0, _EventDialog.useEventDialogContext)();
  const {
    classes
  } = (0, _EventCalendarStyledContext.useEventCalendarStyledContext)();
  const {
    timeGridEvent: TimeGridEvent
  } = (0, _DayTimeGridInternalRenderersContext.useDayTimeGridInternalRenderers)();

  // Ref hooks
  const columnRef = React.useRef(null);

  // Selector hooks
  const isCreatingAnEvent = (0, _store.useStore)(store, _eventCalendarSelectors.eventCalendarOccurrencePlaceholderSelectors.isCreatingInTimeRange, start, end);
  const placeholder = _calendarGrid.CalendarGrid.usePlaceholderInRange({
    start,
    end,
    occurrences,
    maxIndex
  });
  const isLoading = (0, _store.useStore)(store, _schedulerSelectors.schedulerOtherSelectors.isLoading);
  React.useEffect(() => {
    if (!isCreatingAnEvent || !placeholder || !columnRef.current) {
      return;
    }
    startEditing(columnRef, placeholder);
  }, [isCreatingAnEvent, placeholder, startEditing]);
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(DayTimeGridColumnInteractiveLayer, {
    className: classes.dayTimeGridColumnInteractiveLayer,
    ref: columnRef,
    children: [isLoading && /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventSkeleton.EventSkeleton, {
      "data-variant": "time-column"
    }), !isLoading && occurrences.map(occurrence => /*#__PURE__*/(0, _jsxRuntime.jsx)(_EventDialog.EventDialogTrigger, {
      occurrence: occurrence,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(TimeGridEvent, {
        occurrence: occurrence,
        variant: "regular"
      })
    }, occurrence.key)), placeholder != null && /*#__PURE__*/(0, _jsxRuntime.jsx)(TimeGridEvent, {
      occurrence: placeholder,
      variant: "placeholder"
    }), showCurrentTimeIndicator ? /*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridCurrentTimeIndicator, {
      className: classes.dayTimeGridCurrentTimeIndicator,
      "aria-hidden": true,
      children: index === 0 && /*#__PURE__*/(0, _jsxRuntime.jsx)(DayTimeGridCurrentTimeIndicatorCircle, {
        className: classes.dayTimeGridCurrentTimeIndicatorCircle
      })
    }) : null]
  });
}
/**
 * Makes sure any event dropped in the time grid column is turned into a non all-day event.
 */
function addPropertiesToDroppedEvent() {
  return {
    allDay: false
  };
}
