// @mui/x-scheduler-internals/calendar-grid/time-column/CalendarGridTimeColumn.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarGridTimeColumn = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var React = _interopRequireWildcard(require("react"));
var _store = require("@base-ui/utils/store");
var _useRenderElement = require("../../base-ui-copy/utils/useRenderElement");
var _useCompositeListItem = require("../../base-ui-copy/composite/list/useCompositeListItem");
var _CompositeListContext = require("../../base-ui-copy/composite/list/CompositeListContext");
var _useEventCalendarStoreContext = require("../../use-event-calendar-store-context");
var _useAdapterContext = require("../../use-adapter-context");
var _schedulerSelectors = require("../../scheduler-selectors");
var _constants = require("../../constants");
var _useEventCreation = require("../../internals/utils/useEventCreation");
var _useKeyboardEventCreation = require("../../internals/utils/useKeyboardEventCreation");
var _getNavigationTarget = require("../../internals/utils/getNavigationTarget");
var _CalendarGridRootContext = require("../root/CalendarGridRootContext");
var _CalendarGridTimeColumnContext = require("./CalendarGridTimeColumnContext");
var _useTimeDropTarget = require("./useTimeDropTarget");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className", "render", "style", "start", "end", "addPropertiesToDroppedEvent"];
const CalendarGridTimeColumn = exports.CalendarGridTimeColumn = /*#__PURE__*/React.forwardRef(function CalendarGridTimeColumn(componentProps, forwardedRef) {
  const {
      // Rendering props

      // Internal props
      start,
      end,
      addPropertiesToDroppedEvent
      // Props forwarded to the DOM element
    } = componentProps,
    elementProps = (0, _objectWithoutPropertiesLoose2.default)(componentProps, _excluded);
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const {
    focusedCell,
    setFocusedCell,
    rowTypes,
    rowsPerType
  } = (0, _CalendarGridRootContext.useCalendarGridRootContext)();
  const isCurrentDay = (0, _store.useStore)(store, _schedulerSelectors.schedulerNowSelectors.isCurrentDay, start);
  const {
    ref: listItemRef,
    index
  } = (0, _useCompositeListItem.useCompositeListItem)();
  const {
    elementsRef
  } = (0, _CompositeListContext.useCompositeListContext)();
  const cellRef = React.useRef(null);
  const hasFocus = focusedCell?.rowType === 'time-grid' && focusedCell?.rowIndex === 0 && focusedCell?.columnIndex === index;
  const {
    getCursorPositionInElementMs,
    ref: dropTargetRef
  } = (0, _useTimeDropTarget.useTimeDropTarget)({
    start,
    end,
    addPropertiesToDroppedEvent
  });
  const eventCreationProps = (0, _useEventCreation.useEventCreation)(({
    event,
    creationConfig
  }) => {
    const offsetMs = getCursorPositionInElementMs({
      input: {
        clientY: event.clientY
      },
      elementRef: dropTargetRef
    });
    const anchor = adapter.addMilliseconds(start, offsetMs);
    const startDate = adapter.addMinutes(anchor, -(adapter.getMinutes(anchor) % _constants.EVENT_CREATION_PRECISION_MINUTE));
    return {
      surfaceType: 'time-grid',
      start: startDate,
      end: adapter.addMinutes(startDate, creationConfig.duration),
      resourceId: null
    };
  });
  const triggerKeyboardCreation = (0, _useKeyboardEventCreation.useKeyboardEventCreation)(({
    creationConfig
  }) => {
    const noon = adapter.setHours(adapter.setMinutes(start, 0), 12);
    return {
      surfaceType: 'time-grid',
      start: noon,
      end: adapter.addMinutes(noon, creationConfig.duration),
      resourceId: null
    };
  });

  // Apply DOM focus when this cell becomes the focused cell
  React.useEffect(() => {
    if (hasFocus && cellRef.current && !cellRef.current.contains(document.activeElement)) {
      cellRef.current.focus({
        preventScroll: true
      });
    }
  }, [hasFocus]);
  const handleKeyDown = event => {
    const target = (0, _getNavigationTarget.getNavigationTarget)(event.key, 'time-grid', 0, index, {
      columnCount: elementsRef.current.length,
      rowTypes,
      rowsPerType
    });
    if (target) {
      event.preventDefault();
      setFocusedCell(target);
      return;
    }
    if (event.key === 'Enter' && event.target === event.currentTarget && triggerKeyboardCreation) {
      event.preventDefault();
      triggerKeyboardCreation();
    }
  };
  const handleFocus = event => {
    if (event.target === event.currentTarget) {
      setFocusedCell({
        rowType: 'time-grid',
        rowIndex: 0,
        columnIndex: index
      });
    }
  };
  const state = React.useMemo(() => ({
    current: isCurrentDay
  }), [isCurrentDay]);
  const contextValue = React.useMemo(() => ({
    start,
    end,
    index,
    hasFocus,
    getCursorPositionInElementMs
  }), [start, end, index, hasFocus, getCursorPositionInElementMs]);
  const keyboardProps = {
    tabIndex: 0,
    onKeyDown: handleKeyDown,
    onFocus: handleFocus
  };
  const element = (0, _useRenderElement.useRenderElement)('div', componentProps, {
    state,
    ref: [forwardedRef, dropTargetRef, listItemRef, cellRef],
    props: [{
      role: 'gridcell',
      'aria-colindex': index + 1
    }, keyboardProps, eventCreationProps, elementProps]
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_CalendarGridTimeColumnContext.CalendarGridTimeColumnContext.Provider, {
    value: contextValue,
    children: element
  });
});
if (process.env.NODE_ENV !== "production") CalendarGridTimeColumn.displayName = "CalendarGridTimeColumn";
