// @mui/x-charts/BarChart/seriesConfig/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.barSeriesConfig = void 0;
var _extremums = require("./bar/extremums");
var _seriesProcessor = _interopRequireDefault(require("./bar/seriesProcessor"));
var _legend = _interopRequireDefault(require("./bar/legend"));
var _getColor = _interopRequireDefault(require("./bar/getColor"));
var _keyboardFocusHandler = _interopRequireDefault(require("./bar/keyboardFocusHandler"));
var _tooltip = _interopRequireWildcard(require("./bar/tooltip"));
var _tooltipPosition = _interopRequireDefault(require("./bar/tooltipPosition"));
var _getSeriesWithDefaultValues = require("./bar/getSeriesWithDefaultValues");
var _useChartCartesianAxisPosition = require("../../internals/plugins/featurePlugins/useChartCartesianAxis/useChartCartesianAxisPosition.selectors");
var _identifierSerializer = require("../../internals/identifierSerializer");
var _identifierCleaner = require("../../internals/identifierCleaner");
var _descriptionGetter = _interopRequireDefault(require("./bar/descriptionGetter"));
var _useChartHighlight = require("../../internals/plugins/featurePlugins/useChartHighlight");
const barSeriesConfig = exports.barSeriesConfig = {
  seriesProcessor: _seriesProcessor.default,
  colorProcessor: _getColor.default,
  legendGetter: _legend.default,
  tooltipGetter: _tooltip.default,
  tooltipItemPositionGetter: _tooltipPosition.default,
  axisTooltipGetter: _tooltip.axisTooltipGetter,
  xExtremumGetter: _extremums.getExtremumX,
  yExtremumGetter: _extremums.getExtremumY,
  getSeriesWithDefaultValues: _getSeriesWithDefaultValues.getSeriesWithDefaultValues,
  getItemAtPosition: _useChartCartesianAxisPosition.selectorBarItemAtPosition,
  identifierSerializer: _identifierSerializer.identifierSerializerSeriesIdDataIndex,
  identifierCleaner: _identifierCleaner.identifierCleanerSeriesIdDataIndex,
  descriptionGetter: _descriptionGetter.default,
  keyboardFocusHandler: _keyboardFocusHandler.default,
  isHighlightedCreator: _useChartHighlight.createIsHighlighted,
  isFadedCreator: _useChartHighlight.createIsFaded
};
