// @mui/x-charts/BarChart/seriesConfig/index.mjs
import { getExtremumX, getExtremumY } from "./bar/extremums.mjs";
import seriesProcessor from "./bar/seriesProcessor.mjs";
import legendGetter from "./bar/legend.mjs";
import getColor from "./bar/getColor.mjs";
import keyboardFocusHandler from "./bar/keyboardFocusHandler.mjs";
import tooltipGetter, { axisTooltipGetter } from "./bar/tooltip.mjs";
import tooltipItemPositionGetter from "./bar/tooltipPosition.mjs";
import { getSeriesWithDefaultValues } from "./bar/getSeriesWithDefaultValues.mjs";
import { selectorBarItemAtPosition } from "../../internals/plugins/featurePlugins/useChartCartesianAxis/useChartCartesianAxisPosition.selectors.mjs";
import { identifierSerializerSeriesIdDataIndex } from "../../internals/identifierSerializer.mjs";
import { identifierCleanerSeriesIdDataIndex } from "../../internals/identifierCleaner.mjs";
import descriptionGetter from "./bar/descriptionGetter.mjs";
import { createIsHighlighted, createIsFaded } from "../../internals/plugins/featurePlugins/useChartHighlight/index.mjs";
export const barSeriesConfig = {
  seriesProcessor,
  colorProcessor: getColor,
  legendGetter,
  tooltipGetter,
  tooltipItemPositionGetter,
  axisTooltipGetter,
  xExtremumGetter: getExtremumX,
  yExtremumGetter: getExtremumY,
  getSeriesWithDefaultValues,
  getItemAtPosition: selectorBarItemAtPosition,
  identifierSerializer: identifierSerializerSeriesIdDataIndex,
  identifierCleaner: identifierCleanerSeriesIdDataIndex,
  descriptionGetter,
  keyboardFocusHandler,
  isHighlightedCreator: createIsHighlighted,
  isFadedCreator: createIsFaded
};
