// @mui/x-scheduler-internals/constants/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EVENT_DRAG_PRECISION_MS = exports.EVENT_DRAG_PRECISION_MINUTE = exports.EVENT_CREATION_PRECISION_MINUTE = exports.EVENT_COLORS = exports.DEFAULT_EVENT_CREATION_CONFIG = exports.AGENDA_VIEW_DAYS_AMOUNT = exports.AGENDA_MAX_HORIZON_DAYS = void 0;
const EVENT_CREATION_PRECISION_MINUTE = exports.EVENT_CREATION_PRECISION_MINUTE = 30;
const EVENT_COLORS = exports.EVENT_COLORS = ['red', 'pink', 'purple', 'indigo', 'blue', 'teal', 'green', 'lime', 'amber', 'orange', 'grey'];
const EVENT_DRAG_PRECISION_MINUTE = exports.EVENT_DRAG_PRECISION_MINUTE = 15;
const EVENT_DRAG_PRECISION_MS = exports.EVENT_DRAG_PRECISION_MS = EVENT_DRAG_PRECISION_MINUTE * 60 * 1000;

/**
 * Maximum number of days the Agenda view is allowed to scan forward
 * when looking for event occurrences.
 * This acts as a hard limit to prevent excessive iteration
 */
const AGENDA_MAX_HORIZON_DAYS = exports.AGENDA_MAX_HORIZON_DAYS = 180;

// TODO: Create a prop to allow users to customize the number of days in agenda view
const AGENDA_VIEW_DAYS_AMOUNT = exports.AGENDA_VIEW_DAYS_AMOUNT = 12;
const DEFAULT_EVENT_CREATION_CONFIG = exports.DEFAULT_EVENT_CREATION_CONFIG = {
  interaction: 'click',
  duration: 30
};
