// @mui/x-scheduler-internals/calendar-grid/use-placeholder-in-range/useCalendarGridPlaceholderInRange.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCalendarGridPlaceholderInRange = useCalendarGridPlaceholderInRange;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _useStore = require("@base-ui/utils/store/useStore");
var _useEventCalendarStoreContext = require("../../use-event-calendar-store-context");
var _schedulerSelectors = require("../../scheduler-selectors");
var _eventCalendarSelectors = require("../../event-calendar-selectors");
var _processDate = require("../../process-date");
var _useAdapterContext = require("../../use-adapter-context");
var _dragUtils = require("../../internals/utils/drag-utils");
function useCalendarGridPlaceholderInRange(parameters) {
  const {
    start,
    end,
    occurrences,
    maxIndex
  } = parameters;
  const adapter = (0, _useAdapterContext.useAdapterContext)();
  const store = (0, _useEventCalendarStoreContext.useEventCalendarStoreContext)();
  const rawPlaceholder = (0, _useStore.useStore)(store, _eventCalendarSelectors.eventCalendarOccurrencePlaceholderSelectors.placeholderInTimeRange, start, end);
  const originalEventId = (0, _dragUtils.isInternalDragOrResizePlaceholder)(rawPlaceholder) ? rawPlaceholder.eventId : null;
  const originalEvent = (0, _useStore.useStore)(store, _schedulerSelectors.schedulerEventSelectors.processedEvent, originalEventId);
  return React.useMemo(() => {
    if (!rawPlaceholder) {
      return null;
    }
    const startProcessed = (0, _processDate.processDate)(rawPlaceholder.start, adapter);
    const endProcessed = (0, _processDate.processDate)(rawPlaceholder.end, adapter);
    const timezone = adapter.getTimezone(rawPlaceholder.start);
    const sharedProperties = {
      key: 'occurrence-placeholder',
      id: originalEventId ?? 'occurrence-placeholder',
      title: originalEvent ? originalEvent.title : '',
      displayTimezone: {
        start: startProcessed,
        end: endProcessed,
        timezone
      }
    };
    if (rawPlaceholder.type === 'creation') {
      return (0, _extends2.default)({}, sharedProperties, {
        title: '',
        position: {
          firstIndex: 1,
          lastIndex: maxIndex
        }
      });
    }
    if (rawPlaceholder.type === 'external-drag') {
      return (0, _extends2.default)({}, sharedProperties, {
        title: rawPlaceholder.eventData.title ?? '',
        position: {
          firstIndex: 1,
          lastIndex: maxIndex
        }
      });
    }
    const position = occurrences.find(occurrence => occurrence.key === rawPlaceholder.occurrenceKey)?.position ?? {
      firstIndex: 1,
      lastIndex: maxIndex
    };
    return (0, _extends2.default)({}, sharedProperties, {
      position
    });
  }, [rawPlaceholder, adapter, originalEvent, originalEventId, occurrences, maxIndex]);
}
