// @mui/x-internals/disposable/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.disposeSymbol = exports.asyncDisposeSymbol = exports.DisposableStack = exports.AsyncDisposableStack = void 0;
exports.unwrapSuppressedErrors = unwrapSuppressedErrors;
var _disposableStack = _interopRequireDefault(require("core-js-pure/actual/disposable-stack"));
var _asyncDisposableStack = _interopRequireDefault(require("core-js-pure/actual/async-disposable-stack"));
/// <reference path="./core-js-pure.d.ts" />

/**
 * Well-known `Symbol.dispose`. Resolves to the native well-known symbol when
 * the runtime ships the explicit resource management proposal; otherwise falls
 * back to `Symbol.for('Symbol.dispose')`, the same registry symbol the
 * `core-js-pure` ponyfill uses, so `DisposableStack#use` finds the method.
 *
 * Consumers should key dispose methods on this constant —
 * `class Foo { [disposeSymbol]() {} }` — rather than `Symbol.dispose`, so the
 * class works on runtimes without the proposal.
 */
const disposeSymbol = exports.disposeSymbol = typeof Symbol.dispose === 'symbol' ? Symbol.dispose : Symbol.for('Symbol.dispose');

/**
 * Well-known `Symbol.asyncDispose`. See {@link disposeSymbol} for the
 * native/fallback resolution rules.
 */
const asyncDisposeSymbol = exports.asyncDisposeSymbol = typeof Symbol.asyncDispose === 'symbol' ? Symbol.asyncDispose : Symbol.for('Symbol.asyncDispose');

/**
 * Spec-compliant `DisposableStack`. Prefers the platform's native class when
 * present; falls back to the `core-js-pure` ponyfill otherwise.
 */
const DisposableStack = exports.DisposableStack = globalThis.DisposableStack ?? _disposableStack.default;

/**
 * Spec-compliant `AsyncDisposableStack`. Prefers the platform's native class
 * when present; falls back to the `core-js-pure` ponyfill otherwise.
 */
const AsyncDisposableStack = exports.AsyncDisposableStack = globalThis.AsyncDisposableStack ?? _asyncDisposableStack.default;

/**
 * Unwraps a `SuppressedError` chain (as produced by `DisposableStack.dispose()`
 * when multiple disposers throw) into a flat list, outermost failure first.
 * Returns `[error]` unchanged if it isn't a `SuppressedError`.
 */
function unwrapSuppressedErrors(error) {
  const failures = [];
  let current = error;
  while (typeof current === 'object' && current !== null && 'error' in current && 'suppressed' in current) {
    failures.push(current.error);
    current = current.suppressed;
  }
  failures.push(current);
  return failures;
}
