// @mui/x-charts/ChartsLegend/piecewiseColorLegendClasses.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useUtilityClasses = exports.piecewiseColorLegendClasses = void 0;
var _generateUtilityClass = _interopRequireDefault(require("@mui/utils/generateUtilityClass"));
var _generateUtilityClasses = _interopRequireDefault(require("@mui/utils/generateUtilityClasses"));
var _composeClasses = _interopRequireDefault(require("@mui/utils/composeClasses"));
const CORRECT_PREFIX = 'MuiPiecewiseColorLegend';
// TODO v10: remove `MuiPiecewiseColorLegendClasses`. Kept for backwards compatibility with
// users targeting the historically-incorrect prefix.
const LEGACY_PREFIX = 'MuiPiecewiseColorLegendClasses';

// Used by `useUtilityClasses` to also emit the legacy prefix in the rendered className.
function getLegendUtilityClass(slot) {
  return `${(0, _generateUtilityClass.default)(CORRECT_PREFIX, slot)} ${(0, _generateUtilityClass.default)(LEGACY_PREFIX, slot)}`;
}
const useUtilityClasses = props => {
  const {
    classes,
    direction,
    labelPosition
  } = props;
  const slots = {
    root: ['root', direction, labelPosition?.replaceAll(/-(\w)/g, match => match[1].toUpperCase())],
    minLabel: ['minLabel'],
    maxLabel: ['maxLabel'],
    item: ['item'],
    mark: ['mark'],
    label: ['label']
  };
  return (0, _composeClasses.default)(slots, getLegendUtilityClass, classes);
};
exports.useUtilityClasses = useUtilityClasses;
const piecewiseColorLegendClasses = exports.piecewiseColorLegendClasses = (0, _generateUtilityClasses.default)(CORRECT_PREFIX, ['root', 'minLabel', 'maxLabel', 'item', 'vertical', 'horizontal', 'start', 'end', 'extremes', 'inlineStart', 'inlineEnd', 'mark', 'label']);
