// @mui/x-charts/ChartsLegend/piecewiseColorLegendClasses.mjs
import generateUtilityClass from '@mui/utils/generateUtilityClass';
import generateUtilityClasses from '@mui/utils/generateUtilityClasses';
import composeClasses from '@mui/utils/composeClasses';
const CORRECT_PREFIX = 'MuiPiecewiseColorLegend';
// TODO v10: remove `MuiPiecewiseColorLegendClasses`. Kept for backwards compatibility with
// users targeting the historically-incorrect prefix.
const LEGACY_PREFIX = 'MuiPiecewiseColorLegendClasses';

// Used by `useUtilityClasses` to also emit the legacy prefix in the rendered className.
function getLegendUtilityClass(slot) {
  return `${generateUtilityClass(CORRECT_PREFIX, slot)} ${generateUtilityClass(LEGACY_PREFIX, slot)}`;
}
export const useUtilityClasses = props => {
  const {
    classes,
    direction,
    labelPosition
  } = props;
  const slots = {
    root: ['root', direction, labelPosition?.replaceAll(/-(\w)/g, match => match[1].toUpperCase())],
    minLabel: ['minLabel'],
    maxLabel: ['maxLabel'],
    item: ['item'],
    mark: ['mark'],
    label: ['label']
  };
  return composeClasses(slots, getLegendUtilityClass, classes);
};
export const piecewiseColorLegendClasses = generateUtilityClasses(CORRECT_PREFIX, ['root', 'minLabel', 'maxLabel', 'item', 'vertical', 'horizontal', 'start', 'end', 'extremes', 'inlineStart', 'inlineEnd', 'mark', 'label']);
