// @mui/x-scheduler/internals/components/EventCalendarProvider.mjs
'use client';

import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import _extends from "@babel/runtime/helpers/esm/extends";
const _excluded = ["children"];
import * as React from 'react';
import { styled } from '@mui/material/styles';
import { useId } from '@base-ui/utils/useId';
import { EventCalendarProvider as UnstyledEventCalendarProvider } from '@mui/x-scheduler-internals/event-calendar-provider';
import { eventCalendarClasses } from "../../event-calendar/eventCalendarClasses.mjs";
import { EventCalendarStyledContext } from "../../event-calendar/EventCalendarStyledContext.mjs";
import { EventDialogStyledContext } from "./event-dialog/EventDialogStyledContext.mjs";
import { SharedComponentsStyledContext } from "./SharedComponentsStyledContext.mjs";
import { EVENT_CALENDAR_DEFAULT_LOCALE_TEXT } from "../constants/defaultLocaleText.mjs";
import { responsiveTypographyContainerQueries } from "../constants/responsiveTypography.mjs";

/**
 * Root wrapper for standalone views that provides CSS variable tokens.
 * This ensures event colors work correctly outside of EventCalendar.
 *
 * The @container queries spread below redefine the effective typography
 * vars on this slot when the surrounding ResponsiveTypographyContainer
 * (which Standalone*View components wrap their tree in) crosses a tier
 * threshold. display:contents doesn't block custom-property inheritance,
 * so descendants pick up the retargeted values.
 */
import { jsx as _jsx } from "react/jsx-runtime";
const StandaloneViewRoot = styled('div', {
  name: 'MuiEventCalendar',
  slot: 'StandaloneViewRoot'
})(({
  theme
}) => _extends({
  display: 'contents',
  fontFamily: theme.typography.fontFamily,
  boxSizing: 'border-box',
  '*, *::before, *::after': {
    boxSizing: 'inherit'
  }
}, responsiveTypographyContainerQueries));
export function EventCalendarProvider(props) {
  const {
      children
    } = props,
    other = _objectWithoutPropertiesLoose(props, _excluded);
  const schedulerId = useId();
  const calendarStyledValue = React.useMemo(() => ({
    schedulerId,
    classes: eventCalendarClasses,
    localeText: EVENT_CALENDAR_DEFAULT_LOCALE_TEXT
  }), [schedulerId]);
  const dialogStyledValue = React.useMemo(() => ({
    schedulerId,
    classes: eventCalendarClasses,
    localeText: EVENT_CALENDAR_DEFAULT_LOCALE_TEXT
  }), [schedulerId]);
  const sharedComponentsStyledValue = React.useMemo(() => ({
    classes: eventCalendarClasses
  }), []);
  return /*#__PURE__*/_jsx(UnstyledEventCalendarProvider, _extends({}, other, {
    children: /*#__PURE__*/_jsx(EventCalendarStyledContext.Provider, {
      value: calendarStyledValue,
      children: /*#__PURE__*/_jsx(EventDialogStyledContext.Provider, {
        value: dialogStyledValue,
        children: /*#__PURE__*/_jsx(SharedComponentsStyledContext.Provider, {
          value: sharedComponentsStyledValue,
          children: /*#__PURE__*/_jsx(StandaloneViewRoot, {
            children: children
          })
        })
      })
    })
  }));
}
