// @mui/x-scheduler/internals/components/EventCalendarProvider.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EventCalendarProvider = EventCalendarProvider;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _styles = require("@mui/material/styles");
var _useId = require("@base-ui/utils/useId");
var _eventCalendarProvider = require("@mui/x-scheduler-internals/event-calendar-provider");
var _eventCalendarClasses = require("../../event-calendar/eventCalendarClasses");
var _EventCalendarStyledContext = require("../../event-calendar/EventCalendarStyledContext");
var _EventDialogStyledContext = require("./event-dialog/EventDialogStyledContext");
var _SharedComponentsStyledContext = require("./SharedComponentsStyledContext");
var _defaultLocaleText = require("../constants/defaultLocaleText");
var _responsiveTypography = require("../constants/responsiveTypography");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["children"];
/**
 * Root wrapper for standalone views that provides CSS variable tokens.
 * This ensures event colors work correctly outside of EventCalendar.
 *
 * The @container queries spread below redefine the effective typography
 * vars on this slot when the surrounding ResponsiveTypographyContainer
 * (which Standalone*View components wrap their tree in) crosses a tier
 * threshold. display:contents doesn't block custom-property inheritance,
 * so descendants pick up the retargeted values.
 */
const StandaloneViewRoot = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'StandaloneViewRoot'
})(({
  theme
}) => (0, _extends2.default)({
  display: 'contents',
  fontFamily: theme.typography.fontFamily,
  boxSizing: 'border-box',
  '*, *::before, *::after': {
    boxSizing: 'inherit'
  }
}, _responsiveTypography.responsiveTypographyContainerQueries));
function EventCalendarProvider(props) {
  const {
      children
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const schedulerId = (0, _useId.useId)();
  const calendarStyledValue = React.useMemo(() => ({
    schedulerId,
    classes: _eventCalendarClasses.eventCalendarClasses,
    localeText: _defaultLocaleText.EVENT_CALENDAR_DEFAULT_LOCALE_TEXT
  }), [schedulerId]);
  const dialogStyledValue = React.useMemo(() => ({
    schedulerId,
    classes: _eventCalendarClasses.eventCalendarClasses,
    localeText: _defaultLocaleText.EVENT_CALENDAR_DEFAULT_LOCALE_TEXT
  }), [schedulerId]);
  const sharedComponentsStyledValue = React.useMemo(() => ({
    classes: _eventCalendarClasses.eventCalendarClasses
  }), []);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_eventCalendarProvider.EventCalendarProvider, (0, _extends2.default)({}, other, {
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_EventCalendarStyledContext.EventCalendarStyledContext.Provider, {
      value: calendarStyledValue,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_EventDialogStyledContext.EventDialogStyledContext.Provider, {
        value: dialogStyledValue,
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_SharedComponentsStyledContext.SharedComponentsStyledContext.Provider, {
          value: sharedComponentsStyledValue,
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(StandaloneViewRoot, {
            children: children
          })
        })
      })
    })
  }));
}
