// @mui/x-scheduler/internals/components/ResponsiveTypographyContainer.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ResponsiveTypographyContainer = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _styles = require("@mui/material/styles");
var _responsiveTypography = require("../constants/responsiveTypography");
// Establishes the @container ancestor (named EVENT_CALENDAR_CONTAINER_NAME)
// and declares the tier/effective CSS custom properties. Wraps either
// EventCalendarContent (inside EventCalendarRoot) or the view rendered by a
// Standalone*View — its descendants host the @container queries that
// retarget the effective vars at each tier.
const ResponsiveTypographyContainer = exports.ResponsiveTypographyContainer = (0, _styles.styled)('div', {
  name: 'MuiEventCalendar',
  slot: 'ResponsiveTypographyContainer'
})((0, _extends2.default)({
  display: 'flex',
  flexDirection: 'column',
  flex: 1,
  width: '100%',
  height: '100%',
  minHeight: 0,
  minWidth: 0,
  containerType: 'inline-size',
  containerName: _responsiveTypography.EVENT_CALENDAR_CONTAINER_NAME
}, _responsiveTypography.responsiveTokens));
