// @mui/x-charts/BarChart/seriesConfig/bar/descriptionGetter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _getLabel = require("../../../internals/getLabel");
const descriptionGetter = params => {
  const {
    identifier,
    series,
    xAxis,
    yAxis,
    localeText
  } = params;
  const label = (0, _getLabel.getLabel)(series.label, 'tooltip');
  const value = series.data[identifier.dataIndex] ?? null;
  const isHorizontal = series.layout === 'horizontal';
  const categoryAxis = isHorizontal ? yAxis : xAxis;
  const categoryValue = categoryAxis.data?.[identifier.dataIndex] ?? null;
  const formattedValue = series.valueFormatter(value, {
    dataIndex: identifier.dataIndex
  });
  const formattedCategory = categoryAxis.valueFormatter?.(categoryValue, {
    location: 'tooltip',
    scale: categoryAxis.scale
  });
  return localeText.barDescription({
    value,
    formattedValue: formattedValue ?? '',
    categoryValue,
    formattedCategoryValue: formattedCategory ?? '',
    seriesLabel: label
  });
};
var _default = exports.default = descriptionGetter;
